---
title: "Functional enrichment practical"
author: "Olivier Sand"
date: '`r Sys.Date()`'
output:
  html_document:
    self_contained: yes
    fig_caption: yes
    highlight: zenburn
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
    code_folding: "hide"
  ioslides_presentation:
    slide_level: 2
    self_contained: no
    colortheme: dolphin
    fig_caption: yes
    fig_height: 5
    fig_width: 7
    fonttheme: structurebold
    highlight: tango
    smaller: yes
    toc: yes
    widescreen: yes
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  beamer_presentation:
    colortheme: dolphin
    fig_caption: yes
    fig_height: 5
    fig_width: 7
    fonttheme: structurebold
    highlight: tango
    incremental: no
    keep_tex: no
    slide_level: 2
    theme: Montpellier
    toc: yes
  revealjs::revealjs_presentation:
    theme: night
    transition: none
    self_contained: true
    css: ../slides.css
  slidy_presentation:
    smart: no
    slide_level: 2
    self_contained: yes
    fig_caption: yes
    fig_height: 5
    fig_width: 7
    highlight: tango
    incremental: no
    keep_md: yes
    smaller: yes
    theme: cerulean
    toc: yes
    widescreen: yes
  powerpoint_presentation:
    slide_level: 2
    fig_caption: yes
    fig_height: 5
    fig_width: 7
    toc: yes
font-import: http://fonts.googleapis.com/css?family=Risque
subtitle: DUBii 2021
font-family: Garamond
transition: linear
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE, echo=FALSE, eval=TRUE}

options(width = 300)
# options(encoding = 'UTF-8')
knitr::opts_chunk$set(
  fig.width = 7, fig.height = 5, 
  fig.path = 'figures',
  fig.align = "center", 
  size = "tiny", 
  echo = TRUE, 
  eval = TRUE, 
  warning = FALSE, 
  message = FALSE, 
  results = TRUE, 
  comment = "")

options(scipen = 12) ## Max number of digits for non-scientific notation
# knitr::asis_output("\\footnotesize")

```


```{r libraries, echo=FALSE, eval=TRUE}
requiredLib <- c("knitr","gprofiler2")
for (lib in requiredLib) {
  if (!require(lib, character.only = TRUE)) {
    install.packages(lib, )
  }
  require(lib, character.only = TRUE)
}

library(gprofiler2)

```

## Goal of this practical

This practical aims at showing how to analyze functional enrichment on a list of differentially expressed genes with gProfiler.

## Study case : The Cancer Gene Atlas (TCGA) Breast Invasive Cancer (BIC) data

cfr TP4 ?

## References

R package : <https://cran.r-project.org/web/packages/gprofiler2>

Reference manual: <https://cran.r-project.org/web/packages/gprofiler2/gprofiler2.pdf>

Vignette: <https://cran.r-project.org/web/packages/gprofiler2/vignettes/gprofiler2.html>

## Load (clusters of differentially) expressed genes

First, you need to load the gene list.

```{r data_loading}

#local_folder <- "~/Documents/DUBii/DUBii_2019/module-3-Stat-R/stat-R_2020/data/TCGA_BIC_subset"  # on my computer
local_folder <- "/shared/projects/DUBii_2019/module-3-Stat-R/stat-R_2020/data/TCGA_BIC_subset"  # on the nncr
file_name <- "BIC_log2-norm-counts_edgeR_DEG_top_1000.tsv.gz"
BIC.expr.file <- file.path(local_folder, file_name)

# Load expression
message("Loading expression file\t", BIC.expr.file)
BIC.expr <- read.table(file = BIC.expr.file, header = TRUE)

## Size of expression data
dim(BIC.expr)

## Row names correspond to genes
head(rownames(BIC.expr))

## Column names correspond to samples
head(colnames(BIC.expr))

# Get the gene IDs
DEGList <- rownames(BIC.expr)

# Remove isoform number from gene IDs
DEGListClean <- gsub("\\..*$", "", DEGList, perl = TRUE)


```


## Gene list functional enrichment analysis with gost

1. Use the gost function to get the enrichment results
2. Visualize the results as an interactive plot and as a table
3. Get a web link to share the results with your coworkers
4. Make a static plot and highlight the GO:0071840 and HPA:0300000 terms 

````{r gost}

gostres <- gost(query = DEGListClean, 
                organism = "hsapiens", ordered_query = FALSE, 
                multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, 
                measure_underrepresentation = FALSE, evcodes = FALSE, 
                user_threshold = 0.05, correction_method = "g_SCS", 
                domain_scope = "annotated", custom_bg = NULL, 
                numeric_ns = "", sources = NULL, as_short_link = FALSE)

names(gostres)
summary(gostres)
head(gostres$result)
names(gostres$meta)

# Visualization
gostplot(gostres, capped = TRUE, interactive = TRUE)


# With intersection (can be time consuming)
gostres2 <- gost(query = DEGListClean, 
                organism = "hsapiens", ordered_query = FALSE, 
                multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, 
                measure_underrepresentation = FALSE, evcodes = TRUE, 
                user_threshold = 0.05, correction_method = "g_SCS", 
                domain_scope = "annotated", custom_bg = NULL, 
                numeric_ns = "", sources = NULL)

head(gostres2$result)

# Get a web link
gostres_link <- gost(query = DEGListClean, 
                      organism = "hsapiens", ordered_query = FALSE, 
                      multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, 
                      measure_underrepresentation = FALSE, evcodes = TRUE, 
                      user_threshold = 0.05, correction_method = "g_SCS", 
                      domain_scope = "annotated", custom_bg = NULL, 
                      numeric_ns = "", sources = NULL,
                      as_short_link = TRUE)


# Static ggplot object
p <- gostplot(gostres, capped = FALSE, interactive = FALSE)
p

# Highlight terms
pp <- publish_gostplot(p, highlight_terms = c("GO:0071840", "HPA:0300000"), 
                       width = NA, height = NA, filename = NULL )
pp

# Table
publish_gosttable(gostres, highlight_terms = gostres$result[c(1:2,10,50:52,64,69,71),],
                        use_colors = TRUE, 
                        show_columns = c("source", "term_name", "term_size", "intersection_size"),
                        filename = NULL)


````


## Save a memory image of your session

Use the fonction `save.image()` to store an image of your session, in a file `functional_enrichment_memory_image.Rdata` in the  local folder specified above.
This file will contain all the data that you loaded during this session, and enable you to reload everything without having to re-execute all the steps. 

```{r memory_image}

## Define the path to the memory image file
personal_folder <- "/shared/projects/dubii2021/YOURFOLDER"  # adapt to your personal folder
memory_image_file <- file.path(personal_folder, "functional_enrichment_memory_image.Rdata")

## Save the memory image
save.image(file = memory_image_file)

message("Memory image saved in file\n\t", memory_image_file)
```

For a future session, you will be able to reload all the data with a single command:

`load([path_to_your_memory_image])`

 You will need to replace `[path_to_your_memory_image]` by your actual path. 
 In my case, the command becomes:
 
 `load("`r memory_image_file`")`

## Save your session info

For the sake of traceability, store the specifications of your R environment in the report, with the command `sessionInfo()`. This will indicate the version of R as well as of all the libraries used in this notebook. 

```{r session_info}
sessionInfo()
```
