<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jacques van Helden" />

<meta name="date" content="2021-03-11" />

<title>Tutorial : correlation and regression</title>

<script src="tuto_correlation-and-regession_files/header-attrs-2.5/header-attrs.js"></script>
<script src="tuto_correlation-and-regession_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="tuto_correlation-and-regession_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="tuto_correlation-and-regession_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="tuto_correlation-and-regession_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="tuto_correlation-and-regession_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="tuto_correlation-and-regession_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="tuto_correlation-and-regession_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="tuto_correlation-and-regession_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="tuto_correlation-and-regession_files/navigation-1.1/tabsets.js"></script>
<script src="tuto_correlation-and-regession_files/navigation-1.1/codefolding.js"></script>
<link href="tuto_correlation-and-regession_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="tuto_correlation-and-regession_files/anchor-sections-1.0/anchor-sections.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  { color: #cccccc; background-color: #303030; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ffcfaf; } /* Alert */
code span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #dca3a3; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #f0dfaf; } /* ControlFlow */
code span.ch { color: #dca3a3; } /* Char */
code span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code span.co { color: #7f9f7f; } /* Comment */
code span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code span.do { color: #7f9f7f; } /* Documentation */
code span.dt { color: #dfdfbf; } /* DataType */
code span.dv { color: #dcdccc; } /* DecVal */
code span.er { color: #c3bf9f; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #c0bed1; } /* Float */
code span.fu { color: #efef8f; } /* Function */
code span.im { } /* Import */
code span.in { color: #7f9f7f; font-weight: bold; } /* Information */
code span.kw { color: #f0dfaf; } /* Keyword */
code span.op { color: #f0efd0; } /* Operator */
code span.ot { color: #efef8f; } /* Other */
code span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code span.sc { color: #dca3a3; } /* SpecialChar */
code span.ss { color: #cc9393; } /* SpecialString */
code span.st { color: #cc9393; } /* String */
code span.va { } /* Variable */
code span.vs { color: #cc9393; } /* VerbatimString */
code span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Tutorial : correlation and regression</h1>
<h3 class="subtitle">DUBii 2020</h3>
<h4 class="author">Jacques van Helden</h4>
<h4 class="date">2021-03-11</h4>

</div>


<div id="goals-of-this-tutorial" class="section level2">
<h2>Goals of this tutorial</h2>
<p>This tutorial provides a quick reminder of the basic concepts of correlation and regression analysis, illustrated by an empirical study of a multi-omics dataset.</p>
</div>
<div id="study-case-mouse-kidney" class="section level2">
<h2>Study case : mouse kidney</h2>
<p>As study case for this tutorial, we will use multi-omics data from a study published by <a href="https://doi.org/10.1038/s41597-019-0095-5">Pavkovic et al. (2019)</a>, which combines transctiptomics (RNA-seq) and proteomics approaches to understand the molecular mechanisms underlying the kidney fibrosis pathology.</p>
<p>The authors applied two commonly used treatments to induce kidney fibrosis in mouse:</p>
<ul>
<li><p>a reversible chemical-induced injury model, denoted as <strong>FA</strong> for <strong>folic acid</strong> induced nephropathy;</p></li>
<li><p>an irreversible surgically-induced fibrosis model, denoted as <strong>UUO</strong> for <strong>unilateral uretral obstruction</strong>.</p></li>
</ul>
</div>
<div id="references-and-data-sources" class="section level2">
<h2>References and data sources</h2>
<ul>
<li><p><strong>Reference</strong>: Pavkovic, M., Pantano, L., Gerlach, C.V. et al. Multi omics analysis of fibrotic kidneys in two mouse models. Sci Data 6, 92 (2019) <a href="https://doi.org/10.1038/s41597-019-0095-5" class="uri">https://doi.org/10.1038/s41597-019-0095-5</a></p></li>
<li><p><strong>Mouse fibrotic kidney browser:</strong> <a href="http://hbcreports.med.harvard.edu/fmm/" class="uri">http://hbcreports.med.harvard.edu/fmm/</a></p></li>
<li><p>Data on Zenodo: <a href="https://zenodo.org/record/2592516" class="uri">https://zenodo.org/record/2592516</a></p></li>
</ul>
</div>
<div id="data-preparation" class="section level2">
<h2>Data preparation</h2>
<p>A description of the study case can be found in the <em><a href="https://du-bii.github.io/study-cases/#mus-musculus">Mus musculus section</a></em> of the the <a href="https://du-bii.github.io/study-cases/">DUBii study cases repository</a> repository.</p>
<p>We also provide there a detailed explanation of the <a href="https://du-bii.github.io/study-cases/Mus_musculus/pavkovic_2019/Rmd/prepare-data_pavkovic_2019.html">data preparation steps</a>:</p>
<ul>
<li>downloading the data from its original source repository,</li>
<li>exploring the datasets it with various graphical representtions,</li>
<li>computing some descriptive statistics on the different samples,</li>
<li>pre-processing,</li>
<li>storing the results in a memory image.</li>
</ul>
</div>
<div id="data-file-naming" class="section level2">
<h2>Data file naming</h2>
<p>We prepared the data from Pavkovic as a text file with tab-separated values (<strong>tsv</strong> files).</p>
<p>All the files are available on github: - <a href="https://github.com/DU-Bii/module-3-Stat-R/tree/master/stat-R_2021/data/pavkovic_2019" class="uri">https://github.com/DU-Bii/module-3-Stat-R/tree/master/stat-R_2021/data/pavkovic_2019</a></p>
<p>The files are named according to the following convention:</p>
<ul>
<li><p>the prefix indicate the data type</p>
<ul>
<li><strong>fa</strong>: folic acid induced nephropathy (reversible), transcriptome data</li>
<li><strong>pfa</strong>: folic acid induced nephropathy, proteome data</li>
<li><strong>uuo</strong>: unilateral uretral obstruction (irreversible), transcriptome data</li>
<li><strong>puuo</strong>: unilateral uretral obstruction, proteome data</li>
</ul></li>
<li><p>The suffix indicates the data normalisation</p>
<ul>
<li><strong>raw</strong>: transcriptome counts provided by the authors (note: not integer, because their raw data was already somehow normalized)</li>
<li><strong>normalized¨</strong>: transcriptome counts standardized to achieve the same third quartile for each sample</li>
<li><strong>log2</strong>: log2 transformation for proteome data</li>
</ul></li>
<li><p>the <strong>metadata</strong> files contain a short description of each sample (one row per sample). Note that the last column contains a sample-specific color specification in <a href="https://en.wikipedia.org/wiki/Web_colors#Hex_triplet">hexadecimal web color code</a> to facilitate the drawings. Don’t hesitate to chose <a href="https://htmlcolorcodes.com/">other colors</a> according to your personal taste.</p></li>
</ul>
</div>
<div id="approach-for-this-tutorial" class="section level2">
<h2>Approach for this tutorial</h2>
<p>This tutorial will consist of exercises that can be realised in a stepwise way, with alternance of working sessions and live demos of the solutions, in order to make sure that all the trainees acquire each step.</p>
</div>
<div id="data-exploration" class="section level2">
<h2>Data exploration</h2>
<p>Before computing any descriptive parameter on a dataset, I generallyi attempt to get a picture of the whole distribution.</p>
<div id="finding-a-data-file-on-github" class="section level3">
<h3>Finding a data file on github</h3>
<p>We will provide here a magic recipe to download the data from the github repository to your local folder, and to load it in R.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="do">## specify the base URL from which data files can be downloaded </span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>urlBase <span class="ot">&lt;-</span> <span class="st">&quot;https://github.com/DU-Bii/module-3-Stat-R/raw/master/stat-R_2021/data/pavkovic_2019&quot;</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Choose a specific data file</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>dataPrefix <span class="ot">&lt;-</span> <span class="st">&quot;pfa&quot;</span> <span class="do">## proteome data of folic-acid treated mouse</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>dataSuffix <span class="ot">&lt;-</span> <span class="st">&quot;raw&quot;</span> <span class="do">## no prior normalisation</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>dataFile <span class="ot">&lt;-</span> <span class="fu">paste0</span>(dataPrefix, <span class="st">&quot;_&quot;</span>, dataSuffix, <span class="st">&quot;_counts.tsv.gz&quot;</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="do">## Compose the URL to download the file from github</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>url <span class="ot">&lt;-</span> <span class="fu">file.path</span>(urlBase, dataFile)</span></code></pre></div>
</div>
<div id="loading-a-data-file-directly-from-github" class="section level3">
<h3>Loading a data file directly from github</h3>
</div>
<div id="downloading-a-data-file-and-storing-it-locally-once-forever" class="section level3">
<h3>Downloading a data file and storing it locally once forever</h3>
</div>
<div id="loading-the-local-copy-of-your-data-file" class="section level3">
<h3>Loading the local copy of your data file</h3>
</div>
<div id="exercise-writing-a-function-to-download-and-load-a-data-file" class="section level3">
<h3>Exercise: writing a function to download and load a data file</h3>
</div>
<div id="solution-a-function-to-automagically-download-and-load-a-data-file" class="section level3">
<h3>Solution: a function to automagically download and load a data file</h3>
</div>
</div>
<div id="graphical-exploration-of-the-data" class="section level2">
<h2>Graphical exploration of the data</h2>
<div id="histogram" class="section level3">
<h3>Histogram</h3>
<p>Draw an histogram of the <strong>puuo</strong> data, all samples together.</p>
</div>
<div id="box-plot" class="section level3">
<h3>Box plot</h3>
</div>
<div id="scatter-plot" class="section level3">
<h3>Scatter plot</h3>
</div>
</div>
<div id="save-your-session-info" class="section level2">
<h2>Save your session info</h2>
<p>As usual</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>R version 4.0.2 (2020-06-22)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] knitr_1.30

loaded via a namespace (and not attached):
 [1] compiler_4.0.2  magrittr_1.5    tools_4.0.2     htmltools_0.5.0 yaml_2.2.1      stringi_1.5.3   rmarkdown_2.5   stringr_1.4.0   xfun_0.19       digest_0.6.27   rlang_0.4.8     evaluate_0.14  </code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
