---
title: "Statistics with R"
subtitle: "Gene set comparison and enrichment analysis"
author: "Jacques van Helden & Olivier Sand"
date: "`r Sys.Date()`"
output:
  slidy_presentation:
    highlight: default
    incremental: no
    smart: no
    slide_level: 1
    self_contained: no
    fig_caption: no
    fig_height: 5
    fig_width: 5
    keep_md: yes
    smaller: yes
    theme: cerulean
    toc: yes
    widescreen: yes
  revealjs::revealjs_presentation:
    theme: night
    transition: none
    self_contained: true
    slide_level: 1
    css: ../slides.css
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  ioslides_presentation:
    highlight: zenburn
    incremental: no
  html_document:
    fig_caption: yes
    highlight: zenburn
    self_contained: no
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
  beamer_presentation:
    theme: Hannover #Montpellier
    colortheme: beaver
    fonttheme: professionalfonts #structurebold
    highlight: pygments #default
    fig_caption: no
    fig_height: 4
    fig_width: 5
    incremental: no
    keep_tex: no
    slide_level: 1
    toc: yes
font-import: http://fonts.googleapis.com/css?family=Risque
font-family: Garamond
transition: linear
---


```{r settings, include=FALSE, echo=FALSE, eval=TRUE}
## Install required packages
packages <- c("knitr", "kableExtra", "stringr", "limma")

for (pkg in packages) {
  if (!require(pkg, character.only = TRUE)) {
    install.packages(pkg)
  }
}

# library(formattable)

options(width = 300)

# options(encoding = 'UTF-8')

knitr::opts_chunk$set(
  fig.width = 5, fig.height = 5, 
  fig.path = 'figures/irisDeFisher_',
  fig.align = "center", 
  size = "tiny", 
  echo = TRUE, eval = TRUE, 
  warning = FALSE, message = FALSE, 
  results = TRUE, comment = "")

options(scipen = 12) ## Max number of digits for non-scientific notation
```

# Two main approches

- Gene set comparison or over-representation analysis (ORA)
  - Input: a set of functionally related genes
  - Reference: a database of annotated gene functions (GO, pathways, TF targets…)
  - Approach: evaluate the significance of the intersection (over-represented?)
  - Stat: hypergeometric test
- Gene Set Enrichment analysis
  - Input: a sorted list of genes
  - Reference: a database of annotated gene functions (GO, pathways, TF targets…)
  - Approach: evaluate the significance of the rank of the genes belonging to a reference class in the ordered list.
  - Stat: enrichment scores (alternative)

# Gene set comparison (over-representation of the intersection)

- A given organism has 6000 genes, 40 of which are involved in methionine metabolism. 
- A set of 10 genes were reported as co-expressed in an RNAseq experiment. Among them, 6 are related to methionine metabolism. 
- How significant is this observation ? More precisely, what would be the probability to observe such a correspondence by chance alone ?

```{r venn_diagram, include=FALSE, results=TRUE, echo=FALSE, eval=TRUE, out.width="50%"}
# Draw a venn diagram indicating the number of genes in a 
# co-expression cluster that are involved in methionine metobolism
VennDiagram::venn.diagram(
  list(Genome = 1:6000, Methionine_metabolism = 1:40, coexpression_cluster = 35:44),
  height = 400,
  width = 700,
  main.cex = 0.35,
  cex=0.3,
  cat.cex  = 0.25,
  lwd = c(0.5,0.5,0.5),
  cat.pos = c(-10,-45,10),
  filename = "images/venn_yeast_methionine.png", 
  main = "Genes in co-expressed cluster vs methionine metabolism", 
  imagetype = "png", 
  col = c("darkblue", "darkgreen", "darkred"), 
  cat.col = c("darkblue", "darkgreen", "darkred"))

#include_graphics(path = "images/venn_yeast_methionine.png")
```

![Venn diagram](images/venn_yeast_methionine.png)

# The hypergeometric test

- Let us define
  - g = 6000	(number of genes)
  - m = 40	(genes involved in methionine metabolism)
  - n = 5960	(genes not involved in methionine metabolism)
  - k = 10	(number of genes in the cluster)
  - x = 6  (number of methionine genes in the cluster)

- We calculate the number of possibilities for the following selections
  - C1 : 10 distinct genes among 6000

    C1 = $C_{m+n}^{k} = \frac{6000!}{10!5990!} = 1.65e^{31}$
  - C2 : 6 distinct genes among the 40 involved in methionine

    C2 = $C_{m}^{x} = \frac{40!}{6!34!} = 3.8e^{6}$
  - C3 : 4 genes among the 5960 which are not involved in methionine

    C3 = $C_{n}^{k-x} = \frac{5960!}{4!5956!} = 5.2e^{13}$
  - C4 : 6 methionine and 4 non-methionine genes

    C4 = $C2*C3$ = $C_{m}^{x}C_{n}^{k-x} = 2.0e^{20}$

# The hypergeometric test - probabilities

- P(X = 6): probability to have exactly 6 methionine genes within a selection of 10

  P(X=6) = $\frac{C4}{C1}$ = $\frac{C_{m}^{x}C_{n}^{k-x}}{C_{m+n}^{k}} = 1.219e^{11}$

- P(X >= 6) : probability to have at least 6 methionine genes within a selection of 10

  P(X≥6) = $\sum_{i=x}^{k}\frac{C_{m}^{i}C_{n}^{k-i}}{C_{m+n}^{k}} = 1.222e^{11}$

# ORA - precautions to take

- Define your universe (background)
  - set with all the genes susceptible to be part of your analysis
- Not so simple
  - all genes in genomic annotations ?
  - all genes with at least one annotation in the ontology you used ?
  - all coding genes ?
  - genes on a biochip ?
  - genes / proteins detected by an experimental approach (RNAseq, proteomics)
  - genes reachable by your approach (ex : miRNA targets, Godard et al., 2015)
- multiple-test corrections
  - correction choice (adjusted P-values : Bonferroni correction, Benjamini-Hochberg FDR...)
  - inter-test dependancies corrections

# Practical - Annotating a group of differentially expressed genes (DEG)
- DEG:RNA-SEQ analysis of bdf1-Y187F-Y354F mutant strain during yeast sporulation
- tool:g:GOSt from gProfiler (https://biit.cs.ut.ee/gprofiler/gost)
- documentation:https://biit.cs.ut.ee/gprofiler/page/docs
- Goal:
    - detect functions (biological process, pathways, regulation...) associated with the set of DEG
    - interpret the resuls
- What about a negative control ?

# Practical - DEG
- https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE89530
- organism:Saccharomyces cerevisiae
- The aim of this study is to compare NGS-derived yeast transcriptome profiling (RNA-seq) of wild-type and bdf1-Y187F-Y354F mutant strains after sporulation induction (time points: 0h 4h and 8h)
- design:S. cerevisiae wild-type and bdf1-Y187F-Y354F mutant strains were collected 0h, 4h and 8h after sporulation induction in triplicates. mRNA were purifired, prepared and sequenced using Illumina HiSeq 2000 sequencer
- 0h:https://github.com/DU-Bii/module-3-Stat-R/blob/master/stat-R_2019/seance_5/results/bdf1_Y187F_Y354F_mutant_0__vs__Wild_type_0_DESeq2_positive_geneIDs.txt
- 4h:https://github.com/DU-Bii/module-3-Stat-R/blob/master/stat-R_2019/seance_5/results/bdf1_Y187F_Y354F_mutant_4__vs__Wild_type_4_DESeq2_positive_geneIDs.txt

# Practical - Results

Screenshots ?

# Gene Set Enrichment Analysis
- GSEA
- since 2006
- determines whether an a priori defined set of genes shows statistically significant, concordant differences between two biological states
- http://software.broadinstitute.org/gsea/index.jsp
- MSigDB (The Molecular Signatures Database) : collection of annotated gene sets
- R package : https://bioconductor.org/packages/release/bioc/html/GSEABase.html

![GSEA](images/GSEA-homegraphic.gif)

# GSEA principle
- All genes are sorted according to some criterion (e.g. differential expression p-value, correlation of expression with other variables, …).
- Each graph compares the ranked gene list with one reference class (e.g. one biological process).
- Black bars denote genes belonging to the reference class.
- The green curve estimates, at each level i, the degree of over-representation of the reference genes in the i top-ranking genes.

![GSEA screenshot](images/GSEA_screenshot_1.png)

# Practical


