<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
 "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <meta name="author" content="Jacques van Helden &amp; Olivier Sand" />
  <meta name="date" content="2020-05-31" />
  <title>Statistics with R</title>
  <style type="text/css">code{white-space: pre;}</style>
  <script src="cours_enrichment_files/jquery-1.11.3/jquery.min.js"></script>
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <link href="cours_enrichment_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
  <script src="cours_enrichment_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
  <script src="cours_enrichment_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
  <script src="cours_enrichment_files/bootstrap-3.3.5/shim/respond.min.js"></script>
  <link href="cours_enrichment_files/slidy-2/styles/slidy.css" rel="stylesheet" />
  <script src="cours_enrichment_files/slidy-2/scripts/slidy.js"></script>
  <script src="cours_enrichment_files/slidy_shiny-1/slidy_shiny.js"></script>
</head>
<body>
<div class="slide titlepage">
  <h1 class="title">Statistics with R</h1>
  <h1 class="subtitle">Gene set comparison and enrichment analysis</h1>
  <p class="author">
Jacques van Helden &amp; Olivier Sand
  </p>
  <p class="date">2020-05-31</p>
</div>
<div id="two-main-approches" class="slide section level1">
<h1>Two main approches</h1>
<ul>
<li>Gene set comparison or over-representation analysis (ORA)
<ul>
<li>Input: a set of functionally related genes</li>
<li>Reference: a database of annotated gene functions (GO, pathways, TF targets…)</li>
<li>Approach: evaluate the significance of the intersection (over-represented?)</li>
<li>Stat: hypergeometric test</li>
</ul></li>
<li>Gene Set Enrichment analysis
<ul>
<li>Input: a sorted list of genes</li>
<li>Reference: a database of annotated gene functions (GO, pathways, TF targets…)</li>
<li>Approach: evaluate the significance of the rank of the genes belonging to a reference class in the ordered list.</li>
<li>Stat: enrichment scores (alternative)</li>
</ul></li>
</ul>
</div>
<div id="gene-set-comparison-over-representation-of-the-intersection" class="slide section level1">
<h1>Gene set comparison (over-representation of the intersection)</h1>
<ul>
<li>A given organism has 6000 genes, 40 of which are involved in methionine metabolism.</li>
<li>A set of 10 genes were reported as co-expressed in an RNAseq experiment. Among them, 6 are related to methionine metabolism.</li>
<li>How significant is this observation ? More precisely, what would be the probability to observe such a correspondence by chance alone ?</li>
</ul>
<p><img src="images/venn_yeast_methionine.png" alt="Venn diagram" /></p>
</div>
<div id="the-hypergeometric-test" class="slide section level1">
<h1>The hypergeometric test</h1>
<ul>
<li>Let us define
<ul>
<li>g = 6000 (number of genes)</li>
<li>m = 40 (genes involved in methionine metabolism)</li>
<li>n = 5960 (genes not involved in methionine metabolism)</li>
<li>k = 10 (number of genes in the cluster)</li>
<li>x = 6 (number of methionine genes in the cluster)</li>
</ul></li>
<li>We calculate the number of possibilities for the following selections
<ul>
<li><p>C1 : 10 distinct genes among 6000</p>
C1 = <span class="math inline">\(C_{m+n}^{k} = \frac{6000!}{10!5990!} = 1.65e^{31}\)</span></li>
<li><p>C2 : 6 distinct genes among the 40 involved in methionine</p>
C2 = <span class="math inline">\(C_{m}^{x} = \frac{40!}{6!34!} = 3.8e^{6}\)</span></li>
<li><p>C3 : 4 genes among the 5960 which are not involved in methionine</p>
C3 = <span class="math inline">\(C_{n}^{k-x} = \frac{5960!}{4!5956!} = 5.2e^{13}\)</span></li>
<li><p>C4 : 6 methionine and 4 non-methionine genes</p>
<p>C4 = <span class="math inline">\(C2*C3\)</span> = <span class="math inline">\(C_{m}^{x}C_{n}^{k-x} = 2.0e^{20}\)</span></p></li>
</ul></li>
</ul>
</div>
<div id="the-hypergeometric-test---probabilities" class="slide section level1">
<h1>The hypergeometric test - probabilities</h1>
<ul>
<li><p>P(X = 6): probability to have exactly 6 methionine genes within a selection of 10</p>
<p>P(X=6) = <span class="math inline">\(\frac{C4}{C1}\)</span> = <span class="math inline">\(\frac{C_{m}^{x}C_{n}^{k-x}}{C_{m+n}^{k}} = 1.219e^{11}\)</span></p></li>
<li><p>P(X &gt;= 6) : probability to have at least 6 methionine genes within a selection of 10</p>
<p>P(X≥6) = <span class="math inline">\(\sum_{i=x}^{k}\frac{C_{m}^{i}C_{n}^{k-i}}{C_{m+n}^{k}} = 1.222e^{11}\)</span></p></li>
</ul>
</div>
<div id="ora---precautions-to-take" class="slide section level1">
<h1>ORA - precautions to take</h1>
<ul>
<li>Define your universe (background)
<ul>
<li>set with all the genes susceptible to be part of your analysis</li>
</ul></li>
<li>Not so simple
<ul>
<li>all genes in genomic annotations ?</li>
<li>all genes with at least one annotation in the ontology you used ?</li>
<li>all coding genes ?</li>
<li>genes on a biochip ?</li>
<li>genes / proteins detected by an experimental approach (RNAseq, proteomics)</li>
<li>genes reachable by your approach (ex : miRNA targets, Godard et al., 2015)</li>
</ul></li>
<li>multiple-test corrections
<ul>
<li>correction choice (adjusted P-values : Bonferroni correction, Benjamini-Hochberg FDR…)</li>
<li>inter-test dependancies corrections</li>
</ul></li>
</ul>
</div>
<div id="practical---annotating-a-group-of-differentially-expressed-genes-deg" class="slide section level1">
<h1>Practical - Annotating a group of differentially expressed genes (DEG)</h1>
<ul>
<li>DEG:RNA-SEQ analysis of bdf1-Y187F-Y354F mutant strain during yeast sporulation</li>
<li><a href="tool:g:GOSt" class="uri">tool:g:GOSt</a> from gProfiler (<a href="https://biit.cs.ut.ee/gprofiler/gost" class="uri">https://biit.cs.ut.ee/gprofiler/gost</a>)</li>
<li>documentation:<a href="https://biit.cs.ut.ee/gprofiler/page/docs" class="uri">https://biit.cs.ut.ee/gprofiler/page/docs</a></li>
<li>Goal:
<ul>
<li>detect functions (biological process, pathways, regulation…) associated with the set of DEG</li>
<li>interpret the resuls</li>
</ul></li>
<li>What about a negative control ?</li>
</ul>
</div>
<div id="practical---deg" class="slide section level1">
<h1>Practical - DEG</h1>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE89530" class="uri">https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE89530</a></li>
<li>organism:Saccharomyces cerevisiae</li>
<li>The aim of this study is to compare NGS-derived yeast transcriptome profiling (RNA-seq) of wild-type and bdf1-Y187F-Y354F mutant strains after sporulation induction (time points: 0h 4h and 8h)</li>
<li>design:S. cerevisiae wild-type and bdf1-Y187F-Y354F mutant strains were collected 0h, 4h and 8h after sporulation induction in triplicates. mRNA were purifired, prepared and sequenced using Illumina HiSeq 2000 sequencer</li>
<li>0h:<a href="https://github.com/DU-Bii/module-3-Stat-R/blob/master/stat-R_2019/seance_5/results/bdf1_Y187F_Y354F_mutant_0__vs__Wild_type_0_DESeq2_positive_geneIDs.txt" class="uri">https://github.com/DU-Bii/module-3-Stat-R/blob/master/stat-R_2019/seance_5/results/bdf1_Y187F_Y354F_mutant_0__vs__Wild_type_0_DESeq2_positive_geneIDs.txt</a></li>
<li>4h:<a href="https://github.com/DU-Bii/module-3-Stat-R/blob/master/stat-R_2019/seance_5/results/bdf1_Y187F_Y354F_mutant_4__vs__Wild_type_4_DESeq2_positive_geneIDs.txt" class="uri">https://github.com/DU-Bii/module-3-Stat-R/blob/master/stat-R_2019/seance_5/results/bdf1_Y187F_Y354F_mutant_4__vs__Wild_type_4_DESeq2_positive_geneIDs.txt</a></li>
</ul>
</div>
<div id="practical---results" class="slide section level1">
<h1>Practical - Results</h1>
<p>Screenshots ?</p>
</div>
<div id="gene-set-enrichment-analysis" class="slide section level1">
<h1>Gene Set Enrichment Analysis</h1>
<ul>
<li>GSEA</li>
<li>since 2006</li>
<li>determines whether an a priori defined set of genes shows statistically significant, concordant differences between two biological states</li>
<li><a href="http://software.broadinstitute.org/gsea/index.jsp" class="uri">http://software.broadinstitute.org/gsea/index.jsp</a></li>
<li>MSigDB (The Molecular Signatures Database) : collection of annotated gene sets</li>
<li>R package : <a href="https://bioconductor.org/packages/release/bioc/html/GSEABase.html" class="uri">https://bioconductor.org/packages/release/bioc/html/GSEABase.html</a></li>
</ul>
<p><img src="images/GSEA-homegraphic.gif" alt="GSEA" /></p>
</div>
<div id="gsea-principle" class="slide section level1">
<h1>GSEA principle</h1>
<ul>
<li>All genes are sorted according to some criterion (e.g. differential expression p-value, correlation of expression with other variables, …).</li>
<li>Each graph compares the ranked gene list with one reference class (e.g. one biological process).</li>
<li>Black bars denote genes belonging to the reference class.</li>
<li>The green curve estimates, at each level i, the degree of over-representation of the reference genes in the i top-ranking genes.</li>
</ul>
<p><img src="images/GSEA_screenshot_1.png" alt="GSEA screenshot" /></p>
</div>
<div id="practical" class="slide section level1">
<h1>Practical</h1>
</div>

  <!-- dynamically load mathjax for compatibility with self-contained -->
  <script>
    (function () {
      var script = document.createElement("script");
      script.type = "text/javascript";
      script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
      document.getElementsByTagName("head")[0].appendChild(script);
    })();
  </script>

</body>
</html>
