<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Anne Badel &amp; Jacques van Helden" />

<meta name="date" content="2021-03-29" />

<title>Tutorial: machine-learning with TGCA BIC transcriptome</title>

<script src="02_clustering_TCGA-BIC_tuto-tp_files/header-attrs-2.7/header-attrs.js"></script>
<script src="02_clustering_TCGA-BIC_tuto-tp_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="02_clustering_TCGA-BIC_tuto-tp_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="02_clustering_TCGA-BIC_tuto-tp_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="02_clustering_TCGA-BIC_tuto-tp_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="02_clustering_TCGA-BIC_tuto-tp_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="02_clustering_TCGA-BIC_tuto-tp_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="02_clustering_TCGA-BIC_tuto-tp_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="02_clustering_TCGA-BIC_tuto-tp_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="02_clustering_TCGA-BIC_tuto-tp_files/navigation-1.1/tabsets.js"></script>
<script src="02_clustering_TCGA-BIC_tuto-tp_files/navigation-1.1/codefolding.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  { color: #cccccc; background-color: #303030; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ffcfaf; } /* Alert */
code span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #dca3a3; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #f0dfaf; } /* ControlFlow */
code span.ch { color: #dca3a3; } /* Char */
code span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code span.co { color: #7f9f7f; } /* Comment */
code span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code span.do { color: #7f9f7f; } /* Documentation */
code span.dt { color: #dfdfbf; } /* DataType */
code span.dv { color: #dcdccc; } /* DecVal */
code span.er { color: #c3bf9f; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #c0bed1; } /* Float */
code span.fu { color: #efef8f; } /* Function */
code span.im { } /* Import */
code span.in { color: #7f9f7f; font-weight: bold; } /* Information */
code span.kw { color: #f0dfaf; } /* Keyword */
code span.op { color: #f0efd0; } /* Operator */
code span.ot { color: #efef8f; } /* Other */
code span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code span.sc { color: #dca3a3; } /* SpecialChar */
code span.ss { color: #cc9393; } /* SpecialString */
code span.st { color: #cc9393; } /* String */
code span.va { } /* Variable */
code span.vs { color: #cc9393; } /* VerbatimString */
code span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>







<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-sm-12 col-md-4 col-lg-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-sm-12 col-md-8 col-lg-9">




<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Tutorial: machine-learning with TGCA BIC transcriptome</h1>
<h3 class="subtitle">02. Clustering</h3>
<h4 class="author">Anne Badel &amp; Jacques van Helden</h4>
<h4 class="date">2021-03-29</h4>

</div>


<div id="data-loading" class="section level2">
<h2>Data loading</h2>
<p>We provide hereafter a code to load the prepared data from a memory image on the github repository. This image has been generated at the end of the tutorial on <a href="https://du-bii.github.io/module-3-Stat-R/stat-R_2021/tutorials/machine-learning_TCGA-BIC/01_data-loading_TCGA-BIC.html">data loading and exploration</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Reload memory image from github repository ####</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>github_mem_img <span class="ot">&lt;-</span> <span class="st">&quot;https://github.com/DU-Bii/module-3-Stat-R/blob/master/stat-R_2021/data/TCGA_BIC_subset/bic_data.Rdata?raw=true&quot;</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Define local destination folder</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>bic_folder <span class="ot">&lt;-</span> <span class="st">&quot;~/m3-stat-R/TCGA-BIC_analysis&quot;</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Create it if required</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(bic_folder, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="do">## Define local destination for the memory image</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>mem_image <span class="ot">&lt;-</span> <span class="fu">file.path</span>(bic_folder, <span class="st">&quot;bic_data.Rdata&quot;</span>)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">file.exists</span>(mem_image)) {</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="st">&quot;Memory image already there, skipping download&quot;</span>)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="st">&quot;Downloading memory image from</span><span class="sc">\n</span><span class="st">&quot;</span>, github_mem_img)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">download.file</span>(<span class="at">url =</span> github_mem_img, <span class="at">destfile =</span> mem_image)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="st">&quot;Local memory image</span><span class="sc">\t</span><span class="st">&quot;</span>, mem_image)</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="do">## Load the memory image</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="st">&quot;Loading memory image&quot;</span>, mem_image)</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(mem_image)</span></code></pre></div>
<p>The table below indicates the main variables loaded with this memory image.</p>
<table>
<colgroup>
<col width="23%" />
<col width="76%" />
</colgroup>
<thead>
<tr class="header">
<th>Variable name</th>
<th>Data content</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>class_color</code></td>
<td>a vector specifying the color to be associated to each sample class (cancer type)</td>
</tr>
<tr class="even">
<td><code>bic_meta</code></td>
<td>metadata with a few added columns (sample color, estimators of central tendency and dispersion)</td>
</tr>
<tr class="odd">
<td><code>gene_info</code></td>
<td>ID, name and description of the 1000 genes used here</td>
</tr>
<tr class="even">
<td><code>bic_expr</code></td>
<td>non-normalised expression table</td>
</tr>
<tr class="odd">
<td><code>bic_expr_centered</code></td>
<td>median-based expression table</td>
</tr>
<tr class="even">
<td><code>bic_expr_std</code></td>
<td>sample-wise standardised expression table, all samples having the same median and IQR</td>
</tr>
<tr class="odd">
<td><code>bic_expr_labels</code></td>
<td>same content as bic_expr_std but with row names replaced by human-readable gene names, and column names by human-readable sample labels</td>
</tr>
</tbody>
</table>
<p>Use the command <code>View()</code> or <code>head()</code> or any other convenient way of your choice to check the content of these variables.</p>
</div>
<div id="subset-selection" class="section level2">
<h2>Subset selection</h2>
<p>The original data table contains 819 samples (columns) x 1000 genes (rows). These dimensions are not convenient for visualisation, so we will start by selecting a smaller set of samples and genes.</p>
<p>Note that this smaller data subset will serve only for didactic purposes; at the end of the practical, we will apply the clustering to all the genes with all the samples, in order to get relevant results.</p>
<div id="exercise-subset-selection" class="section level3">
<h3>Exercise: subset selection</h3>
<ol style="list-style-type: decimal">
<li>In the metadata table, count the number of samples per cancer type.</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Samples per cancer type ####</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(<span class="fu">sort</span>(<span class="fu">table</span>(bic_meta<span class="sc">$</span>cancer.type), <span class="at">decreasing =</span> <span class="cn">TRUE</span>),</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">caption =</span> <span class="st">&quot;Number of samples per cancer type&quot;</span>)</span></code></pre></div>
<table>
<caption>Number of samples per cancer type</caption>
<thead>
<tr class="header">
<th align="left">Var1</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Luminal.A</td>
<td align="right">422</td>
</tr>
<tr class="even">
<td align="left">Basal.like</td>
<td align="right">131</td>
</tr>
<tr class="odd">
<td align="left">Luminal.B</td>
<td align="right">118</td>
</tr>
<tr class="even">
<td align="left">Unclassified</td>
<td align="right">107</td>
</tr>
<tr class="odd">
<td align="left">HER2pos</td>
<td align="right">41</td>
</tr>
</tbody>
</table>
<ol start="2" style="list-style-type: decimal">
<li>Generate a vector named <code>sample_subset</code> that will contain 20 sample IDs from each cancer type <em>except the unclassified</em>.</li>
</ol>
<p>Tip: use the commands <code>subset()</code> and <code>sample()</code>. If you are not familiar with these commands, use <code>help()</code> to read their doc.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Sample subset selection ####</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>samples_per_class <span class="ot">&lt;-</span> <span class="dv">20</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Get the list of cancer types</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>cancer_types <span class="ot">&lt;-</span> <span class="fu">unique</span>(bic_meta<span class="sc">$</span>cancer.type)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="do">## Initialise the result vector</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>sample_subset <span class="ot">&lt;-</span> <span class="fu">vector</span>()</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="do">## Select samples for each class</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (type <span class="cf">in</span> cancer_types) {</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (type <span class="sc">!=</span> <span class="st">&quot;Unclassified&quot;</span>) { <span class="do">## Skip Unclassified samples</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>    <span class="do">## Select the samples of the current cancer type</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>    type_sample <span class="ot">&lt;-</span> <span class="fu">subset</span>(<span class="at">x =</span> bic_meta, cancer.type <span class="sc">==</span> type, </span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>                          <span class="at">select =</span> <span class="st">&quot;label&quot;</span>)</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>    sample_subset <span class="ot">&lt;-</span> <span class="fu">append</span>(</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>      sample_subset, </span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>      <span class="fu">sample</span>(<span class="fu">row.names</span>(type_sample), </span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> samples_per_class,</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>             <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Create a data.frame named <code>subse_meta</code> with the metadata corresponding to this random selection and check the number of samples per class.</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Subset metadata ####</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>subset_meta <span class="ot">&lt;-</span> bic_meta[sample_subset, ]</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(<span class="at">x =</span> <span class="fu">table</span>(subset_meta<span class="sc">$</span>cancer.type), </span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">caption =</span> <span class="st">&quot;Number of samples per class in the selected subset&quot;</span>)</span></code></pre></div>
<table>
<caption>Number of samples per class in the selected subset</caption>
<thead>
<tr class="header">
<th align="left">Var1</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Basal.like</td>
<td align="right">20</td>
</tr>
<tr class="even">
<td align="left">HER2pos</td>
<td align="right">20</td>
</tr>
<tr class="odd">
<td align="left">Luminal.A</td>
<td align="right">20</td>
</tr>
<tr class="even">
<td align="left">Luminal.B</td>
<td align="right">20</td>
</tr>
</tbody>
</table>
<ol start="3" style="list-style-type: decimal">
<li><p>Generate an expression table named <code>subset_expr_labels</code> from the expression table <code>bic_expr_labels</code> by keeping only</p>
<ul>
<li>the columns corresponding to the selected samples (beware, you will need to use labels instead of row names from the subset metadata)</li>
<li>the 100 top-ranking genes</li>
</ul></li>
</ol>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Select a subset of the expression table ####</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>subset_gene_nb <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>subset_labels <span class="ot">&lt;-</span> subset_meta<span class="sc">$</span>label</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>subset_expr_labels <span class="ot">&lt;-</span> bic_expr_labels[<span class="dv">1</span><span class="sc">:</span>subset_gene_nb, subset_labels]</span></code></pre></div>
</div>
</div>
<div id="distance-computation" class="section level2">
<h2>Distance computation</h2>
<p>Read the help page of the enhanced distance function of the package <code>factoextra</code> (<code>factoextra::get_dist</code>).</p>
<p>Use this method to compute 3 dissimilarity matrices, using respectively Euclidian, Pearson and Spearman distances. Name these matrices respectively <code>sample_dist_euclidian</code>, <code>sample_dist_pearson</code>, and <code>sample_dist_spearman</code>,</p>
<p>Tips:</p>
<ul>
<li>Beware: <code>get_dist()</code> computes distances betwee the <em>rows</em> of a data matrix. For sample clustering we will thus need to transpose the data.frame with the function <code>t()</code>.</li>
</ul>
<div id="example-euclidian-distance" class="section level3">
<h3>Example: Euclidian distance</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Inter-sample distances ####</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(factoextra)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ?factoextra::get_dist</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Euclidian distance</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>sample_dist_euclidian <span class="ot">&lt;-</span> factoextra<span class="sc">::</span><span class="fu">get_dist</span>(</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">t</span>(subset_expr_labels), <span class="at">method =</span> <span class="st">&quot;euclidian&quot;</span>)</span></code></pre></div>
</div>
<div id="exercise" class="section level3">
<h3>Exercise</h3>
<p>Use <code>get_dist()</code> to compute correlation-based distances with respectiely Pearson and Spearman correlations. Save the results in variables named <code>sample_dist_pearson</code> and <code>sample_dist_spearman</code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Pearson distance</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>sample_dist_pearson <span class="ot">&lt;-</span> factoextra<span class="sc">::</span><span class="fu">get_dist</span>(</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">t</span>(subset_expr_labels), <span class="at">method =</span> <span class="st">&quot;pearson&quot;</span>)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Spearman distance</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>sample_dist_spearman <span class="ot">&lt;-</span> factoextra<span class="sc">::</span><span class="fu">get_dist</span>(</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">t</span>(subset_expr_labels), <span class="at">method =</span> <span class="st">&quot;spearman&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="distance-heatmaps" class="section level2">
<h2>Distance heatmaps</h2>
<p>Use the function <code>pheatmap()</code> (from the package of the same name) to plot the distance matrices. Use the <code>annotation</code> option to display colors above the columns indicating the cancer type and the status of the three gene markers (Her2, PR1 and ER1).</p>
<div id="example-euclidian-distance-heatmap" class="section level3">
<h3>Example: Euclidian distance heatmap</h3>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Heatmap of the Euclidian distance ####s</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>subset_annotation <span class="ot">&lt;-</span> subset_meta[, <span class="fu">c</span>(<span class="st">&quot;cancer.type&quot;</span>, <span class="st">&quot;ER1&quot;</span>, <span class="st">&quot;PR1&quot;</span>, <span class="st">&quot;Her2&quot;</span>)]</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(subset_annotation) <span class="ot">&lt;-</span> subset_meta<span class="sc">$</span>label</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(sample_dist_euclidian, </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">cluster_rows =</span> <span class="cn">FALSE</span>, </span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">cluster_cols =</span> <span class="cn">FALSE</span>,</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">main =</span> <span class="st">&quot;Euclidian distance&quot;</span>, </span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_col =</span> subset_annotation)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/tcga-bic_dist_euclidian_heatmap-1.png" alt="Heatmap of the Euclidian distance" width="100%" />
<p class="caption">
Heatmap of the Euclidian distance
</p>
</div>
</div>
<div id="exercise-distance-heatmaps" class="section level3">
<h3>Exercise: distance heatmaps</h3>
<p>Run the same analysis for the Pearson and Spearman distance matrices. Interpret the results: do some of these measures seem more relevant to highlight the similarities between samples?</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Pearson distance heatmap ####</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Spearman distance heatmap ####</span></span></code></pre></div>
</div>
<div id="interpretation-dissimilarity-measures" class="section level3">
<h3>Interpretation: dissimilarity measures</h3>
<p>Comparing the three distance heatmaps above, can you already conclude something about the most relevant measure of dissimilarity between samples?</p>
<ul>
<li>TO BE COMPLETED</li>
</ul>
</div>
</div>
<div id="computing-a-tree-from-a-distance-matrix" class="section level2">
<h2>Computing a tree from a distance matrix</h2>
<p>Use the function <code>hclust()</code> to get trees from the previously computed distance matrices.</p>
<div id="example-euclidian-matrix-complete-linkage" class="section level3">
<h3>Example: Euclidian matrix, complete linkage</h3>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Sample tree with Euclidian distance, complete linkage ####</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>sample_tree_euclidian_complete <span class="ot">&lt;-</span> <span class="fu">hclust</span>(</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">d =</span> sample_dist_euclidian, <span class="at">method =</span> <span class="st">&quot;complete&quot;</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sample_tree_euclidian_complete, </span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="cn">NA</span>, <span class="at">ylab =</span> <span class="cn">NA</span>, <span class="at">hang =</span> <span class="sc">-</span><span class="dv">1</span>)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/tcga-bic_sample_tree_euclidian_complete-1.png" alt="Sample tree of the expression matrix with Euclidian distance and complete linkage. " width="100%" />
<p class="caption">
Sample tree of the expression matrix with Euclidian distance and complete linkage.
</p>
</div>
<p>We can also use some fancy library (e.g. <code>ClassDiscovery</code>) to color the tree labels. Other librairies can be used to generate nice tree drawings, don’t hesitate to google and improve.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ClassDiscovery)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>ClassDiscovery<span class="sc">::</span><span class="fu">plotColoredClusters</span>(</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  sample_tree_euclidian_complete, </span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">labs =</span> subset_meta<span class="sc">$</span>cancer.type, <span class="do">## class labels</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="st">&quot;grey&quot;</span>, <span class="do">## tree color</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">cols =</span> subset_meta<span class="sc">$</span>color, </span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">cex =</span> <span class="fl">0.6</span>,</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">&quot;Euclidian distance, Ward linkage&quot;</span>)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;topright&quot;</span>,</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">names</span>(class_color), </span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> class_color,</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>       <span class="at">text.col =</span> class_color, <span class="at">cex=</span><span class="fl">0.9</span>)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/tcga-bic_sample_tree_euclidian_complete_colored-1.png" alt="Sample tree of the expression matrix with Euclidian distance and complete linkage. " width="100%" />
<p class="caption">
Sample tree of the expression matrix with Euclidian distance and complete linkage.
</p>
</div>
</div>
<div id="exercises-other-matrices-and-linkage-rules" class="section level3">
<h3>Exercises: other matrices and linkage rules</h3>
<p>Test the different agglomeration rules (single, average, complete, ward) and the different disimilarity measures (Euclidian, Spearman, Pearson).</p>
</div>
<div id="interpretation-agglomeration-rule" class="section level3">
<h3>Interpretation: agglomeration rule</h3>
<p>Based on the analyses above, would you consider that some agglomeration rules are generally more relevant than other ones with this dataset? If so, is it systematic or does the best agglomeration rule vary depending on the dissimilarity measure?</p>
</div>
</div>
<div id="tree-cut" class="section level2">
<h2>Tree cut</h2>
<ol style="list-style-type: decimal">
<li><p>Cut the tree to obtain 4 samples clusters from the tree obtained with Euclidian distance and complete linkage.</p></li>
<li><p>Create a confusion matrix to compare the sample compositon of the clusters with the cancer types annotated in the metadata.</p></li>
<li><p>Compute the ARI from this confusion matrix.</p></li>
</ol>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Cluster - cancer type comparison ####</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Cut the tree</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>nclust <span class="ot">&lt;-</span> <span class="dv">4</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>clusters_euclidian_complete <span class="ot">&lt;-</span> <span class="fu">cutree</span>(</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">tree =</span> sample_tree_euclidian_complete, <span class="at">k =</span> nclust)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>confusion_euclidian_complete <span class="ot">&lt;-</span> <span class="fu">table</span>(</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  clusters_euclidian_complete,</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>  subset_meta<span class="sc">$</span>cancer.type)</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(confusion_euclidian_complete)</span></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">Basal.like</th>
<th align="right">HER2pos</th>
<th align="right">Luminal.A</th>
<th align="right">Luminal.B</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">16</td>
<td align="right">16</td>
<td align="right">0</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td align="right">3</td>
<td align="right">1</td>
<td align="right">3</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="right">0</td>
<td align="right">3</td>
<td align="right">17</td>
<td align="right">17</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Compute Adjusted Rand Index (ARI)</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>ari_euclidian_complete <span class="ot">&lt;-</span> <span class="fu">round</span>(</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">digits =</span> <span class="dv">3</span>, </span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  aricode<span class="sc">::</span><span class="fu">ARI</span>(</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">c1 =</span> clusters_euclidian_complete, </span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">c2 =</span> subset_meta<span class="sc">$</span>cancer.type))</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="st">&quot;ARI for Euclidian distance and complete linkage: &quot;</span>,</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>        ari_euclidian_complete)</span></code></pre></div>
<p>With Euclidian distance and the complete lin,kage, we obtain an Adjusted Rank Index (ARI) of 0.312.</p>
<p>The next question is of course: can we obtain better results with other distances and agglomeration rules? This will be left as an exercise.</p>
</div>
<div id="heatmap-with-biclustering-samples-genes" class="section level2">
<h2>Heatmap with biclustering (samples + genes)</h2>
<p>Biclustering amounts to cluster both the individuals (patient samples in our case) and variables (genes in our case).</p>
<p>We will illustrate this with the Spearman correlation as distance and Ward as agglomeration rule.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Biclustering ####</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="co"># left as an exercise</span></span></code></pre></div>
</div>
<div id="personal-work-chosing-the-most-relevant-parameters" class="section level2">
<h2>Personal work: chosing the most relevant parameters</h2>
<p>These analyses can be led collectively: each trainee could take one combination of parameters.</p>
<ol style="list-style-type: decimal">
<li><p>Based on the examples above, find the most relevant parameters (dissimilarity matrix, agglomeration rule) to cluster the samples of the subset expression matrix. Choose these parameters based on the confusion matrix and the ARI.</p></li>
<li><p>Draw different graphical representations of the best result:</p>
<ul>
<li>heatmap of the distance matrix</li>
<li>sample tree with coloring of the annotated cancer type</li>
<li>heatmap (samples x genes) with coloring of the metadata</li>
</ul></li>
<li><p>Use these optimal parameters to run clustering with all the samples of the BIC dataset and the `1000 genes instead of the subset of 100. Export a table with the sample clusters and another one with the gene clusters.</p></li>
</ol>
</div>
<div id="session-info" class="section level2">
<h2>Session info</h2>
<p>As usual, we write the session info in the report for the sake of traceability.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>R version 4.0.2 (2020-06-22)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] ClassDiscovery_3.3.13 oompaBase_3.2.9       cluster_2.1.1         scales_1.1.1          pheatmap_1.0.12       aricode_1.0.0         factoextra_1.0.7      ggplot2_3.3.3         knitr_1.31           

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.6         oompaData_3.1.1    highr_0.8          RColorBrewer_1.1-2 bslib_0.2.4        compiler_4.0.2     pillar_1.5.1       jquerylib_0.1.3    tools_4.0.2        mclust_5.4.7       digest_0.6.27      lattice_0.20-41    jsonlite_1.7.2     evaluate_0.14      lifecycle_1.0.0   
[16] tibble_3.1.0       gtable_0.3.0       pkgconfig_2.0.3    rlang_0.4.10       Matrix_1.3-2       DBI_1.1.1          ggrepel_0.9.1      yaml_2.2.1         xfun_0.22          withr_2.4.1        stringr_1.4.0      dplyr_1.0.5        generics_0.1.0     sass_0.3.1         vctrs_0.3.6       
[31] grid_4.0.2         tidyselect_1.1.0   glue_1.4.2         R6_2.5.0           fansi_0.4.2        rmarkdown_2.7      farver_2.1.0       purrr_0.3.4        magrittr_2.0.1     ellipsis_0.3.1     htmltools_0.5.1.1  assertthat_0.2.1   colorspace_2.0-0   utf8_1.2.1         stringi_1.5.3     
[46] munsell_0.5.0      crayon_1.4.1      </code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
