<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jacques van Helden" />

<meta name="date" content="2021-04-13" />

<title>Tutorial: machine-learning with TGCA BIC transcriptome</title>

<script src="04_supervised-classification_TCGA-BIC_tuto-tp_files/header-attrs-2.7/header-attrs.js"></script>
<script src="04_supervised-classification_TCGA-BIC_tuto-tp_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="04_supervised-classification_TCGA-BIC_tuto-tp_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="04_supervised-classification_TCGA-BIC_tuto-tp_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="04_supervised-classification_TCGA-BIC_tuto-tp_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="04_supervised-classification_TCGA-BIC_tuto-tp_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="04_supervised-classification_TCGA-BIC_tuto-tp_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="04_supervised-classification_TCGA-BIC_tuto-tp_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="04_supervised-classification_TCGA-BIC_tuto-tp_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="04_supervised-classification_TCGA-BIC_tuto-tp_files/navigation-1.1/tabsets.js"></script>
<script src="04_supervised-classification_TCGA-BIC_tuto-tp_files/navigation-1.1/codefolding.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  { color: #cccccc; background-color: #303030; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ffcfaf; } /* Alert */
code span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #dca3a3; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #f0dfaf; } /* ControlFlow */
code span.ch { color: #dca3a3; } /* Char */
code span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code span.co { color: #7f9f7f; } /* Comment */
code span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code span.do { color: #7f9f7f; } /* Documentation */
code span.dt { color: #dfdfbf; } /* DataType */
code span.dv { color: #dcdccc; } /* DecVal */
code span.er { color: #c3bf9f; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #c0bed1; } /* Float */
code span.fu { color: #efef8f; } /* Function */
code span.im { } /* Import */
code span.in { color: #7f9f7f; font-weight: bold; } /* Information */
code span.kw { color: #f0dfaf; } /* Keyword */
code span.op { color: #f0efd0; } /* Operator */
code span.ot { color: #efef8f; } /* Other */
code span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code span.sc { color: #dca3a3; } /* SpecialChar */
code span.ss { color: #cc9393; } /* SpecialString */
code span.st { color: #cc9393; } /* String */
code span.va { } /* Variable */
code span.vs { color: #cc9393; } /* VerbatimString */
code span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>







<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-sm-12 col-md-4 col-lg-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-sm-12 col-md-8 col-lg-9">




<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Tutorial: machine-learning with TGCA BIC transcriptome</h1>
<h3 class="subtitle">04. Supervised classification</h3>
<h4 class="author">Jacques van Helden</h4>
<h4 class="date">2021-04-13</h4>

</div>


<div id="data-loading" class="section level2">
<h2>Data loading</h2>
<p>We provide hereafter a code to load the prepared data from a memory image on the github repository. This image has been generated at the end of the tutorial on <a href="https://du-bii.github.io/module-3-Stat-R/stat-R_2021/tutorials/machine-learning_TCGA-BIC/01_data-loading_TCGA-BIC.html">data loading and exploration</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Reload memory image from github repository ####</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>github_mem_img <span class="ot">&lt;-</span> <span class="st">&quot;https://github.com/DU-Bii/module-3-Stat-R/blob/master/stat-R_2021/data/TCGA_BIC_subset/bic_data.Rdata?raw=true&quot;</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Define local destination folder</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>bic_folder <span class="ot">&lt;-</span> <span class="st">&quot;~/m3-stat-R/TCGA-BIC_analysis&quot;</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Create it if required</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(bic_folder, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="do">## Define local destination for the memory image</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>mem_image <span class="ot">&lt;-</span> <span class="fu">file.path</span>(bic_folder, <span class="st">&quot;bic_data.Rdata&quot;</span>)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">file.exists</span>(mem_image)) {</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="st">&quot;Memory image already there, skipping download&quot;</span>)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="st">&quot;Downloading memory image from</span><span class="sc">\n</span><span class="st">&quot;</span>, github_mem_img)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">download.file</span>(<span class="at">url =</span> github_mem_img, <span class="at">destfile =</span> mem_image)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="st">&quot;Local memory image</span><span class="sc">\t</span><span class="st">&quot;</span>, mem_image)</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="do">## Load the memory image</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="st">&quot;Loading memory image&quot;</span>, mem_image)</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(mem_image)</span></code></pre></div>
<p>The table below indicates the main variables loaded with this memory image.</p>
<table>
<colgroup>
<col width="23%" />
<col width="76%" />
</colgroup>
<thead>
<tr class="header">
<th>Variable name</th>
<th>Data content</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>class_color</code></td>
<td>a vector specifying the color to be associated to each sample class (cancer type)</td>
</tr>
<tr class="even">
<td><code>bic_meta</code></td>
<td>metadata with a few added columns (sample color, estimators of central tendency and dispersion)</td>
</tr>
<tr class="odd">
<td><code>gene_info</code></td>
<td>ID, name and description of the 1000 genes used here</td>
</tr>
<tr class="even">
<td><code>bic_expr</code></td>
<td>non-normalised expression table</td>
</tr>
<tr class="odd">
<td><code>bic_expr_centered</code></td>
<td>median-based expression table</td>
</tr>
<tr class="even">
<td><code>bic_expr_std</code></td>
<td>sample-wise standardised expression table, all samples having the same median and IQR</td>
</tr>
<tr class="odd">
<td><code>bic_expr_labels</code></td>
<td>same content as bic_expr_std but with row names replaced by human-readable gene names, and column names by human-readable sample labels</td>
</tr>
</tbody>
</table>
<p>Use the command <code>View()</code> or <code>head()</code> or any other convenient way of your choice to check the content of these variables.</p>
</div>
<div id="discard-the-unclassified-samples" class="section level2">
<h2>Discard the unclassified samples</h2>
<p>Generate an expression matrix and a metadata table without the unclassified samples, since these ones cannot be used to train a classifier.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Discard unclassified ####</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>bic_meta_ok <span class="ot">&lt;-</span> bic_meta[bic_meta<span class="sc">$</span>cancer.type <span class="sc">!=</span> <span class="st">&quot;Unclassified&quot;</span>,]</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Check that the metadata table has only the right classes</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(bic_meta_ok)</span></code></pre></div>
<pre><code>[1] 712</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(bic_meta_ok<span class="sc">$</span>cancer.type)</span></code></pre></div>
<pre><code>
Basal.like    HER2pos  Luminal.A  Luminal.B 
       131         41        422        118 </code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Select the corresponding columns (samples) of the expression table</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="do">## We also transpose the data table for the supoervised</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>bic_expr_class_ok <span class="ot">&lt;-</span> <span class="fu">t</span>(bic_expr_labels[, bic_meta_ok<span class="sc">$</span>label])</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(bic_expr_class_ok)</span></code></pre></div>
<pre><code>[1]  712 1000</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>bic_classes_ok <span class="ot">&lt;-</span> bic_meta_ok<span class="sc">$</span>cancer.type</span></code></pre></div>
</div>
<div id="split-the-dataset-into-a-testing-and-a-training-set" class="section level2">
<h2>Split the dataset into a testing and a training set</h2>
<p>Split the sample set into - a training (2/3 of the samples) - a testing set (the other 1/3 of the samples)</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Split samples into training and testing sets ####</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Count the number of samples (total, train and test)</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>nsamples <span class="ot">&lt;-</span> <span class="fu">nrow</span>(bic_expr_class_ok)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>nsamples_train <span class="ot">&lt;-</span> <span class="fu">round</span>(nsamples <span class="sc">*</span> <span class="dv">2</span><span class="sc">/</span><span class="dv">3</span>)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>nsamples_test <span class="ot">&lt;-</span> nsamples <span class="sc">-</span> nsamples_train</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="do">## Perfoirm a random sampling of indices</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>resampled_indices <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span>nsamples, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>train_indices <span class="ot">&lt;-</span> resampled_indices[<span class="dv">1</span><span class="sc">:</span>nsamples_train]</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>test_indices <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(resampled_indices, train_indices)</span></code></pre></div>
</div>
<div id="training-a-classifier-with-the-train-set" class="section level2">
<h2>Training a classifier with the train set</h2>
<p>Use the <code>svm()</code>funcvtion to train a support vector machine with the training set.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Train the SVM with the training subset ####</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>svm_kernel <span class="ot">=</span> <span class="st">&quot;radial&quot;</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Define training set: expression matrix and class labels</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>training_expr <span class="ot">&lt;-</span> bic_expr_class_ok[train_indices, ]</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>training_classes <span class="ot">&lt;-</span> bic_meta_ok[train_indices, <span class="st">&quot;cancer.type&quot;</span>]</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(training_classes)</span></code></pre></div>
<pre><code>training_classes
Basal.like    HER2pos  Luminal.A  Luminal.B 
        87         27        279         82 </code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Train the SVM</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>svm_model <span class="ot">&lt;-</span> <span class="fu">svm</span>(<span class="at">x =</span> training_expr, </span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">y =</span> <span class="fu">as.factor</span>(training_classes), </span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">type =</span> <span class="st">&quot;C-classification&quot;</span>, </span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">kernel =</span> svm_kernel)</span></code></pre></div>
</div>
<div id="predicting-the-classes-of-the-testing-set" class="section level2">
<h2>Predicting the classes of the testing set</h2>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Predict the classes of the testing set ####</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>testing_expr <span class="ot">&lt;-</span> bic_expr_class_ok[test_indices, ]</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>testing_pred <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(<span class="fu">predict</span>(svm_model,  testing_expr))</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Check the number of predicted samples per class</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(testing_pred)</span></code></pre></div>
<pre><code>testing_pred
Basal.like    HER2pos  Luminal.A  Luminal.B 
        44          2        187          4 </code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Evaluate the performances of the classifier ####</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Generate a contingency table comparing </span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="do">## the known and predicted classes for the testing set</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Get the annotated classes for the testing set</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>testing_classes <span class="ot">&lt;-</span> bic_meta_ok[test_indices, <span class="st">&quot;cancer.type&quot;</span>]</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(testing_classes)</span></code></pre></div>
<pre><code>testing_classes
Basal.like    HER2pos  Luminal.A  Luminal.B 
        44         14        143         36 </code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Compare the annotated and  predicted cancer types</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>contingency <span class="ot">&lt;-</span> <span class="fu">table</span>(testing_classes, testing_pred)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(contingency, <span class="at">row.names =</span> <span class="cn">TRUE</span>, <span class="at">caption =</span> )</span></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">Basal.like</th>
<th align="right">HER2pos</th>
<th align="right">Luminal.A</th>
<th align="right">Luminal.B</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Basal.like</td>
<td align="right">40</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">HER2pos</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">10</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">Luminal.A</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">142</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">Luminal.B</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">31</td>
<td align="right">4</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Compute the number of correct predictions</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>errors <span class="ot">&lt;-</span> testing_pred <span class="sc">!=</span> testing_classes</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Compute the misclassification error rate (MER)</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>mer <span class="ot">&lt;-</span> <span class="fu">sum</span>(errors) <span class="sc">/</span> <span class="fu">length</span>(errors)</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="st">&quot;MER = &quot;</span>, mer)</span></code></pre></div>
<p>In this training / testing experiment we obtained a misclassification error rate (MER) of <span class="math inline">\(MER = 0.2067511\)</span>. However, we just performed a single random sampling, so we can expect to obtain different results if we repaeat this experiment.</p>
<p>Several strategies are classically used to obtain a more reliable estimation of the performances :</p>
<ol style="list-style-type: decimal">
<li><p><strong>K-fold cross-validation:</strong> split the dataset into <span class="math inline">\(k\)</span> subsets, each of which is used as test for one train/test experiment (and the <span class="math inline">\(k-1\)</span> remaining ones are then used for training). We could for example run a 10-fold cross-validation with this dataset.</p></li>
<li><p><strong>Leave-one-out</strong> (<strong>LOO</strong>, also called <strong>jack-knife</strong>) is a specific case of k-fold cross-validation where <span class="math inline">\(k = n\)</span>, i.e. each individual (biological sample in our case) is in turn discarded from the data set (“left out”), a classifier is trained with the <span class="math inline">\(n - 1\)</span> remaining individuals, and the trained classifier (“model”) is used to predict the class of the left out individual.</p></li>
<li><p><strong>Iterative subsampling</strong> consists in repeating the above procedure, where we select a given proportion of the individuals for training (e.g. 2/3) and the rest for testing.</p></li>
</ol>
<p>Each of these methods of performance estimation has its pros and cons.</p>
<p>For this course, we will run a <strong>collective iterative subsampling</strong>: each trainee will run a trainig/test experiment and we will write the results in a shared spreadsheet, which will then enable us to compute some statistics on the MER values (mean, standard deviation, min, max).</p>
</div>
<div id="exercise-impact-of-the-svm-kernel" class="section level2">
<h2>Exercise: impact of the SVM kernel</h2>
<p>Based on the above example, test the 4 SVM kernels (linear, radial, sigmoid, polynomial) and compare the performances. Which kernel gives the best results.</p>
</div>
<div id="exercise-impact-of-the-number-of-variables-genes" class="section level2">
<h2>Exercise: impact of the number of variables (genes)</h2>
<p>Use the best kernel defined above, and estimate the MER with the 100, 200, 300, 500, 1000 top genes respectively.</p>
</div>
<div id="svm-tuning" class="section level2">
<h2>SVM tuning</h2>
<p>use the <code>tune.svm()</code> function to find the optimal SVM parameters.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="do">#### Tune SVM ####</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>svm_tuning_result <span class="ot">&lt;-</span> <span class="fu">tune.svm</span>(</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> training_expr, </span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> <span class="fu">as.factor</span>(training_classes))</span></code></pre></div>
</div>
<div id="session-info" class="section level2">
<h2>Session info</h2>
<p>As usual, we write the session info in the report for the sake of traceability.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>R version 4.0.2 (2020-06-22)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] caret_6.0-86    ggplot2_3.3.3   lattice_0.20-41 e1071_1.7-6     knitr_1.31     

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.6           lubridate_1.7.10     class_7.3-18         assertthat_0.2.1     digest_0.6.27        ipred_0.9-11         foreach_1.5.1        utf8_1.2.1           R6_2.5.0             plyr_1.8.6           stats4_4.0.2         evaluate_0.14        highr_0.8            pillar_1.5.1        
[15] rlang_0.4.10         data.table_1.14.0    jquerylib_0.1.3      rpart_4.1-15         Matrix_1.3-2         rmarkdown_2.7        splines_4.0.2        gower_0.2.2          stringr_1.4.0        munsell_0.5.0        proxy_0.4-25         compiler_4.0.2       xfun_0.22            pkgconfig_2.0.3     
[29] htmltools_0.5.1.1    nnet_7.3-15          tidyselect_1.1.0     tibble_3.1.0         prodlim_2019.11.13   codetools_0.2-18     fansi_0.4.2          crayon_1.4.1         dplyr_1.0.5          withr_2.4.1          MASS_7.3-53.1        recipes_0.1.15       ModelMetrics_1.2.2.2 grid_4.0.2          
[43] nlme_3.1-152         jsonlite_1.7.2       gtable_0.3.0         lifecycle_1.0.0      DBI_1.1.1            magrittr_2.0.1       pROC_1.17.0.1        scales_1.1.1         stringi_1.5.3        reshape2_1.4.4       timeDate_3043.102    bslib_0.2.4          ellipsis_0.3.1       generics_0.1.0      
[57] vctrs_0.3.6          lava_1.6.9           iterators_1.0.13     tools_4.0.2          glue_1.4.2           purrr_0.3.4          survival_3.2-10      yaml_2.2.1           colorspace_2.0-0     sass_0.3.1          </code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
