<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Prénom Nom" />

<meta name="date" content="2021-04-13" />

<title>Mini-projet 2021 - Exploration des données de Pavkovic</title>

<script src="Prenom-NOM_m3-stat-R_mini-projet_files/header-attrs-2.7/header-attrs.js"></script>
<script src="Prenom-NOM_m3-stat-R_mini-projet_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="Prenom-NOM_m3-stat-R_mini-projet_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="Prenom-NOM_m3-stat-R_mini-projet_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="Prenom-NOM_m3-stat-R_mini-projet_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="Prenom-NOM_m3-stat-R_mini-projet_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="Prenom-NOM_m3-stat-R_mini-projet_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="Prenom-NOM_m3-stat-R_mini-projet_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="Prenom-NOM_m3-stat-R_mini-projet_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="Prenom-NOM_m3-stat-R_mini-projet_files/navigation-1.1/tabsets.js"></script>
<script src="Prenom-NOM_m3-stat-R_mini-projet_files/navigation-1.1/codefolding.js"></script>
<script src="Prenom-NOM_m3-stat-R_mini-projet_files/navigation-1.1/sourceembed.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  { color: #cccccc; background-color: #303030; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ffcfaf; } /* Alert */
code span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #dca3a3; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #f0dfaf; } /* ControlFlow */
code span.ch { color: #dca3a3; } /* Char */
code span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code span.co { color: #7f9f7f; } /* Comment */
code span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code span.do { color: #7f9f7f; } /* Documentation */
code span.dt { color: #dfdfbf; } /* DataType */
code span.dv { color: #dcdccc; } /* DecVal */
code span.er { color: #c3bf9f; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #c0bed1; } /* Float */
code span.fu { color: #efef8f; } /* Function */
code span.im { } /* Import */
code span.in { color: #7f9f7f; font-weight: bold; } /* Information */
code span.kw { color: #f0dfaf; } /* Keyword */
code span.op { color: #f0efd0; } /* Operator */
code span.ot { color: #efef8f; } /* Other */
code span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code span.sc { color: #dca3a3; } /* SpecialChar */
code span.ss { color: #cc9393; } /* SpecialString */
code span.st { color: #cc9393; } /* String */
code span.va { } /* Variable */
code span.vs { color: #cc9393; } /* VerbatimString */
code span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>


<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-sm-12 col-md-4 col-lg-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-sm-12 col-md-8 col-lg-9">




<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
<li role="separator" class="divider"></li>
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Mini-projet 2021 - Exploration des données de Pavkovic</h1>
<h4 class="author">Prénom Nom</h4>
<h4 class="date">2021-04-13</h4>

</div>


<div id="synopsis-du-projet" class="section level2">
<h2>Synopsis du projet</h2>
<p>Le but du projet est de mettre en oeuvre les méthodes vues dans le module 3 “R et statistiques” pour explorer le jeu de données de Pavokovic, et de fournir un rapport d’analyse au format R markdown.</p>
<p>Nous fournissons ci-dessous une trame avec les principales sections attendus. Sentez-vous libres d’adapter cette trame ou d’y ajouter des analyses complémentaires si elles vous aident à interpréter vos résultats.</p>
</div>
<div id="remise-du-rapport" class="section level2">
<h2>Remise du rapport</h2>
<p>Date: <strong>le 10 mai 2021 minuit</strong>. Si vous anticipez un problème pour remettre le rapport à cette date contactez-nous aussi rapidement que possible pour que nous puissions prévoir une remise plus tardive.</p>
<ul>
<li><p>Commencez par renommer le fichier Rmd en remplaçant Prenom-NOM par vos nom et prénom.</p></li>
<li><p>Le rapport est attendu en formats Rmd + HTML (en activant l’option self_contained de l’en-tête).</p></li>
<li><p>Déposez les fichiers dans un sous-dossier de vote compte du cluster. Attention, veillez à respecter précisément cette structure de chemin car nous nous baserons dessus pour récupérer vos résultats.</p>
<p><code>/shared/projects/dubii2021/[login]/m3-stat-R/mini-projet</code></p></li>
</ul>
</div>
<div id="critères-dévaluation" class="section level2">
<h2>Critères d’évaluation</h2>
<ul>
<li>Reproductibilité des analyses: nous tenterons de regénérer le rapport HTML à partir de votre Rmd, en partant de notre compte sur le serveur IFB.</li>
<li>Manipulation des objets R</li>
<li>Mobilisation des méthodes statistiques vues au cours</li>
<li>Pertinence des interprétations statistiques</li>
<li>Pertinence des interprétations biologiques</li>
<li>Clarté de la rédaction</li>
<li>Clarté des illustrations (figures et tableaux): graphismes, légendes …</li>
</ul>
<p>Nous vous encourageons à assurer la lisibilpité de votre code (syntaxe, nommage des variables, commentaires de code)</p>
</div>
<div id="chargement-des-données-brutes" class="section level2">
<h2>Chargement des données brutes</h2>
<p>Le bloc suivant contient une fonction qui permet de télécharger un fichier dans l’espace de travail, sauf s’il est déjà présent. Nous l’utiliserons ensuite pour télécharger les données à analyser en évitant de refaire le transfert à chaque exécution de l’analyse.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @title Download a file only if it is not yet here</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @author Jacques van Helden email{Jacques.van-Helden@@france-bioinformatique.fr}</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @param url_base base of the URL, that will be prepended to the file name</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @param file_name name of the file (should not contain any path)</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @param local_folder path of a local folder where the file should be stored</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @return the function returns the path of the local file, built from local_folder and file_name</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @export©</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>download_only_once <span class="ot">&lt;-</span> <span class="cf">function</span>(</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>  url_base, </span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>  file_name,</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>  local_folder) {</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Define the source URL  </span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>  url <span class="ot">&lt;-</span> <span class="fu">file.path</span>(url_base, file_name)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="st">&quot;Source URL</span><span class="sc">\n\t</span><span class="st">&quot;</span>,  url)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Define the local file</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>  local_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(local_folder, file_name)</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Create the local data folder if it does not exist</span></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir.create</span>(local_folder, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Download the file ONLY if it is not already there</span></span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(local_file)) {</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">message</span>(<span class="st">&quot;Downloading file from source URL to local file</span><span class="sc">\n\t</span><span class="st">&quot;</span>, </span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>            local_file)</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">download.file</span>(<span class="at">url =</span> url, <span class="at">destfile =</span> local_file)</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">message</span>(<span class="st">&quot;Local file already exists, no need to download</span><span class="sc">\n\t</span><span class="st">&quot;</span>, </span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>            local_file)</span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(local_file)</span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>Nous téléchargeons deux fichiers dans un dossier local, et les chargeons dans les data.frames suivantes:</p>
<ul>
<li>Données brutes de transcriptome: <code>fa_expr_raw</code></li>
<li>Métadonnées: <code>fa_meta</code></li>
</ul>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Define the remote URL and local folder</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>pavkovic_url <span class="ot">&lt;-</span> <span class="st">&quot;https://github.com/DU-Bii/module-3-Stat-R/raw/master/stat-R_2021/data/pavkovic_2019/&quot;</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Define the local folder for this analysis (where the data will be downloaded and the results generated)</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>pavkovic_folder <span class="ot">&lt;-</span> <span class="st">&quot;~/m3-stat-R/pavkovic_analysis&quot;</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Define a sub-folder for the data</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>pavkovic_data_folder <span class="ot">&lt;-</span> <span class="fu">file.path</span>(pavkovic_folder, <span class="st">&quot;data&quot;</span>)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="do">## Download and load the expression data table</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="do">## Note: we use check.names=FALSE to avoid replacing hyphens by dots</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="do">## in sample names, because we want to keep them as in the </span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="do">## original data files. </span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="st">&quot;Downloading FA transcriptome file</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="st">&quot;fa_raw_counts.tsv.gz&quot;</span>,</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;</span><span class="sc">\n\t</span><span class="st">from</span><span class="sc">\t</span><span class="st">&quot;</span>, pavkovic_url)</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>fa_expr_file <span class="ot">&lt;-</span> <span class="fu">download_only_once</span>(</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">url_base =</span> pavkovic_url, </span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">file_name =</span> <span class="st">&quot;fa_raw_counts.tsv.gz&quot;</span>,</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">local_folder =</span> pavkovic_data_folder)</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="st">&quot;Loading FA transcriptome data from</span><span class="sc">\n\t</span><span class="st">&quot;</span>, fa_expr_file)</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>fa_expr_raw <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="at">file =</span> fa_expr_file, </span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>                       <span class="at">header =</span> <span class="cn">TRUE</span>, </span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>                       <span class="at">row.names =</span> <span class="dv">1</span>)</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a><span class="co"># colnames(fa_expr_raw)</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a><span class="co"># dim(fa_expr_raw)</span></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a><span class="co"># View(head(fa_expr_raw))</span></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a><span class="do">## Download the metadata file</span></span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="st">&quot;Downloading FA metadata file</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="st">&quot;fa_transcriptome_metadata.tsv&quot;</span>,</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;</span><span class="sc">\n\t</span><span class="st">from</span><span class="sc">\t</span><span class="st">&quot;</span>, pavkovic_url)</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>fa_meta_file <span class="ot">&lt;-</span> <span class="fu">download_only_once</span>(</span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>  <span class="at">url_base =</span> pavkovic_url, </span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>  <span class="at">file_name =</span> <span class="st">&quot;fa_transcriptome_metadata.tsv&quot;</span>,</span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>  <span class="at">local_folder =</span> pavkovic_data_folder)</span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="st">&quot;Loading FA metadata from</span><span class="sc">\n\t</span><span class="st">&quot;</span>, fa_meta_file)</span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a>fa_meta <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="at">file =</span> fa_meta_file, </span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a>                       <span class="at">header =</span> <span class="cn">TRUE</span>, </span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a>                       <span class="at">row.names =</span> <span class="dv">1</span>)</span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-40"><a href="#cb2-40" aria-hidden="true" tabindex="-1"></a><span class="co"># View(fa_meta)</span></span>
<span id="cb2-41"><a href="#cb2-41" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(fa_meta, <span class="at">caption =</span> <span class="st">&quot;Metdata for Pavkovoc FA transcriptome&quot;</span>)</span></code></pre></div>
<table>
<caption>Metdata for Pavkovoc FA transcriptome</caption>
<thead>
<tr class="header">
<th align="left">dataType</th>
<th align="left">sampleName</th>
<th align="left">condition</th>
<th align="right">sampleNumber</th>
<th align="left">color</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">transcriptome</td>
<td align="left">day14_1</td>
<td align="left">day14</td>
<td align="right">1</td>
<td align="left">#FF4400</td>
</tr>
<tr class="even">
<td align="left">transcriptome</td>
<td align="left">day14_2</td>
<td align="left">day14</td>
<td align="right">2</td>
<td align="left">#FF4400</td>
</tr>
<tr class="odd">
<td align="left">transcriptome</td>
<td align="left">day14_3</td>
<td align="left">day14</td>
<td align="right">3</td>
<td align="left">#FF4400</td>
</tr>
<tr class="even">
<td align="left">transcriptome</td>
<td align="left">day1_1</td>
<td align="left">day1</td>
<td align="right">1</td>
<td align="left">#BBD7FF</td>
</tr>
<tr class="odd">
<td align="left">transcriptome</td>
<td align="left">day1_2</td>
<td align="left">day1</td>
<td align="right">2</td>
<td align="left">#BBD7FF</td>
</tr>
<tr class="even">
<td align="left">transcriptome</td>
<td align="left">day1_3</td>
<td align="left">day1</td>
<td align="right">3</td>
<td align="left">#BBD7FF</td>
</tr>
<tr class="odd">
<td align="left">transcriptome</td>
<td align="left">day2_1</td>
<td align="left">day2</td>
<td align="right">1</td>
<td align="left">#F0BBFF</td>
</tr>
<tr class="even">
<td align="left">transcriptome</td>
<td align="left">day2_2</td>
<td align="left">day2</td>
<td align="right">2</td>
<td align="left">#F0BBFF</td>
</tr>
<tr class="odd">
<td align="left">transcriptome</td>
<td align="left">day2_3</td>
<td align="left">day2</td>
<td align="right">3</td>
<td align="left">#F0BBFF</td>
</tr>
<tr class="even">
<td align="left">transcriptome</td>
<td align="left">day3_1</td>
<td align="left">day3</td>
<td align="right">1</td>
<td align="left">#FFFFDD</td>
</tr>
<tr class="odd">
<td align="left">transcriptome</td>
<td align="left">day3_2</td>
<td align="left">day3</td>
<td align="right">2</td>
<td align="left">#FFFFDD</td>
</tr>
<tr class="even">
<td align="left">transcriptome</td>
<td align="left">day3_3</td>
<td align="left">day3</td>
<td align="right">3</td>
<td align="left">#FFFFDD</td>
</tr>
<tr class="odd">
<td align="left">transcriptome</td>
<td align="left">day7_1</td>
<td align="left">day7</td>
<td align="right">1</td>
<td align="left">#FFDD88</td>
</tr>
<tr class="even">
<td align="left">transcriptome</td>
<td align="left">day7_2</td>
<td align="left">day7</td>
<td align="right">2</td>
<td align="left">#FFDD88</td>
</tr>
<tr class="odd">
<td align="left">transcriptome</td>
<td align="left">day7_3</td>
<td align="left">day7</td>
<td align="right">3</td>
<td align="left">#FFDD88</td>
</tr>
<tr class="even">
<td align="left">transcriptome</td>
<td align="left">normal_1</td>
<td align="left">normal</td>
<td align="right">1</td>
<td align="left">#BBFFBB</td>
</tr>
<tr class="odd">
<td align="left">transcriptome</td>
<td align="left">normal_2</td>
<td align="left">normal</td>
<td align="right">2</td>
<td align="left">#BBFFBB</td>
</tr>
<tr class="even">
<td align="left">transcriptome</td>
<td align="left">normal_3</td>
<td align="left">normal</td>
<td align="right">3</td>
<td align="left">#BBFFBB</td>
</tr>
</tbody>
</table>
</div>
<div id="transformation-log2" class="section level2">
<h2>Transformation log2</h2>
<p>Appliquez une transformation log2 des données brutes, après avoir ajouté un epsilon <span class="math inline">\(\epsilon = 1\)</span> (les valeurs nulles seront donc représentées par un log2(counts) valant <span class="math inline">\(0\)</span>. Stockez le résultat dans une data.frame nommée <code>fa_expr_log2</code>.</p>
<p>Affchez un fragment des tableaux <code>fa_expr_raw</code> et <code>fa_expr_log2</code> en sélectionnant les lignes 100 à 109 et les colonnes 5 à 10, afin de vous assurer que la transformation log2 a bien fonctionné.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Log2 transformation of the transcriptome data</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>epsilon <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>fa_expr_log2 <span class="ot">&lt;-</span> <span class="fu">log2</span>(fa_expr_raw <span class="sc">+</span> epsilon)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># dim(fa_expr_log2)</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># View(head(fa_expr_log2))</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Display of a fragment of the data before and after log2 transformation</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(fa_expr_raw[<span class="dv">100</span><span class="sc">:</span><span class="dv">109</span>, <span class="dv">5</span><span class="sc">:</span><span class="dv">10</span>], <span class="at">caption =</span> <span class="st">&quot;Fragment des données transcriptomiques brutes&quot;</span>)</span></code></pre></div>
<table style="width:100%;">
<caption>Fragment des données transcriptomiques brutes</caption>
<colgroup>
<col width="21%" />
<col width="13%" />
<col width="13%" />
<col width="13%" />
<col width="13%" />
<col width="12%" />
<col width="12%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">day14_2</th>
<th align="right">day14_3</th>
<th align="right">day2_1</th>
<th align="right">day2_2</th>
<th align="right">day2_3</th>
<th align="right">day3_1</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ENSMUSG00000000567</td>
<td align="right">51.004801</td>
<td align="right">177.973522</td>
<td align="right">1052.689447</td>
<td align="right">106.995584</td>
<td align="right">347.13842</td>
<td align="right">479.59911</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000568</td>
<td align="right">871.981145</td>
<td align="right">530.136492</td>
<td align="right">818.257157</td>
<td align="right">116.136417</td>
<td align="right">3406.45030</td>
<td align="right">766.09722</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000579</td>
<td align="right">728.517368</td>
<td align="right">468.254668</td>
<td align="right">473.399472</td>
<td align="right">36.258005</td>
<td align="right">410.57787</td>
<td align="right">347.05054</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000581</td>
<td align="right">488.131157</td>
<td align="right">510.212365</td>
<td align="right">546.260344</td>
<td align="right">87.788535</td>
<td align="right">319.12679</td>
<td align="right">461.45732</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000594</td>
<td align="right">2331.419915</td>
<td align="right">1663.654899</td>
<td align="right">2283.051270</td>
<td align="right">1115.588250</td>
<td align="right">1491.61384</td>
<td align="right">1576.68451</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000600</td>
<td align="right">2021.209338</td>
<td align="right">2405.746829</td>
<td align="right">3674.188108</td>
<td align="right">589.840068</td>
<td align="right">1399.10751</td>
<td align="right">3446.01856</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000605</td>
<td align="right">235.765748</td>
<td align="right">293.247867</td>
<td align="right">489.579657</td>
<td align="right">77.008213</td>
<td align="right">256.00200</td>
<td align="right">282.80077</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000606</td>
<td align="right">1.978342</td>
<td align="right">3.749163</td>
<td align="right">3.970399</td>
<td align="right">0.000000</td>
<td align="right">138.48677</td>
<td align="right">0.00000</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000617</td>
<td align="right">11.780528</td>
<td align="right">22.842788</td>
<td align="right">30.228506</td>
<td align="right">6.670500</td>
<td align="right">18.27493</td>
<td align="right">13.41152</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000627</td>
<td align="right">3.099033</td>
<td align="right">9.776030</td>
<td align="right">46.297517</td>
<td align="right">2.878275</td>
<td align="right">17.03638</td>
<td align="right">15.45854</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(fa_expr_log2[<span class="dv">100</span><span class="sc">:</span><span class="dv">109</span>, <span class="dv">5</span><span class="sc">:</span><span class="dv">10</span>], <span class="at">caption =</span> <span class="st">&quot;Fragment des données transcriptomiques après transformation log2&quot;</span>)</span></code></pre></div>
<table>
<caption>Fragment des données transcriptomiques après transformation log2</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">day14_2</th>
<th align="right">day14_3</th>
<th align="right">day2_1</th>
<th align="right">day2_2</th>
<th align="right">day2_3</th>
<th align="right">day3_1</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ENSMUSG00000000567</td>
<td align="right">5.700573</td>
<td align="right">7.483602</td>
<td align="right">10.041234</td>
<td align="right">6.754829</td>
<td align="right">8.443517</td>
<td align="right">8.908690</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000568</td>
<td align="right">9.769807</td>
<td align="right">9.052939</td>
<td align="right">9.678173</td>
<td align="right">6.872046</td>
<td align="right">11.734477</td>
<td align="right">9.583266</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000579</td>
<td align="right">9.510798</td>
<td align="right">8.874227</td>
<td align="right">8.889959</td>
<td align="right">5.219479</td>
<td align="right">8.685022</td>
<td align="right">8.443153</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000581</td>
<td align="right">8.934078</td>
<td align="right">8.997779</td>
<td align="right">9.096084</td>
<td align="right">6.472302</td>
<td align="right">8.322500</td>
<td align="right">8.853176</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000594</td>
<td align="right">11.187612</td>
<td align="right">10.701007</td>
<td align="right">11.157379</td>
<td align="right">10.124882</td>
<td align="right">10.543625</td>
<td align="right">10.623593</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000600</td>
<td align="right">10.981717</td>
<td align="right">11.232869</td>
<td align="right">11.843602</td>
<td align="right">9.206624</td>
<td align="right">10.451322</td>
<td align="right">11.751133</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000605</td>
<td align="right">7.887317</td>
<td align="right">8.200888</td>
<td align="right">8.938344</td>
<td align="right">6.285554</td>
<td align="right">8.005636</td>
<td align="right">8.148735</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000606</td>
<td align="right">1.574509</td>
<td align="right">2.247673</td>
<td align="right">2.313362</td>
<td align="right">0.000000</td>
<td align="right">7.123984</td>
<td align="right">0.000000</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000617</td>
<td align="right">3.675876</td>
<td align="right">4.575481</td>
<td align="right">4.964792</td>
<td align="right">2.939321</td>
<td align="right">4.268654</td>
<td align="right">3.849151</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000627</td>
<td align="right">2.035284</td>
<td align="right">3.429754</td>
<td align="right">5.563693</td>
<td align="right">1.955415</td>
<td align="right">4.172838</td>
<td align="right">4.040765</td>
</tr>
</tbody>
</table>
</div>
<div id="statistiques-descriptives" class="section level2">
<h2>Statistiques descriptives</h2>
<p>Nous vous demandons de créer un data.frame qui collectera les statistiques par gène et par échantillon.</p>
<div id="claire-prépare-les-questions" class="section level3">
<h3>(Claire): prépare les questions</h3>
</div>
<div id="statistiques-par-échantillon-avant-normalisation" class="section level3">
<h3>Statistiques par échantillon avant normalisation</h3>
<p>Créez une data.frame nommée <code>gene_stat_prenorm</code> qui comportera une ligne par gène et une colonne par statistique, et calculez les statistiques suivantes sur les valeurs log2 de chaque gène.</p>
<ul>
<li>moyenne</li>
<li>médiane</li>
<li>écart-type</li>
<li>premier quartile</li>
<li>troisième quartile</li>
<li>maximum</li>
<li>nombre de valeurs nulles</li>
<li>intervalle inter-quartiles</li>
</ul>
<p>Ces résultats seront sotckés dans un data.frame avec 1 ligne par échantillon et 1 colonne par statistique, qui sera affiché avec la fonction <code>kable()</code> (n’oubliez pas la légende).</p>
<p>Affichez les lignes 100 à 109 de ce tableau de statistiques.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Gene-wise statistics before normalisation</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>gene_stat_prenorm <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">mean =</span> <span class="fu">apply</span>(fa_expr_raw, <span class="dv">1</span>, mean),</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">median =</span> <span class="fu">apply</span>(fa_expr_raw, <span class="dv">1</span>, median),</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">sd =</span> <span class="fu">apply</span>(fa_expr_raw, <span class="dv">1</span>, sd),</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">Q1 =</span> <span class="fu">apply</span>(fa_expr_raw, <span class="dv">1</span>, quantile, <span class="at">p =</span> <span class="fl">0.25</span>),</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">Q3 =</span> <span class="fu">apply</span>(fa_expr_raw, <span class="dv">1</span>, quantile, <span class="at">p =</span> <span class="fl">0.75</span>),</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">max =</span> <span class="fu">apply</span>(fa_expr_raw, <span class="dv">1</span>, max),</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">null =</span> <span class="fu">apply</span>(fa_expr_raw <span class="sc">==</span> <span class="dv">0</span>, <span class="dv">1</span>, sum),</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">iqr =</span> <span class="fu">apply</span>(fa_expr_raw, <span class="dv">1</span>, IQR)</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(gene_stat_prenorm[<span class="dv">100</span><span class="sc">:</span><span class="dv">109</span>, ], <span class="at">caption =</span> <span class="st">&quot;Gene-wise statistics before normalisation.&quot;</span>)</span></code></pre></div>
<table style="width:100%;">
<caption>Gene-wise statistics before normalisation.</caption>
<colgroup>
<col width="17%" />
<col width="11%" />
<col width="11%" />
<col width="11%" />
<col width="11%" />
<col width="10%" />
<col width="10%" />
<col width="4%" />
<col width="11%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">mean</th>
<th align="right">median</th>
<th align="right">sd</th>
<th align="right">Q1</th>
<th align="right">Q3</th>
<th align="right">max</th>
<th align="right">null</th>
<th align="right">iqr</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ENSMUSG00000000567</td>
<td align="right">320.336744</td>
<td align="right">261.659973</td>
<td align="right">286.399084</td>
<td align="right">91.046357</td>
<td align="right">455.40866</td>
<td align="right">1052.68945</td>
<td align="right">0</td>
<td align="right">364.362298</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000568</td>
<td align="right">857.348825</td>
<td align="right">711.096105</td>
<td align="right">694.432189</td>
<td align="right">575.571199</td>
<td align="right">858.55015</td>
<td align="right">3406.45030</td>
<td align="right">0</td>
<td align="right">282.978949</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000579</td>
<td align="right">422.075702</td>
<td align="right">447.572742</td>
<td align="right">162.290415</td>
<td align="right">340.417017</td>
<td align="right">498.14962</td>
<td align="right">728.51737</td>
<td align="right">0</td>
<td align="right">157.732606</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000581</td>
<td align="right">498.978227</td>
<td align="right">471.269224</td>
<td align="right">248.612029</td>
<td align="right">320.599085</td>
<td align="right">579.92342</td>
<td align="right">942.51104</td>
<td align="right">0</td>
<td align="right">259.324338</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000594</td>
<td align="right">2113.558334</td>
<td align="right">1868.718762</td>
<td align="right">1237.521694</td>
<td align="right">1400.226182</td>
<td align="right">2337.79729</td>
<td align="right">6408.96044</td>
<td align="right">0</td>
<td align="right">937.571105</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000600</td>
<td align="right">2299.978163</td>
<td align="right">2093.572929</td>
<td align="right">911.222891</td>
<td align="right">1719.874980</td>
<td align="right">3200.06756</td>
<td align="right">3674.18811</td>
<td align="right">0</td>
<td align="right">1480.192582</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000605</td>
<td align="right">325.552118</td>
<td align="right">282.608205</td>
<td align="right">169.456624</td>
<td align="right">221.735904</td>
<td align="right">417.61546</td>
<td align="right">791.31256</td>
<td align="right">0</td>
<td align="right">195.879561</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000606</td>
<td align="right">9.929258</td>
<td align="right">1.934744</td>
<td align="right">32.151350</td>
<td align="right">1.178311</td>
<td align="right">3.91509</td>
<td align="right">138.48677</td>
<td align="right">4</td>
<td align="right">2.736779</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000617</td>
<td align="right">18.804116</td>
<td align="right">18.150391</td>
<td align="right">9.470819</td>
<td align="right">12.188276</td>
<td align="right">25.84298</td>
<td align="right">36.00235</td>
<td align="right">0</td>
<td align="right">13.654701</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000627</td>
<td align="right">15.623306</td>
<td align="right">10.064895</td>
<td align="right">14.341852</td>
<td align="right">4.780913</td>
<td align="right">17.08904</td>
<td align="right">46.29752</td>
<td align="right">0</td>
<td align="right">12.308127</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="filtrage-et-normalisation-des-données" class="section level2">
<h2>Filtrage et normalisation des données</h2>
<p>Nous fournissons ci-dessous le code pour normaliser les données, en standardisant le 3ème quantile.</p>
<p>La méthode choisie ici consiste à</p>
<ul>
<li><p>écarter les gènes “non-détectés”, c’est -à-dire ceux ayant des valeurs nulles dans au moins 90% des échantillons;</p></li>
<li><p>écarter les gènes à peine exprimés, c’est-à-dire ceux ayant une valeur moyenne &lt; 10 (arbitrairement);</p></li>
<li><p>standardiser les échantillons sur le 3ème quartile des valeurs non-nulles. Nous vous rappelons que d’autres méthodes plus élaborées ont été vues dans les modules 4 et 5, mais nous avons choisi ici une solution simple tout en étant robuste.</p></li>
</ul>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Data filtering: genes having at least 90% null values</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>undetected_genes <span class="ot">&lt;-</span> gene_stat_prenorm<span class="sc">$</span>null <span class="sc">&gt;=</span> <span class="fu">ncol</span>(fa_expr_raw) <span class="sc">*</span> <span class="fl">0.9</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">&quot;Undetected genes (null in &gt;= 90% samples): &quot;</span>, <span class="fu">sum</span>(undetected_genes)))</span></code></pre></div>
<pre><code>[1] &quot;Undetected genes (null in &gt;= 90% samples): 14380&quot;</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Data filtering: genes having a mean expression &lt; 10</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>barely_expressed_genes <span class="ot">&lt;-</span> gene_stat_prenorm<span class="sc">$</span>mean <span class="sc">&lt;</span> <span class="dv">10</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">&quot;Barely expressed genes (mean &lt; 10): &quot;</span>, <span class="fu">sum</span>(barely_expressed_genes)))</span></code></pre></div>
<pre><code>[1] &quot;Barely expressed genes (mean &lt; 10): 26286&quot;</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Apply filtering on both criteria</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>discarded_genes <span class="ot">&lt;-</span> undetected_genes <span class="sc">|</span> barely_expressed_genes</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">&quot;Discarded genes: &quot;</span>, <span class="fu">sum</span>(discarded_genes)))</span></code></pre></div>
<pre><code>[1] &quot;Discarded genes: 26288&quot;</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>kept_genes <span class="ot">&lt;-</span> <span class="sc">!</span>discarded_genes</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">&quot;Kept genes: &quot;</span>, <span class="fu">sum</span>(kept_genes)))</span></code></pre></div>
<pre><code>[1] &quot;Kept genes: 20391&quot;</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Genes after filtering</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>fa_expr_log2_filtered <span class="ot">&lt;-</span> fa_expr_log2[kept_genes, ]</span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Generate a data frame where null values are replaced by NA</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>fa_expr_nonull <span class="ot">&lt;-</span> fa_expr_log2_filtered</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>fa_expr_nonull[fa_expr_log2_filtered <span class="sc">&lt;=</span> <span class="dv">0</span>] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">is.na</span>(fa_expr_nonull))</span></code></pre></div>
<pre><code>[1] 5598</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Compute the 3rd quartile of non-null values for each sample</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>sample_q3_nonull <span class="ot">&lt;-</span> <span class="fu">apply</span>(fa_expr_nonull, <span class="dv">2</span>, quantile, <span class="at">prob =</span> <span class="fl">0.75</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co"># print(sample_q3_nonull)</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Compute the A3 for all the values, which will serve as target value for the standardised sample Q3</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>all_q3_nonull <span class="ot">&lt;-</span> <span class="fu">quantile</span>(<span class="fu">unlist</span>(fa_expr_nonull), <span class="at">prob =</span> <span class="fl">0.75</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="co"># print(all_q3_nonull)</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="do">## Standardise expression on  the third quartile of non-null values</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a><span class="do">## Beware : for this standardization we keep the null values</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a><span class="do">## Trick : I transpose the table to apply the ratio sample per sample, </span></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a><span class="do">## and then transpose the result to get the genes in rows and samples in columns</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>fa_expr_log2_standard <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">t</span>(fa_expr_log2_filtered) <span class="sc">/</span> sample_q3_nonull <span class="sc">*</span> all_q3_nonull)</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a><span class="co"># quantile(unlist(fa_expr_log2_standard), probs = 0.75, na.rm = TRUE)</span></span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a><span class="do">## I also compute the values for the &quot;nonull&quot; table for </span></span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a><span class="do">## the sake of comparison and to check that the third quantiles of non-null </span></span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a><span class="do">## values are well identical across samples.</span></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a>fa_expr_log2_standard_nonull <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">t</span>(fa_expr_nonull) <span class="sc">/</span> sample_q3_nonull <span class="sc">*</span> all_q3_nonull)</span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a><span class="co"># quantile(unlist(fa_expr_log2_standard_nonull), probs = 0.75, na.rm = TRUE)</span></span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true" tabindex="-1"></a>standardisation_impact <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">before_all =</span> <span class="fu">apply</span>(fa_expr_log2_filtered, <span class="dv">2</span>, quantile, <span class="at">prob =</span>  <span class="fl">0.75</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb17-25"><a href="#cb17-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">before_nonull =</span> <span class="fu">apply</span>(fa_expr_nonull, <span class="dv">2</span>, quantile, <span class="at">prob =</span>  <span class="fl">0.75</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb17-26"><a href="#cb17-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">after_nonul =</span> <span class="fu">apply</span>(fa_expr_log2_standard_nonull, <span class="dv">2</span>, quantile, <span class="at">prob =</span>  <span class="fl">0.75</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb17-27"><a href="#cb17-27" aria-hidden="true" tabindex="-1"></a>  <span class="at">after_all =</span> <span class="fu">apply</span>(fa_expr_log2_standard, <span class="dv">2</span>, quantile, <span class="at">prob =</span>  <span class="fl">0.75</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-28"><a href="#cb17-28" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-29"><a href="#cb17-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-30"><a href="#cb17-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-31"><a href="#cb17-31" aria-hidden="true" tabindex="-1"></a><span class="do">## Note: after standardization the Q3 show some variations because we compute them here with the null values</span></span>
<span id="cb17-32"><a href="#cb17-32" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(standardisation_impact, <span class="at">caption =</span> <span class="st">&quot;Impact of standardization on the third quantile (Q3) per sample. Third quantiles are computed before and after standardisation, with either all the values of the filtered table, or only the non-null values. &quot;</span>)</span></code></pre></div>
<table>
<caption>Impact of standardization on the third quantile (Q3) per sample. Third quantiles are computed before and after standardisation, with either all the values of the filtered table, or only the non-null values.</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">before_all</th>
<th align="right">before_nonull</th>
<th align="right">after_nonul</th>
<th align="right">after_all</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">day1_1</td>
<td align="right">8.275875</td>
<td align="right">8.310843</td>
<td align="right">8.546124</td>
<td align="right">8.510165</td>
</tr>
<tr class="even">
<td align="left">day1_2</td>
<td align="right">9.215737</td>
<td align="right">9.236819</td>
<td align="right">8.546124</td>
<td align="right">8.526618</td>
</tr>
<tr class="odd">
<td align="left">day1_3</td>
<td align="right">9.339495</td>
<td align="right">9.356787</td>
<td align="right">8.546124</td>
<td align="right">8.530330</td>
</tr>
<tr class="even">
<td align="left">day14_1</td>
<td align="right">8.181662</td>
<td align="right">8.194184</td>
<td align="right">8.546124</td>
<td align="right">8.533064</td>
</tr>
<tr class="odd">
<td align="left">day14_2</td>
<td align="right">8.339087</td>
<td align="right">8.364105</td>
<td align="right">8.546124</td>
<td align="right">8.520562</td>
</tr>
<tr class="even">
<td align="left">day14_3</td>
<td align="right">8.313689</td>
<td align="right">8.334105</td>
<td align="right">8.546124</td>
<td align="right">8.525188</td>
</tr>
<tr class="odd">
<td align="left">day2_1</td>
<td align="right">9.372222</td>
<td align="right">9.391459</td>
<td align="right">8.546124</td>
<td align="right">8.528618</td>
</tr>
<tr class="even">
<td align="left">day2_2</td>
<td align="right">6.401122</td>
<td align="right">6.563217</td>
<td align="right">8.546124</td>
<td align="right">8.335056</td>
</tr>
<tr class="odd">
<td align="left">day2_3</td>
<td align="right">8.428494</td>
<td align="right">8.462042</td>
<td align="right">8.546124</td>
<td align="right">8.512242</td>
</tr>
<tr class="even">
<td align="left">day3_1</td>
<td align="right">8.600914</td>
<td align="right">8.618327</td>
<td align="right">8.546124</td>
<td align="right">8.528857</td>
</tr>
<tr class="odd">
<td align="left">day3_2</td>
<td align="right">8.449052</td>
<td align="right">8.476090</td>
<td align="right">8.546124</td>
<td align="right">8.518862</td>
</tr>
<tr class="even">
<td align="left">day3_3</td>
<td align="right">8.929730</td>
<td align="right">8.945829</td>
<td align="right">8.546124</td>
<td align="right">8.530744</td>
</tr>
<tr class="odd">
<td align="left">day7_1</td>
<td align="right">8.637420</td>
<td align="right">8.652095</td>
<td align="right">8.546124</td>
<td align="right">8.531628</td>
</tr>
<tr class="even">
<td align="left">day7_2</td>
<td align="right">8.457495</td>
<td align="right">8.478531</td>
<td align="right">8.546124</td>
<td align="right">8.524920</td>
</tr>
<tr class="odd">
<td align="left">day7_3</td>
<td align="right">8.568555</td>
<td align="right">8.585032</td>
<td align="right">8.546124</td>
<td align="right">8.529721</td>
</tr>
<tr class="even">
<td align="left">normal_1</td>
<td align="right">7.892971</td>
<td align="right">7.929471</td>
<td align="right">8.546124</td>
<td align="right">8.506785</td>
</tr>
<tr class="odd">
<td align="left">normal_2</td>
<td align="right">9.095259</td>
<td align="right">9.120100</td>
<td align="right">8.546124</td>
<td align="right">8.522846</td>
</tr>
<tr class="even">
<td align="left">normal_3</td>
<td align="right">7.690496</td>
<td align="right">7.728764</td>
<td align="right">8.546124</td>
<td align="right">8.503808</td>
</tr>
</tbody>
</table>
<div id="statistiques-par-gène-après-normalisation-et-annotation-des-gènes" class="section level3">
<h3>Statistiques par gène après normalisation, et annotation des gènes</h3>
<p>Générez un data.frame avec une ligne par gène à partir du tableau de données normalisées, avec les statistiques suivantes (une statistique par colonne):</p>
<ul>
<li>moyenne</li>
<li>écart-type</li>
<li>coefficient de variation</li>
<li>minimum</li>
<li>médiane</li>
<li>maximum</li>
</ul>
<p>Ajoutez une colonne avec les annotations biomart de chaque gène/</p>
<p>Affichez les lignes 100 à 109 de ce tableau de statistiques.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Stat per gene after normalisation</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Gene-wise statistics before normalisation</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>gene_stat_norm <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">mean =</span> <span class="fu">apply</span>(fa_expr_log2_standard, <span class="dv">1</span>, mean),</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">sd =</span> <span class="fu">apply</span>(fa_expr_log2_standard, <span class="dv">1</span>, sd))</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>gene_stat_norm<span class="sc">$</span>V <span class="ot">=</span> gene_stat_norm<span class="sc">$</span>sd <span class="sc">/</span> gene_stat_norm<span class="sc">$</span>mean</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>gene_stat_norm<span class="sc">$</span>min <span class="ot">=</span> <span class="fu">apply</span>(fa_expr_log2_standard, <span class="dv">1</span>, min)</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>gene_stat_norm<span class="sc">$</span>median <span class="ot">=</span> <span class="fu">apply</span>(fa_expr_log2_standard, <span class="dv">1</span>, median)</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>gene_stat_norm<span class="sc">$</span>max <span class="ot">=</span> <span class="fu">apply</span>(fa_expr_log2_standard, <span class="dv">1</span>, max)</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(gene_stat_norm[<span class="dv">100</span><span class="sc">:</span><span class="dv">109</span>, ], <span class="at">caption =</span> <span class="st">&quot;Gene-wise statistics after normalisation.&quot;</span>)</span></code></pre></div>
<table>
<caption>Gene-wise statistics after normalisation.</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">mean</th>
<th align="right">sd</th>
<th align="right">V</th>
<th align="right">min</th>
<th align="right">median</th>
<th align="right">max</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ENSMUSG00000000711</td>
<td align="right">9.938709</td>
<td align="right">0.9412004</td>
<td align="right">0.0947005</td>
<td align="right">9.232061</td>
<td align="right">9.727752</td>
<td align="right">13.292801</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000724</td>
<td align="right">5.595006</td>
<td align="right">0.4441668</td>
<td align="right">0.0793863</td>
<td align="right">4.816940</td>
<td align="right">5.632564</td>
<td align="right">6.426584</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000731</td>
<td align="right">4.936591</td>
<td align="right">0.7660500</td>
<td align="right">0.1551779</td>
<td align="right">3.715427</td>
<td align="right">4.984400</td>
<td align="right">6.247104</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000732</td>
<td align="right">5.543710</td>
<td align="right">1.6080743</td>
<td align="right">0.2900719</td>
<td align="right">2.534509</td>
<td align="right">5.736471</td>
<td align="right">8.250127</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000738</td>
<td align="right">8.426883</td>
<td align="right">0.5240653</td>
<td align="right">0.0621897</td>
<td align="right">7.477862</td>
<td align="right">8.330412</td>
<td align="right">9.484120</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000740</td>
<td align="right">10.820838</td>
<td align="right">0.3647941</td>
<td align="right">0.0337122</td>
<td align="right">10.156481</td>
<td align="right">10.850259</td>
<td align="right">11.622910</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000743</td>
<td align="right">7.619693</td>
<td align="right">1.0860052</td>
<td align="right">0.1425261</td>
<td align="right">6.110121</td>
<td align="right">7.398740</td>
<td align="right">10.866373</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000751</td>
<td align="right">7.368976</td>
<td align="right">0.7229245</td>
<td align="right">0.0981038</td>
<td align="right">5.984878</td>
<td align="right">7.241989</td>
<td align="right">9.010795</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000753</td>
<td align="right">2.830650</td>
<td align="right">1.7230619</td>
<td align="right">0.6087160</td>
<td align="right">0.000000</td>
<td align="right">3.131414</td>
<td align="right">5.177214</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000759</td>
<td align="right">8.617902</td>
<td align="right">0.5502123</td>
<td align="right">0.0638453</td>
<td align="right">7.904374</td>
<td align="right">8.497898</td>
<td align="right">10.039692</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="distribution-des-données" class="section level2">
<h2>Distribution des données</h2>
<ul>
<li>Dessinez la distribution des valeurs après normalisation (tous échantillons confondus)</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">unlist</span>(fa_expr_log2_standard), <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">0</span>, <span class="at">to =</span> <span class="fu">max</span>(fa_expr_log2_standard) <span class="sc">+</span> <span class="dv">1</span>, <span class="at">by =</span> <span class="fl">0.25</span>),</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">&quot;log2(counts) after standardisation&quot;</span>, </span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">&quot;number of genes afterfiltering&quot;</span>,</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="st">&quot;#BBDDFF&quot;</span>,</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">las =</span> <span class="dv">1</span>, <span class="at">cex.axis =</span> <span class="fl">0.8</span>,</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">&quot;distribution after standardisation&quot;</span>)</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> <span class="fu">mean</span>(fa_expr_log2_standard), <span class="at">col =</span> <span class="st">&quot;darkgreen&quot;</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/tcga-bic_fa_expr_norm_distrib-1.png" alt="Distribution of expression values (log2 counts) after gene filtering and standardisation on the sample-wise third-quartile of non-null values. The vertical line highlights the mean value. " width="60%" />
<p class="caption">
Distribution of expression values (log2 counts) after gene filtering and standardisation on the sample-wise third-quartile of non-null values. The vertical line highlights the mean value.
</p>
</div>
<ul>
<li>Dessinez un box plot des échantillons avant et après normalisation, et commentez la façon dont l’effet de la normalisation apparaît sur ces graphiques.</li>
</ul>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span></code></pre></div>
</div>
<div id="filtrage" class="section level2">
<h2>Filtrage</h2>
<ul>
<li>Sélectionnez les gènes ayant un niveau log2 moyen minimal supérieur à 4 et un coefficient de variation</li>
</ul>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="do">## (Jacques): prépare  ##</span></span></code></pre></div>
</div>
<div id="acp" class="section level2">
<h2>ACP</h2>
<ul>
<li>plot des gènes: Jacques? Anne et Claire ne sont pas sûres que ce soient pertinent!</li>
<li>plot des échantillons en colrant par condition (time point) en prenant les gènes comme variables</li>
</ul>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="do">## (Jacques ou Anne): prépare  ##</span></span></code></pre></div>
</div>
<div id="clustering" class="section level2">
<h2>Clustering</h2>
<ol style="list-style-type: decimal">
<li>matrice de distance</li>
<li>hclust dans les 2 sens</li>
<li>heatmap</li>
<li>représenter sur l’ACP les lcusters d’échantillon</li>
</ol>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="do">## (Anne): prépare  ##</span></span></code></pre></div>
</div>

<div id="rmd-source-code">---
title: "Mini-projet 2021 - Exploration des données de Pavkovic"
author: "Prénom Nom"
date: '`r Sys.Date()`'
output:
  html_document:
    self_contained: no
    code_download: yes
    fig_caption: yes
    highlight: zenburn
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
    code_folding: "hide"
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
editor_options: 
  chunk_output_type: console
---




```{r settings, include=FALSE, echo=FALSE, eval=TRUE}

options(width = 300)
# options(encoding = 'UTF-8')
knitr::opts_chunk$set(
  fig.width = 7, fig.height = 5, 
  fig.path = 'figures/tcga-bic_',
  fig.align = "center", 
  size = "tiny", 
  echo = TRUE, 
  eval = TRUE, 
  warning = FALSE, 
  message = FALSE, 
  results = TRUE, 
  comment = "")

options(scipen = 12) ## Max number of digits for non-scientific notation
# knitr::asis_output("\\footnotesize")

```



```{r libraries, echo=FALSE, eval=TRUE}
#### Load required R libraries ####

## CRAN libraries
requiredLib <- c("knitr")
for (lib in requiredLib) {
  if (!require(lib, character.only = TRUE)) {
    install.packages(lib, )
  }
  require(lib, character.only = TRUE)
}



```


## Synopsis du projet


Le but du projet est de mettre en oeuvre les méthodes vues dans le module 3 "R et statistiques" pour explorer le jeu de données de Pavokovic, et de fournir un rapport d'analyse au format R markdown. 

Nous fournissons ci-dessous une trame avec les principales sections attendus. Sentez-vous libres d'adapter cette trame ou d'y ajouter des analyses complémentaires si elles vous aident à interpréter vos résultats. 

## Remise du rapport

Date: **le 10 mai 2021 minuit**.  Si vous anticipez un problème pour remettre le rapport à cette date contactez-nous aussi rapidement que possible pour que nous puissions prévoir une remise plus tardive. 

- Commencez par renommer le fichier Rmd en remplaçant Prenom-NOM par vos nom et prénom. 
- Le rapport est attendu en formats Rmd + HTML (en activant l'option self_contained de l'en-tête). 
- Déposez les fichiers dans un sous-dossier de vote compte du cluster. Attention, veillez à respecter précisément cette structure de chemin car nous nous baserons dessus pour récupérer vos résultats. 

    `/shared/projects/dubii2021/[login]/m3-stat-R/mini-projet` 


## Critères d'évaluation

- Reproductibilité des analyses: nous tenterons de regénérer le rapport HTML à partir de votre Rmd, en partant de notre compte sur le serveur IFB. 
- Manipulation des objets R
- Mobilisation des méthodes statistiques vues au cours
- Pertinence des interprétations statistiques
- Pertinence des interprétations biologiques
- Clarté de la rédaction
- Clarté des illustrations (figures et tableaux): graphismes, légendes ...

Nous vous encourageons à assurer la lisibilpité de votre code (syntaxe, nommage des variables, commentaires de code)

## Chargement des données brutes

Le bloc suivant contient une  fonction qui permet de télécharger un fichier dans l'espace de travail, sauf s'il est déjà présent. Nous l'utiliserons ensuite pour télécharger les données à analyser en évitant de refaire le transfert à chaque exécution de l'analyse. 

```{r function_download_only_once}
#' @title Download a file only if it is not yet here
#' @author Jacques van Helden email{Jacques.van-Helden@@france-bioinformatique.fr}
#' @param url_base base of the URL, that will be prepended to the file name
#' @param file_name name of the file (should not contain any path)
#' @param local_folder path of a local folder where the file should be stored
#' @return the function returns the path of the local file, built from local_folder and file_name
#' @export©
download_only_once <- function(
  url_base, 
  file_name,
  local_folder) {

  ## Define the source URL  
  url <- file.path(url_base, file_name)
  message("Source URL\n\t",  url)

  ## Define the local file
  local_file <- file.path(local_folder, file_name)
  
  ## Create the local data folder if it does not exist
  dir.create(local_folder, showWarnings = FALSE, recursive = TRUE)
  
  ## Download the file ONLY if it is not already there
  if (!file.exists(local_file)) {
    message("Downloading file from source URL to local file\n\t", 
            local_file)
    download.file(url = url, destfile = local_file)
  } else {
    message("Local file already exists, no need to download\n\t", 
            local_file)
  }
  
  return(local_file)
}
```

Nous téléchargeons deux fichiers dans un dossier local, et les chargeons dans les data.frames suivantes: 

- Données brutes de transcriptome: `fa_expr_raw`
- Métadonnées: `fa_meta`

```{r download_and_load}
## Define the remote URL and local folder
pavkovic_url <- "https://github.com/DU-Bii/module-3-Stat-R/raw/master/stat-R_2021/data/pavkovic_2019/"

## Define the local folder for this analysis (where the data will be downloaded and the results generated)
pavkovic_folder <- "~/m3-stat-R/pavkovic_analysis"

## Define a sub-folder for the data
pavkovic_data_folder <- file.path(pavkovic_folder, "data")

## Download and load the expression data table
## Note: we use check.names=FALSE to avoid replacing hyphens by dots
## in sample names, because we want to keep them as in the 
## original data files. 
message("Downloading FA transcriptome file\t", "fa_raw_counts.tsv.gz",
  "\n\tfrom\t", pavkovic_url)
fa_expr_file <- download_only_once(
  url_base = pavkovic_url, 
  file_name = "fa_raw_counts.tsv.gz",
  local_folder = pavkovic_data_folder)
message("Loading FA transcriptome data from\n\t", fa_expr_file)
fa_expr_raw <- read.delim(file = fa_expr_file, 
                       header = TRUE, 
                       row.names = 1)
# colnames(fa_expr_raw)
# dim(fa_expr_raw)
# View(head(fa_expr_raw))

## Download the metadata file
message("Downloading FA metadata file\t", "fa_transcriptome_metadata.tsv",
  "\n\tfrom\t", pavkovic_url)
fa_meta_file <- download_only_once(
  url_base = pavkovic_url, 
  file_name = "fa_transcriptome_metadata.tsv",
  local_folder = pavkovic_data_folder)
message("Loading FA metadata from\n\t", fa_meta_file)
fa_meta <- read.delim(file = fa_meta_file, 
                       header = TRUE, 
                       row.names = 1)

# View(fa_meta)
kable(fa_meta, caption = "Metdata for Pavkovoc FA transcriptome")
```




## Transformation log2

Appliquez une transformation log2 des données brutes, après avoir ajouté un epsilon $\epsilon = 1$ (les valeurs nulles seront donc représentées par un log2(counts) valant $0$. Stockez le résultat dans une data.frame nommée `fa_expr_log2`.

Affchez un fragment des tableaux `fa_expr_raw` et `fa_expr_log2` en sélectionnant les lignes 100 à 109 et les colonnes 5 à 10, afin de vous assurer que la transformation log2 a bien fonctionné. 

```{r log2_transform}
## Log2 transformation of the transcriptome data
epsilon <- 1
fa_expr_log2 <- log2(fa_expr_raw + epsilon)
# dim(fa_expr_log2)
# View(head(fa_expr_log2))

## Display of a fragment of the data before and after log2 transformation
kable(fa_expr_raw[100:109, 5:10], caption = "Fragment des données transcriptomiques brutes")
kable(fa_expr_log2[100:109, 5:10], caption = "Fragment des données transcriptomiques après transformation log2")
```


## Statistiques descriptives

Nous vous demandons de créer un data.frame qui collectera les statistiques par gène et par échantillon. 

### (Claire): prépare les questions ###


### Statistiques par échantillon avant normalisation

Créez une data.frame nommée `gene_stat_prenorm` qui comportera une ligne par gène et une colonne par statistique, et calculez les statistiques suivantes sur les valeurs log2 de chaque gène. 

- moyenne
- médiane
- écart-type
- premier quartile
- troisième quartile
- maximum
- nombre de valeurs nulles
- intervalle inter-quartiles

Ces résultats seront sotckés dans un data.frame avec 1 ligne par échantillon et 1 colonne par statistique, qui sera affiché avec la fonction `kable()` (n'oubliez pas la légende). 

Affichez les lignes 100 à 109 de ce tableau de statistiques. 

```{r gene_stat_pre_norm}
## Gene-wise statistics before normalisation
gene_stat_prenorm <- data.frame(
  mean = apply(fa_expr_raw, 1, mean),
  median = apply(fa_expr_raw, 1, median),
  sd = apply(fa_expr_raw, 1, sd),
  Q1 = apply(fa_expr_raw, 1, quantile, p = 0.25),
  Q3 = apply(fa_expr_raw, 1, quantile, p = 0.75),
  max = apply(fa_expr_raw, 1, max),
  null = apply(fa_expr_raw == 0, 1, sum),
  iqr = apply(fa_expr_raw, 1, IQR)
)

kable(gene_stat_prenorm[100:109, ], caption = "Gene-wise statistics before normalisation.")

```




## Filtrage et normalisation des données

Nous fournissons ci-dessous le code pour normaliser les données, en standardisant le 3ème quantile. 

La méthode choisie ici consiste à 

- écarter les gènes "non-détectés", c'est -à-dire ceux ayant des valeurs nulles dans au moins 90% des échantillons;

- écarter les gènes à peine exprimés, c'est-à-dire ceux ayant une valeur moyenne < 10 (arbitrairement);

- standardiser les échantillons sur le 3ème quartile des valeurs non-nulles. Nous vous rappelons que d'autres méthodes plus élaborées ont été vues dans les modules 4 et 5, mais nous avons choisi ici une solution simple tout en étant robuste. 

```{r gene_filtering}
## Data filtering: genes having at least 90% null values
undetected_genes <- gene_stat_prenorm$null >= ncol(fa_expr_raw) * 0.9
print(paste0("Undetected genes (null in >= 90% samples): ", sum(undetected_genes)))

## Data filtering: genes having a mean expression < 10
barely_expressed_genes <- gene_stat_prenorm$mean < 10
print(paste0("Barely expressed genes (mean < 10): ", sum(barely_expressed_genes)))

## Apply filtering on both criteria
discarded_genes <- undetected_genes | barely_expressed_genes
print(paste0("Discarded genes: ", sum(discarded_genes)))
kept_genes <- !discarded_genes
print(paste0("Kept genes: ", sum(kept_genes)))

## Genes after filtering
fa_expr_log2_filtered <- fa_expr_log2[kept_genes, ]

```


```{r sample_standardisation}
## Generate a data frame where null values are replaced by NA
fa_expr_nonull <- fa_expr_log2_filtered
fa_expr_nonull[fa_expr_log2_filtered <= 0] <- NA
sum(is.na(fa_expr_nonull))

## Compute the 3rd quartile of non-null values for each sample
sample_q3_nonull <- apply(fa_expr_nonull, 2, quantile, prob = 0.75, na.rm = TRUE)
# print(sample_q3_nonull)

## Compute the A3 for all the values, which will serve as target value for the standardised sample Q3
all_q3_nonull <- quantile(unlist(fa_expr_nonull), prob = 0.75, na.rm = TRUE)
# print(all_q3_nonull)

## Standardise expression on  the third quartile of non-null values
## Beware : for this standardization we keep the null values
##
## Trick : I transpose the table to apply the ratio sample per sample, 
## and then transpose the result to get the genes in rows and samples in columns
fa_expr_log2_standard <- t(t(fa_expr_log2_filtered) / sample_q3_nonull * all_q3_nonull)
# quantile(unlist(fa_expr_log2_standard), probs = 0.75, na.rm = TRUE)

## I also compute the values for the "nonull" table for 
## the sake of comparison and to check that the third quantiles of non-null 
## values are well identical across samples.
fa_expr_log2_standard_nonull <- t(t(fa_expr_nonull) / sample_q3_nonull * all_q3_nonull)
# quantile(unlist(fa_expr_log2_standard_nonull), probs = 0.75, na.rm = TRUE)

standardisation_impact <- data.frame(
  before_all = apply(fa_expr_log2_filtered, 2, quantile, prob =  0.75, na.rm = TRUE),
  before_nonull = apply(fa_expr_nonull, 2, quantile, prob =  0.75, na.rm = TRUE),
  after_nonul = apply(fa_expr_log2_standard_nonull, 2, quantile, prob =  0.75, na.rm = TRUE),
  after_all = apply(fa_expr_log2_standard, 2, quantile, prob =  0.75, na.rm = TRUE)
)


## Note: after standardization the Q3 show some variations because we compute them here with the null values
kable(standardisation_impact, caption = "Impact of standardization on the third quantile (Q3) per sample. Third quantiles are computed before and after standardisation, with either all the values of the filtered table, or only the non-null values. ")

```


### Statistiques par gène après normalisation, et annotation des gènes


Générez un data.frame avec une ligne par gène à partir du tableau de données normalisées, avec les statistiques suivantes (une statistique par colonne):

- moyenne
- écart-type
- coefficient de variation
- minimum
- médiane
- maximum

Ajoutez une colonne avec les annotations biomart de chaque gène/ 

Affichez les lignes 100 à 109 de ce tableau de statistiques. 


```{r gene_stat_post_norm}
## Stat per gene after normalisation
## Gene-wise statistics before normalisation
gene_stat_norm <- data.frame(
  mean = apply(fa_expr_log2_standard, 1, mean),
  sd = apply(fa_expr_log2_standard, 1, sd))
gene_stat_norm$V = gene_stat_norm$sd / gene_stat_norm$mean
gene_stat_norm$min = apply(fa_expr_log2_standard, 1, min)
gene_stat_norm$median = apply(fa_expr_log2_standard, 1, median)
gene_stat_norm$max = apply(fa_expr_log2_standard, 1, max)

kable(gene_stat_norm[100:109, ], caption = "Gene-wise statistics after normalisation.")


```


## Distribution des données

- Dessinez la distribution des valeurs après normalisation (tous échantillons confondus)

```{r fa_expr_norm_distrib, fig.width=8, fig.height=5, out.width="60%", fig.cap="Distribution of expression values (log2 counts) after gene filtering and standardisation on the sample-wise third-quartile of non-null values. The vertical line highlights the mean value. "}
hist(unlist(fa_expr_log2_standard), breaks = seq(from = 0, to = max(fa_expr_log2_standard) + 1, by = 0.25),
     xlab = "log2(counts) after standardisation", 
     ylab = "number of genes afterfiltering",
     col = "#BBDDFF",
     las = 1, cex.axis = 0.8,
     main = "distribution after standardisation")
abline(v = mean(fa_expr_log2_standard), col = "darkgreen", lwd = 2)

```


- Dessinez un box plot des échantillons avant et après normalisation, et commentez la façon dont l'effet de la normalisation apparaît sur ces graphiques. 

```{r }
## 

```

## Filtrage

- Sélectionnez les gènes ayant un niveau log2 moyen minimal supérieur à 4 et un coefficient de variation 


```{r }
## (Jacques): prépare  ##

```

## ACP

- plot des gènes: Jacques? Anne et Claire ne sont pas sûres que ce soient pertinent!
- plot des échantillons en colrant par condition (time point) en prenant les gènes comme variables

```{r }
## (Jacques ou Anne): prépare  ##

```

## Clustering

1. matrice de distance
2. hclust dans les 2 sens
3. heatmap
4. représenter sur l'ACP les lcusters d'échantillon

```{r }
## (Anne): prépare  ##

```



</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Prenom-NOM_m3-stat-R_mini-projet.Rmd");
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
