<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Claire Vandiedonck, Jacques van Helden, Anne Badel &amp; Olivier Sand" />

<meta name="date" content="2021-04-22" />

<title>Mini-projet 2021 - Exploration des données de Pavkovic</title>

<script src="solutions_m3-stat-R_mini-projet_files/jquery-3.5.1/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="solutions_m3-stat-R_mini-projet_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="solutions_m3-stat-R_mini-projet_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="solutions_m3-stat-R_mini-projet_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="solutions_m3-stat-R_mini-projet_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="solutions_m3-stat-R_mini-projet_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="solutions_m3-stat-R_mini-projet_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="solutions_m3-stat-R_mini-projet_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="solutions_m3-stat-R_mini-projet_files/navigation-1.1/tabsets.js"></script>
<script src="solutions_m3-stat-R_mini-projet_files/navigation-1.1/codefolding.js"></script>
<script src="solutions_m3-stat-R_mini-projet_files/navigation-1.1/sourceembed.js"></script>
<script src="solutions_m3-stat-R_mini-projet_files/htmlwidgets-1.5.3/htmlwidgets.js"></script>
<script src="solutions_m3-stat-R_mini-projet_files/plotly-binding-4.9.3/plotly.js"></script>
<script src="solutions_m3-stat-R_mini-projet_files/typedarray-0.1/typedarray.min.js"></script>
<link href="solutions_m3-stat-R_mini-projet_files/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet" />
<script src="solutions_m3-stat-R_mini-projet_files/crosstalk-1.1.1/js/crosstalk.min.js"></script>
<link href="solutions_m3-stat-R_mini-projet_files/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="solutions_m3-stat-R_mini-projet_files/plotly-main-1.57.1/plotly-latest.min.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  { color: #cccccc; background-color: #303030; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ffcfaf; } /* Alert */
code span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #dca3a3; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #f0dfaf; } /* ControlFlow */
code span.ch { color: #dca3a3; } /* Char */
code span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code span.co { color: #7f9f7f; } /* Comment */
code span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code span.do { color: #7f9f7f; } /* Documentation */
code span.dt { color: #dfdfbf; } /* DataType */
code span.dv { color: #dcdccc; } /* DecVal */
code span.er { color: #c3bf9f; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #c0bed1; } /* Float */
code span.fu { color: #efef8f; } /* Function */
code span.im { } /* Import */
code span.in { color: #7f9f7f; font-weight: bold; } /* Information */
code span.kw { color: #f0dfaf; } /* Keyword */
code span.op { color: #f0efd0; } /* Operator */
code span.ot { color: #efef8f; } /* Other */
code span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code span.sc { color: #dca3a3; } /* SpecialChar */
code span.ss { color: #cc9393; } /* SpecialString */
code span.st { color: #cc9393; } /* String */
code span.va { } /* Variable */
code span.vs { color: #cc9393; } /* VerbatimString */
code span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>


<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-sm-12 col-md-4 col-lg-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-sm-12 col-md-8 col-lg-9">




<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
<li role="separator" class="divider"></li>
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Mini-projet 2021 - Exploration des données de Pavkovic</h1>
<h4 class="author">Claire Vandiedonck, Jacques van Helden, Anne Badel &amp; Olivier Sand</h4>
<h4 class="date">2021-04-22</h4>

</div>


<table>
<caption>Loaded required libraries</caption>
<thead>
<tr class="header">
<th align="left">libraries</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">knitr</td>
</tr>
<tr class="even">
<td align="left">FactoMineR</td>
</tr>
<tr class="odd">
<td align="left">factoextra</td>
</tr>
<tr class="even">
<td align="left">gprofiler2</td>
</tr>
<tr class="odd">
<td align="left">pheatmap</td>
</tr>
<tr class="even">
<td align="left">biomaRt</td>
</tr>
</tbody>
</table>
<div id="synopsis-du-projet" class="section level2">
<h2>Synopsis du projet</h2>
<div id="travail-demandé" class="section level3">
<h3>Travail demandé</h3>
<p>Le but de ce travail est de mettre en oeuvre les méthodes vues dans le module 3 “R et statistiques” pour explorer le jeu de données de Pavkovic, et de rendre un rapport d’analyse au format <code>.Rmd</code>.</p>
<p>Nous fournissons ci-dessous une trame avec les principales sections attendues. Certaines contiennent déjà du code. Vous devrez en compléter d’autres. Sentez-vous libres d’adapter cette trame ou d’y ajouter des analyses complémentaires si elles vous aident à interpréter vos résultats.</p>
</div>
<div id="remise-du-rapport" class="section level3">
<h3>Remise du rapport</h3>
<p>Date: <strong>le 22 mai 2021 à minuit</strong>. Si vous anticipez un problème pour remettre le rapport à cette date contactez-nous aussi rapidement que possible pour que nous puissions prévoir une remise plus tardive.</p>
<ul>
<li>Commencez par renommer le fichier .Rmd en remplaçant Prenom-NOM par vos nom et prénom.</li>
<li>Le rapport est attendu en formats .Rmd + .HTML (en gardant l’option self_contained de l’en-tête activée).</li>
<li><p>Déposez les fichiers dans un sous-dossier de vote compte du cluster. Attention, veillez à respecter précisément cette structure de chemin car nous nous baserons dessus pour récupérer vos résultats.</p>
<p><code>/shared/projects/dubii2021/[login]/m3-stat-R/mini-projet</code></p></li>
</ul>
</div>
<div id="critères-dévaluation" class="section level3">
<h3>Critères d’évaluation</h3>
<ul>
<li>Reproductibilité des analyses: nous tenterons de regénérer le rapport HTML à partir de votre Rmd, en partant de notre compte sur le serveur IFB.</li>
<li>Manipulation des objets R</li>
<li>Mobilisation des méthodes statistiques vues au cours</li>
<li>Pertinence des interprétations statistiques</li>
<li>Pertinence des interprétations biologiques</li>
<li>Clarté de la rédaction</li>
<li>Clarté des illustrations (figures et tableaux): graphismes, légendes …</li>
</ul>
<p>Nous vous encourageons à assurer la lisibilité de votre code (syntaxe, nommage des variables, commentaires de code)</p>
</div>
<div id="objectifs-scientifiques" class="section level3">
<h3>Objectifs scientifiques</h3>
<p>Nous partons du même jeu de données <em>Fil Rouge</em> de ce module issues de la publication Pavkovic, M., Pantano, L., Gerlach, C.V. et al. Multi omics analysis of fibrotic kidneys in two mouse models. Sci Data 6, 92 (2019). <a href="https://doi.org/10.1038/s41597-019-0095-5" class="uri">https://doi.org/10.1038/s41597-019-0095-5</a></p>
<p><strong>Rappel sur les échantillons:</strong></p>
<p>Deux modèles de fibrose rénale chez la souris sont étudiés:</p>
<ol style="list-style-type: decimal">
<li><p>Le premier est un modèle de néphropathie réversible induite par l’acide folique (folic acid (FA)). Les souris ont été sacrifiées avant le traitement (normal), puis à jour 1, 2, 7 et 14 (day1,…) après une seule injection d’acide folique.</p></li>
<li><p>Le second est un modèle irréversible induit chrirurgicalement (unilateral ureteral obstruction (UUO)). les souris ont été sacrifiées avant obstruction (day 0) et à 3, 7 et 14 jours après obstruction par ligation de l’uretère du rein gauche.</p></li>
</ol>
<p>A partir de ces extraits de rein, l’ARN messager total et les petits ARNs ont été séquencés et les protéines caratérisées par spectrométrie de masse en tandem (TMT).</p>
<p><strong>But scientifique:</strong> Dans le tutoriel sur les dataframes, vous avez travaillé sur les données de <strong><em>transcriptome du modèle UUO</em></strong>. Dans ce mini-projet, vous allez travailler sur les données du <strong><em>transcriptome du modèle FA</em></strong> afin de regrouper les observations (échantillon) et les gènes selon des profils d’expression similaires.</p>
<p><strong>Votre projet se décompose en 5 parties dont 3 seront à réaliser par:</strong></p>
<ol style="list-style-type: decimal">
<li>Statitiques descriptives des données brutes: commandes fournies</li>
<li>Normalisation des données : commandes fournies</li>
<li>Statistiques descriptives des données normalisées: à vous de jouer</li>
<li>Analyse de regroupement des données: à vous de jouer</li>
<li>Analyse d’enrichissement fonctionnel: à vous de jouer</li>
</ol>
</div>
</div>
<div id="les-données-brutes" class="section level2">
<h2>1. Les données brutes</h2>
<p><strong><em>Vous n’avez rien à coder ici. Le code est fourni.</em></strong></p>
<div id="chargement-des-données-brutes" class="section level3">
<h3>Chargement des données brutes</h3>
<p>Le bloc suivant contient une fonction qui permet de télécharger un fichier dans l’espace de travail, sauf s’il est déjà présent. Nous l’utiliserons ensuite pour télécharger les données à analyser en évitant de refaire le transfert à chaque exécution de l’analyse.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co">#&#39; @title Download a file only if it is not yet here</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="co">#&#39; @author Jacques van Helden email{Jacques.van-Helden@@france-bioinformatique.fr}</span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="co">#&#39; @param url_base base of the URL, that will be prepended to the file name</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="co">#&#39; @param file_name name of the file (should not contain any path)</span></a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="co">#&#39; @param local_folder path of a local folder where the file should be stored</span></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="co">#&#39; @return the function returns the path of the local file, built from local_folder and file_name</span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="co">#&#39; @export©</span></a>
<a class="sourceLine" id="cb1-8" data-line-number="8">download_only_once &lt;-<span class="st"> </span><span class="cf">function</span>(</a>
<a class="sourceLine" id="cb1-9" data-line-number="9">  url_base, </a>
<a class="sourceLine" id="cb1-10" data-line-number="10">  file_name,</a>
<a class="sourceLine" id="cb1-11" data-line-number="11">  local_folder) {</a>
<a class="sourceLine" id="cb1-12" data-line-number="12"></a>
<a class="sourceLine" id="cb1-13" data-line-number="13">  <span class="co">## Define the source URL  </span></a>
<a class="sourceLine" id="cb1-14" data-line-number="14">  url &lt;-<span class="st"> </span><span class="kw">file.path</span>(url_base, file_name)</a>
<a class="sourceLine" id="cb1-15" data-line-number="15">  <span class="kw">message</span>(<span class="st">&quot;Source URL</span><span class="ch">\n\t</span><span class="st">&quot;</span>,  url)</a>
<a class="sourceLine" id="cb1-16" data-line-number="16"></a>
<a class="sourceLine" id="cb1-17" data-line-number="17">  <span class="co">## Define the local file</span></a>
<a class="sourceLine" id="cb1-18" data-line-number="18">  local_file &lt;-<span class="st"> </span><span class="kw">file.path</span>(local_folder, file_name)</a>
<a class="sourceLine" id="cb1-19" data-line-number="19">  </a>
<a class="sourceLine" id="cb1-20" data-line-number="20">  <span class="co">## Create the local data folder if it does not exist</span></a>
<a class="sourceLine" id="cb1-21" data-line-number="21">  <span class="kw">dir.create</span>(local_folder, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>, <span class="dt">recursive =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb1-22" data-line-number="22">  </a>
<a class="sourceLine" id="cb1-23" data-line-number="23">  <span class="co">## Download the file ONLY if it is not already there</span></a>
<a class="sourceLine" id="cb1-24" data-line-number="24">  <span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(local_file)) {</a>
<a class="sourceLine" id="cb1-25" data-line-number="25">    <span class="kw">message</span>(<span class="st">&quot;Downloading file from source URL to local file</span><span class="ch">\n\t</span><span class="st">&quot;</span>, </a>
<a class="sourceLine" id="cb1-26" data-line-number="26">            local_file)</a>
<a class="sourceLine" id="cb1-27" data-line-number="27">    <span class="kw">download.file</span>(<span class="dt">url =</span> url, <span class="dt">destfile =</span> local_file)</a>
<a class="sourceLine" id="cb1-28" data-line-number="28">  } <span class="cf">else</span> {</a>
<a class="sourceLine" id="cb1-29" data-line-number="29">    <span class="kw">message</span>(<span class="st">&quot;Local file already exists, no need to download</span><span class="ch">\n\t</span><span class="st">&quot;</span>, </a>
<a class="sourceLine" id="cb1-30" data-line-number="30">            local_file)</a>
<a class="sourceLine" id="cb1-31" data-line-number="31">  }</a>
<a class="sourceLine" id="cb1-32" data-line-number="32">  </a>
<a class="sourceLine" id="cb1-33" data-line-number="33">  <span class="kw">return</span>(local_file)</a>
<a class="sourceLine" id="cb1-34" data-line-number="34">}</a></code></pre></div>
<p>Nous téléchargeons deux fichiers dans un dossier local <code>~/m3-stat-R/pavkovic_analysis</code> <strong>(vous pouvez changer le nom ou chemin dans le chunk ci-dessous)</strong>, et les chargeons dans les data.frames suivants:</p>
<ul>
<li>Données brutes de transcriptome: <code>fa_expr_raw</code></li>
<li>Métadonnées: <code>fa_meta</code></li>
</ul>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co">## Define the remote URL and local folder</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2">pavkovic_url &lt;-<span class="st"> &quot;https://github.com/DU-Bii/module-3-Stat-R/raw/master/stat-R_2021/data/pavkovic_2019/&quot;</span></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="co">## Define the local folder for this analysis (where the data will be downloaded and the results generated)</span></a>
<a class="sourceLine" id="cb2-5" data-line-number="5">pavkovic_folder &lt;-<span class="st"> &quot;~/m3-stat-R/pavkovic_analysis&quot;</span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"></a>
<a class="sourceLine" id="cb2-7" data-line-number="7"><span class="co">## Define a sub-folder for the data</span></a>
<a class="sourceLine" id="cb2-8" data-line-number="8">pavkovic_data_folder &lt;-<span class="st"> </span><span class="kw">file.path</span>(pavkovic_folder, <span class="st">&quot;data&quot;</span>)</a>
<a class="sourceLine" id="cb2-9" data-line-number="9"></a>
<a class="sourceLine" id="cb2-10" data-line-number="10"><span class="co">## Download and load the expression data table</span></a>
<a class="sourceLine" id="cb2-11" data-line-number="11"><span class="co">## Note: we use check.names=FALSE to avoid replacing hyphens by dots</span></a>
<a class="sourceLine" id="cb2-12" data-line-number="12"><span class="co">## in sample names, because we want to keep them as in the </span></a>
<a class="sourceLine" id="cb2-13" data-line-number="13"><span class="co">## original data files. </span></a>
<a class="sourceLine" id="cb2-14" data-line-number="14"><span class="kw">message</span>(<span class="st">&quot;Downloading FA transcriptome file</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="st">&quot;fa_raw_counts.tsv.gz&quot;</span>,</a>
<a class="sourceLine" id="cb2-15" data-line-number="15">  <span class="st">&quot;</span><span class="ch">\n\t</span><span class="st">from</span><span class="ch">\t</span><span class="st">&quot;</span>, pavkovic_url)</a>
<a class="sourceLine" id="cb2-16" data-line-number="16">fa_expr_file &lt;-<span class="st"> </span><span class="kw">download_only_once</span>(</a>
<a class="sourceLine" id="cb2-17" data-line-number="17">  <span class="dt">url_base =</span> pavkovic_url, </a>
<a class="sourceLine" id="cb2-18" data-line-number="18">  <span class="dt">file_name =</span> <span class="st">&quot;fa_raw_counts.tsv.gz&quot;</span>,</a>
<a class="sourceLine" id="cb2-19" data-line-number="19">  <span class="dt">local_folder =</span> pavkovic_data_folder)</a>
<a class="sourceLine" id="cb2-20" data-line-number="20"></a>
<a class="sourceLine" id="cb2-21" data-line-number="21"><span class="co">## Load the expresdsion table</span></a>
<a class="sourceLine" id="cb2-22" data-line-number="22"><span class="kw">message</span>(<span class="st">&quot;Loading FA transcriptome data from</span><span class="ch">\n\t</span><span class="st">&quot;</span>, fa_expr_file)</a>
<a class="sourceLine" id="cb2-23" data-line-number="23">fa_expr_raw &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="dt">file =</span> fa_expr_file, </a>
<a class="sourceLine" id="cb2-24" data-line-number="24">                       <span class="dt">header =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb2-25" data-line-number="25">                       <span class="dt">row.names =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb2-26" data-line-number="26"></a>
<a class="sourceLine" id="cb2-27" data-line-number="27"><span class="co">## Download the metadata file</span></a>
<a class="sourceLine" id="cb2-28" data-line-number="28"><span class="kw">message</span>(<span class="st">&quot;Downloading FA metadata file</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="st">&quot;fa_transcriptome_metadata.tsv&quot;</span>,</a>
<a class="sourceLine" id="cb2-29" data-line-number="29">  <span class="st">&quot;</span><span class="ch">\n\t</span><span class="st">from</span><span class="ch">\t</span><span class="st">&quot;</span>, pavkovic_url)</a>
<a class="sourceLine" id="cb2-30" data-line-number="30">fa_meta_file &lt;-<span class="st"> </span><span class="kw">download_only_once</span>(</a>
<a class="sourceLine" id="cb2-31" data-line-number="31">  <span class="dt">url_base =</span> pavkovic_url, </a>
<a class="sourceLine" id="cb2-32" data-line-number="32">  <span class="dt">file_name =</span> <span class="st">&quot;fa_transcriptome_metadata.tsv&quot;</span>,</a>
<a class="sourceLine" id="cb2-33" data-line-number="33">  <span class="dt">local_folder =</span> pavkovic_data_folder)</a>
<a class="sourceLine" id="cb2-34" data-line-number="34"></a>
<a class="sourceLine" id="cb2-35" data-line-number="35"><span class="co">## Load the metadata</span></a>
<a class="sourceLine" id="cb2-36" data-line-number="36"><span class="kw">message</span>(<span class="st">&quot;Loading FA metadata from</span><span class="ch">\n\t</span><span class="st">&quot;</span>, fa_meta_file)</a>
<a class="sourceLine" id="cb2-37" data-line-number="37">fa_meta &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="dt">file =</span> fa_meta_file, </a>
<a class="sourceLine" id="cb2-38" data-line-number="38">                       <span class="dt">header =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb2-39" data-line-number="39">                       <span class="dt">row.names =</span> <span class="dv">1</span>)</a></code></pre></div>
<p>Nous regardons la structure de chaque dataframe.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">str</span>(fa_expr_raw)</a></code></pre></div>
<pre><code>&#39;data.frame&#39;:   46679 obs. of  18 variables:
 $ day1_1  : num  2278.8 0 36.3 13.2 0 ...
 $ day1_2  : num  1786.5 0 22.15 7.15 27.9 ...
 $ day1_3  : num  2368.62 0 39.48 1.12 6.9 ...
 $ day14_1 : num  627.758 0 14.471 0.867 5.692 ...
 $ day14_2 : num  559.2 0 10.2 0 1.9 ...
 $ day14_3 : num  611.434 0 31.691 0 0.655 ...
 $ day2_1  : num  2145.22 0 300.56 1.71 57.38 ...
 $ day2_2  : num  262.45 0 4.77 0 0 ...
 $ day2_3  : num  745.84 0 123.9 5.26 38.9 ...
 $ day3_1  : num  987.185 0 51.856 0.802 8.931 ...
 $ day3_2  : num  1077.65 0 8.43 0 6.97 ...
 $ day3_3  : num  1335.1 0 69.9 0 0 ...
 $ day7_1  : num  1096.08 0 6.67 0 7.94 ...
 $ day7_2  : num  1035.846 0 6.955 0.849 101.648 ...
 $ day7_3  : num  1090.04 0 42.58 1.71 0.65 ...
 $ normal_1: num  483.23 0 7.35 0.86 32.06 ...
 $ normal_2: num  1842.1 0 11.2 0 10.4 ...
 $ normal_3: num  475.7 0 1.03 0 0 ...</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">str</span>(fa_meta)</a></code></pre></div>
<pre><code>&#39;data.frame&#39;:   18 obs. of  5 variables:
 $ dataType    : chr  &quot;transcriptome&quot; &quot;transcriptome&quot; &quot;transcriptome&quot; &quot;transcriptome&quot; ...
 $ sampleName  : chr  &quot;day14_1&quot; &quot;day14_2&quot; &quot;day14_3&quot; &quot;day1_1&quot; ...
 $ condition   : chr  &quot;day14&quot; &quot;day14&quot; &quot;day14&quot; &quot;day1&quot; ...
 $ sampleNumber: int  1 2 3 1 2 3 1 2 3 1 ...
 $ color       : chr  &quot;#FF4400&quot; &quot;#FF4400&quot; &quot;#FF4400&quot; &quot;#BBD7FF&quot; ...</code></pre>
<p>Les deux fichiers ne donnent pas les observations de l’échantillon dans le même ordre:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">fa_meta<span class="op">$</span>sampleName <span class="op">==</span><span class="st"> </span><span class="kw">names</span>(fa_expr_raw)</a></code></pre></div>
<pre><code> [1] FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE</code></pre>
<p>Nous les réorganisons les échantillons dans l’ordre de l’expérience: condition normale, puis day 1 à 14 avec les 3 réplicats.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">sample_order &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">paste</span>(<span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&quot;normal&quot;</span>, <span class="st">&quot;day1&quot;</span>, <span class="st">&quot;day2&quot;</span>, <span class="st">&quot;day3&quot;</span>, <span class="st">&quot;day7&quot;</span>, <span class="st">&quot;day14&quot;</span>), <span class="dt">each =</span> <span class="dv">3</span>),</a>
<a class="sourceLine" id="cb9-2" data-line-number="2">                        <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="dt">sep =</span> <span class="st">&quot;_&quot;</span>))</a>
<a class="sourceLine" id="cb9-3" data-line-number="3"></a>
<a class="sourceLine" id="cb9-4" data-line-number="4">fa_expr_raw &lt;-<span class="st"> </span>fa_expr_raw[,sample_order]</a>
<a class="sourceLine" id="cb9-5" data-line-number="5">fa_meta &lt;-<span class="st"> </span>fa_meta[<span class="kw">match</span>(sample_order, fa_meta<span class="op">$</span>sampleName),]</a>
<a class="sourceLine" id="cb9-6" data-line-number="6"></a>
<a class="sourceLine" id="cb9-7" data-line-number="7"><span class="co"># View(fa_meta)</span></a>
<a class="sourceLine" id="cb9-8" data-line-number="8"><span class="kw">kable</span>(fa_meta, <span class="dt">caption =</span> <span class="st">&quot;Metdata for Pavkovoc FA transcriptome&quot;</span>)</a></code></pre></div>
<table>
<caption>Metdata for Pavkovoc FA transcriptome</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">dataType</th>
<th align="left">sampleName</th>
<th align="left">condition</th>
<th align="right">sampleNumber</th>
<th align="left">color</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">16</td>
<td align="left">transcriptome</td>
<td align="left">normal_1</td>
<td align="left">normal</td>
<td align="right">1</td>
<td align="left">#BBFFBB</td>
</tr>
<tr class="even">
<td align="left">17</td>
<td align="left">transcriptome</td>
<td align="left">normal_2</td>
<td align="left">normal</td>
<td align="right">2</td>
<td align="left">#BBFFBB</td>
</tr>
<tr class="odd">
<td align="left">18</td>
<td align="left">transcriptome</td>
<td align="left">normal_3</td>
<td align="left">normal</td>
<td align="right">3</td>
<td align="left">#BBFFBB</td>
</tr>
<tr class="even">
<td align="left">4</td>
<td align="left">transcriptome</td>
<td align="left">day1_1</td>
<td align="left">day1</td>
<td align="right">1</td>
<td align="left">#BBD7FF</td>
</tr>
<tr class="odd">
<td align="left">5</td>
<td align="left">transcriptome</td>
<td align="left">day1_2</td>
<td align="left">day1</td>
<td align="right">2</td>
<td align="left">#BBD7FF</td>
</tr>
<tr class="even">
<td align="left">6</td>
<td align="left">transcriptome</td>
<td align="left">day1_3</td>
<td align="left">day1</td>
<td align="right">3</td>
<td align="left">#BBD7FF</td>
</tr>
<tr class="odd">
<td align="left">7</td>
<td align="left">transcriptome</td>
<td align="left">day2_1</td>
<td align="left">day2</td>
<td align="right">1</td>
<td align="left">#F0BBFF</td>
</tr>
<tr class="even">
<td align="left">8</td>
<td align="left">transcriptome</td>
<td align="left">day2_2</td>
<td align="left">day2</td>
<td align="right">2</td>
<td align="left">#F0BBFF</td>
</tr>
<tr class="odd">
<td align="left">9</td>
<td align="left">transcriptome</td>
<td align="left">day2_3</td>
<td align="left">day2</td>
<td align="right">3</td>
<td align="left">#F0BBFF</td>
</tr>
<tr class="even">
<td align="left">10</td>
<td align="left">transcriptome</td>
<td align="left">day3_1</td>
<td align="left">day3</td>
<td align="right">1</td>
<td align="left">#FFFFDD</td>
</tr>
<tr class="odd">
<td align="left">11</td>
<td align="left">transcriptome</td>
<td align="left">day3_2</td>
<td align="left">day3</td>
<td align="right">2</td>
<td align="left">#FFFFDD</td>
</tr>
<tr class="even">
<td align="left">12</td>
<td align="left">transcriptome</td>
<td align="left">day3_3</td>
<td align="left">day3</td>
<td align="right">3</td>
<td align="left">#FFFFDD</td>
</tr>
<tr class="odd">
<td align="left">13</td>
<td align="left">transcriptome</td>
<td align="left">day7_1</td>
<td align="left">day7</td>
<td align="right">1</td>
<td align="left">#FFDD88</td>
</tr>
<tr class="even">
<td align="left">14</td>
<td align="left">transcriptome</td>
<td align="left">day7_2</td>
<td align="left">day7</td>
<td align="right">2</td>
<td align="left">#FFDD88</td>
</tr>
<tr class="odd">
<td align="left">15</td>
<td align="left">transcriptome</td>
<td align="left">day7_3</td>
<td align="left">day7</td>
<td align="right">3</td>
<td align="left">#FFDD88</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="left">transcriptome</td>
<td align="left">day14_1</td>
<td align="left">day14</td>
<td align="right">1</td>
<td align="left">#FF4400</td>
</tr>
<tr class="odd">
<td align="left">2</td>
<td align="left">transcriptome</td>
<td align="left">day14_2</td>
<td align="left">day14</td>
<td align="right">2</td>
<td align="left">#FF4400</td>
</tr>
<tr class="even">
<td align="left">3</td>
<td align="left">transcriptome</td>
<td align="left">day14_3</td>
<td align="left">day14</td>
<td align="right">3</td>
<td align="left">#FF4400</td>
</tr>
</tbody>
</table>
<p>=&gt; Ainsi, nous avons un jeu de données avec un échantillon de 18 observations et des données d’expression de 46679 gènes.</p>
</div>
<div id="statistiques-descriptives" class="section level3">
<h3>Statistiques descriptives</h3>
<p>Dans le tutorial sur les dataframes sur le jeu de données “uuo” (relisez le corrigé), nous vous avons demandé de créer un data.frame qui collecte les statistiques par gène et par échantillon. Nous vous demandons de réaliser une étude similaire sur les données “FA” avant et après normalisation inter-échantillons des données. Le code de la partie avant normalisation est donné.</p>
<div id="par-échantillon-avant-normalisation" class="section level4">
<h4>Par échantillon avant normalisation</h4>
<p>Nous créons un data.frame nommé <code>sample_stat_prenorm</code> qui comporte une ligne par échantillon et une colonne par statistique. Nous calculons les statistiques suivantes sur les valeurs log2 d’expression de chaque échantillon:</p>
<ul>
<li>moyenne</li>
<li>écart-type</li>
<li>intervalle inter-quartiles</li>
<li>premier quartile</li>
<li>médiane</li>
<li>troisième quartile</li>
<li>maximum</li>
<li>nombre de valeurs nulles</li>
</ul>
<p>Il est affiché avec la fonction <code>kable()</code>.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="kw">message</span>(<span class="st">&quot;Computing sample-wise statistics on raw counts&quot;</span>)</a>
<a class="sourceLine" id="cb10-2" data-line-number="2">sample_stat_prenorm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</a>
<a class="sourceLine" id="cb10-3" data-line-number="3">  <span class="dt">mean =</span> <span class="kw">apply</span>(fa_expr_raw, <span class="dv">2</span>, mean, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb10-4" data-line-number="4">  <span class="dt">sd =</span> <span class="kw">apply</span>(fa_expr_raw, <span class="dv">2</span>, sd, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb10-5" data-line-number="5">  <span class="dt">iqr =</span> <span class="kw">apply</span>(fa_expr_raw, <span class="dv">2</span>, IQR, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb10-6" data-line-number="6">  <span class="dt">Q1 =</span> <span class="kw">apply</span>(fa_expr_raw, <span class="dv">2</span>, quantile, <span class="dt">p =</span> <span class="fl">0.25</span>, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb10-7" data-line-number="7">  <span class="dt">median =</span> <span class="kw">apply</span>(fa_expr_raw, <span class="dv">2</span>, median, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb10-8" data-line-number="8">  <span class="dt">Q3 =</span> <span class="kw">apply</span>(fa_expr_raw, <span class="dv">2</span>, quantile, <span class="dt">p =</span> <span class="fl">0.75</span>, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb10-9" data-line-number="9">  <span class="dt">max =</span> <span class="kw">apply</span>(fa_expr_raw, <span class="dv">2</span>, max, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb10-10" data-line-number="10">  <span class="dt">null =</span> <span class="kw">apply</span>(fa_expr_raw <span class="op">==</span><span class="st"> </span><span class="dv">0</span>, <span class="dv">2</span>, sum, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb10-11" data-line-number="11">)</a>
<a class="sourceLine" id="cb10-12" data-line-number="12"></a>
<a class="sourceLine" id="cb10-13" data-line-number="13"><span class="kw">kable</span>(sample_stat_prenorm, <span class="dt">caption =</span> <span class="st">&quot;Sample-wise statistics before normalisation.&quot;</span>)</a></code></pre></div>
<table>
<caption>Sample-wise statistics before normalisation.</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">mean</th>
<th align="right">sd</th>
<th align="right">iqr</th>
<th align="right">Q1</th>
<th align="right">median</th>
<th align="right">Q3</th>
<th align="right">max</th>
<th align="right">null</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">normal_1</td>
<td align="right">576.2992</td>
<td align="right">57643.81</td>
<td align="right">46.38417</td>
<td align="right">0</td>
<td align="right">1.166694</td>
<td align="right">46.38417</td>
<td align="right">11777395</td>
<td align="right">21415</td>
</tr>
<tr class="even">
<td align="left">normal_2</td>
<td align="right">1172.5615</td>
<td align="right">111509.82</td>
<td align="right">103.13749</td>
<td align="right">0</td>
<td align="right">2.975840</td>
<td align="right">103.13749</td>
<td align="right">22658521</td>
<td align="right">20203</td>
</tr>
<tr class="odd">
<td align="left">normal_3</td>
<td align="right">531.8620</td>
<td align="right">57203.96</td>
<td align="right">41.80724</td>
<td align="right">0</td>
<td align="right">1.081680</td>
<td align="right">41.80724</td>
<td align="right">11636735</td>
<td align="right">21660</td>
</tr>
<tr class="even">
<td align="left">day1_1</td>
<td align="right">663.0580</td>
<td align="right">65219.36</td>
<td align="right">58.76393</td>
<td align="right">0</td>
<td align="right">1.105540</td>
<td align="right">58.76393</td>
<td align="right">13174979</td>
<td align="right">21700</td>
</tr>
<tr class="odd">
<td align="left">day1_2</td>
<td align="right">1224.5149</td>
<td align="right">123140.72</td>
<td align="right">111.71166</td>
<td align="right">0</td>
<td align="right">3.053561</td>
<td align="right">111.71166</td>
<td align="right">25118595</td>
<td align="right">20152</td>
</tr>
<tr class="even">
<td align="left">day1_3</td>
<td align="right">1368.6144</td>
<td align="right">133450.04</td>
<td align="right">122.38187</td>
<td align="right">0</td>
<td align="right">3.766666</td>
<td align="right">122.38187</td>
<td align="right">27255572</td>
<td align="right">19621</td>
</tr>
<tr class="odd">
<td align="left">day2_1</td>
<td align="right">1161.7628</td>
<td align="right">103598.87</td>
<td align="right">127.04744</td>
<td align="right">0</td>
<td align="right">3.501480</td>
<td align="right">127.04744</td>
<td align="right">21036079</td>
<td align="right">19978</td>
</tr>
<tr class="even">
<td align="left">day2_2</td>
<td align="right">223.8844</td>
<td align="right">20431.97</td>
<td align="right">14.54242</td>
<td align="right">0</td>
<td align="right">0.000000</td>
<td align="right">14.54242</td>
<td align="right">4037963</td>
<td align="right">24446</td>
</tr>
<tr class="odd">
<td align="left">day2_3</td>
<td align="right">628.4406</td>
<td align="right">57480.66</td>
<td align="right">63.36112</td>
<td align="right">0</td>
<td align="right">1.478198</td>
<td align="right">63.36112</td>
<td align="right">11662570</td>
<td align="right">21455</td>
</tr>
<tr class="even">
<td align="left">day3_1</td>
<td align="right">649.8389</td>
<td align="right">55210.88</td>
<td align="right">77.84371</td>
<td align="right">0</td>
<td align="right">2.076066</td>
<td align="right">77.84371</td>
<td align="right">11170233</td>
<td align="right">20772</td>
</tr>
<tr class="odd">
<td align="left">day3_2</td>
<td align="right">514.8649</td>
<td align="right">42768.34</td>
<td align="right">74.16479</td>
<td align="right">0</td>
<td align="right">1.941705</td>
<td align="right">74.16479</td>
<td align="right">8612320</td>
<td align="right">21142</td>
</tr>
<tr class="even">
<td align="left">day3_3</td>
<td align="right">854.0997</td>
<td align="right">80565.29</td>
<td align="right">104.95124</td>
<td align="right">0</td>
<td align="right">2.975772</td>
<td align="right">104.95124</td>
<td align="right">16484449</td>
<td align="right">20287</td>
</tr>
<tr class="odd">
<td align="left">day7_1</td>
<td align="right">576.2437</td>
<td align="right">53112.82</td>
<td align="right">90.51017</td>
<td align="right">0</td>
<td align="right">2.256164</td>
<td align="right">90.51017</td>
<td align="right">10690493</td>
<td align="right">20865</td>
</tr>
<tr class="even">
<td align="left">day7_2</td>
<td align="right">510.2741</td>
<td align="right">47373.18</td>
<td align="right">80.57478</td>
<td align="right">0</td>
<td align="right">2.537673</td>
<td align="right">80.57478</td>
<td align="right">9682527</td>
<td align="right">20467</td>
</tr>
<tr class="odd">
<td align="left">day7_3</td>
<td align="right">604.9701</td>
<td align="right">59072.56</td>
<td align="right">83.50019</td>
<td align="right">0</td>
<td align="right">2.874706</td>
<td align="right">83.50019</td>
<td align="right">12164986</td>
<td align="right">20045</td>
</tr>
<tr class="even">
<td align="left">day14_1</td>
<td align="right">540.2640</td>
<td align="right">54481.09</td>
<td align="right">65.01270</td>
<td align="right">0</td>
<td align="right">2.025707</td>
<td align="right">65.01270</td>
<td align="right">11178198</td>
<td align="right">20496</td>
</tr>
<tr class="odd">
<td align="left">day14_2</td>
<td align="right">681.9987</td>
<td align="right">74825.35</td>
<td align="right">69.05744</td>
<td align="right">0</td>
<td align="right">2.032894</td>
<td align="right">69.05744</td>
<td align="right">15430799</td>
<td align="right">20870</td>
</tr>
<tr class="even">
<td align="left">day14_3</td>
<td align="right">562.5697</td>
<td align="right">56026.62</td>
<td align="right">75.11681</td>
<td align="right">0</td>
<td align="right">2.247271</td>
<td align="right">75.11681</td>
<td align="right">11575718</td>
<td align="right">20396</td>
</tr>
</tbody>
</table>
</div>
<div id="par-gène-avant-normalisation" class="section level4">
<h4>Par gène avant normalisation</h4>
<p>Nous créons ci-dessous un data.frame nommé <code>gene_stat_prenorm</code> qui comporte une ligne par gène et une colonne par statistique. Nous calculons les statistiques suivantes sur les valeurs log2 de chaque gène.</p>
<ul>
<li>moyenne</li>
<li>médiane</li>
<li>écart-type</li>
<li>premier quartile</li>
<li>troisième quartile</li>
<li>maximum</li>
<li>nombre de valeurs nulles</li>
<li>intervalle inter-quartiles</li>
</ul>
<p>Ces résultats sont stockés dans un data.frame avec 1 ligne par échantillon et 1 colonne par statistique. Nous affichons les lignes 100 à 109 de ce tableau de statistiques avec la fonction <code>kable()</code>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="co">## Gene-wise statistics for the raw counts (will be used for normalisation)</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;Computing gene-wise statistics on raw counts&quot;</span>)</a>
<a class="sourceLine" id="cb11-3" data-line-number="3">gene_stat_prenorm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</a>
<a class="sourceLine" id="cb11-4" data-line-number="4">  <span class="dt">mean =</span> <span class="kw">apply</span>(fa_expr_raw, <span class="dv">1</span>, mean, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb11-5" data-line-number="5">  <span class="dt">sd =</span> <span class="kw">apply</span>(fa_expr_raw, <span class="dv">1</span>, sd, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb11-6" data-line-number="6">  <span class="dt">iqr =</span> <span class="kw">apply</span>(fa_expr_raw, <span class="dv">1</span>, IQR, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb11-7" data-line-number="7">  <span class="dt">Q1 =</span> <span class="kw">apply</span>(fa_expr_raw, <span class="dv">1</span>, quantile, <span class="dt">p =</span> <span class="fl">0.25</span>, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb11-8" data-line-number="8">  <span class="dt">median =</span> <span class="kw">apply</span>(fa_expr_raw, <span class="dv">1</span>, median, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb11-9" data-line-number="9">  <span class="dt">Q3 =</span> <span class="kw">apply</span>(fa_expr_raw, <span class="dv">1</span>, quantile, <span class="dt">p =</span> <span class="fl">0.75</span>, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb11-10" data-line-number="10">  <span class="dt">max =</span> <span class="kw">apply</span>(fa_expr_raw, <span class="dv">1</span>, max, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb11-11" data-line-number="11">  <span class="dt">null =</span> <span class="kw">apply</span>(fa_expr_raw <span class="op">==</span><span class="st"> </span><span class="dv">0</span>, <span class="dv">1</span>, sum, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb11-12" data-line-number="12">)</a>
<a class="sourceLine" id="cb11-13" data-line-number="13"></a>
<a class="sourceLine" id="cb11-14" data-line-number="14"><span class="kw">kable</span>(gene_stat_prenorm[<span class="dv">100</span><span class="op">:</span><span class="dv">109</span>, ], <span class="dt">caption =</span> <span class="st">&quot;Gene-wise statistics before normalisation&quot;</span>)</a></code></pre></div>
<table style="width:100%;">
<caption>Gene-wise statistics before normalisation</caption>
<colgroup>
<col width="17%" />
<col width="11%" />
<col width="11%" />
<col width="11%" />
<col width="11%" />
<col width="11%" />
<col width="10%" />
<col width="10%" />
<col width="4%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">mean</th>
<th align="right">sd</th>
<th align="right">iqr</th>
<th align="right">Q1</th>
<th align="right">median</th>
<th align="right">Q3</th>
<th align="right">max</th>
<th align="right">null</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ENSMUSG00000000567</td>
<td align="right">320.336744</td>
<td align="right">286.399084</td>
<td align="right">364.362298</td>
<td align="right">91.046357</td>
<td align="right">261.659973</td>
<td align="right">455.40866</td>
<td align="right">1052.68945</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000568</td>
<td align="right">857.348825</td>
<td align="right">694.432189</td>
<td align="right">282.978949</td>
<td align="right">575.571199</td>
<td align="right">711.096105</td>
<td align="right">858.55015</td>
<td align="right">3406.45030</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000579</td>
<td align="right">422.075702</td>
<td align="right">162.290415</td>
<td align="right">157.732606</td>
<td align="right">340.417017</td>
<td align="right">447.572742</td>
<td align="right">498.14962</td>
<td align="right">728.51737</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000581</td>
<td align="right">498.978227</td>
<td align="right">248.612029</td>
<td align="right">259.324338</td>
<td align="right">320.599085</td>
<td align="right">471.269224</td>
<td align="right">579.92342</td>
<td align="right">942.51104</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000594</td>
<td align="right">2113.558334</td>
<td align="right">1237.521694</td>
<td align="right">937.571105</td>
<td align="right">1400.226182</td>
<td align="right">1868.718762</td>
<td align="right">2337.79729</td>
<td align="right">6408.96044</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000600</td>
<td align="right">2299.978163</td>
<td align="right">911.222891</td>
<td align="right">1480.192582</td>
<td align="right">1719.874980</td>
<td align="right">2093.572929</td>
<td align="right">3200.06756</td>
<td align="right">3674.18811</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000605</td>
<td align="right">325.552118</td>
<td align="right">169.456624</td>
<td align="right">195.879561</td>
<td align="right">221.735904</td>
<td align="right">282.608205</td>
<td align="right">417.61546</td>
<td align="right">791.31256</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000606</td>
<td align="right">9.929258</td>
<td align="right">32.151350</td>
<td align="right">2.736779</td>
<td align="right">1.178311</td>
<td align="right">1.934744</td>
<td align="right">3.91509</td>
<td align="right">138.48677</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000617</td>
<td align="right">18.804116</td>
<td align="right">9.470819</td>
<td align="right">13.654701</td>
<td align="right">12.188276</td>
<td align="right">18.150391</td>
<td align="right">25.84298</td>
<td align="right">36.00235</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000627</td>
<td align="right">15.623306</td>
<td align="right">14.341852</td>
<td align="right">12.308127</td>
<td align="right">4.780913</td>
<td align="right">10.064895</td>
<td align="right">17.08904</td>
<td align="right">46.29752</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="filtrage-et-normalisation-des-données" class="section level2">
<h2>2. Filtrage et normalisation des données</h2>
<p><strong><em>Vous n’avez rien à coder ici. Le code est fourni.</em></strong></p>
<p>Il existe plusieurs façons de normaliser les données de transcriptome vues dans les modules 4 et 5 (cf. total counts, quantiles, TMM, RLE, limma voom,…), mais nous avons choisi ici une solution simple tout en étant robuste pour normaliser les données en standardisant le 3ème quantile.</p>
<p>La méthode choisie ici consiste à :</p>
<ol style="list-style-type: decimal">
<li><p><strong>Ecarter les gènes “non-détectés”</strong>, c’est-à-dire ceux ayant des valeurs nulles dans au moins 90% des échantillons.</p></li>
<li><p><strong>Ecarter les gènes à peine exprimés</strong>, c’est-à-dire ceux ayant une valeur moyenne &lt; 10 (arbitrairement).</p></li>
<li><p><strong>Standardiser les échantillons </strong>sur le 3ème quartile des gènes restants: on divise les comptages bruts par le 3ème quartile de l’échantillon et on multiplie par le 3ème quartile de l’ensemble des échantillons.</p></li>
<li><p><strong>Normaliser les comptages</strong> (au sens propre, c’est-à-dire rapprocher leur distribution de la distribution gaussienne) par une transformation logarithmique (log2).</p></li>
</ol>
<p>Nous fournissons ci-dessous le code.</p>
<div id="filtrage-élimination-des-gènes-non-détectés-ou-à-peine-exprimés" class="section level3">
<h3>Filtrage : élimination des gènes non détectés ou à peine exprimés</h3>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="co">## Data filtering: genes having at least 90% null values</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;Filtering undetected genes&quot;</span>)</a>
<a class="sourceLine" id="cb12-3" data-line-number="3">undetected_genes &lt;-<span class="st"> </span>gene_stat_prenorm<span class="op">$</span>null <span class="op">&gt;=</span><span class="st"> </span><span class="kw">ncol</span>(fa_expr_raw) <span class="op">*</span><span class="st"> </span><span class="fl">0.9</span></a>
<a class="sourceLine" id="cb12-4" data-line-number="4"><span class="kw">print</span>(<span class="kw">paste0</span>(<span class="st">&quot;Undetected genes (null in &gt;= 90% samples): &quot;</span>, <span class="kw">sum</span>(undetected_genes)))</a></code></pre></div>
<pre><code>[1] &quot;Undetected genes (null in &gt;= 90% samples): 14380&quot;</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="co">## Data filtering: genes having a mean expression &lt; 10</span></a>
<a class="sourceLine" id="cb14-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;Filtering barely expressed genes&quot;</span>)</a>
<a class="sourceLine" id="cb14-3" data-line-number="3">barely_expressed_genes &lt;-<span class="st"> </span>gene_stat_prenorm<span class="op">$</span>mean <span class="op">&lt;</span><span class="st"> </span><span class="dv">10</span></a>
<a class="sourceLine" id="cb14-4" data-line-number="4"><span class="kw">print</span>(<span class="kw">paste0</span>(<span class="st">&quot;Barely expressed genes (mean &lt; 10): &quot;</span>, <span class="kw">sum</span>(barely_expressed_genes)))</a></code></pre></div>
<pre><code>[1] &quot;Barely expressed genes (mean &lt; 10): 26286&quot;</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="co">## Apply filtering on both criteria</span></a>
<a class="sourceLine" id="cb16-2" data-line-number="2">discarded_genes &lt;-<span class="st"> </span>undetected_genes <span class="op">|</span><span class="st"> </span>barely_expressed_genes</a>
<a class="sourceLine" id="cb16-3" data-line-number="3"><span class="kw">print</span>(<span class="kw">paste0</span>(<span class="st">&quot;Discarded genes: &quot;</span>, <span class="kw">sum</span>(discarded_genes)))</a></code></pre></div>
<pre><code>[1] &quot;Discarded genes: 26288&quot;</code></pre>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">kept_genes &lt;-<span class="st"> </span><span class="op">!</span>discarded_genes</a>
<a class="sourceLine" id="cb18-2" data-line-number="2"><span class="kw">print</span>(<span class="kw">paste0</span>(<span class="st">&quot;Kept genes: &quot;</span>, <span class="kw">sum</span>(kept_genes)))</a></code></pre></div>
<pre><code>[1] &quot;Kept genes: 20391&quot;</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="co">## Genes after filtering</span></a>
<a class="sourceLine" id="cb20-2" data-line-number="2">fa_expr_filtered &lt;-<span class="st"> </span>fa_expr_raw[kept_genes, ]</a></code></pre></div>
</div>
<div id="standardisation-entre-échantillons" class="section level3">
<h3>Standardisation entre échantillons</h3>
<p>Nous appliquons ici une méthode simple mais efficace de standardisation</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="co">#### Inter-sample standardisation on the Q3 of raw counts ####</span></a>
<a class="sourceLine" id="cb21-2" data-line-number="2">total_q3 &lt;-<span class="st"> </span><span class="kw">quantile</span>(<span class="kw">unlist</span>(fa_expr_filtered), <span class="dt">probs =</span> <span class="fl">0.75</span>)</a>
<a class="sourceLine" id="cb21-3" data-line-number="3">sample_stat_prenorm<span class="op">$</span>Q3_filterd &lt;-<span class="st"> </span><span class="kw">apply</span>(fa_expr_filtered, <span class="dv">2</span>, quantile, <span class="dt">probs =</span> <span class="fl">0.75</span>)</a>
<a class="sourceLine" id="cb21-4" data-line-number="4">sample_stat_prenorm<span class="op">$</span>scale_factor &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">/</span><span class="st"> </span>sample_stat_prenorm<span class="op">$</span>Q3_filterd <span class="op">*</span><span class="st"> </span>total_q3</a>
<a class="sourceLine" id="cb21-5" data-line-number="5"></a>
<a class="sourceLine" id="cb21-6" data-line-number="6"><span class="co">## Apply standardisation</span></a>
<a class="sourceLine" id="cb21-7" data-line-number="7">fa_expr_standard &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">t</span>(fa_expr_filtered) <span class="op">*</span><span class="st"> </span><span class="kw">unlist</span>(sample_stat_prenorm<span class="op">$</span>scale_factor))</a>
<a class="sourceLine" id="cb21-8" data-line-number="8"><span class="co">## Check 3rd quantile after standardisation</span></a>
<a class="sourceLine" id="cb21-9" data-line-number="9"><span class="kw">kable</span>(<span class="kw">apply</span>(fa_expr_standard, <span class="dv">2</span>, quantile, <span class="dt">probs =</span> <span class="fl">0.75</span>), </a>
<a class="sourceLine" id="cb21-10" data-line-number="10">      <span class="dt">col.names =</span> <span class="st">&quot;Q3_standardised&quot;</span>, </a>
<a class="sourceLine" id="cb21-11" data-line-number="11">      <span class="dt">caption =</span> <span class="st">&quot;Third quartile of the filtered counts after inter-sample standardisation of the third quartiles. &quot;</span>)</a></code></pre></div>
<table>
<caption>Third quartile of the filtered counts after inter-sample standardisation of the third quartiles.</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">Q3_standardised</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">normal_1</td>
<td align="right">364.9332</td>
</tr>
<tr class="even">
<td align="left">normal_2</td>
<td align="right">364.9332</td>
</tr>
<tr class="odd">
<td align="left">normal_3</td>
<td align="right">364.9332</td>
</tr>
<tr class="even">
<td align="left">day1_1</td>
<td align="right">364.9332</td>
</tr>
<tr class="odd">
<td align="left">day1_2</td>
<td align="right">364.9332</td>
</tr>
<tr class="even">
<td align="left">day1_3</td>
<td align="right">364.9332</td>
</tr>
<tr class="odd">
<td align="left">day2_1</td>
<td align="right">364.9332</td>
</tr>
<tr class="even">
<td align="left">day2_2</td>
<td align="right">364.9332</td>
</tr>
<tr class="odd">
<td align="left">day2_3</td>
<td align="right">364.9332</td>
</tr>
<tr class="even">
<td align="left">day3_1</td>
<td align="right">364.9332</td>
</tr>
<tr class="odd">
<td align="left">day3_2</td>
<td align="right">364.9332</td>
</tr>
<tr class="even">
<td align="left">day3_3</td>
<td align="right">364.9332</td>
</tr>
<tr class="odd">
<td align="left">day7_1</td>
<td align="right">364.9332</td>
</tr>
<tr class="even">
<td align="left">day7_2</td>
<td align="right">364.9332</td>
</tr>
<tr class="odd">
<td align="left">day7_3</td>
<td align="right">364.9332</td>
</tr>
<tr class="even">
<td align="left">day14_1</td>
<td align="right">364.9332</td>
</tr>
<tr class="odd">
<td align="left">day14_2</td>
<td align="right">364.9332</td>
</tr>
<tr class="even">
<td align="left">day14_3</td>
<td align="right">364.9332</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="co"># boxplot(fa_expr_standard, horizontal = TRUE)</span></a></code></pre></div>
</div>
<div id="transformation-log2" class="section level3">
<h3>Transformation log2</h3>
<p>Nous appliquons une transformation en log2 des données brutes, après avoir ajouté un epsilon <span class="math inline">\(\epsilon = 1\)</span> (les valeurs nulles seront donc représentées par un log2(counts) valant <span class="math inline">\(0\)</span>. Nous stockons le résultat dans un data.frame nommé <code>fa_expr_log2</code>.</p>
<p>Nous affichons un fragment des tableaux <code>fa_expr_raw</code> et <code>fa_expr_log2</code> en sélectionnant les lignes 100 à 109 et les colonnes 5 à 10, afin de nous assurer que la transformation en log2 a bien fonctionné.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="co">## Log2 transformation of the transcriptome data</span></a>
<a class="sourceLine" id="cb23-2" data-line-number="2">epsilon &lt;-<span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb23-3" data-line-number="3">fa_expr_log2 &lt;-<span class="st"> </span><span class="kw">log2</span>(fa_expr_standard <span class="op">+</span><span class="st"> </span>epsilon)</a>
<a class="sourceLine" id="cb23-4" data-line-number="4"><span class="co"># dim(fa_expr_log2)</span></a>
<a class="sourceLine" id="cb23-5" data-line-number="5"><span class="co"># View(head(fa_expr_log2))</span></a>
<a class="sourceLine" id="cb23-6" data-line-number="6"></a>
<a class="sourceLine" id="cb23-7" data-line-number="7"><span class="co">## Display of a fragment of the data before and after log2 transformation</span></a>
<a class="sourceLine" id="cb23-8" data-line-number="8"><span class="kw">kable</span>(fa_expr_raw[<span class="dv">100</span><span class="op">:</span><span class="dv">109</span>, <span class="dv">5</span><span class="op">:</span><span class="dv">10</span>], <span class="dt">caption =</span> <span class="st">&quot;Fragment des données transcriptomiques brutes&quot;</span>)</a></code></pre></div>
<table style="width:100%;">
<caption>Fragment des données transcriptomiques brutes</caption>
<colgroup>
<col width="21%" />
<col width="13%" />
<col width="13%" />
<col width="13%" />
<col width="13%" />
<col width="12%" />
<col width="12%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">day1_2</th>
<th align="right">day1_3</th>
<th align="right">day2_1</th>
<th align="right">day2_2</th>
<th align="right">day2_3</th>
<th align="right">day3_1</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ENSMUSG00000000567</td>
<td align="right">599.648075</td>
<td align="right">304.093177</td>
<td align="right">1052.689447</td>
<td align="right">106.995584</td>
<td align="right">347.13842</td>
<td align="right">479.59911</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000568</td>
<td align="right">1008.026306</td>
<td align="right">1349.126716</td>
<td align="right">818.257157</td>
<td align="right">116.136417</td>
<td align="right">3406.45030</td>
<td align="right">766.09722</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000579</td>
<td align="right">599.832586</td>
<td align="right">500.735597</td>
<td align="right">473.399472</td>
<td align="right">36.258005</td>
<td align="right">410.57787</td>
<td align="right">347.05054</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000581</td>
<td align="right">942.511039</td>
<td align="right">744.646735</td>
<td align="right">546.260344</td>
<td align="right">87.788535</td>
<td align="right">319.12679</td>
<td align="right">461.45732</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000594</td>
<td align="right">3063.845705</td>
<td align="right">2743.404374</td>
<td align="right">2283.051270</td>
<td align="right">1115.588250</td>
<td align="right">1491.61384</td>
<td align="right">1576.68451</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000600</td>
<td align="right">3341.854530</td>
<td align="right">3561.805180</td>
<td align="right">3674.188108</td>
<td align="right">589.840068</td>
<td align="right">1399.10751</td>
<td align="right">3446.01856</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000605</td>
<td align="right">791.312561</td>
<td align="right">558.710943</td>
<td align="right">489.579657</td>
<td align="right">77.008213</td>
<td align="right">256.00200</td>
<td align="right">282.80077</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000606</td>
<td align="right">4.785252</td>
<td align="right">6.566374</td>
<td align="right">3.970399</td>
<td align="right">0.000000</td>
<td align="right">138.48677</td>
<td align="right">0.00000</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000617</td>
<td align="right">15.839486</td>
<td align="right">29.138902</td>
<td align="right">30.228506</td>
<td align="right">6.670500</td>
<td align="right">18.27493</td>
<td align="right">13.41152</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000627</td>
<td align="right">36.673777</td>
<td align="right">35.967747</td>
<td align="right">46.297517</td>
<td align="right">2.878275</td>
<td align="right">17.03638</td>
<td align="right">15.45854</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1"><span class="kw">kable</span>(fa_expr_log2[<span class="dv">100</span><span class="op">:</span><span class="dv">109</span>, <span class="dv">5</span><span class="op">:</span><span class="dv">10</span>], <span class="dt">caption =</span> <span class="st">&quot;Fragment des données transcriptomiques après transformation log2&quot;</span>)</a></code></pre></div>
<table>
<caption>Fragment des données transcriptomiques après transformation log2</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">day1_2</th>
<th align="right">day1_3</th>
<th align="right">day2_1</th>
<th align="right">day2_2</th>
<th align="right">day2_3</th>
<th align="right">day3_1</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ENSMUSG00000000711</td>
<td align="right">9.709367</td>
<td align="right">9.861557</td>
<td align="right">9.287726</td>
<td align="right">12.335149</td>
<td align="right">9.755002</td>
<td align="right">10.013232</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000724</td>
<td align="right">5.904466</td>
<td align="right">6.218882</td>
<td align="right">4.464461</td>
<td align="right">5.782181</td>
<td align="right">5.701477</td>
<td align="right">5.803482</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000731</td>
<td align="right">5.203571</td>
<td align="right">5.404039</td>
<td align="right">4.963615</td>
<td align="right">5.970908</td>
<td align="right">4.299021</td>
<td align="right">4.505637</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000732</td>
<td align="right">5.398925</td>
<td align="right">8.209062</td>
<td align="right">2.126857</td>
<td align="right">3.751910</td>
<td align="right">6.948597</td>
<td align="right">4.970987</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000738</td>
<td align="right">8.520431</td>
<td align="right">8.430289</td>
<td align="right">8.263532</td>
<td align="right">7.909466</td>
<td align="right">8.250468</td>
<td align="right">7.455809</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000740</td>
<td align="right">11.228789</td>
<td align="right">11.320488</td>
<td align="right">11.117466</td>
<td align="right">10.630164</td>
<td align="right">11.595716</td>
<td align="right">10.156658</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000743</td>
<td align="right">7.882486</td>
<td align="right">7.268069</td>
<td align="right">8.646853</td>
<td align="right">10.469226</td>
<td align="right">6.135917</td>
<td align="right">7.385994</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000751</td>
<td align="right">7.302103</td>
<td align="right">8.404643</td>
<td align="right">7.851161</td>
<td align="right">6.677037</td>
<td align="right">8.189071</td>
<td align="right">7.123055</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000753</td>
<td align="right">0.000000</td>
<td align="right">2.885227</td>
<td align="right">2.310527</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">4.605696</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000759</td>
<td align="right">8.296064</td>
<td align="right">8.620921</td>
<td align="right">7.830522</td>
<td align="right">9.377892</td>
<td align="right">9.333861</td>
<td align="right">10.038888</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="statistiques-descriptives-sur-les-données-normalisées" class="section level2">
<h2>3. Statistiques descriptives sur les données normalisées</h2>
<p><strong><em>A vous de jouer!</em></strong></p>
<div id="statistiques-par-gène-après-normalisation" class="section level3">
<h3>Statistiques par gène après normalisation</h3>
<p>Générez un data.frame nommée <code>gene_stat_norm</code> avec une ligne par gène à partir du tableau de données normalisées, avec les statistiques suivantes (une statistique par colonne):</p>
<ul>
<li>moyenne</li>
<li>variance</li>
<li>écart-type</li>
<li>coefficient de variation (écart-type divisé par la moyenne)</li>
<li>intervalle inter-quartiles</li>
<li>minimum</li>
<li>médiane</li>
<li>maximum</li>
</ul>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1"><span class="co">## Gene-wise statistics after normalisation</span></a>
<a class="sourceLine" id="cb25-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;Computing gene-wise statistics on log2-transformed and normalised counts&quot;</span>)</a>
<a class="sourceLine" id="cb25-3" data-line-number="3">gene_stat_norm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</a>
<a class="sourceLine" id="cb25-4" data-line-number="4">  <span class="dt">mean =</span> <span class="kw">apply</span>(fa_expr_log2, <span class="dv">1</span>, mean, <span class="dt">na.rm =</span> F),</a>
<a class="sourceLine" id="cb25-5" data-line-number="5">  <span class="dt">var =</span> <span class="kw">apply</span>(fa_expr_log2, <span class="dv">1</span>, var, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb25-6" data-line-number="6">  <span class="dt">sd =</span> <span class="kw">apply</span>(fa_expr_log2, <span class="dv">1</span>, sd, <span class="dt">na.rm =</span> F),</a>
<a class="sourceLine" id="cb25-7" data-line-number="7">  <span class="dt">CV =</span> <span class="ot">NA</span>,</a>
<a class="sourceLine" id="cb25-8" data-line-number="8">  <span class="dt">min =</span> <span class="kw">apply</span>(fa_expr_log2, <span class="dv">1</span>, min, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb25-9" data-line-number="9">  <span class="dt">Q1 =</span> <span class="kw">apply</span>(fa_expr_log2, <span class="dv">1</span>, quantile, <span class="dt">p =</span> <span class="fl">0.25</span>, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb25-10" data-line-number="10">  <span class="dt">median =</span> <span class="kw">apply</span>(fa_expr_log2, <span class="dv">1</span>, median, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb25-11" data-line-number="11">  <span class="dt">Q3 =</span> <span class="kw">apply</span>(fa_expr_log2, <span class="dv">1</span>, quantile, <span class="dt">p =</span> <span class="fl">0.75</span>, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb25-12" data-line-number="12">  <span class="dt">max =</span> <span class="kw">apply</span>(fa_expr_log2, <span class="dv">1</span>, max, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb25-13" data-line-number="13">  <span class="dt">null =</span> <span class="kw">apply</span>(fa_expr_log2 <span class="op">==</span><span class="st"> </span><span class="dv">0</span>, <span class="dv">1</span>, sum, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb25-14" data-line-number="14">  )</a>
<a class="sourceLine" id="cb25-15" data-line-number="15">gene_stat_norm<span class="op">$</span>CV &lt;-<span class="st"> </span>gene_stat_norm<span class="op">$</span>sd <span class="op">/</span><span class="st"> </span>gene_stat_norm<span class="op">$</span>mean</a></code></pre></div>
</div>
<div id="annotation-des-gènes" class="section level3">
<h3>Annotation des gènes</h3>
<p>Chaque gène étant donné par son identifiant dans la base de données ENSEMBL vous utiliserez le <code>paquet biomaRt de bioconductor</code> pour ajouter des annotations : symbole, chromosome, coordonnées génomiques, brin. Suivez pas à pas la méthode proposée (<strong><em>certaines étapes peuvent prendre quelques minutes</em></strong>):</p>
<ul>
<li><p>chargez le paquet biomaRt, voire installer-le uniquement si nécessaire. Indiquez le code à l’emplacement adéquat dans de .Rmd.</p></li>
<li><p>sélectionnez la base de données ENSEMBL avec la fonction <code>useMart()</code>. Attention à choisir le bon génome avec l’agument <code>dataset</code>: <code>mmusculus_gene_ensembl</code></p></li>
<li><p>avec la fonction <code>getBM()</code> récupérez de la base de données ENSEMBL les champs demandés (<strong><em>pour symbole utilisez external_gene_name</em></strong>) en appliquant “ensembl_geneid” pour l’agument <code>filter</code> et en indiquant pour l’argument <code>values</code> le vecteur des identifiants des gènes présents dans le dataframe <code>gene_stat_norm</code>. Vous obtenez un dataframe.</p></li>
</ul>
<p>A présent, ajoutez au dataframe <code>gene_stat_norm</code> en 1ères colonnes les annotations retrouvées grâce à biomaRt. Attention, certains gènes ne sont pas retrouvés dans la version d’ENSEMBL sur biomaRt donc laissez des NA comme données manquantes dans ce cas. Nous vous recommandons d’utiliser la function <code>merge()</code> de R base ou bien <code>left_join()</code> de <code>dplyr</code> pour fusionner les deux dataframes en un seul.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="co">### Gene annotations ####</span></a>
<a class="sourceLine" id="cb26-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;Getting gene annotations&quot;</span>)</a>
<a class="sourceLine" id="cb26-3" data-line-number="3"></a>
<a class="sourceLine" id="cb26-4" data-line-number="4"><span class="co">## Open a connection to Ensembl MART</span></a>
<a class="sourceLine" id="cb26-5" data-line-number="5"><span class="kw">message</span>(<span class="st">&quot;Opening connection to Ensembl MART&quot;</span>)</a>
<a class="sourceLine" id="cb26-6" data-line-number="6">ensembl &lt;-<span class="st"> </span><span class="kw">useMart</span>(<span class="st">&quot;ENSEMBL_MART_ENSEMBL&quot;</span>, </a>
<a class="sourceLine" id="cb26-7" data-line-number="7">                   <span class="dt">host =</span> <span class="st">&quot;www.ensembl.org&quot;</span>, </a>
<a class="sourceLine" id="cb26-8" data-line-number="8">                   <span class="dt">dataset =</span> <span class="st">&quot;mmusculus_gene_ensembl&quot;</span>)</a>
<a class="sourceLine" id="cb26-9" data-line-number="9"></a>
<a class="sourceLine" id="cb26-10" data-line-number="10"><span class="co">## Get gene annotations</span></a>
<a class="sourceLine" id="cb26-11" data-line-number="11"><span class="kw">message</span>(<span class="st">&quot;Getting gene annotations&quot;</span>)</a>
<a class="sourceLine" id="cb26-12" data-line-number="12">genes &lt;-<span class="st"> </span><span class="kw">getBM</span>(<span class="dt">attributes =</span> <span class="kw">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, </a>
<a class="sourceLine" id="cb26-13" data-line-number="13">                              <span class="st">&quot;external_gene_name&quot;</span>, </a>
<a class="sourceLine" id="cb26-14" data-line-number="14">                              <span class="st">&quot;chromosome_name&quot;</span>,</a>
<a class="sourceLine" id="cb26-15" data-line-number="15">                              <span class="st">&quot;start_position&quot;</span>, </a>
<a class="sourceLine" id="cb26-16" data-line-number="16">                              <span class="st">&quot;end_position&quot;</span>, </a>
<a class="sourceLine" id="cb26-17" data-line-number="17">                              <span class="st">&quot;strand&quot;</span>), </a>
<a class="sourceLine" id="cb26-18" data-line-number="18">               <span class="dt">filter =</span> <span class="st">&quot;ensembl_gene_id&quot;</span>,</a>
<a class="sourceLine" id="cb26-19" data-line-number="19">               <span class="dt">values =</span> <span class="kw">row.names</span>(gene_stat_norm),</a>
<a class="sourceLine" id="cb26-20" data-line-number="20">               <span class="dt">mart =</span> ensembl)</a>
<a class="sourceLine" id="cb26-21" data-line-number="21">  </a>
<a class="sourceLine" id="cb26-22" data-line-number="22"><span class="co">## Merge gene annotations and expression statistics</span></a>
<a class="sourceLine" id="cb26-23" data-line-number="23">gene_stat_norm &lt;-<span class="st"> </span><span class="kw">merge</span>(genes, gene_stat_norm, <span class="dt">by.x =</span> <span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="dt">by.y =</span> <span class="dv">0</span>, <span class="dt">sort =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb26-24" data-line-number="24"></a>
<a class="sourceLine" id="cb26-25" data-line-number="25"><span class="kw">kable</span>(gene_stat_norm[<span class="dv">100</span><span class="op">:</span><span class="dv">109</span>, ], <span class="dt">caption =</span> <span class="st">&quot;Gene-wise statistics after normalisation&quot;</span>)</a></code></pre></div>
<p><strong>Challenge falcultatif:</strong></p>
<p>Réordonnez les gènes par position génomique et affichez les lignes 5 premières et 5 dernières lignes de ce tableau de statistiques.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="co">#### Sorting genes by chromosome ####</span></a>
<a class="sourceLine" id="cb27-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;Sorting genes by chromosome&quot;</span>)</a>
<a class="sourceLine" id="cb27-3" data-line-number="3"></a>
<a class="sourceLine" id="cb27-4" data-line-number="4"></a>
<a class="sourceLine" id="cb27-5" data-line-number="5">gene_stat_norm &lt;-<span class="st"> </span>gene_stat_norm[<span class="kw">order</span>(gene_stat_norm<span class="op">$</span>chromosome_name, gene_stat_norm<span class="op">$</span>start_position),]</a>
<a class="sourceLine" id="cb27-6" data-line-number="6"></a>
<a class="sourceLine" id="cb27-7" data-line-number="7"><span class="kw">kable</span>(gene_stat_norm[<span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, (<span class="kw">nrow</span>(gene_stat_norm) <span class="op">-</span><span class="st"> </span><span class="dv">4</span>)<span class="op">:</span><span class="kw">nrow</span>(gene_stat_norm)), ], </a>
<a class="sourceLine" id="cb27-8" data-line-number="8">      <span class="dt">caption =</span> <span class="st">&quot;Gene-wise statistics after normalisation.&quot;</span>)</a></code></pre></div>
</div>
<div id="distribution-des-données" class="section level3">
<h3>Distribution des données</h3>
<ul>
<li>Dessinez sous forme d’un histogramme la distribution des valeurs après normalisation (tous échantillons confondus)</li>
</ul>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1"><span class="kw">hist</span>(<span class="kw">unlist</span>(fa_expr_log2), </a>
<a class="sourceLine" id="cb28-2" data-line-number="2">     <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">0</span>, <span class="dt">to =</span> <span class="kw">max</span>(fa_expr_log2) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, <span class="dt">by =</span> <span class="fl">0.25</span>),</a>
<a class="sourceLine" id="cb28-3" data-line-number="3">     <span class="dt">xlab =</span> <span class="st">&quot;log2(counts) after standardisation&quot;</span>, </a>
<a class="sourceLine" id="cb28-4" data-line-number="4">     <span class="dt">ylab =</span> <span class="st">&quot;number of genes after filtering&quot;</span>,</a>
<a class="sourceLine" id="cb28-5" data-line-number="5">     <span class="dt">col =</span> <span class="st">&quot;#BBDDFF&quot;</span>,</a>
<a class="sourceLine" id="cb28-6" data-line-number="6">     <span class="dt">las =</span> <span class="dv">1</span>, <span class="dt">cex.axis =</span> <span class="fl">0.8</span>,</a>
<a class="sourceLine" id="cb28-7" data-line-number="7">     <span class="dt">main =</span> <span class="st">&quot;distribution after standardisation&quot;</span>)</a>
<a class="sourceLine" id="cb28-8" data-line-number="8"><span class="kw">abline</span>(<span class="dt">v =</span> <span class="kw">mean</span>(fa_expr_log2), <span class="dt">col =</span> <span class="st">&quot;darkgreen&quot;</span>, <span class="dt">lwd =</span> <span class="dv">2</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/mini-projet_fa_expr_norm_distrib-1.png" alt="Distribution of expression values (log2 counts) after gene filtering and standardisation on the sample-wise third-quartile of non-null values. The vertical line highlights the mean value. " width="70%" />
<p class="caption">
Distribution of expression values (log2 counts) after gene filtering and standardisation on the sample-wise third-quartile of non-null values. The vertical line highlights the mean value.
</p>
</div>
<ul>
<li>Dessinez un box plot par échantillon avant et après normalisation, et commentez la façon dont l’effet de la normalisation apparaît sur ces graphiques.</li>
</ul>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1"><span class="co">#### Box plots to show normalisation impact ####</span></a>
<a class="sourceLine" id="cb29-2" data-line-number="2"><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">4</span>,<span class="dv">1</span>)) <span class="co">## Set the margins</span></a>
<a class="sourceLine" id="cb29-3" data-line-number="3"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))</a>
<a class="sourceLine" id="cb29-4" data-line-number="4"><span class="kw">boxplot</span>(fa_expr_raw, </a>
<a class="sourceLine" id="cb29-5" data-line-number="5">        <span class="dt">horizontal =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb29-6" data-line-number="6">        <span class="dt">xlab =</span> <span class="st">&quot;counts&quot;</span>, </a>
<a class="sourceLine" id="cb29-7" data-line-number="7">        <span class="dt">las =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb29-8" data-line-number="8">        <span class="dt">col =</span> fa_meta<span class="op">$</span>color, </a>
<a class="sourceLine" id="cb29-9" data-line-number="9">        <span class="dt">main =</span> <span class="st">&quot;Raw counts, all genes&quot;</span>)</a>
<a class="sourceLine" id="cb29-10" data-line-number="10"><span class="kw">boxplot</span>(fa_expr_filtered, </a>
<a class="sourceLine" id="cb29-11" data-line-number="11">        <span class="dt">horizontal =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb29-12" data-line-number="12">        <span class="dt">xlab =</span> <span class="st">&quot;counts&quot;</span>, </a>
<a class="sourceLine" id="cb29-13" data-line-number="13">        <span class="dt">las =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb29-14" data-line-number="14">        <span class="dt">col =</span> fa_meta<span class="op">$</span>color, </a>
<a class="sourceLine" id="cb29-15" data-line-number="15">        <span class="dt">main =</span> <span class="st">&quot;Raw counts, filtered genes&quot;</span>)</a>
<a class="sourceLine" id="cb29-16" data-line-number="16"><span class="kw">boxplot</span>(fa_expr_standard, </a>
<a class="sourceLine" id="cb29-17" data-line-number="17">        <span class="dt">horizontal =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb29-18" data-line-number="18">        <span class="dt">xlab =</span> <span class="st">&quot;counts&quot;</span>, </a>
<a class="sourceLine" id="cb29-19" data-line-number="19">        <span class="dt">las =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb29-20" data-line-number="20">        <span class="dt">col =</span> fa_meta<span class="op">$</span>color, </a>
<a class="sourceLine" id="cb29-21" data-line-number="21">        <span class="dt">main =</span> <span class="st">&quot;Standardised counts (sample Q3)&quot;</span>)</a>
<a class="sourceLine" id="cb29-22" data-line-number="22"><span class="kw">boxplot</span>(fa_expr_log2, </a>
<a class="sourceLine" id="cb29-23" data-line-number="23">        <span class="dt">xlab =</span> <span class="st">&quot;log2(counts)&quot;</span>, </a>
<a class="sourceLine" id="cb29-24" data-line-number="24">        <span class="dt">las =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb29-25" data-line-number="25">        <span class="dt">horizontal =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb29-26" data-line-number="26">        <span class="dt">col =</span> fa_meta<span class="op">$</span>color, </a>
<a class="sourceLine" id="cb29-27" data-line-number="27">        <span class="dt">main =</span> <span class="st">&quot;Normalised counts&quot;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/mini-projet_boxplots_normalisation_impact-1.png" alt="Box plots showing the impact of normalisation" width="100%" />
<p class="caption">
Box plots showing the impact of normalisation
</p>
</div>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</a>
<a class="sourceLine" id="cb30-2" data-line-number="2"><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">5</span>,<span class="dv">1</span>))</a></code></pre></div>
</div>
</div>
<div id="analyse-de-regroupement-des-données" class="section level2">
<h2>4. Analyse de regroupement des données</h2>
<p><strong><em>A vous de jouer!</em></strong></p>
<div id="sélection-de-gènes-dexpression-élevée-et-variable" class="section level3">
<h3>Sélection de gènes d’expression élevée et variable</h3>
<p>Pour réduire le nombre de gènes, nous allons écarter les gènes faiblement exprimés (log2 moyen inférieur à 4), et ne retenir que ceux qui montrent des variations importantes entre échantillons. Pour ce dernier critère, nous nous basons sur la variance.</p>
<!-- Sélectionnez les gènes ayant un niveau log2 moyen minimal supérieur à 3 ($m > 3$) et un coefficient de variation supérieur à 0.5 ($CV > 0.5$). Note: ces valeurs sont parfaitement arbitraires, elles ont été choisies pour obtenir un nombre raisonnable de gènes.  -->
<p>Sélectionnez les gènes ayant un niveau log2 moyen minimal supérieur à 5 (<span class="math inline">\(m &gt; 5\)</span>) et une variance supérieure à 2 (<span class="math inline">\(s^2 &gt; 2\)</span>). Note: ces valeurs sont parfaitement arbitraires, elles ont été choisies pour obtenir un nombre raisonnable de gènes.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" data-line-number="1"><span class="co">#### Selection of a subset of genes with igh expression and variance ####</span></a>
<a class="sourceLine" id="cb31-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;Selecting genes with high expression and variance&quot;</span>)</a>
<a class="sourceLine" id="cb31-3" data-line-number="3"></a>
<a class="sourceLine" id="cb31-4" data-line-number="4"><span class="co">## Compute a Boolean vector indicating whether each gene passes or not the expression level threshold</span></a>
<a class="sourceLine" id="cb31-5" data-line-number="5">high_expression &lt;-<span class="st"> </span>gene_stat_norm<span class="op">$</span>mean <span class="op">&gt;</span><span class="st"> </span><span class="dv">5</span></a>
<a class="sourceLine" id="cb31-6" data-line-number="6"><span class="co"># table(high_expression) # count number of genes with high/weak expression</span></a>
<a class="sourceLine" id="cb31-7" data-line-number="7"></a>
<a class="sourceLine" id="cb31-8" data-line-number="8"><span class="co">## Compute a Boolean vector indicating whether each gene passes or not the variance threshold</span></a>
<a class="sourceLine" id="cb31-9" data-line-number="9">high_variance &lt;-<span class="st"> </span>gene_stat_norm<span class="op">$</span>var <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span></a>
<a class="sourceLine" id="cb31-10" data-line-number="10"><span class="co"># table(high_variance) # count number of genes with weak high variance</span></a>
<a class="sourceLine" id="cb31-11" data-line-number="11"></a>
<a class="sourceLine" id="cb31-12" data-line-number="12"><span class="co">## Select genes having both a high mean expression and a high variance</span></a>
<a class="sourceLine" id="cb31-13" data-line-number="13">selected_genes &lt;-<span class="st"> </span>high_variance <span class="op">&amp;</span><span class="st"> </span>high_expression</a>
<a class="sourceLine" id="cb31-14" data-line-number="14"><span class="co"># table(selected_genes) # count number of genes with weak high coeffficient of variation</span></a>
<a class="sourceLine" id="cb31-15" data-line-number="15"><span class="kw">print</span>(<span class="kw">paste0</span>(<span class="st">&quot;Selected genes: &quot;</span>, <span class="kw">sum</span>(selected_genes)))</a></code></pre></div>
<pre><code>[1] &quot;Selected genes: 690&quot;</code></pre>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1"><span class="co">## Create a data frame with the expression of the selected genes</span></a>
<a class="sourceLine" id="cb33-2" data-line-number="2">fa_expr_selected &lt;-<span class="st"> </span>fa_expr_log2[selected_genes, ]</a></code></pre></div>
<p>Dessinez des histogrammes des valeurs d’expression avant et après cette sélection de gènes, et commentez les différences.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1"><span class="co">#### Histograms of expression before and after gene selection ####</span></a>
<a class="sourceLine" id="cb34-2" data-line-number="2"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">1</span>))</a>
<a class="sourceLine" id="cb34-3" data-line-number="3"><span class="kw">hist</span>(<span class="kw">unlist</span>(fa_expr_log2), </a>
<a class="sourceLine" id="cb34-4" data-line-number="4">     <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">0</span>, <span class="dt">to =</span> <span class="kw">max</span>(fa_expr_log2) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, <span class="dt">by =</span> <span class="fl">0.25</span>),</a>
<a class="sourceLine" id="cb34-5" data-line-number="5">     <span class="dt">las =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb34-6" data-line-number="6">     <span class="dt">cex.axis =</span> <span class="fl">0.8</span>, </a>
<a class="sourceLine" id="cb34-7" data-line-number="7">     <span class="dt">main =</span> <span class="st">&quot;Standardized values before gene selection&quot;</span>,</a>
<a class="sourceLine" id="cb34-8" data-line-number="8">     <span class="dt">col =</span>  <span class="st">&quot;#DDBBFF&quot;</span>)</a>
<a class="sourceLine" id="cb34-9" data-line-number="9"></a>
<a class="sourceLine" id="cb34-10" data-line-number="10"><span class="kw">hist</span>(<span class="kw">unlist</span>(fa_expr_selected), </a>
<a class="sourceLine" id="cb34-11" data-line-number="11">     <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">0</span>, <span class="dt">to =</span> <span class="kw">max</span>(fa_expr_log2) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, <span class="dt">by =</span> <span class="fl">0.25</span>),</a>
<a class="sourceLine" id="cb34-12" data-line-number="12">     <span class="dt">las =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb34-13" data-line-number="13">     <span class="dt">cex.axis =</span> <span class="fl">0.8</span>, </a>
<a class="sourceLine" id="cb34-14" data-line-number="14">     <span class="dt">main =</span> <span class="st">&quot;Standardized values after gene selection&quot;</span>,</a>
<a class="sourceLine" id="cb34-15" data-line-number="15">     <span class="dt">col =</span>  <span class="st">&quot;#FFDDBB&quot;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/mini-projet_hist_expr_selected_genes-1.png" alt="Distribution of expression values before and after gene selection" width="100%" />
<p class="caption">
Distribution of expression values before and after gene selection
</p>
</div>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" data-line-number="1"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</a></code></pre></div>
<p>Dessinez un box plot par échantillon des valeurs d’expression avant et après sélection des gènes, et commentez le résultat.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1"><span class="co">#### Boxplots of expression before and after gene selection ####</span></a>
<a class="sourceLine" id="cb36-2" data-line-number="2"></a>
<a class="sourceLine" id="cb36-3" data-line-number="3"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</a>
<a class="sourceLine" id="cb36-4" data-line-number="4"></a>
<a class="sourceLine" id="cb36-5" data-line-number="5"><span class="kw">boxplot</span>(fa_expr_log2, </a>
<a class="sourceLine" id="cb36-6" data-line-number="6">        <span class="dt">horizontal =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb36-7" data-line-number="7">        <span class="dt">xlab =</span> <span class="st">&quot;log2(counts)&quot;</span>, </a>
<a class="sourceLine" id="cb36-8" data-line-number="8">        <span class="dt">las =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb36-9" data-line-number="9">        <span class="dt">col =</span> fa_meta<span class="op">$</span>color, </a>
<a class="sourceLine" id="cb36-10" data-line-number="10">        <span class="dt">main =</span> <span class="st">&quot;Before gene selection&quot;</span>)</a>
<a class="sourceLine" id="cb36-11" data-line-number="11"><span class="kw">boxplot</span>(fa_expr_selected, </a>
<a class="sourceLine" id="cb36-12" data-line-number="12">        <span class="dt">horizontal =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb36-13" data-line-number="13">        <span class="dt">xlab =</span> <span class="st">&quot;log2(counts)&quot;</span>, </a>
<a class="sourceLine" id="cb36-14" data-line-number="14">        <span class="dt">las =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb36-15" data-line-number="15">        <span class="dt">col =</span> fa_meta<span class="op">$</span>color, </a>
<a class="sourceLine" id="cb36-16" data-line-number="16">        <span class="dt">main =</span> <span class="st">&quot;After gene selection&quot;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/mini-projet_boxplots_expr_selected_genes-1.png" alt="Box plots of standardised expression values before and after gene selection. " width="60%" />
<p class="caption">
Box plots of standardised expression values before and after gene selection.
</p>
</div>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb37-1" data-line-number="1"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</a></code></pre></div>
</div>
<div id="acp" class="section level3">
<h3>ACP</h3>
<p>Dessinez un plot ACP des échantillons en les colorant par condition avant et après normalisation.</p>
<ul>
<li>avec les comptages bruts de la matrice d’expression initiale (<span class="math inline">\(fa_expr\)</span>)</li>
</ul>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1"><span class="co">#### PCA from raw counts, all genes ####</span></a>
<a class="sourceLine" id="cb38-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;Running PCA with raw counts, all genes&quot;</span>)</a>
<a class="sourceLine" id="cb38-3" data-line-number="3"></a>
<a class="sourceLine" id="cb38-4" data-line-number="4"><span class="co">## Raw expression values, all genes</span></a>
<a class="sourceLine" id="cb38-5" data-line-number="5">ma_pca_raw_tt &lt;-<span class="st"> </span><span class="kw">PCA</span>(<span class="kw">t</span>(fa_expr_raw), </a>
<a class="sourceLine" id="cb38-6" data-line-number="6">                      <span class="dt">scale.unit =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb38-7" data-line-number="7">                      <span class="dt">graph =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb38-8" data-line-number="8"><span class="co"># plot(ma_pca_raw_tt, choix = &quot;ind&quot;)</span></a>
<a class="sourceLine" id="cb38-9" data-line-number="9"><span class="kw">fviz_pca_ind</span>(ma_pca_raw_tt, <span class="dt">col.ind =</span> fa_meta[,<span class="st">&quot;condition&quot;</span>])</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/mini-projet_acp_raw_all_genes-1.png" alt="PC plot of the samples from the raw expression values of all genes. " width="60%" />
<p class="caption">
PC plot of the samples from the raw expression values of all genes.
</p>
</div>
<ul>
<li>avec la matrice de valeurs normalisées des gènes filtrés</li>
</ul>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" data-line-number="1"><span class="co">#### PCA from normalised counts, all genes ####</span></a>
<a class="sourceLine" id="cb39-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;Running PCA with normalised counts, all genes&quot;</span>)</a>
<a class="sourceLine" id="cb39-3" data-line-number="3"></a>
<a class="sourceLine" id="cb39-4" data-line-number="4">ma_pca_filtered &lt;-<span class="st"> </span><span class="kw">PCA</span>(<span class="kw">t</span>(fa_expr_log2), <span class="dt">scale.unit =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb39-5" data-line-number="5">                  <span class="dt">graph =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb39-6" data-line-number="6"><span class="co"># plot(ma_pca_filtered, choix = &quot;var&quot;)</span></a>
<a class="sourceLine" id="cb39-7" data-line-number="7"><span class="co"># plot(ma_pca_sel, choix = &quot;ind&quot;)</span></a>
<a class="sourceLine" id="cb39-8" data-line-number="8"><span class="kw">fviz_pca_ind</span>(ma_pca_filtered, <span class="dt">col.ind =</span> fa_meta[, <span class="st">&quot;condition&quot;</span>])</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/mini-projet_acp_norm_filtered_genes-1.png" alt="PC plot of the samples from normalised counts, before gene selection. " width="60%" />
<p class="caption">
PC plot of the samples from normalised counts, before gene selection.
</p>
</div>
<ul>
<li>avec la matrice finale (transformation log2, filtre des gènes non-détectés, standardisation et sélection des gènes fortement exprimés et à haut coefficient de variation)</li>
</ul>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1"><span class="co">#### PCA from normalised counts, selected genes ####</span></a>
<a class="sourceLine" id="cb40-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;Running PCA with normalised counts, selected genes&quot;</span>)</a>
<a class="sourceLine" id="cb40-3" data-line-number="3"></a>
<a class="sourceLine" id="cb40-4" data-line-number="4">ma_pca_sel &lt;-<span class="st"> </span><span class="kw">PCA</span>(<span class="kw">t</span>(fa_expr_selected), <span class="dt">scale.unit =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb40-5" data-line-number="5">                  <span class="dt">graph =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb40-6" data-line-number="6"><span class="co"># plot(ma_pca_sel, choix = &quot;var&quot;)</span></a>
<a class="sourceLine" id="cb40-7" data-line-number="7"><span class="co"># plot(ma_pca_sel, choix = &quot;ind&quot;)</span></a>
<a class="sourceLine" id="cb40-8" data-line-number="8"><span class="kw">fviz_pca_ind</span>(ma_pca_sel, <span class="dt">col.ind =</span> fa_meta[, <span class="st">&quot;condition&quot;</span>])</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/mini-projet_acp_norm_selected_genes-1.png" alt="PC plot of the samples from normalised values, after gene selection. " width="60%" />
<p class="caption">
PC plot of the samples from normalised values, after gene selection.
</p>
</div>
</div>
<div id="clustering" class="section level3">
<h3>Clustering</h3>
<ul>
<li>Calculez les matrices de distance entre échantillons, en utilisant respectivement les distances euclidienne (<code>dist()</code>), coefficient de Pearson (<code>cor(, method = &quot;pearson&quot;)</code>) et de Spearman (<code>cor(, method = &quot;spearman&quot;)</code>).</li>
</ul>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" data-line-number="1"><span class="co">#### Sample distances ####</span></a>
<a class="sourceLine" id="cb41-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;Computing inter-sample distances&quot;</span>)</a>
<a class="sourceLine" id="cb41-3" data-line-number="3"></a>
<a class="sourceLine" id="cb41-4" data-line-number="4">dist_euc_samples &lt;-<span class="st"> </span><span class="kw">dist</span>(<span class="kw">t</span>(fa_expr_selected))</a>
<a class="sourceLine" id="cb41-5" data-line-number="5">cor_pearson_samples &lt;-<span class="st"> </span><span class="kw">as.dist</span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span><span class="kw">cor</span>(fa_expr_selected))</a>
<a class="sourceLine" id="cb41-6" data-line-number="6">cor_spearman_samples &lt;-<span class="st"> </span><span class="kw">as.dist</span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span><span class="kw">cor</span>(fa_expr_selected, <span class="dt">method =</span> <span class="st">&quot;spearman&quot;</span>))</a></code></pre></div>
<ul>
<li>Effectuez un clustering hiérarchique des échantillons, en utilisant le critère “complete” pour l’agglomération. Comparez les arbres d’échantillons obtenus avec ces trois métriques et choisissez celle qui vous paraît la plus pertinente.</li>
</ul>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" data-line-number="1"><span class="co">#### Sample clustering ####</span></a>
<a class="sourceLine" id="cb42-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;Sample clustering&quot;</span>)</a>
<a class="sourceLine" id="cb42-3" data-line-number="3"></a>
<a class="sourceLine" id="cb42-4" data-line-number="4"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</a>
<a class="sourceLine" id="cb42-5" data-line-number="5"><span class="kw">plot</span>(<span class="kw">hclust</span>(dist_euc_samples), <span class="dt">hang =</span> <span class="dv">-1</span>,</a>
<a class="sourceLine" id="cb42-6" data-line-number="6">     <span class="dt">main =</span> <span class="st">&quot;euclidean distance&quot;</span>)</a>
<a class="sourceLine" id="cb42-7" data-line-number="7"><span class="kw">plot</span>(<span class="kw">hclust</span>(cor_pearson_samples), <span class="dt">hang =</span> <span class="dv">-1</span>,</a>
<a class="sourceLine" id="cb42-8" data-line-number="8">     <span class="dt">main =</span> <span class="st">&quot;pearson&quot;</span>)</a>
<a class="sourceLine" id="cb42-9" data-line-number="9"><span class="kw">plot</span>(<span class="kw">hclust</span>(cor_spearman_samples), <span class="dt">hang =</span> <span class="dv">-1</span>,</a>
<a class="sourceLine" id="cb42-10" data-line-number="10">     <span class="dt">main =</span> <span class="st">&quot;spearman&quot;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/mini-projet_sample_clustering-1.png" alt="Sample tree with three alternative distance metrics: Euclidiant distance (left), Pearson correlation (center), Spearman correlation (right).? " width="100%" />
<p class="caption">
Sample tree with three alternative distance metrics: Euclidiant distance (left), Pearson correlation (center), Spearman correlation (right).?
</p>
</div>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" data-line-number="1"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</a></code></pre></div>
<ul>
<li>Effectuez un clustering hiérarchique des gènes en utilisant la distance basée sur le coefficient de Pearson et la règle d’agglomération complète</li>
</ul>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb44-1" data-line-number="1"><span class="co">#### Gene tree with Pearson correlation ####</span></a>
<a class="sourceLine" id="cb44-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;Drawing gene tree&quot;</span>)</a>
<a class="sourceLine" id="cb44-3" data-line-number="3"></a>
<a class="sourceLine" id="cb44-4" data-line-number="4"></a>
<a class="sourceLine" id="cb44-5" data-line-number="5">clustering_method &lt;-<span class="st"> &quot;complete&quot;</span> <span class="co">## Choose a clustering method</span></a>
<a class="sourceLine" id="cb44-6" data-line-number="6">cor_pearson_genes &lt;-<span class="st">  </span><span class="kw">as.dist</span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span><span class="kw">cor</span>(<span class="kw">t</span>(fa_expr_selected)))</a>
<a class="sourceLine" id="cb44-7" data-line-number="7"></a>
<a class="sourceLine" id="cb44-8" data-line-number="8"><span class="co">## Build a gene tree</span></a>
<a class="sourceLine" id="cb44-9" data-line-number="9">gene_tree_pearson &lt;-<span class="st"> </span><span class="kw">hclust</span>(cor_pearson_genes, <span class="dt">method =</span> clustering_method)</a>
<a class="sourceLine" id="cb44-10" data-line-number="10"></a>
<a class="sourceLine" id="cb44-11" data-line-number="11"><span class="co">## Plot the gene tree</span></a>
<a class="sourceLine" id="cb44-12" data-line-number="12"><span class="kw">plot</span>(gene_tree_pearson, <span class="dt">hang =</span> <span class="dv">-1</span>, <span class="dt">las =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb44-13" data-line-number="13">     <span class="dt">main =</span> <span class="kw">paste0</span>(<span class="st">&quot;Gene tree; Pearson coefficient; &quot;</span>, clustering_method, <span class="st">&quot; clustering&quot;</span>), </a>
<a class="sourceLine" id="cb44-14" data-line-number="14">     <span class="dt">labels =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p><img src="figures/mini-projet_gene_tree-1.png" width="672" style="display: block; margin: auto;" /></p>
<ul>
<li>Dessinez un arbre avec le résultat du clustering des gènes et commentez sa structure. Si vous deviez choisir de façon arbitraire un nombre de clusters, que choisiriez-vous ? Pourquoi ? Pas de panique, nous pouvons assumer ici que la réponse comporte une part de subjectivité.</li>
</ul>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" data-line-number="1"><span class="co">#### Plot the gene tree with boxes to denote the clusters ####</span></a>
<a class="sourceLine" id="cb45-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;Plotting gene tree&quot;</span>)</a>
<a class="sourceLine" id="cb45-3" data-line-number="3"></a>
<a class="sourceLine" id="cb45-4" data-line-number="4"><span class="kw">plot</span>(gene_tree_pearson, <span class="dt">hang =</span> <span class="dv">-1</span>,</a>
<a class="sourceLine" id="cb45-5" data-line-number="5">     <span class="dt">main =</span> <span class="kw">paste0</span>(<span class="st">&quot;Gene tree with 7 clusters; Pearson coefficient; &quot;</span>, clustering_method, <span class="st">&quot; clustering&quot;</span>), </a>
<a class="sourceLine" id="cb45-6" data-line-number="6">     <span class="dt">labels =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb45-7" data-line-number="7"><span class="kw">rect.hclust</span>(gene_tree_pearson, <span class="dt">k =</span> <span class="dv">7</span>)</a></code></pre></div>
<p><img src="figures/mini-projet_gene_tree_with_boxes-1.png" width="672" style="display: block; margin: auto;" /></p>
<ul>
<li>Dessinez une heatmap du résultat, en sélectionnant les deux résultats de clustering ci-dessus pour les gènes et les échantillons.</li>
</ul>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb46-1" data-line-number="1"><span class="co">#### Heatmap with biclustering ####</span></a>
<a class="sourceLine" id="cb46-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;heatmap with biclustering&quot;</span>)</a>
<a class="sourceLine" id="cb46-3" data-line-number="3"></a>
<a class="sourceLine" id="cb46-4" data-line-number="4"><span class="kw">pheatmap</span>(<span class="kw">t</span>(fa_expr_selected), </a>
<a class="sourceLine" id="cb46-5" data-line-number="5">         <span class="dt">main =</span> <span class="kw">paste0</span>(<span class="st">&quot;Biclustering; &quot;</span>, <span class="st">&quot;Pearson correlation; &quot;</span>, clustering_method, <span class="st">&quot; clustering&quot;</span>),</a>
<a class="sourceLine" id="cb46-6" data-line-number="6">         <span class="dt">show_colnames =</span> <span class="ot">FALSE</span>, <span class="dt">scale =</span> <span class="st">&quot;none&quot;</span>,</a>
<a class="sourceLine" id="cb46-7" data-line-number="7">         <span class="dt">clustering_method =</span> clustering_method, </a>
<a class="sourceLine" id="cb46-8" data-line-number="8">         <span class="dt">clustering_distance_cols =</span> <span class="st">&quot;correlation&quot;</span>, </a>
<a class="sourceLine" id="cb46-9" data-line-number="9">         <span class="dt">clustering_distance_rows =</span> <span class="st">&quot;correlation&quot;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/mini-projet_heatmap_biclustering-1.png" alt="Heatmap showing the biclustering of genes and samples. " width="90%" />
<p class="caption">
Heatmap showing the biclustering of genes and samples.
</p>
</div>
<ul>
<li>Dessinez une heatmap du résultat, en affichant un arbre sur les gènes mais pas sur les échantillons</li>
</ul>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb47-1" data-line-number="1"><span class="co">#### Heatmap with clustering on genes only ####</span></a>
<a class="sourceLine" id="cb47-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;Heatmap with gene clustering&quot;</span>)</a>
<a class="sourceLine" id="cb47-3" data-line-number="3"></a>
<a class="sourceLine" id="cb47-4" data-line-number="4"><span class="kw">pheatmap</span>(<span class="kw">t</span>(fa_expr_selected), </a>
<a class="sourceLine" id="cb47-5" data-line-number="5">         <span class="dt">main =</span> <span class="kw">paste0</span>(<span class="st">&quot;Gene clustering; &quot;</span>, <span class="st">&quot;Pearson correlation; &quot;</span>, clustering_method, <span class="st">&quot; clustering&quot;</span>),</a>
<a class="sourceLine" id="cb47-6" data-line-number="6">         <span class="dt">show_colnames =</span> <span class="ot">FALSE</span>, <span class="dt">scale =</span> <span class="st">&quot;none&quot;</span>,</a>
<a class="sourceLine" id="cb47-7" data-line-number="7">         <span class="dt">clustering_method =</span> clustering_method, </a>
<a class="sourceLine" id="cb47-8" data-line-number="8">         <span class="dt">clustering_distance_cols =</span> <span class="st">&quot;correlation&quot;</span>, </a>
<a class="sourceLine" id="cb47-9" data-line-number="9">         <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p><img src="figures/mini-projet_biclustering-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Interprétez les résultats en quelques phrases.</p>
</div>
</div>
<div id="enrichissement-fonctionnel" class="section level2">
<h2>5. Enrichissement fonctionnel</h2>
<p><strong><em>A vous de jouer!</em></strong></p>
<p>Effectuez une analyse d’enrichissement fonctionnel avec les principaux clusters obtenus dans la section précédente.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb48-1" data-line-number="1"><span class="co">#### Run enrichment analysis with gost() ####</span></a>
<a class="sourceLine" id="cb48-2" data-line-number="2"><span class="kw">message</span>(<span class="st">&quot;Running enrichment analysis with gost&quot;</span>)</a>
<a class="sourceLine" id="cb48-3" data-line-number="3"></a>
<a class="sourceLine" id="cb48-4" data-line-number="4"><span class="co">## Get the list of IDs for the selected genes ##</span></a>
<a class="sourceLine" id="cb48-5" data-line-number="5">gene_ids &lt;-<span class="st"> </span><span class="kw">rownames</span>(gene_stat_norm)[selected_genes]</a>
<a class="sourceLine" id="cb48-6" data-line-number="6"></a>
<a class="sourceLine" id="cb48-7" data-line-number="7"><span class="co">## Compute enrichment</span></a>
<a class="sourceLine" id="cb48-8" data-line-number="8">gostres &lt;-<span class="st"> </span><span class="kw">gost</span>(<span class="dt">query =</span> gene_ids, </a>
<a class="sourceLine" id="cb48-9" data-line-number="9">                <span class="dt">organism =</span> <span class="st">&quot;mmusculus&quot;</span>, </a>
<a class="sourceLine" id="cb48-10" data-line-number="10">                <span class="dt">ordered_query =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb48-11" data-line-number="11">                <span class="dt">multi_query =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb48-12" data-line-number="12">                <span class="dt">significant =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb48-13" data-line-number="13">                <span class="dt">exclude_iea =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb48-14" data-line-number="14">                <span class="dt">measure_underrepresentation =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb48-15" data-line-number="15">                <span class="dt">evcodes =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb48-16" data-line-number="16">                <span class="dt">user_threshold =</span> <span class="fl">0.05</span>, </a>
<a class="sourceLine" id="cb48-17" data-line-number="17">                <span class="dt">correction_method =</span> <span class="st">&quot;fdr&quot;</span>, </a>
<a class="sourceLine" id="cb48-18" data-line-number="18">                <span class="dt">domain_scope =</span> <span class="st">&quot;annotated&quot;</span>, </a>
<a class="sourceLine" id="cb48-19" data-line-number="19">                <span class="dt">custom_bg =</span> <span class="ot">NULL</span>, </a>
<a class="sourceLine" id="cb48-20" data-line-number="20">                <span class="dt">numeric_ns =</span> <span class="st">&quot;&quot;</span>, </a>
<a class="sourceLine" id="cb48-21" data-line-number="21">                <span class="dt">sources =</span> <span class="ot">NULL</span>, </a>
<a class="sourceLine" id="cb48-22" data-line-number="22">                <span class="dt">as_short_link =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb48-23" data-line-number="23"></a>
<a class="sourceLine" id="cb48-24" data-line-number="24"></a>
<a class="sourceLine" id="cb48-25" data-line-number="25"><span class="co">## Check the structure of the result</span></a>
<a class="sourceLine" id="cb48-26" data-line-number="26"><span class="kw">names</span>(gostres)</a></code></pre></div>
<pre><code>[1] &quot;result&quot; &quot;meta&quot;  </code></pre>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb50-1" data-line-number="1"><span class="kw">names</span>(gostres<span class="op">$</span>result)</a></code></pre></div>
<pre><code> [1] &quot;query&quot;                 &quot;significant&quot;           &quot;p_value&quot;               &quot;term_size&quot;             &quot;query_size&quot;            &quot;intersection_size&quot;     &quot;precision&quot;             &quot;recall&quot;                &quot;term_id&quot;               &quot;source&quot;                &quot;term_name&quot;             &quot;effective_domain_size&quot;
[13] &quot;source_order&quot;          &quot;parents&quot;              </code></pre>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb52-1" data-line-number="1"><span class="kw">kable</span>(<span class="kw">summary</span>(gostres))</a></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">Length</th>
<th align="left">Class</th>
<th align="left">Mode</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">result</td>
<td align="left">14</td>
<td align="left">data.frame</td>
<td align="left">list</td>
</tr>
<tr class="even">
<td align="left">meta</td>
<td align="left">5</td>
<td align="left">-none-</td>
<td align="left">list</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb53-1" data-line-number="1"><span class="co"># Visualization</span></a>
<a class="sourceLine" id="cb53-2" data-line-number="2"><span class="kw">gostplot</span>(gostres, <span class="dt">capped =</span> <span class="ot">TRUE</span>, <span class="dt">interactive =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<div id="htmlwidget-dfd4fbe32852d9d950d2" style="width:80%;height:480px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-dfd4fbe32852d9d950d2">{"x":{"data":[{"x":[55.3363921460187,50.7402533597022,65.5006926289088,50.4417387095979,86.1031376125532,55.3388592092427,85.0620369320242,51.4705040740069,55.3462603989148,86.0982034861052,58.0254910601814,85.0595698688002,89.7938641956609,67.0006670691023,55.7508587676512,50.4392716463739,69.3246406261126,99.9260928566389,55.3437933356907,69.3295747525606,51.4655699475589,50.9178819118304,65.3502017722447,62.2046961616415,76.5482017459917],"y":[7.40392287980855,6.79081588492956,3.67321269327742,3.67321269327742,3.37250785379766,3.28390687172796,3.28390687172796,2.99920821804602,2.92672484802273,2.44220193317336,2.28614787000941,2.23333985251944,1.80943010930107,1.58121705239003,1.58121705239003,1.58121705239003,1.50604358526077,1.50604358526077,1.50604358526077,1.50604358526077,1.50604358526077,1.50604358526077,1.50604358526077,1.50604358526077,1.50604358526077],"text":["GO:0006955 (272) <br> immune response <br> 3.945e-08","GO:0002376 (331) <br> immune system process <br> 1.619e-07","GO:0019882 (20) <br> antigen processing and presentation <br> 2.122e-04","GO:0002253 (47) <br> activation of immune response <br> 2.122e-04","GO:0050778 (53) <br> positive regulation of immune response <br> 4.241e-04","GO:0006956 (24) <br> complement activation <br> 5.201e-04","GO:0048584 (97) <br> positive regulation of response to stimulus <br> 5.201e-04","GO:0002684 (61) <br> positive regulation of immune system process <br> 1.002e-03","GO:0006959 (28) <br> humoral immune response <br> 1.184e-03","GO:0050776 (72) <br> regulation of immune response <br> 3.612e-03","GO:0009605 (219) <br> response to external stimulus <br> 5.174e-03","GO:0048583 (307) <br> regulation of response to stimulus <br> 5.843e-03","GO:0055114 (546) <br> obsolete oxidation-reduction process <br> 1.551e-02","GO:0022610 (279) <br> biological adhesion <br> 2.623e-02","GO:0007155 (279) <br> cell adhesion <br> 2.623e-02","GO:0002252 (63) <br> immune effector process <br> 2.623e-02","GO:0032101 (51) <br> regulation of response to external stimulus <br> 3.119e-02","GO:0080134 (67) <br> regulation of response to stress <br> 3.119e-02","GO:0006958 (7) <br> complement activation, classical pathway <br> 3.119e-02","GO:0032103 (25) <br> positive regulation of response to external stimulus <br> 3.119e-02","GO:0002682 (103) <br> regulation of immune system process <br> 3.119e-02","GO:0002455 (7) <br> humoral immune response mediated by circulating immunoglobulin <br> 3.119e-02","GO:0019724 (7) <br> B cell mediated immunity <br> 3.119e-02","GO:0016064 (7) <br> immunoglobulin mediated immune response <br> 3.119e-02","GO:0040011 (139) <br> locomotion <br> 3.119e-02"],"key":["GO:0006955","GO:0002376","GO:0019882","GO:0002253","GO:0050778","GO:0006956","GO:0048584","GO:0002684","GO:0006959","GO:0050776","GO:0009605","GO:0048583","GO:0055114","GO:0022610","GO:0007155","GO:0002252","GO:0032101","GO:0080134","GO:0006958","GO:0032103","GO:0002682","GO:0002455","GO:0019724","GO:0016064","GO:0040011"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(255,153,0,1)","opacity":0.8,"size":[17.3514527457023,17.6438736137172,12.5846608408581,14.3836828754516,14.6127217468526,12.9980622117105,15.6983838573381,14.8747209292772,13.3336428773766,15.1759662688234,17.0211266573712,17.5324900122574,18.3628828869183,17.3896554947286,17.3896554947286,14.9339766116786,14.539921566973,15.0461794599383,9.66398721317626,13.0881080176599,15.8008384614794,9.66398721317626,9.66398721317626,9.66398721317626,16.2999793914258],"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(255,153,0,1)"}},"hoveron":"points","set":"SharedData9d23eecd","name":"GO:BP","legendgroup":"GO:BP","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[42.4676501727968,39.8027121626118,39.7977770922226,35.9903702869305,43.6915476293262,35.4055644458066,41.3498567296359,35.3142656436058,37.7818008382215,37.4240082350023,37.6830994304369,35.9928378221252,35.3167331788005],"y":[5.18583428117637,5.18583428117637,5.02982881805163,4.14347547845657,2.88636337182863,2.34792427025139,2.34792427025139,2.34792427025139,2.085269588924,1.89873772644427,1.89873772644427,1.89873772644427,1.58628846968052],"text":["GO:0071944 (627) <br> cell periphery <br> 6.519e-06","GO:0042613 (10) <br> MHC class II protein complex <br> 6.519e-06","GO:0042611 (11) <br> MHC protein complex <br> 9.336e-06","GO:0005886 (559) <br> plasma membrane <br> 7.187e-05","GO:0098797 (180) <br> plasma membrane protein complex <br> 1.299e-03","GO:0005615 (208) <br> extracellular space <br> 4.488e-03","GO:0062023 (31) <br> collagen-containing extracellular matrix <br> 4.488e-03","GO:0005576 (641) <br> extracellular region <br> 4.488e-03","GO:0031226 (244) <br> intrinsic component of plasma membrane <br> 8.217e-03","GO:0030312 (57) <br> external encapsulating structure <br> 1.263e-02","GO:0031012 (57) <br> extracellular matrix <br> 1.263e-02","GO:0005887 (237) <br> integral component of plasma membrane <br> 1.263e-02","GO:0005577 (3) <br> fibrinogen complex <br> 2.592e-02"],"key":["GO:0071944","GO:0042613","GO:0042611","GO:0005886","GO:0098797","GO:0005615","GO:0062023","GO:0005576","GO:0031226","GO:0030312","GO:0031012","GO:0005887","GO:0005577"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(16,150,24,1)","opacity":0.8,"size":[18.5554316374665,10.7940300405511,11.0663872976286,18.395815836283,16.7149533486033,16.9413625410532,13.5489039044932,18.5859387661536,17.1868975478868,14.7491028635723,14.7491028635723,17.1424612987914,3.77952755905512],"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(16,150,24,1)"}},"hoveron":"points","set":"SharedData9d23eecd","name":"GO:CC","legendgroup":"GO:CC","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[6.03137799459199,18.4462458457468,6.76413152237278,2.76976128170912,6.64570670980215,20.1337994248783,18.7694468967208,13.5957628975412,5.51080225600024,11.1310464859149,6.03384517818721,20.1609384444257,13.5168130224941,7.50922096812969,20.1510697100448],"y":[3.35061872560591,2.73880072505832,2.73880072505832,1.56888446057284,1.56888446057284,1.46168901908776,1.46168901908776,1.3630169942712,1.3630169942712,1.3630169942712,1.36030953963472,1.36030953963472,1.34858261122988,1.30247778421846,1.30247778421846],"text":["GO:0005125 (73) <br> cytokine activity <br> 4.460e-04","GO:0042379 (39) <br> chemokine receptor binding <br> 1.825e-03","GO:0008009 (39) <br> chemokine activity <br> 1.825e-03","GO:0001664 (60) <br> G protein-coupled receptor binding <br> 2.698e-02","GO:0005506 (165) <br> iron ion binding <br> 2.698e-02","GO:0046872 (1546) <br> metal ion binding <br> 3.454e-02","GO:0043169 (1571) <br> cation binding <br> 3.454e-02","GO:0022857 (746) <br> transmembrane transporter activity <br> 4.335e-02","GO:0004857 (145) <br> enzyme inhibitor activity <br> 4.335e-02","GO:0016712 (40) <br> oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen, reduced flavin or flavoprotein as one donor, and incorporation of one atom of oxygen <br> 4.335e-02","GO:0005126 (127) <br> cytokine receptor binding <br> 4.362e-02","GO:0046914 (762) <br> transition metal ion binding <br> 4.362e-02","GO:0020037 (150) <br> heme binding <br> 4.481e-02","GO:0008509 (134) <br> anion transmembrane transporter activity <br> 4.983e-02","GO:0046906 (154) <br> tetrapyrrole binding <br> 4.983e-02"],"key":["GO:0005125","GO:0042379","GO:0008009","GO:0001664","GO:0005506","GO:0046872","GO:0043169","GO:0022857","GO:0004857","GO:0016712","GO:0005126","GO:0046914","GO:0020037","GO:0008509","GO:0046906"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(220,57,18,1)","opacity":0.8,"size":[15.2006707836727,14.0178294703604,14.0178294703604,14.8442375070661,16.5767659764539,19.7548004804641,19.7753056172727,18.7938287637891,16.3687595792522,14.0682349199298,16.1517514857845,18.8226816852294,16.4236657789979,16.2400480980646,16.4661248743845],"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(220,57,18,1)"}},"hoveron":"points","set":"SharedData9d23eecd","name":"GO:MF","legendgroup":"GO:MF","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[167.024416079295,162.936367920288,159.553921314375,165.741504164579,167.056488877162,176.147893503559],"y":[3.84141137829545,3.55580219313319,2.88410462338594,2.7507106755661,2.25442766855734,1.77111439518386],"text":["HP:0005339 (30) <br> Abnormality of complement system <br> 1.441e-04","HP:0002725 (34) <br> Systemic lupus erythematosus <br> 2.781e-04","HP:0000793 (7) <br> Membranoproliferative glomerulonephritis <br> 1.306e-03","HP:0004431 (26) <br> Complement deficiency <br> 1.775e-03","HP:0005356 (5) <br> Decreased serum complement factor I <br> 5.566e-03","HP:0010881 (36) <br> Abnormality of the umbilical cord <br> 1.694e-02"],"key":["HP:0005339","HP:0002725","HP:0000793","HP:0004431","HP:0005356","HP:0010881"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(153,0,153,1)","opacity":0.8,"size":[13.4800780169523,13.7402389859433,9.66398721317626,13.1738087423678,8.34856662821513,13.8568133155523],"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(153,0,153,1)"}},"hoveron":"points","set":"SharedData9d23eecd","name":"HP","legendgroup":"HP","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[126.601229292094,126.890397781795,126.65560285426,126.999144906127,126.514725897738,126.586400138775,126.887926256242,126.588871664329,126.660545905366,126.875568628477,126.882983205136,126.989258803915,126.440580131148,126.892869307348,126.922527613984,126.507311321079,126.583928613222,126.996673380574,126.633359124283,126.994201855021,126.868154051818,126.880511679583,126.870625577371,126.438108605595,127.018917110551,126.771764555251,126.87804015403,126.596286240988,126.445523182254,126.443051656701,126.816252015205,126.885454730689,126.022892312691,126.905226935113,126.45046623336,125.9067306117],"y":[9.16579872210702,6.67222224502946,4.75144531384717,4.75144531384717,3.6795974182128,3.24185995761033,3.11561737255213,3.00859709631807,3.00859709631807,3.00859709631807,3.00859709631807,2.839282053565,2.69340532935971,2.69340532935971,2.63014773552814,2.38878471986835,2.24643918842378,2.24643918842378,2.17734426170881,2.17734426170881,2.17734426170881,2.00862265069925,1.58868993520472,1.4983560281783,1.4983560281783,1.43126355054717,1.39050871453031,1.36598168049342,1.36598168049342,1.34536440562325,1.33815833698155,1.32780031079437,1.32780031079437,1.32780031079437,1.30711519023461,1.30711519023461],"text":["KEGG:04610 (92) <br> Complement and coagulation cascades <br> 6.827e-10","KEGG:05150 (116) <br> Staphylococcus aureus infection <br> 2.127e-07","KEGG:04668 (113) <br> TNF signaling pathway <br> 1.772e-05","KEGG:05323 (86) <br> Rheumatoid arthritis <br> 1.772e-05","KEGG:04151 (359) <br> PI3K-Akt signaling pathway <br> 2.091e-04","KEGG:04512 (88) <br> ECM-receptor interaction <br> 5.730e-04","KEGG:05146 (107) <br> Amoebiasis <br> 7.663e-04","KEGG:04514 (165) <br> Cell adhesion molecules <br> 9.804e-04","KEGG:04672 (42) <br> Intestinal immune network for IgA production <br> 9.804e-04","KEGG:05140 (69) <br> Leishmaniasis <br> 9.804e-04","KEGG:05144 (55) <br> Malaria <br> 9.804e-04","KEGG:05310 (24) <br> Asthma <br> 1.448e-03","KEGG:04061 (92) <br> Viral protein interaction with cytokine and cytokine receptor <br> 2.026e-03","KEGG:05152 (179) <br> Tuberculosis <br> 2.026e-03","KEGG:05171 (241) <br> Coronavirus disease - COVID-19 <br> 2.343e-03","KEGG:04145 (173) <br> Phagosome <br> 4.085e-03","KEGG:04510 (200) <br> Focal adhesion <br> 5.670e-03","KEGG:05322 (142) <br> Systemic lupus erythematosus <br> 5.670e-03","KEGG:04640 (93) <br> Hematopoietic cell lineage <br> 6.647e-03","KEGG:05321 (61) <br> Inflammatory bowel disease <br> 6.647e-03","KEGG:05133 (77) <br> Pertussis <br> 6.647e-03","KEGG:05143 (37) <br> African trypanosomiasis <br> 9.803e-03","KEGG:05134 (60) <br> Legionellosis <br> 2.578e-02","KEGG:04060 (288) <br> Cytokine-cytokine receptor interaction <br> 3.174e-02","KEGG:05416 (80) <br> Viral myocarditis <br> 3.174e-02","KEGG:04933 (101) <br> AGE-RAGE signaling pathway in diabetic complications <br> 3.705e-02","KEGG:05142 (103) <br> Chagas disease <br> 4.069e-02","KEGG:04540 (86) <br> Gap junction <br> 4.305e-02","KEGG:04064 (105) <br> NF-kappa B signaling pathway <br> 4.305e-02","KEGG:04062 (189) <br> Chemokine signaling pathway <br> 4.515e-02","KEGG:04978 (53) <br> Mineral absorption <br> 4.590e-02","KEGG:05145 (109) <br> Toxoplasmosis <br> 4.701e-02","KEGG:00480 (71) <br> Glutathione metabolism <br> 4.701e-02","KEGG:05164 (171) <br> Influenza A <br> 4.701e-02","KEGG:04068 (131) <br> FoxO signaling pathway <br> 4.930e-02","KEGG:00140 (91) <br> Steroid hormone biosynthesis <br> 4.930e-02"],"key":["KEGG:04610","KEGG:05150","KEGG:04668","KEGG:05323","KEGG:04151","KEGG:04512","KEGG:05146","KEGG:04514","KEGG:04672","KEGG:05140","KEGG:05144","KEGG:05310","KEGG:04061","KEGG:05152","KEGG:05171","KEGG:04145","KEGG:04510","KEGG:05322","KEGG:04640","KEGG:05321","KEGG:05133","KEGG:05143","KEGG:05134","KEGG:04060","KEGG:05416","KEGG:04933","KEGG:05142","KEGG:04540","KEGG:04064","KEGG:04062","KEGG:04978","KEGG:05145","KEGG:00480","KEGG:05164","KEGG:04068","KEGG:00140"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(221,68,119,1)","opacity":0.8,"size":[15.6073056347349,16.0012076732765,15.9573205377134,15.4902143033667,17.7630424559913,15.5302597061109,15.865426685159,16.5767659764539,14.1646817029125,15.0994000952643,14.6823655835536,12.9980622117105,15.6073056347349,16.7061499407379,17.1680296192606,16.6521428423289,16.8803310190161,16.3347798817969,15.6259678810295,14.8747209292772,15.2957172237801,13.9122178835638,14.8442375070661,17.4372385444402,15.3633363543141,15.7674618455403,15.8008384614794,15.4902143033667,15.8334834111015,16.7917974529234,14.6127217468526,15.8966964992365,15.1508614227806,16.6336715011763,16.2028617660777,15.5884094471941],"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(221,68,119,1)"}},"hoveron":"points","set":"SharedData9d23eecd","name":"KEGG","legendgroup":"KEGG","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[130.787000787264,132.682760982559,132.273000315347,131.337462406471,132.825930372308,132.24584750005,132.240910624541,130.66357889955,130.569778264887,132.663013480525],"y":[3.28789191382907,3.28789191382907,3.207152535556,2.75301824349226,2.50959530169931,2.11802320350184,1.95691945521327,1.55257285913237,1.48358809633785,1.43393484591606],"text":["REAC:R-MMU-1474244 (258) <br> Extracellular matrix organization <br> 5.154e-04","REAC:R-MMU-381426 (114) <br> Regulation of Insulin-like Growth Factor (IGF) transport and uptake by Insulin-like Growth Factor Binding Proteins (IGFBPs) <br> 5.154e-04","REAC:R-MMU-8957275 (108) <br> Post-translational protein phosphorylation <br> 6.207e-04","REAC:R-MMU-216083 (66) <br> Integrin cell surface interactions <br> 1.766e-03","REAC:R-MMU-76005 (113) <br> Response to elevated platelet cytosolic Ca2+ <br> 3.093e-03","REAC:R-MMU-114608 (109) <br> Platelet degranulation  <br> 7.620e-03","REAC:R-MMU-76002 (235) <br> Platelet activation, signaling and aggregation <br> 1.104e-02","REAC:R-MMU-3000178 (40) <br> ECM proteoglycans <br> 2.802e-02","REAC:R-MMU-1474228 (131) <br> Degradation of the extracellular matrix <br> 3.284e-02","REAC:R-MMU-977606 (134) <br> Regulation of Complement cascade <br> 3.682e-02"],"key":["REAC:R-MMU-1474244","REAC:R-MMU-381426","REAC:R-MMU-8957275","REAC:R-MMU-216083","REAC:R-MMU-76005","REAC:R-MMU-114608","REAC:R-MMU-76002","REAC:R-MMU-3000178","REAC:R-MMU-1474228","REAC:R-MMU-977606"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(51,102,204,1)","opacity":0.8,"size":[17.271659271527,15.9720953529563,15.8811440767522,15.0188730035346,15.9573205377134,15.8966964992365,17.1294961774852,14.0682349199298,16.2028617660777,16.2400480980646],"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(51,102,204,1)"}},"hoveron":"points","set":"SharedData9d23eecd","name":"REAC","legendgroup":"REAC","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[139.621400685866,142.034689639624,145.447346718403,146.212295159778,146.412168784783,141.262338471397,142.037157215241,145.454749445255,146.204892432926,145.452281869638,141.277143925101,145.778001851127,143.268477448293,142.728078388096,146.214762735396,137.750978367923,146.202424857309,145.780469426744,137.331490512976],"y":[1.68476651010331,1.68476651010331,1.68476651010331,1.68476651010331,1.59516475804706,1.59516475804706,1.59516475804706,1.59516475804706,1.59516475804706,1.53091892912104,1.52798899906536,1.49791267437232,1.47132201810011,1.47132201810011,1.47132201810011,1.47132201810011,1.47132201810011,1.45501655971177,1.31366975034508],"text":["TF:M07289_1 (10998) <br> Factor: GKLF; motif: NNNRGGNGNGGSN; match class: 1 <br> 2.066e-02","TF:M00208_1 (429) <br> Factor: NF-kappaB; motif: NGGGACTTTCCA; match class: 1 <br> 2.066e-02","TF:M07436 (7486) <br> Factor: WT1; motif: NNGGGNGGGSGN <br> 2.066e-02","TF:M03944 (7312) <br> Factor: Zfp740; motif: NCCCCCCCAC <br> 2.066e-02","TF:M02941_1 (14351) <br> Factor: ZIC3; motif: NNNCACAGCAKGNNN; match class: 1 <br> 2.540e-02","TF:M00491 (3859) <br> Factor: MAZR; motif: NSGGGGGGGGMCN <br> 2.540e-02","TF:M00194 (2366) <br> Factor: NF-kappaB; motif: NGGGGAMTTTCCNN <br> 2.540e-02","TF:M01118_1 (4603) <br> Factor: WT1; motif: SMCNCCNSC; match class: 1 <br> 2.540e-02","TF:M02834_1 (3827) <br> Factor: Zfp740; motif: NNNCCCCCCCCMNNNN; match class: 1 <br> 2.540e-02","TF:M01118 (8128) <br> Factor: WT1; motif: SMCNCCNSC <br> 2.945e-02","TF:M02023 (7957) <br> Factor: MAZ; motif: NKGGGAGGGGRGGR <br> 2.965e-02","TF:M02831 (3971) <br> Factor: Zfp281; motif: NNCCCCCCCCCCMYC <br> 3.178e-02","TF:M09720_1 (1095) <br> Factor: RelA-p65; motif: GGAAWTTCCMNNNTTNCNNMAN; match class: 1 <br> 3.378e-02","TF:M00378 (18716) <br> Factor: Pax-4; motif: NNNNNYCACCCB <br> 3.378e-02","TF:M03944_1 (4460) <br> Factor: Zfp740; motif: NCCCCCCCAC; match class: 1 <br> 3.378e-02","TF:M03545_1 (983) <br> Factor: c-Rel; motif: NGGGAATYTCCN; match class: 1 <br> 3.378e-02","TF:M02834 (5204) <br> Factor: Zfp740; motif: NNNCCCCCCCCMNNNN <br> 3.378e-02","TF:M02831_1 (2494) <br> Factor: Zfp281; motif: NNCCCCCCCCCCMYC; match class: 1 <br> 3.507e-02","TF:M00109_1 (485) <br> Factor: C/EBPbeta; motif: RNRTKDNGMAAKNN; match class: 1 <br> 4.857e-02"],"key":["TF:M07289_1","TF:M00208_1","TF:M07436","TF:M03944","TF:M02941_1","TF:M00491","TF:M00194","TF:M01118_1","TF:M02834_1","TF:M01118","TF:M02023","TF:M02831","TF:M09720_1","TF:M00378","TF:M03944_1","TF:M03545_1","TF:M02834","TF:M02831_1","TF:M00109_1"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(85,116,166,1)","opacity":0.8,"size":[22.0932636960451,18.0209665664018,21.6589041012392,21.6320062404956,22.387805309763,20.884856507207,20.2901194479747,21.0941805876648,20.8749064843696,21.7526933118531,21.7285032904283,20.9189992420618,19.3073573758497,22.6771653543307,21.0568958383802,19.1647114840457,21.2384057805259,20.3551973454088,18.1959309800866],"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(85,116,166,1)"}},"hoveron":"points","set":"SharedData9d23eecd","name":"TF","legendgroup":"TF","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[199.628052070567,199.531348881441,199.575981122576,199.811540173009,199.841295000432,199.960314310125,199.764428362923,199.955355172221],"y":[4.52293226699225,4.21085203596126,1.74858042072352,1.74858042072352,1.74858042072352,1.74858042072352,1.70942829171692,1.60897937512676],"text":["WP:WP3626 (41) <br> Microglia Pathogen Phagocytosis Pathway <br> 3.000e-05","WP:WP200 (16) <br> Complement Activation, Classical Pathway <br> 6.154e-05","WP:WP1496 (41) <br> Oxidative Damage <br> 1.784e-02","WP:WP460 (20) <br> Blood Clotting Cascade <br> 1.784e-02","WP:WP3632 (61) <br> Lung fibrosis <br> 1.784e-02","WP:WP85 (185) <br> Focal Adhesion <br> 1.784e-02","WP:WP2841 (320) <br> Focal Adhesion-PI3K-Akt-mTOR-signaling pathway <br> 1.952e-02","WP:WP3625 (58) <br> TYROBP Causal Network <br> 2.460e-02"],"key":["WP:WP3626","WP:WP200","WP:WP1496","WP:WP460","WP:WP3632","WP:WP85","WP:WP2841","WP:WP3625"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,153,198,1)","opacity":0.8,"size":[14.1171589209574,12.0506262808884,14.1171589209574,12.5846608408581,14.8747209292772,16.7581625162145,17.5939721841541,14.7814518346967],"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,153,198,1)"}},"hoveron":"points","set":"SharedData9d23eecd","name":"WP","legendgroup":"WP","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[2,32.2378021430351],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.3385826771654,"color":"rgba(220,57,18,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[34.7047844505358,45.714926488911],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.3385826771654,"color":"rgba(16,150,24,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[48.1819087964117,123.397732369798],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.3385826771654,"color":"rgba(255,153,0,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[125.864714677299,127.206753052579],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.3385826771654,"color":"rgba(221,68,119,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[129.67373536008,133.857737353601],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.3385826771654,"color":"rgba(51,102,204,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[136.324719661101,146.584899077997],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.3385826771654,"color":"rgba(85,116,166,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[149.051881385497,151.422651383005],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.3385826771654,"color":"rgba(34,170,153,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[153.889633690506,155.564714677299],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.3385826771654,"color":"rgba(102,170,0,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[158.031696984799,197.049489160229],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.3385826771654,"color":"rgba(153,0,153,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[199.51647146773,200.002466982307],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.3385826771654,"color":"rgba(0,153,198,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[180],"y":[16.2],"text":"values above this threshold are capped","hovertext":"","textfont":{"size":7.55905511811024,"color":"rgba(190,190,190,1)"},"type":"scatter","mode":"text","hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0,210],"y":[16,16],"text":"","type":"scatter","mode":"lines","line":{"width":0.755905511811024,"color":"rgba(190,190,190,1)","dash":"dash"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":29.2835201328352,"r":6.6417600664176,"b":55.8781661388202,"l":61.0377750103778},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0,210],"tickmode":"array","ticktext":["GO:MF","GO:CC","GO:BP","KEGG","REAC","TF","MIRNA","CORUM","HP","WP"],"tickvals":[17.1189010715176,40.2098554697234,85.7898205831049,126.535733864939,131.76573635684,141.454809369549,150.237266384251,154.727174183902,177.540593072514,199.759469225019],"categoryorder":"array","categoryarray":["GO:MF","GO:CC","GO:BP","KEGG","REAC","TF","MIRNA","CORUM","HP","WP"],"nticks":null,"ticks":"","tickcolor":null,"ticklen":3.65296803652968,"tickwidth":0,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":10.6268161062682},"tickangle":-45,"showline":true,"linecolor":"rgba(190,190,190,1)","linewidth":0.132835201328352,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":"","hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[-1,18.5],"tickmode":"array","ticktext":["0","2","4","6","8","10","12","14",">16"],"tickvals":[0,2,4,6,8,10,12,14,16],"categoryorder":"array","categoryarray":["0","2","4","6","8","10","12","14",">16"],"nticks":null,"ticks":"outside","tickcolor":"rgba(190,190,190,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":true,"linecolor":"rgba(190,190,190,1)","linewidth":0.66417600664176,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":"","hoverformat":".2f"},"annotations":[{"text":"-log10(p-adj)","x":-0.0414121646998359,"y":0.5,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":13.2835201328352},"xref":"paper","yref":"paper","textangle":-90,"xanchor":"right","yanchor":"center","annotationType":"axis"},{"text":"query_1","x":0.5,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(169,169,169,1)","family":"","size":13.2835201328352},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"}],"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":24.9730178497302,"yanchor":1,"ysizemode":"pixel"}],"showlegend":false,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895}},"hovermode":"closest","barmode":"relative","dragmode":"zoom"},"config":{"doubleClick":"reset","showSendToCloud":false},"source":"A","attrs":{"e55c36a72592":{"colour":{},"size":{},"alpha":{},"key":{},"x":{},"y":{},"text":{},"type":"scatter"},"e55c1793b046":{"x":{},"xend":{},"y":{},"yend":{}},"e55c6119cfc1":{"x":{},"xend":{},"y":{},"yend":{}},"e55c6596b9b4":{"x":{},"xend":{},"y":{},"yend":{}},"e55ca7a0c87":{"x":{},"xend":{},"y":{},"yend":{}},"e55c52cc7870":{"x":{},"xend":{},"y":{},"yend":{}},"e55c6df72b87":{"x":{},"xend":{},"y":{},"yend":{}},"e55c7e4ae477":{"x":{},"xend":{},"y":{},"yend":{}},"e55c66dd8167":{"x":{},"xend":{},"y":{},"yend":{}},"e55c585ac5f3":{"x":{},"xend":{},"y":{},"yend":{}},"e55c2f7a01d6":{"x":{},"xend":{},"y":{},"yend":{}},"e55c730ea0f3":{"x":{},"y":{}},"e55c4568ec88":{"yintercept":{}}},"cur_data":"e55c36a72592","visdat":{"e55c36a72592":["function (y) ","x"],"e55c1793b046":["function (y) ","x"],"e55c6119cfc1":["function (y) ","x"],"e55c6596b9b4":["function (y) ","x"],"e55ca7a0c87":["function (y) ","x"],"e55c52cc7870":["function (y) ","x"],"e55c6df72b87":["function (y) ","x"],"e55c7e4ae477":["function (y) ","x"],"e55c66dd8167":["function (y) ","x"],"e55c585ac5f3":["function (y) ","x"],"e55c2f7a01d6":["function (y) ","x"],"e55c730ea0f3":["function (y) ","x"],"e55c4568ec88":["function (y) ","x"]},"highlight":{"on":"plotly_click","off":"plotly_doubleclick","persistent":false,"dynamic":false,"color":null,"selectize":false,"defaultValues":null,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0,"ctGroups":["SharedData9d23eecd"]},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
<p class="caption">
Dynamic map of the enrichment analysis ofr the selected genes. Position the cursor on a circle to see the enriched term.
</p>
</div>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb54-1" data-line-number="1"><span class="co">## Check the most significant results, formating for kable</span></a>
<a class="sourceLine" id="cb54-2" data-line-number="2">enrich_order &lt;-<span class="st"> </span><span class="kw">order</span>(gostres<span class="op">$</span>result<span class="op">$</span>p_value, <span class="dt">decreasing =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb54-3" data-line-number="3">sorted_result &lt;-<span class="st"> </span>gostres<span class="op">$</span>result[enrich_order, ]</a>
<a class="sourceLine" id="cb54-4" data-line-number="4"><span class="kw">kable</span>(<span class="kw">head</span>(sorted_result, <span class="dt">n =</span> <span class="dv">10</span>), </a>
<a class="sourceLine" id="cb54-5" data-line-number="5">      <span class="dt">digits =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">15</span>, <span class="dv">7</span>, <span class="dv">7</span>, <span class="dv">7</span>, <span class="dv">3</span>, <span class="dv">3</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">7</span>, <span class="dv">7</span>, <span class="dv">0</span>, <span class="dv">0</span>), </a>
<a class="sourceLine" id="cb54-6" data-line-number="6">      <span class="dt">caption =</span> <span class="st">&quot;Top 10 most significant enriched functional classes&quot;</span>)</a></code></pre></div>
<table style="width:100%;">
<caption>Top 10 most significant enriched functional classes</caption>
<colgroup>
<col width="1%" />
<col width="3%" />
<col width="5%" />
<col width="8%" />
<col width="4%" />
<col width="5%" />
<col width="8%" />
<col width="4%" />
<col width="3%" />
<col width="5%" />
<col width="3%" />
<col width="18%" />
<col width="10%" />
<col width="6%" />
<col width="10%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">query</th>
<th align="left">significant</th>
<th align="right">p_value</th>
<th align="right">term_size</th>
<th align="right">query_size</th>
<th align="right">intersection_size</th>
<th align="right">precision</th>
<th align="right">recall</th>
<th align="left">term_id</th>
<th align="left">source</th>
<th align="left">term_name</th>
<th align="right">effective_domain_size</th>
<th align="right">source_order</th>
<th align="left">parents</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">60</td>
<td align="left">query_1</td>
<td align="left">TRUE</td>
<td align="right">0.000000000682655</td>
<td align="right">92</td>
<td align="right">266</td>
<td align="right">20</td>
<td align="right">0.075</td>
<td align="right">0.217</td>
<td align="left">KEGG:04610</td>
<td align="left">KEGG</td>
<td align="left">Complement and coagulation cascades</td>
<td align="right">8787</td>
<td align="right">299</td>
<td align="left">KEGG:00000</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="left">query_1</td>
<td align="left">TRUE</td>
<td align="right">0.000000039452735</td>
<td align="right">272</td>
<td align="right">320</td>
<td align="right">31</td>
<td align="right">0.097</td>
<td align="right">0.114</td>
<td align="left"><a href="GO:0006955" class="uri">GO:0006955</a></td>
<td align="left"><a href="GO:BP" class="uri">GO:BP</a></td>
<td align="left">immune response</td>
<td align="right">11240</td>
<td align="right">2901</td>
<td align="left"><a href="GO:0002376" class="uri">GO:0002376</a>, <a href="GO:0050896" class="uri">GO:0050896</a></td>
</tr>
<tr class="odd">
<td align="left">2</td>
<td align="left">query_1</td>
<td align="left">TRUE</td>
<td align="right">0.000000161876615</td>
<td align="right">331</td>
<td align="right">320</td>
<td align="right">33</td>
<td align="right">0.103</td>
<td align="right">0.100</td>
<td align="left"><a href="GO:0002376" class="uri">GO:0002376</a></td>
<td align="left"><a href="GO:BP" class="uri">GO:BP</a></td>
<td align="left">immune system process</td>
<td align="right">11240</td>
<td align="right">1038</td>
<td align="left"><a href="GO:0008150" class="uri">GO:0008150</a></td>
</tr>
<tr class="even">
<td align="left">61</td>
<td align="left">query_1</td>
<td align="left">TRUE</td>
<td align="right">0.000000212705027</td>
<td align="right">116</td>
<td align="right">266</td>
<td align="right">19</td>
<td align="right">0.071</td>
<td align="right">0.164</td>
<td align="left">KEGG:05150</td>
<td align="left">KEGG</td>
<td align="left">Staphylococcus aureus infection</td>
<td align="right">8787</td>
<td align="right">416</td>
<td align="left">KEGG:00000</td>
</tr>
<tr class="odd">
<td align="left">26</td>
<td align="left">query_1</td>
<td align="left">TRUE</td>
<td align="right">0.000006518770910</td>
<td align="right">627</td>
<td align="right">204</td>
<td align="right">42</td>
<td align="right">0.206</td>
<td align="right">0.067</td>
<td align="left"><a href="GO:0071944" class="uri">GO:0071944</a></td>
<td align="left"><a href="GO:CC" class="uri">GO:CC</a></td>
<td align="left">cell periphery</td>
<td align="right">7218</td>
<td align="right">3147</td>
<td align="left"><a href="GO:0110165" class="uri">GO:0110165</a></td>
</tr>
<tr class="even">
<td align="left">27</td>
<td align="left">query_1</td>
<td align="left">TRUE</td>
<td align="right">0.000006518770910</td>
<td align="right">10</td>
<td align="right">204</td>
<td align="right">6</td>
<td align="right">0.029</td>
<td align="right">0.600</td>
<td align="left"><a href="GO:0042613" class="uri">GO:0042613</a></td>
<td align="left"><a href="GO:CC" class="uri">GO:CC</a></td>
<td align="left">MHC class II protein complex</td>
<td align="right">7218</td>
<td align="right">2067</td>
<td align="left"><a href="GO:0042611" class="uri">GO:0042611</a></td>
</tr>
<tr class="odd">
<td align="left">28</td>
<td align="left">query_1</td>
<td align="left">TRUE</td>
<td align="right">0.000009336222258</td>
<td align="right">11</td>
<td align="right">204</td>
<td align="right">6</td>
<td align="right">0.029</td>
<td align="right">0.545</td>
<td align="left"><a href="GO:0042611" class="uri">GO:0042611</a></td>
<td align="left"><a href="GO:CC" class="uri">GO:CC</a></td>
<td align="left">MHC protein complex</td>
<td align="right">7218</td>
<td align="right">2065</td>
<td align="left"><a href="GO:0098797" class="uri">GO:0098797</a></td>
</tr>
<tr class="even">
<td align="left">62</td>
<td align="left">query_1</td>
<td align="left">TRUE</td>
<td align="right">0.000017723712072</td>
<td align="right">113</td>
<td align="right">266</td>
<td align="right">16</td>
<td align="right">0.060</td>
<td align="right">0.142</td>
<td align="left">KEGG:04668</td>
<td align="left">KEGG</td>
<td align="left">TNF signaling pathway</td>
<td align="right">8787</td>
<td align="right">321</td>
<td align="left">KEGG:00000</td>
</tr>
<tr class="odd">
<td align="left">63</td>
<td align="left">query_1</td>
<td align="left">TRUE</td>
<td align="right">0.000017723712072</td>
<td align="right">86</td>
<td align="right">266</td>
<td align="right">14</td>
<td align="right">0.053</td>
<td align="right">0.163</td>
<td align="left">KEGG:05323</td>
<td align="left">KEGG</td>
<td align="left">Rheumatoid arthritis</td>
<td align="right">8787</td>
<td align="right">460</td>
<td align="left">KEGG:00000</td>
</tr>
<tr class="even">
<td align="left">125</td>
<td align="left">query_1</td>
<td align="left">TRUE</td>
<td align="right">0.000029996303079</td>
<td align="right">41</td>
<td align="right">174</td>
<td align="right">11</td>
<td align="right">0.063</td>
<td align="right">0.268</td>
<td align="left">WP:WP3626</td>
<td align="left">WP</td>
<td align="left">Microglia Pathogen Phagocytosis Pathway</td>
<td align="right">4526</td>
<td align="right">46</td>
<td align="left">WP:000000</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb55-1" data-line-number="1"><span class="kw">names</span>(gostres<span class="op">$</span>meta)</a></code></pre></div>
<pre><code>[1] &quot;query_metadata&quot;  &quot;result_metadata&quot; &quot;genes_metadata&quot;  &quot;timestamp&quot;       &quot;version&quot;        </code></pre>
<p><strong>Challenge falcultatif:</strong></p>
<p>Effectuez une analyse d’enrichissement sur chacun des clusters obtenus à partir de l’arbre des gènes.</p>
</div>
<div id="conclusions-générales" class="section level2">
<h2>Conclusions générales</h2>
<p>Résumez en quelques phrases vos conclusions à partir des résultats obtenus.</p>
</div>
<div id="session-info" class="section level2">
<h2>Session info</h2>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb57-1" data-line-number="1"><span class="co">#### Session info ####</span></a>
<a class="sourceLine" id="cb57-2" data-line-number="2"><span class="kw">sessionInfo</span>()</a></code></pre></div>
<pre><code>R version 4.0.2 (2020-06-22)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] biomaRt_2.44.4   pheatmap_1.0.12  gprofiler2_0.2.0 factoextra_1.0.7 ggplot2_3.3.3    FactoMineR_2.4   knitr_1.32      

loaded via a namespace (and not attached):
  [1] bitops_1.0-6         bit64_4.0.5          RColorBrewer_1.1-2   progress_1.2.2       httr_1.4.2           tools_4.0.2          backports_1.2.1      bslib_0.2.4          utf8_1.2.1           R6_2.5.0             DT_0.18              DBI_1.1.1            lazyeval_0.2.2       BiocGenerics_0.34.0 
 [15] colorspace_2.0-0     withr_2.4.2          tidyselect_1.1.0     prettyunits_1.1.1    bit_4.0.4            curl_4.3             compiler_4.0.2       Biobase_2.48.0       flashClust_1.01-2    xml2_1.3.2           plotly_4.9.3         labeling_0.4.2       sass_0.3.1           scales_1.1.1        
 [29] askpass_1.1          rappdirs_0.3.3       stringr_1.4.0        digest_0.6.27        foreign_0.8-81       rmarkdown_2.7        rio_0.5.26           pkgconfig_2.0.3      htmltools_0.5.1.1    dbplyr_2.1.1         fastmap_1.1.0        highr_0.9            readxl_1.3.1         htmlwidgets_1.5.3   
 [43] rlang_0.4.10         RSQLite_2.2.6        shiny_1.6.0          farver_2.1.0         jquerylib_0.1.3      generics_0.1.0       jsonlite_1.7.2       crosstalk_1.1.1      zip_2.1.1            dplyr_1.0.5          car_3.0-10           RCurl_1.98-1.3       magrittr_2.0.1       leaps_3.1           
 [57] Rcpp_1.0.6           munsell_0.5.0        S4Vectors_0.26.1     fansi_0.4.2          abind_1.4-5          lifecycle_1.0.0      scatterplot3d_0.3-41 stringi_1.5.3        yaml_2.2.1           carData_3.0-4        MASS_7.3-53.1        BiocFileCache_1.12.1 grid_4.0.2           blob_1.2.1          
 [71] promises_1.2.0.1     parallel_4.0.2       ggrepel_0.9.1        forcats_0.5.1        crayon_1.4.1         lattice_0.20-41      haven_2.4.0          hms_1.0.0            pillar_1.6.0         ggpubr_0.4.0         ggsignif_0.6.1       stats4_4.0.2         XML_3.99-0.6         glue_1.4.2          
 [85] evaluate_0.14        data.table_1.14.0    httpuv_1.5.5         vctrs_0.3.7          cellranger_1.1.0     gtable_0.3.0         openssl_1.4.3        purrr_0.3.4          tidyr_1.1.3          assertthat_0.2.1     cachem_1.0.4         openxlsx_4.2.3       xfun_0.22            mime_0.10           
 [99] xtable_1.8-4         broom_0.7.6          later_1.1.0.1        rstatix_0.7.0        viridisLite_0.4.0    tibble_3.1.1         AnnotationDbi_1.50.3 memoise_2.0.0        IRanges_2.22.2       cluster_2.1.2        ellipsis_0.3.1      </code></pre>
</div>

<div id="rmd-source-code">---
title: "Mini-projet 2021 - Exploration des données de Pavkovic"
author: "Claire Vandiedonck, Jacques van Helden, Anne Badel & Olivier Sand"
date: '`r Sys.Date()`'
output:
  html_document:
    self_contained: no
    code_download: true
    fig_caption: yes
    highlight: zenburn
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
    code_folding: "hide"
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
editor_options: 
  chunk_output_type: console
---


```{r settings, include=FALSE, echo=FALSE, eval=TRUE}
options(width = 300)
# options(encoding = 'UTF-8')
knitr::opts_chunk$set(
  fig.width = 7, fig.height = 5, 
  fig.path = 'figures/mini-projet_',
  fig.align = "center", 
  size = "tiny", 
  echo = TRUE, 
  eval = TRUE, 
  warning = FALSE, 
  message = FALSE, 
  results = TRUE, 
  comment = "")

options(scipen = 12) ## Max number of digits for non-scientific notation
# knitr::asis_output("\\footnotesize")
```

```{r libraries, echo=FALSE, eval=TRUE}
#### Required libraries ####

# Load required CRAN R libraries
required_cranLib <- c("knitr", 
                      "FactoMineR", 
                      "factoextra", 
                      "gprofiler2",
                      "pheatmap")
for (lib in required_cranLib) {
  if (!require(lib, character.only = TRUE)) {
    install.packages(lib)
  }
  require(lib, character.only = TRUE)
}

required_biocLib <- c("biomaRt")
for (lib in required_biocLib) {
  if (!require(lib, character.only = TRUE)) {
    BiocManager::install(lib)
  }
  require(lib, character.only = TRUE)
}

kable(as.data.frame(c(required_cranLib, required_biocLib)),
      col.names = "libraries",
      caption = "Loaded required libraries"
    )

```

## Synopsis du projet

### Travail demandé

Le but de ce travail est de mettre en oeuvre les méthodes vues dans le module 3 "R et statistiques" pour explorer le jeu de données de Pavkovic, et de rendre un rapport d'analyse au format `.Rmd`. 

Nous fournissons ci-dessous une trame avec les principales sections attendues. Certaines contiennent déjà du code. Vous devrez en compléter d'autres. Sentez-vous libres d'adapter cette trame ou d'y ajouter des analyses complémentaires si elles vous aident à interpréter vos résultats. 

### Remise du rapport

Date: **le 22 mai 2021 à minuit**.  Si vous anticipez un problème pour remettre le rapport à cette date contactez-nous aussi rapidement que possible pour que nous puissions prévoir une remise plus tardive. 

- Commencez par renommer le fichier .Rmd en remplaçant Prenom-NOM par vos nom et prénom. 
- Le rapport est attendu en formats .Rmd + .HTML (en gardant l'option self_contained de l'en-tête activée). 
- Déposez les fichiers dans un sous-dossier de vote compte du cluster. Attention, veillez à respecter précisément cette structure de chemin car nous nous baserons dessus pour récupérer vos résultats. 

    `/shared/projects/dubii2021/[login]/m3-stat-R/mini-projet` 

### Critères d'évaluation

- Reproductibilité des analyses: nous tenterons de regénérer le rapport HTML à partir de votre Rmd, en partant de notre compte sur le serveur IFB. 
- Manipulation des objets R
- Mobilisation des méthodes statistiques vues au cours
- Pertinence des interprétations statistiques
- Pertinence des interprétations biologiques
- Clarté de la rédaction
- Clarté des illustrations (figures et tableaux): graphismes, légendes ...

Nous vous encourageons à assurer la lisibilité de votre code (syntaxe, nommage des variables, commentaires de code)

### Objectifs scientifiques

Nous partons du même jeu de données *Fil Rouge* de ce module issues de la publication Pavkovic, M., Pantano, L., Gerlach, C.V. et al. Multi omics analysis of fibrotic kidneys in two mouse models. Sci Data 6, 92 (2019). https://doi.org/10.1038/s41597-019-0095-5

**Rappel sur les échantillons:**

Deux modèles de fibrose rénale chez la souris sont étudiés:

1. Le premier est un modèle de néphropathie réversible induite par l'acide folique (folic acid (FA)). Les souris ont été sacrifiées avant le traitement (normal), puis à jour 1, 2, 7 et 14 (day1,...) après une seule injection d'acide folique.

2. Le second est un modèle irréversible induit chrirurgicalement (unilateral ureteral obstruction (UUO)). les souris ont été sacrifiées avant obstruction (day 0) et à 3, 7 et 14 jours après obstruction par ligation de l'uretère du rein gauche.

A partir de ces extraits de rein, l'ARN messager total et les petits ARNs ont été séquencés et les protéines caratérisées par spectrométrie de masse en tandem (TMT).

**But scientifique:** Dans le tutoriel sur les dataframes, vous avez travaillé sur les données de ***transcriptome du modèle UUO***. Dans ce mini-projet, vous allez travailler sur les données du ***transcriptome du modèle FA*** afin de regrouper les observations (échantillon) et les gènes selon des profils d'expression similaires.

**Votre projet se décompose en 5 parties dont 3 seront à réaliser par:**

1. Statitiques descriptives des données brutes: commandes fournies
2. Normalisation des données : commandes fournies
3. Statistiques descriptives des données normalisées: à vous de jouer
4. Analyse de regroupement des données: à vous de jouer
5. Analyse d'enrichissement fonctionnel: à vous de jouer

## 1. Les données brutes

***Vous n'avez rien à coder ici. Le code est fourni.***

### Chargement des données brutes

Le bloc suivant contient une fonction qui permet de télécharger un fichier dans l'espace de travail, sauf s'il est déjà présent. Nous l'utiliserons ensuite pour télécharger les données à analyser en évitant de refaire le transfert à chaque exécution de l'analyse. 

```{r function_download_only_once}
#' @title Download a file only if it is not yet here
#' @author Jacques van Helden email{Jacques.van-Helden@@france-bioinformatique.fr}
#' @param url_base base of the URL, that will be prepended to the file name
#' @param file_name name of the file (should not contain any path)
#' @param local_folder path of a local folder where the file should be stored
#' @return the function returns the path of the local file, built from local_folder and file_name
#' @export©
download_only_once <- function(
  url_base, 
  file_name,
  local_folder) {

  ## Define the source URL  
  url <- file.path(url_base, file_name)
  message("Source URL\n\t",  url)

  ## Define the local file
  local_file <- file.path(local_folder, file_name)
  
  ## Create the local data folder if it does not exist
  dir.create(local_folder, showWarnings = FALSE, recursive = TRUE)
  
  ## Download the file ONLY if it is not already there
  if (!file.exists(local_file)) {
    message("Downloading file from source URL to local file\n\t", 
            local_file)
    download.file(url = url, destfile = local_file)
  } else {
    message("Local file already exists, no need to download\n\t", 
            local_file)
  }
  
  return(local_file)
}
```

Nous téléchargeons deux fichiers dans un dossier local `~/m3-stat-R/pavkovic_analysis` **(vous pouvez changer le nom ou chemin dans le chunk ci-dessous)**, et les chargeons dans les data.frames suivants: 

- Données brutes de transcriptome: `fa_expr_raw`
- Métadonnées: `fa_meta`

```{r download_and_load}
## Define the remote URL and local folder
pavkovic_url <- "https://github.com/DU-Bii/module-3-Stat-R/raw/master/stat-R_2021/data/pavkovic_2019/"

## Define the local folder for this analysis (where the data will be downloaded and the results generated)
pavkovic_folder <- "~/m3-stat-R/pavkovic_analysis"

## Define a sub-folder for the data
pavkovic_data_folder <- file.path(pavkovic_folder, "data")

## Download and load the expression data table
## Note: we use check.names=FALSE to avoid replacing hyphens by dots
## in sample names, because we want to keep them as in the 
## original data files. 
message("Downloading FA transcriptome file\t", "fa_raw_counts.tsv.gz",
  "\n\tfrom\t", pavkovic_url)
fa_expr_file <- download_only_once(
  url_base = pavkovic_url, 
  file_name = "fa_raw_counts.tsv.gz",
  local_folder = pavkovic_data_folder)

## Load the expresdsion table
message("Loading FA transcriptome data from\n\t", fa_expr_file)
fa_expr_raw <- read.delim(file = fa_expr_file, 
                       header = TRUE, 
                       row.names = 1)

## Download the metadata file
message("Downloading FA metadata file\t", "fa_transcriptome_metadata.tsv",
  "\n\tfrom\t", pavkovic_url)
fa_meta_file <- download_only_once(
  url_base = pavkovic_url, 
  file_name = "fa_transcriptome_metadata.tsv",
  local_folder = pavkovic_data_folder)

## Load the metadata
message("Loading FA metadata from\n\t", fa_meta_file)
fa_meta <- read.delim(file = fa_meta_file, 
                       header = TRUE, 
                       row.names = 1)
```

Nous regardons la structure de chaque dataframe.

```{r insepct data}
str(fa_expr_raw)
str(fa_meta)
```

Les deux fichiers ne donnent pas les observations de l'échantillon dans le même ordre:

```{r check data order}
fa_meta$sampleName == names(fa_expr_raw)
```

Nous les réorganisons les échantillons dans l'ordre de l'expérience: condition normale, puis day 1 à 14 avec les 3 réplicats.

```{r reoder data}
sample_order <- c(paste(rep(c("normal", "day1", "day2", "day3", "day7", "day14"), each = 3),
                        1:3, sep = "_"))

fa_expr_raw <- fa_expr_raw[,sample_order]
fa_meta <- fa_meta[match(sample_order, fa_meta$sampleName),]

# View(fa_meta)
kable(fa_meta, caption = "Metdata for Pavkovoc FA transcriptome")
```

=> Ainsi, nous avons un jeu de données avec un échantillon de `r nrow(fa_meta)` observations et des données d'expression de `r nrow(fa_expr_raw)` gènes.


### Statistiques descriptives

Dans le tutorial sur les dataframes sur le jeu de données "uuo" (relisez le corrigé), nous vous avons demandé de créer un data.frame qui collecte les statistiques par gène et par échantillon. Nous vous demandons de réaliser une étude similaire sur les données "FA" avant et après normalisation inter-échantillons des données. Le code de la partie avant normalisation est donné.

#### Par échantillon avant normalisation

Nous créons un data.frame nommé `sample_stat_prenorm` qui comporte une ligne par échantillon et une colonne par statistique. Nous calculons les statistiques suivantes sur les valeurs log2 d'expression de chaque échantillon:

- moyenne
- écart-type
- intervalle inter-quartiles
- premier quartile
- médiane
- troisième quartile
- maximum
- nombre de valeurs nulles

Il est affiché avec la fonction `kable()`. 

```{r sample_stat_pre_norm}
message("Computing sample-wise statistics on raw counts")
sample_stat_prenorm <- data.frame(
  mean = apply(fa_expr_raw, 2, mean, na.rm = TRUE),
  sd = apply(fa_expr_raw, 2, sd, na.rm = TRUE),
  iqr = apply(fa_expr_raw, 2, IQR, na.rm = TRUE),
  Q1 = apply(fa_expr_raw, 2, quantile, p = 0.25, na.rm = TRUE),
  median = apply(fa_expr_raw, 2, median, na.rm = TRUE),
  Q3 = apply(fa_expr_raw, 2, quantile, p = 0.75, na.rm = TRUE),
  max = apply(fa_expr_raw, 2, max, na.rm = TRUE),
  null = apply(fa_expr_raw == 0, 2, sum, na.rm = TRUE)
)

kable(sample_stat_prenorm, caption = "Sample-wise statistics before normalisation.")
```

#### Par gène avant normalisation

Nous créons ci-dessous un data.frame nommé `gene_stat_prenorm` qui comporte une ligne par gène et une colonne par statistique. Nous calculons les statistiques suivantes sur les valeurs log2 de chaque gène.

- moyenne
- médiane
- écart-type
- premier quartile
- troisième quartile
- maximum
- nombre de valeurs nulles
- intervalle inter-quartiles

Ces résultats sont stockés dans un data.frame avec 1 ligne par échantillon et 1 colonne par statistique. Nous affichons les lignes 100 à 109 de ce tableau de statistiques avec la fonction `kable()`.

```{r gene_stat_pre_norm}
## Gene-wise statistics for the raw counts (will be used for normalisation)
message("Computing gene-wise statistics on raw counts")
gene_stat_prenorm <- data.frame(
  mean = apply(fa_expr_raw, 1, mean, na.rm = TRUE),
  sd = apply(fa_expr_raw, 1, sd, na.rm = TRUE),
  iqr = apply(fa_expr_raw, 1, IQR, na.rm = TRUE),
  Q1 = apply(fa_expr_raw, 1, quantile, p = 0.25, na.rm = TRUE),
  median = apply(fa_expr_raw, 1, median, na.rm = TRUE),
  Q3 = apply(fa_expr_raw, 1, quantile, p = 0.75, na.rm = TRUE),
  max = apply(fa_expr_raw, 1, max, na.rm = TRUE),
  null = apply(fa_expr_raw == 0, 1, sum, na.rm = TRUE)
)

kable(gene_stat_prenorm[100:109, ], caption = "Gene-wise statistics before normalisation")
```

## 2. Filtrage et normalisation des données

***Vous n'avez rien à coder ici. Le code est fourni.***

Il existe plusieurs façons de normaliser les données de transcriptome  vues dans les modules 4 et 5 (cf. total counts, quantiles, TMM, RLE, limma voom,...), mais nous avons choisi ici une solution simple tout en étant robuste pour normaliser les données en standardisant le 3ème quantile. 

La méthode choisie ici consiste à :

1. **Ecarter les gènes "non-détectés"**, c'est-à-dire ceux ayant des valeurs nulles dans au moins 90% des échantillons.

2. **Ecarter les gènes à peine exprimés**, c'est-à-dire ceux ayant une valeur moyenne < 10 (arbitrairement).

3. **Standardiser les échantillons **sur le 3ème quartile des gènes restants: on divise les comptages bruts par le 3ème quartile de l'échantillon et on multiplie par le 3ème quartile de l'ensemble des échantillons.

4. **Normaliser les comptages** (au sens propre, c'est-à-dire rapprocher leur distribution de la distribution gaussienne) par une transformation logarithmique (log2). 

Nous fournissons ci-dessous le code.

### Filtrage : élimination des gènes non détectés ou à peine exprimés

```{r gene_filtering}
## Data filtering: genes having at least 90% null values
message("Filtering undetected genes")
undetected_genes <- gene_stat_prenorm$null >= ncol(fa_expr_raw) * 0.9
print(paste0("Undetected genes (null in >= 90% samples): ", sum(undetected_genes)))

## Data filtering: genes having a mean expression < 10
message("Filtering barely expressed genes")
barely_expressed_genes <- gene_stat_prenorm$mean < 10
print(paste0("Barely expressed genes (mean < 10): ", sum(barely_expressed_genes)))

## Apply filtering on both criteria
discarded_genes <- undetected_genes | barely_expressed_genes
print(paste0("Discarded genes: ", sum(discarded_genes)))
kept_genes <- !discarded_genes
print(paste0("Kept genes: ", sum(kept_genes)))

## Genes after filtering
fa_expr_filtered <- fa_expr_raw[kept_genes, ]

```





### Standardisation entre échantillons

Nous appliquons ici une méthode simple mais efficace de standardisation

```{r normalisation_q3}
#### Inter-sample standardisation on the Q3 of raw counts ####
total_q3 <- quantile(unlist(fa_expr_filtered), probs = 0.75)
sample_stat_prenorm$Q3_filterd <- apply(fa_expr_filtered, 2, quantile, probs = 0.75)
sample_stat_prenorm$scale_factor <- 1 / sample_stat_prenorm$Q3_filterd * total_q3

## Apply standardisation
fa_expr_standard <- t(t(fa_expr_filtered) * unlist(sample_stat_prenorm$scale_factor))
## Check 3rd quantile after standardisation
kable(apply(fa_expr_standard, 2, quantile, probs = 0.75), 
      col.names = "Q3_standardised", 
      caption = "Third quartile of the filtered counts after inter-sample standardisation of the third quartiles. ")
# boxplot(fa_expr_standard, horizontal = TRUE)
```




### Transformation log2

Nous appliquons une transformation en log2 des données brutes, après avoir ajouté un epsilon $\epsilon = 1$ (les valeurs nulles seront donc représentées par un log2(counts) valant $0$. Nous stockons le résultat dans un data.frame nommé `fa_expr_log2`.

Nous affichons un fragment des tableaux `fa_expr_raw` et `fa_expr_log2` en sélectionnant les lignes 100 à 109 et les colonnes 5 à 10, afin de nous assurer que la transformation en log2 a bien fonctionné. 

```{r log2_transform}
## Log2 transformation of the transcriptome data
epsilon <- 1
fa_expr_log2 <- log2(fa_expr_standard + epsilon)
# dim(fa_expr_log2)
# View(head(fa_expr_log2))

## Display of a fragment of the data before and after log2 transformation
kable(fa_expr_raw[100:109, 5:10], caption = "Fragment des données transcriptomiques brutes")
kable(fa_expr_log2[100:109, 5:10], caption = "Fragment des données transcriptomiques après transformation log2")
```

## 3. Statistiques descriptives sur les données normalisées

***A vous de jouer!***

### Statistiques par gène après normalisation

Générez un data.frame nommée `gene_stat_norm` avec une ligne par gène à partir du tableau de données normalisées, avec les statistiques suivantes (une statistique par colonne):

- moyenne
- variance
- écart-type
- coefficient de variation (écart-type divisé par la moyenne)
- intervalle inter-quartiles
- minimum
- médiane
- maximum

```{r gene_stat_post_norm}
## Gene-wise statistics after normalisation
message("Computing gene-wise statistics on log2-transformed and normalised counts")
gene_stat_norm <- data.frame(
  mean = apply(fa_expr_log2, 1, mean, na.rm = F),
  var = apply(fa_expr_log2, 1, var, na.rm = TRUE),
  sd = apply(fa_expr_log2, 1, sd, na.rm = F),
  CV = NA,
  min = apply(fa_expr_log2, 1, min, na.rm = TRUE),
  Q1 = apply(fa_expr_log2, 1, quantile, p = 0.25, na.rm = TRUE),
  median = apply(fa_expr_log2, 1, median, na.rm = TRUE),
  Q3 = apply(fa_expr_log2, 1, quantile, p = 0.75, na.rm = TRUE),
  max = apply(fa_expr_log2, 1, max, na.rm = TRUE),
  null = apply(fa_expr_log2 == 0, 1, sum, na.rm = TRUE)
  )
gene_stat_norm$CV <- gene_stat_norm$sd / gene_stat_norm$mean
```

### Annotation des gènes

Chaque gène étant donné par son identifiant dans la base de données ENSEMBL vous utiliserez le `paquet biomaRt de bioconductor` pour ajouter des annotations : symbole, chromosome, coordonnées génomiques, brin. 
Suivez pas à pas la méthode proposée (***certaines étapes peuvent prendre quelques minutes***):

 - chargez le paquet biomaRt, voire installer-le uniquement si nécessaire. Indiquez le code à l'emplacement adéquat dans de .Rmd.

 - sélectionnez la base de données ENSEMBL avec la fonction `useMart()`. Attention à choisir le bon génome avec l'agument `dataset`: `mmusculus_gene_ensembl`
 
 - avec la fonction `getBM()` récupérez de la base de données ENSEMBL les champs demandés (***pour symbole utilisez external_gene_name***) en appliquant "ensembl_geneid" pour l'agument `filter` et en indiquant pour l'argument `values` le vecteur des identifiants des gènes présents dans le dataframe `gene_stat_norm`. Vous obtenez un dataframe.
 
A présent, ajoutez au dataframe `gene_stat_norm` en 1ères colonnes les annotations retrouvées grâce à biomaRt. Attention, certains gènes ne sont pas retrouvés dans la version d'ENSEMBL sur biomaRt donc laissez des NA comme données manquantes dans ce cas. Nous vous recommandons d'utiliser la function `merge()` de R base ou bien `left_join()` de `dplyr` pour fusionner les deux dataframes en un seul.
 
```{r gene_annotations, eval=FALSE}
### Gene annotations ####
message("Getting gene annotations")

## Open a connection to Ensembl MART
message("Opening connection to Ensembl MART")
ensembl <- useMart("ENSEMBL_MART_ENSEMBL", 
                   host = "www.ensembl.org", 
                   dataset = "mmusculus_gene_ensembl")

## Get gene annotations
message("Getting gene annotations")
genes <- getBM(attributes = c("ensembl_gene_id", 
                              "external_gene_name", 
                              "chromosome_name",
                              "start_position", 
                              "end_position", 
                              "strand"), 
               filter = "ensembl_gene_id",
               values = row.names(gene_stat_norm),
               mart = ensembl)
  
## Merge gene annotations and expression statistics
gene_stat_norm <- merge(genes, gene_stat_norm, by.x = "ensembl_gene_id", by.y = 0, sort = FALSE)

kable(gene_stat_norm[100:109, ], caption = "Gene-wise statistics after normalisation")
```

**Challenge falcultatif:**

Réordonnez les gènes par position génomique et affichez les lignes 5 premières et  5 dernières lignes de ce tableau de statistiques. 

```{r sort_genes_by_chrom, eval=FALSE}
#### Sorting genes by chromosome ####
message("Sorting genes by chromosome")


gene_stat_norm <- gene_stat_norm[order(gene_stat_norm$chromosome_name, gene_stat_norm$start_position),]

kable(gene_stat_norm[c(1:5, (nrow(gene_stat_norm) - 4):nrow(gene_stat_norm)), ], 
      caption = "Gene-wise statistics after normalisation.")
```


### Distribution des données

- Dessinez sous forme d'un histogramme la distribution des valeurs après normalisation (tous échantillons confondus)

```{r fa_expr_norm_distrib, fig.width=8, fig.height=5, out.width="70%", fig.cap="Distribution of expression values (log2 counts) after gene filtering and standardisation on the sample-wise third-quartile of non-null values. The vertical line highlights the mean value. "}
hist(unlist(fa_expr_log2), 
     breaks = seq(from = 0, to = max(fa_expr_log2) + 1, by = 0.25),
     xlab = "log2(counts) after standardisation", 
     ylab = "number of genes after filtering",
     col = "#BBDDFF",
     las = 1, cex.axis = 0.8,
     main = "distribution after standardisation")
abline(v = mean(fa_expr_log2), col = "darkgreen", lwd = 2)
```

- Dessinez un box plot par échantillon avant et après normalisation, et commentez la façon dont l'effet de la normalisation apparaît sur ces graphiques. 

```{r boxplots_normalisation_impact, fig.width=10, fig.height=12, out.width="100%", fig.cap="Box plots showing the impact of normalisation"}
#### Box plots to show normalisation impact ####
par(mar = c(4,6,4,1)) ## Set the margins
par(mfrow = c(2,2))
boxplot(fa_expr_raw, 
        horizontal = TRUE,
        xlab = "counts", 
        las = 1, 
        col = fa_meta$color, 
        main = "Raw counts, all genes")
boxplot(fa_expr_filtered, 
        horizontal = TRUE, 
        xlab = "counts", 
        las = 1, 
        col = fa_meta$color, 
        main = "Raw counts, filtered genes")
boxplot(fa_expr_standard, 
        horizontal = TRUE, 
        xlab = "counts", 
        las = 1, 
        col = fa_meta$color, 
        main = "Standardised counts (sample Q3)")
boxplot(fa_expr_log2, 
        xlab = "log2(counts)", 
        las = 1, 
        horizontal = TRUE, 
        col = fa_meta$color, 
        main = "Normalised counts")
par(mfrow = c(1, 1))
par(mar = c(4,5,5,1))
```

## 4. Analyse de regroupement des données

***A vous de jouer!***

### Sélection de gènes d'expression élevée et variable

Pour réduire le nombre de gènes, nous allons écarter les gènes faiblement exprimés (log2 moyen inférieur à 4), et ne retenir que ceux qui montrent des variations importantes entre échantillons. Pour ce dernier critère, nous nous basons sur la variance. 

<!-- Sélectionnez les gènes ayant un niveau log2 moyen minimal supérieur à 3 ($m > 3$) et un coefficient de variation supérieur à 0.5 ($CV > 0.5$). Note: ces valeurs sont parfaitement arbitraires, elles ont été choisies pour obtenir un nombre raisonnable de gènes.  -->

Sélectionnez les gènes ayant un niveau log2 moyen minimal supérieur à 5 ($m > 5$) et une variance supérieure à 2 ($s^2 > 2$). Note: ces valeurs sont parfaitement arbitraires, elles ont été choisies pour obtenir un nombre raisonnable de gènes. 

```{r gene_selection}
#### Selection of a subset of genes with igh expression and variance ####
message("Selecting genes with high expression and variance")

## Compute a Boolean vector indicating whether each gene passes or not the expression level threshold
high_expression <- gene_stat_norm$mean > 5
# table(high_expression) # count number of genes with high/weak expression

## Compute a Boolean vector indicating whether each gene passes or not the variance threshold
high_variance <- gene_stat_norm$var > 2
# table(high_variance) # count number of genes with weak high variance

## Select genes having both a high mean expression and a high variance
selected_genes <- high_variance & high_expression
# table(selected_genes) # count number of genes with weak high coeffficient of variation
print(paste0("Selected genes: ", sum(selected_genes)))

## Create a data frame with the expression of the selected genes
fa_expr_selected <- fa_expr_log2[selected_genes, ]
```

Dessinez des histogrammes des valeurs d'expression avant et après cette sélection de gènes, et commentez les différences. 

```{r hist_expr_selected_genes, fig.width=8, fig.height=6, out.width="100%", fig.cap="Distribution of expression values before and after gene selection"}
#### Histograms of expression before and after gene selection ####
par(mfrow = c(2,1))
hist(unlist(fa_expr_log2), 
     breaks = seq(from = 0, to = max(fa_expr_log2) + 1, by = 0.25),
     las = 1, 
     cex.axis = 0.8, 
     main = "Standardized values before gene selection",
     col =  "#DDBBFF")

hist(unlist(fa_expr_selected), 
     breaks = seq(from = 0, to = max(fa_expr_log2) + 1, by = 0.25),
     las = 1, 
     cex.axis = 0.8, 
     main = "Standardized values after gene selection",
     col =  "#FFDDBB")

par(mfrow = c(1,1))

```

Dessinez un box plot par échantillon des valeurs d'expression avant et après sélection des gènes, et commentez le résultat. 

```{r boxplots_expr_selected_genes, fig.width=10, fig.height=5, out.width="60%", fig.cap="Box plots of standardised expression values before and after gene selection. "}
#### Boxplots of expression before and after gene selection ####

par(mfrow = c(1,2))

boxplot(fa_expr_log2, 
        horizontal = TRUE,
        xlab = "log2(counts)", 
        las = 1, 
        col = fa_meta$color, 
        main = "Before gene selection")
boxplot(fa_expr_selected, 
        horizontal = TRUE,
        xlab = "log2(counts)", 
        las = 1, 
        col = fa_meta$color, 
        main = "After gene selection")

par(mfrow = c(1,1))
```


### ACP

Dessinez un plot ACP des échantillons en les colorant par condition avant et après normalisation.

- avec les comptages bruts de la matrice d'expression initiale ($fa_expr$)

```{r acp_raw_all_genes, fig.width=8, fig.height=8, out.width="60%", fig.cap="PC plot of the samples from the raw expression values of all genes. "}
#### PCA from raw counts, all genes ####
message("Running PCA with raw counts, all genes")

## Raw expression values, all genes
ma_pca_raw_tt <- PCA(t(fa_expr_raw), 
                      scale.unit = FALSE, 
                      graph = FALSE)
# plot(ma_pca_raw_tt, choix = "ind")
fviz_pca_ind(ma_pca_raw_tt, col.ind = fa_meta[,"condition"])
```


- avec la matrice de valeurs normalisées des gènes filtrés

```{r acp_norm_filtered_genes, fig.width=8, fig.height=8, out.width="60%", fig.cap="PC plot of the samples from normalised counts, before gene selection. "}
#### PCA from normalised counts, all genes ####
message("Running PCA with normalised counts, all genes")

ma_pca_filtered <- PCA(t(fa_expr_log2), scale.unit = FALSE,
                  graph = FALSE)
# plot(ma_pca_filtered, choix = "var")
# plot(ma_pca_sel, choix = "ind")
fviz_pca_ind(ma_pca_filtered, col.ind = fa_meta[, "condition"])
```

- avec la matrice finale (transformation log2, filtre des gènes non-détectés, standardisation et sélection des gènes fortement exprimés et à haut coefficient de variation)

```{r acp_norm_selected_genes, fig.width=8, fig.height=8, out.width="60%", fig.cap="PC plot of the samples from normalised values, after gene selection. "}
#### PCA from normalised counts, selected genes ####
message("Running PCA with normalised counts, selected genes")

ma_pca_sel <- PCA(t(fa_expr_selected), scale.unit = FALSE,
                  graph = FALSE)
# plot(ma_pca_sel, choix = "var")
# plot(ma_pca_sel, choix = "ind")
fviz_pca_ind(ma_pca_sel, col.ind = fa_meta[, "condition"])
```

### Clustering

- Calculez les matrices de distance entre échantillons, en utilisant respectivement les distances euclidienne (`dist()`), coefficient de Pearson (`cor(, method = "pearson")`)  et de Spearman (`cor(, method = "spearman")`).


```{r sample_distances}
#### Sample distances ####
message("Computing inter-sample distances")

dist_euc_samples <- dist(t(fa_expr_selected))
cor_pearson_samples <- as.dist(1 - cor(fa_expr_selected))
cor_spearman_samples <- as.dist(1 - cor(fa_expr_selected, method = "spearman"))
```

- Effectuez un clustering hiérarchique des échantillons, en utilisant le critère "complete" pour l'agglomération. Comparez les arbres d'échantillons obtenus avec ces trois métriques et choisissez celle qui vous paraît la plus pertinente.

```{r sample_clustering, fig.width=12, plot.height=5, out.width="100%", fig.cap="Sample tree with three alternative distance metrics: Euclidiant distance (left), Pearson correlation (center), Spearman correlation (right).? "}
#### Sample clustering ####
message("Sample clustering")

par(mfrow = c(1,3))
plot(hclust(dist_euc_samples), hang = -1,
     main = "euclidean distance")
plot(hclust(cor_pearson_samples), hang = -1,
     main = "pearson")
plot(hclust(cor_spearman_samples), hang = -1,
     main = "spearman")
par(mfrow = c(1,1))
```

- Effectuez un clustering hiérarchique des gènes en utilisant la distance basée sur le coefficient de Pearson et la règle d'agglomération complète

```{r gene_tree}
#### Gene tree with Pearson correlation ####
message("Drawing gene tree")


clustering_method <- "complete" ## Choose a clustering method
cor_pearson_genes <-  as.dist(1 - cor(t(fa_expr_selected)))

## Build a gene tree
gene_tree_pearson <- hclust(cor_pearson_genes, method = clustering_method)

## Plot the gene tree
plot(gene_tree_pearson, hang = -1, las = 1, 
     main = paste0("Gene tree; Pearson coefficient; ", clustering_method, " clustering"), 
     labels = FALSE)
```

- Dessinez un arbre avec le résultat du clustering des gènes et commentez sa structure. Si vous deviez choisir de façon arbitraire un nombre de clusters, que choisiriez-vous ? Pourquoi ? Pas de panique, nous pouvons assumer ici que la réponse comporte une part de subjectivité. 

```{r gene_tree_with_boxes}
#### Plot the gene tree with boxes to denote the clusters ####
message("Plotting gene tree")

plot(gene_tree_pearson, hang = -1,
     main = paste0("Gene tree with 7 clusters; Pearson coefficient; ", clustering_method, " clustering"), 
     labels = FALSE)
rect.hclust(gene_tree_pearson, k = 7)
```

- Dessinez une heatmap du résultat, en sélectionnant les deux résultats de clustering ci-dessus pour les gènes et les échantillons. 

```{r heatmap_biclustering, fig.width=10, fig.height=8, out.width="90%", fig.cap="Heatmap showing the biclustering of genes and samples. "}
#### Heatmap with biclustering ####
message("heatmap with biclustering")

pheatmap(t(fa_expr_selected), 
         main = paste0("Biclustering; ", "Pearson correlation; ", clustering_method, " clustering"),
         show_colnames = FALSE, scale = "none",
         clustering_method = clustering_method, 
         clustering_distance_cols = "correlation", 
         clustering_distance_rows = "correlation")
```

- Dessinez une heatmap du résultat, en affichant un arbre sur les gènes mais pas sur les échantillons

```{r biclustering}
#### Heatmap with clustering on genes only ####
message("Heatmap with gene clustering")

pheatmap(t(fa_expr_selected), 
         main = paste0("Gene clustering; ", "Pearson correlation; ", clustering_method, " clustering"),
         show_colnames = FALSE, scale = "none",
         clustering_method = clustering_method, 
         clustering_distance_cols = "correlation", 
         cluster_rows = FALSE)
```


Interprétez les résultats en quelques phrases. 


## 5. Enrichissement fonctionnel

***A vous de jouer!***

Effectuez une analyse d'enrichissement fonctionnel avec les principaux clusters obtenus dans la section précédente. 

```{r functional_enrichment, fig.width=7, fig.height=5, out.width="80%", fig.cap="Dynamic map of the enrichment analysis ofr the selected genes. Position the cursor on a circle to see the enriched term. "}
#### Run enrichment analysis with gost() ####
message("Running enrichment analysis with gost")

## Get the list of IDs for the selected genes ##
gene_ids <- rownames(gene_stat_norm)[selected_genes]

## Compute enrichment
gostres <- gost(query = gene_ids, 
                organism = "mmusculus", 
                ordered_query = FALSE, 
                multi_query = FALSE, 
                significant = TRUE, 
                exclude_iea = FALSE, 
                measure_underrepresentation = FALSE, 
                evcodes = FALSE, 
                user_threshold = 0.05, 
                correction_method = "fdr", 
                domain_scope = "annotated", 
                custom_bg = NULL, 
                numeric_ns = "", 
                sources = NULL, 
                as_short_link = FALSE)


## Check the structure of the result
names(gostres)
names(gostres$result)
kable(summary(gostres))

# Visualization
gostplot(gostres, capped = TRUE, interactive = TRUE)

## Check the most significant results, formating for kable
enrich_order <- order(gostres$result$p_value, decreasing = FALSE)
sorted_result <- gostres$result[enrich_order, ]
kable(head(sorted_result, n = 10), 
      digits = c(0, 0, 15, 7, 7, 7, 3, 3, 0, 0, 7, 7, 0, 0), 
      caption = "Top 10 most significant enriched functional classes")
names(gostres$meta)
```


**Challenge falcultatif:**

Effectuez une analyse d'enrichissement sur chacun des clusters obtenus à partir de l'arbre des gènes. 

```{r functional_enrichment_clusters}

```

## Conclusions générales


Résumez en quelques phrases vos conclusions à partir des résultats obtenus. 


## Session info


```{r session_info}
#### Session info ####
sessionInfo()

```

</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("solutions_m3-stat-R_mini-projet.Rmd");
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
