<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Anne Badel, Frederic Guyon &amp; Jacques van Helden" />

<meta name="date" content="2019-02-26" />

<title>TP de la séance 4, Clustering</title>

<script src="TP_clustering_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="TP_clustering_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="TP_clustering_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="TP_clustering_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="TP_clustering_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="TP_clustering_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="TP_clustering_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="TP_clustering_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="TP_clustering_files/navigation-1.1/tabsets.js"></script>
<script src="TP_clustering_files/navigation-1.1/codefolding.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #303030; color: #cccccc; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; }
td.sourceCode { padding-left: 5px; }
pre, code { color: #cccccc; background-color: #303030; }
code > span.kw { color: #f0dfaf; } /* Keyword */
code > span.dt { color: #dfdfbf; } /* DataType */
code > span.dv { color: #dcdccc; } /* DecVal */
code > span.bn { color: #dca3a3; } /* BaseN */
code > span.fl { color: #c0bed1; } /* Float */
code > span.ch { color: #dca3a3; } /* Char */
code > span.st { color: #cc9393; } /* String */
code > span.co { color: #7f9f7f; } /* Comment */
code > span.ot { color: #efef8f; } /* Other */
code > span.al { color: #ffcfaf; } /* Alert */
code > span.fu { color: #efef8f; } /* Function */
code > span.er { color: #c3bf9f; } /* Error */
code > span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */
code > span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code > span.sc { color: #dca3a3; } /* SpecialChar */
code > span.vs { color: #cc9393; } /* VerbatimString */
code > span.ss { color: #cc9393; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { } /* Variable */
code > span.cf { color: #f0dfaf; } /* ControlFlow */
code > span.op { color: #f0efd0; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code > span.at { } /* Attribute */
code > span.do { color: #7f9f7f; } /* Documentation */
code > span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code > span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code > span.in { color: #7f9f7f; font-weight: bold; } /* Information */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<div class="container-fluid main-container">

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">TP de la séance 4, Clustering</h1>
<h3 class="subtitle"><em>Diplôme Interuniversitaire en Bioinformatique intégrative (DU-Bii 2019)</em></h3>
<h4 class="author"><em>Anne Badel, Frederic Guyon &amp; Jacques van Helden</em></h4>
<h4 class="date"><em>2019-02-26</em></h4>

</div>


<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Location of the shared folder on the current install
data.folder &lt;-<span class="st"> &quot;~/TCGA_import/data/BIC/&quot;</span> <span class="co"># Just for preparation</span>
<span class="co">#data.folder &lt;- &quot;/shared/projects/du_bii_2019/data/module3/seance4/BIC/&quot; # on the IFB-cluster-core</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Load libraries
<span class="kw">message</span>(<span class="st">&quot;Loading libraries&quot;</span>)
<span class="kw">library</span>(knitr)
<span class="co">#library(kableExtra) ## Note: kableExtra has some side effect on kable: column padding is null, so all numbers seem to be mixed up</span>
<span class="kw">library</span>(FactoMineR)
<span class="kw">library</span>(clues)
<span class="kw">library</span>(RColorBrewer)
<span class="kw">library</span>(ComplexHeatmap)

<span class="co"># library(corrplot)</span>
<span class="kw">library</span>(FactoMineR)
<span class="kw">library</span>(ClassDiscovery)
<span class="co"># library(formattable)</span></code></pre></div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<div id="but-de-ce-tp" class="section level3">
<h3>But de ce TP</h3>
<p>Le tutoriel ci-dessous vous guidera pas-à-pas dans l’utilisation de fonctions <strong>R</strong> pour effectuer un clustering sur des profils transcriptomiques RNA-seq.</p>
</div>
<div id="source-des-donnees" class="section level3">
<h3>Source des données</h3>
<p>Les données sont issues de la base Recount2 (<a href="https://jhubiostatistics.shinyapps.io/recount/" class="uri">https://jhubiostatistics.shinyapps.io/recount/</a>). Nous avons sélectionné l’étude <strong>TCGA</strong> (The Cancer Genome Atlas; <a href="https://cancergenome.nih.gov/" class="uri">https://cancergenome.nih.gov/</a>), regroupant des données RNA-seq pour plus de 12.000 patients souffrant de différents types de cancer. Nous nous intéressons ici uniquement aux données <strong>Breast Invasive Cancer</strong> (<strong>BIC</strong>) concernant le cancer du sein.</p>
<p>Les données ont été préparées pour vous, selon la procédure détaillée au cours sur l’analyse différentielle de données RNA-seq.</p>
<ol style="list-style-type: decimal">
<li><p>Filtrage des gènes à variance nulle et de ceux ccontenant trop de zéros.</p></li>
<li><p>Normalisation (méthode robuste aux outliers)</p></li>
<li><p>Analyse différentielle multi-groupes (en utilisant le package Bioconductor <code>edgeR</code>).</p></li>
<li><p>Correction des P-valeurs nomiables pour tenir compte des tests multiples (nous avons testé ici ~20.000 gènes). Nous estimons le le False Discovery Rate (FDR) selon la méthode de Benjamini-Hochberg (fonction R <code>p.adjust(all.pvalues, method=&quot;fdr&quot;)</code>).</p></li>
<li><p>Sélection de gènes différentiellement exprimés sur base d’un seuil <span class="math inline">\(\alpha = 0.05\)</span> appliqué au FDR.</p></li>
</ol>
</div>
</div>
<div id="choisir-son-environnement-de-travail" class="section level2">
<h2>Choisir son environnement de travail</h2>
<p>Vous pouvez choisir de travailler soir sur le cluster core de l’IFB soit sur les ordinateurs de Paris-Diderot.</p>
<ol style="list-style-type: decimal">
<li><p>Sur le <a href="https://rstudio.cluster.france-bioinformatique.fr/"><strong>cluster de l’IFB</strong></a></p>
<ul>
<li>ouvrez une connexion au serveur RStudio <a href="https://rstudio.cluster.france-bioinformatique.fr/" class="uri">https://rstudio.cluster.france-bioinformatique.fr/</a> et identifiez-vous</li>
</ul></li>
<li><p>Sur les machines de la <strong>salle d’ordinateurs de Paris-Diderot</strong></p>
<ul>
<li>vous devez avoir la commande suivante dans votre <code>.bashrc</code> :</li>
</ul></li>
</ol>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="bu">source</span> /opt/sdv/anaconda/etc/profile.d/conda.sh<span class="kw">`</span></code></pre></div>
<ul>
<li>puis vous devez <strong>lancer l’environnement conda</strong> adéquat :</li>
</ul>
<p><code>conda activate</code></p>
<ul>
<li>et enfin lancer le serveur Rstudio au moyen de la commande bash: <code>rstudio</code></li>
</ul>
</div>
<div id="dossier-partage-contenant-les-donnees" class="section level2">
<h2>Dossier partagé contenant les données</h2>
<p>Les données sont dans un répertoire partagé, dont le chemin dépend du serveur auquel vous êtes connectézs. Nous allons définir une variable <code>data.folder</code> qui indiquera le chemin de ce dossier partagé, en fonction du serveur.</p>
<p>Sur le <strong>serveur Rstudio de l’IFB-core-cluster</strong>, les données sont dans le répertoire <code>/shared/projects/du_bii_2019/data/module3/seance4/BIC/</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Liste des fichiers de données
data.folder &lt;-<span class="st"> &quot;/shared/projects/du_bii_2019/data/module3/seance4/BIC/&quot;</span></code></pre></div>
<!--
2. Sur les machines de Paris-Diderot, elles sont dans le répertoire `/home/sdv/dubii/data-m3/s4` 


```r
## Liste des fichiers de données
data.folder <- "/home/sdv/dubii/data-m3/s4/"
```

-->
</div>
<div id="contenu-du-dossier-de-donnees" class="section level2">
<h2>Contenu du dossier de données</h2>
<p>Utilisez les commandes R suivantes:</p>
<ul>
<li><code>list.files()</code> pour vérifier le contenu du dosser <code>data.folder</code>,</li>
<li><code>file.size()</code> pour calculer la taille de ces fichiers.</li>
</ul>
<p>Astuces:</p>
<ul>
<li><code>list.files()</code> retourne par défaut le nom de fichier, mais avec l’option <code>full.names=TRUE</code> vous obtiendrez le chemin complet.</li>
<li>Calculez la taille des fichiers en bytes et en Megabytes (<span class="math inline">\(1Mb = 1024 \cot 1024 \cdot b\)</span>), sachant que pour chaque conversion il faut diviser par 1024.</li>
<li>Vous pouvez consulter notre solution à l’aide du code suivant (cliquer sur <strong><em>Code</em></strong> pour l’afficher).</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Return file sizes (in bytes)</span>
<span class="kw">message</span>(<span class="st">&quot;Listing files in data folder: &quot;</span>, data.folder)

data.files &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path =</span> data.folder) <span class="co"># List the data files</span>
<span class="kw">print</span>(data.files)</code></pre></div>
<pre><code> [1] &quot;BIC_counts_all-genes.tsv.gz&quot;                    &quot;BIC_counts_filtered-genes.tsv.gz&quot;               &quot;BIC_edgeR_DEG_table.tsv&quot;                        &quot;BIC_edgeR_DEG_table.tsv.gz&quot;                     &quot;BIC_log2-norm-counts_all-genes.tsv.gz&quot;          &quot;BIC_log2-norm-counts_edgeR_DEG_fdr_0.01.tsv&quot;   
 [7] &quot;BIC_log2-norm-counts_edgeR_DEG_fdr_0.01.tsv.gz&quot; &quot;BIC_log2-norm-counts_edgeR_DEG_top_1000.tsv.gz&quot; &quot;BIC_log2-norm-counts_filtered-genes.tsv&quot;        &quot;BIC_log2-norm-counts_filtered-genes.tsv.gz&quot;     &quot;BIC_pheno.tsv.gz&quot;                               &quot;BIC_sample-classes.tsv.gz&quot;                     </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Full path
data.path &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path =</span> data.folder, <span class="dt">full.names =</span> <span class="ot">TRUE</span>) <span class="co"># List the data files</span>
<span class="kw">print</span>(data.files)</code></pre></div>
<pre><code> [1] &quot;BIC_counts_all-genes.tsv.gz&quot;                    &quot;BIC_counts_filtered-genes.tsv.gz&quot;               &quot;BIC_edgeR_DEG_table.tsv&quot;                        &quot;BIC_edgeR_DEG_table.tsv.gz&quot;                     &quot;BIC_log2-norm-counts_all-genes.tsv.gz&quot;          &quot;BIC_log2-norm-counts_edgeR_DEG_fdr_0.01.tsv&quot;   
 [7] &quot;BIC_log2-norm-counts_edgeR_DEG_fdr_0.01.tsv.gz&quot; &quot;BIC_log2-norm-counts_edgeR_DEG_top_1000.tsv.gz&quot; &quot;BIC_log2-norm-counts_filtered-genes.tsv&quot;        &quot;BIC_log2-norm-counts_filtered-genes.tsv.gz&quot;     &quot;BIC_pheno.tsv.gz&quot;                               &quot;BIC_sample-classes.tsv.gz&quot;                     </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data.file.sizes &lt;-<span class="st"> </span><span class="kw">file.size</span>(data.path) <span class="co"># Get the size of the data files, in bytes</span>

## Add file names 
## (for some strange reason, file.size returns a vector
## with no names, which is not very convenient)
<span class="kw">names</span>(data.file.sizes) &lt;-<span class="st"> </span>data.files

## Compute file sizes in megabytes
data.file.Mb &lt;-<span class="st"> </span><span class="kw">signif</span>(<span class="dt">digits =</span> <span class="dv">2</span>, data.file.sizes <span class="op">/</span><span class="st"> </span>(<span class="dv">1024</span><span class="op">^</span><span class="dv">2</span>))
<span class="kw">kable</span>(<span class="kw">data.frame</span>(data.file.Mb))</code></pre></div>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">data.file.Mb</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>BIC_counts_all-genes.tsv.gz</td>
<td align="right">79.000</td>
</tr>
<tr class="even">
<td>BIC_counts_filtered-genes.tsv.gz</td>
<td align="right">47.000</td>
</tr>
<tr class="odd">
<td>BIC_edgeR_DEG_table.tsv</td>
<td align="right">1.600</td>
</tr>
<tr class="even">
<td>BIC_edgeR_DEG_table.tsv.gz</td>
<td align="right">1.700</td>
</tr>
<tr class="odd">
<td>BIC_log2-norm-counts_all-genes.tsv.gz</td>
<td align="right">120.000</td>
</tr>
<tr class="even">
<td>BIC_log2-norm-counts_edgeR_DEG_fdr_0.01.tsv</td>
<td align="right">85.000</td>
</tr>
<tr class="odd">
<td>BIC_log2-norm-counts_edgeR_DEG_fdr_0.01.tsv.gz</td>
<td align="right">85.000</td>
</tr>
<tr class="even">
<td>BIC_log2-norm-counts_edgeR_DEG_top_1000.tsv.gz</td>
<td align="right">6.000</td>
</tr>
<tr class="odd">
<td>BIC_log2-norm-counts_filtered-genes.tsv</td>
<td align="right">120.000</td>
</tr>
<tr class="even">
<td>BIC_log2-norm-counts_filtered-genes.tsv.gz</td>
<td align="right">120.000</td>
</tr>
<tr class="odd">
<td>BIC_pheno.tsv.gz</td>
<td align="right">0.950</td>
</tr>
<tr class="even">
<td>BIC_sample-classes.tsv.gz</td>
<td align="right">0.019</td>
</tr>
</tbody>
</table>
</div>
<div id="lire-le-tableau-de-valeurs-dexpression" class="section level2">
<h2>Lire le tableau de valeurs d’expression</h2>
<p>Nous allons maintenant lire le fichier d’expression. Pour cela, nous concaténons le chemin du dossier de données et le nom du fichier d’expressiion (<code>BIC_log2-norm-counts_edgeR_DEG_top_1000.tsv.gz</code>).</p>
<p>Ce fichier contient le comptages de lectures RNA-seq par gène, avec une sélection des gènes déclarés positifs pour letest de comparaison de moyennes multiples (voir-ci-dessus). Par ailleurs, nous avons arbitrairement appliqué un seuil supplémentaire en n’exportant que les 1000 gènes les plus significatifs, pour éviter de passer trop de temps sur le clustering hiérachique (complexité quadratique).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Define the path of the expression file, by concatenating the data folder and the file name</span>
<span class="co"># Note that file.path is convenient because it automatically used the appropriate parameter to separate the elements of a path (/ on Unix, \\ on Windows)</span>
BIC.expr.file &lt;-<span class="st"> </span><span class="kw">file.path</span>(data.folder, <span class="st">&quot;BIC_log2-norm-counts_edgeR_DEG_top_1000.tsv.gz&quot;</span>)

<span class="co"># Load expression</span>
<span class="kw">message</span>(<span class="st">&quot;Loading expression file</span><span class="ch">\t</span><span class="st">&quot;</span>, BIC.expr.file)
BIC.expr &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">file =</span> BIC.expr.file, <span class="dt">header =</span> <span class="ot">TRUE</span>)</code></pre></div>
</div>
<div id="mesure-de-la-taille-des-donnees" class="section level2">
<h2>Mesure de la taille des données</h2>
<p>Prenez le temps d’identifier</p>
<ul>
<li>la taille du jeu de données</li>
<li>le nombre d’individus</li>
<li>le nombre de variables</li>
</ul>
<p><strong>Remarque :</strong> Classiquement, en analyse de données, les individus sont les lignes du tableau de données, les colonnes sont les variables.</p>
<p>Pour des raisons historiques, en analyse transcriptomique les données sont toujours fournies avec</p>
<ul>
<li>1 ligne = 1 gène</li>
<li>1 échantillon biologique = 1 colonne</li>
</ul>
<p>Cette convention a été établie en 1997, lors des toutes premières publications sur le transcriptome de la levure. Dans ces études, l’objet d’intérêt (l’“éindividu”) était le gène, et les variables étaient ses mesures d’expression dans les différentes conditions testées.</p>
<p>Pour l’anlayse de tissus cancéreux, on considère au contraire que l’“objet” d’intérêt est l’échantillon prélevé sur le patient, et les variables sont les mesures d’expression des différents gènes chez un patient.</p>
<p>Ce qui implique de faire attention, et éventuellement de travailler sur la matrice transposée (fonction <code>t</code> en R) pour utiliser correctement les fonctions classiques.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Les données d&#39;expression
<span class="kw">dim</span>(BIC.expr)</code></pre></div>
<pre><code>[1] 1000  819</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Les noms des lignes correspondent à des gènes
<span class="kw">head</span>(<span class="kw">rownames</span>(BIC.expr))</code></pre></div>
<pre><code>[1] &quot;ENSG00000001617.11&quot; &quot;ENSG00000002834.17&quot; &quot;ENSG00000003989.16&quot; &quot;ENSG00000004838.13&quot; &quot;ENSG00000005302.17&quot; &quot;ENSG00000006016.10&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Les noms des colonnes correspondent à des échantillons
<span class="kw">head</span>(<span class="kw">colnames</span>(BIC.expr))</code></pre></div>
<pre><code>[1] &quot;X1AB92ADA.637E.4A42.A39A.70CEEEA41AE3&quot; &quot;DA98A67C.F11F.41D3.8223.1161EBFF8B58&quot;  &quot;X06CCFD0F.7FB8.471E.B823.C7876582D6FC&quot; &quot;A33B2F42.6EC6.4FB2.8BE5.542407A0382E&quot;  &quot;D021A258.8713.4383.9DCA.45E2F54A0411&quot;  &quot;C705FA90.D9AA.4949.BACA.1C022A14CB03&quot; </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## transposons la table d&#39;expression
<span class="kw">message</span>(<span class="st">&quot;Transcribing expression matrix&quot;</span>)
BIC.expr.transposed &lt;-<span class="st"> </span><span class="kw">t</span>(BIC.expr)</code></pre></div>
</div>
<div id="charger-les-etiquettes-de-classes-des-echantillons" class="section level2">
<h2>Charger les étiquettes de classes des échantillons</h2>
<p>Le fichier <code>BIC_sample-classes.tsv.gz</code>contient les étiquettes de classes des échantillons.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">BIC.sample.classes &lt;-<span class="st"> </span><span class="kw">read.table</span>(
  <span class="kw">file.path</span>(data.folder, <span class="st">&quot;BIC_sample-classes.tsv.gz&quot;</span>),
  <span class="dt">header =</span> <span class="ot">TRUE</span>)

<span class="kw">kable</span>(BIC.sample.classes[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,])</code></pre></div>
<table>
<thead>
<tr class="header">
<th></th>
<th align="left">cancer.type</th>
<th align="left">ER1</th>
<th align="left">PR1</th>
<th align="left">Her2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1AB92ADA-637E-4A42-A39A-70CEEEA41AE3</td>
<td align="left">Luminal.A</td>
<td align="left">Positive</td>
<td align="left">Positive</td>
<td align="left">Negative</td>
</tr>
<tr class="even">
<td>DA98A67C-F11F-41D3-8223-1161EBFF8B58</td>
<td align="left">Unclassified</td>
<td align="left">Positive</td>
<td align="left">Negative</td>
<td align="left">Negative</td>
</tr>
<tr class="odd">
<td>06CCFD0F-7FB8-471E-B823-C7876582D6FC</td>
<td align="left">HER2pos</td>
<td align="left">Negative</td>
<td align="left">Negative</td>
<td align="left">Positive</td>
</tr>
<tr class="even">
<td>A33B2F42-6EC6-4FB2-8BE5-542407A0382E</td>
<td align="left">Unclassified</td>
<td align="left">Positive</td>
<td align="left">Negative</td>
<td align="left">Negative</td>
</tr>
<tr class="odd">
<td>D021A258-8713-4383-9DCA-45E2F54A0411</td>
<td align="left">Luminal.A</td>
<td align="left">Positive</td>
<td align="left">Positive</td>
<td align="left">Negative</td>
</tr>
<tr class="even">
<td>C705FA90-D9AA-4949-BACA-1C022A14CB03</td>
<td align="left">Luminal.A</td>
<td align="left">Positive</td>
<td align="left">Positive</td>
<td align="left">Negative</td>
</tr>
<tr class="odd">
<td>85380A2D-9951-4D4B-A2A4-6F5F2AFC54E3</td>
<td align="left">Luminal.A</td>
<td align="left">Positive</td>
<td align="left">Positive</td>
<td align="left">Negative</td>
</tr>
<tr class="even">
<td>F53A9C63-1AF7-4CBC-B8B7-4AA7AAED3364</td>
<td align="left">Luminal.A</td>
<td align="left">Positive</td>
<td align="left">Positive</td>
<td align="left">Negative</td>
</tr>
<tr class="odd">
<td>13EF5323-EAD9-4BC7-8AC4-33875BF12E17</td>
<td align="left">Luminal.B</td>
<td align="left">Positive</td>
<td align="left">Positive</td>
<td align="left">Positive</td>
</tr>
<tr class="even">
<td>079EACA1-0319-4B54-B20B-673F4576C69D</td>
<td align="left">Basal.like</td>
<td align="left">Negative</td>
<td align="left">Negative</td>
<td align="left">Negative</td>
</tr>
</tbody>
</table>
<p>Chaque échantillon a été assigné à une classe selon la combinaison de 3 marqueurs immunologiques:</p>
<ul>
<li>Estrogen Receptor 1 (ER1)</li>
<li>Progesterone Receptor 1 (PR1)</li>
<li>Human epidermal growth factor receptor 2 (Her2)</li>
</ul>
<p>Utilisez</p>
<ul>
<li><p>La fonction R <code>summary()</code> pour compter le nombre de patientes positives / négatives pour chacun de ces trois marqueurs.</p></li>
<li><p>La fonction R <code>table()</code> pour calculer le nombre d’échantillons de chaque type de cancer.</p></li>
<li><p>La fonction R <code>table()</code> pour calculer une table de contiingence des marqueurs ER1 et PR1</p></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(BIC.sample.classes)</code></pre></div>
<pre><code>       cancer.type        ER1            PR1            Her2    
 Basal.like  :131   Negative:184   Negative:267   Negative:631  
 HER2pos     : 41   Positive:635   Positive:552   Positive:188  
 Luminal.A   :422                                               
 Luminal.B   :118                                               
 Unclassified:107                                               </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(BIC.sample.classes<span class="op">$</span>cancer.type)</code></pre></div>
<pre><code>
  Basal.like      HER2pos    Luminal.A    Luminal.B Unclassified 
         131           41          422          118          107 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(BIC.sample.classes<span class="op">$</span>ER1, BIC.sample.classes<span class="op">$</span>PR1)</code></pre></div>
<pre><code>          
           Negative Positive
  Negative      172       12
  Positive       95      540</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(BIC.sample.classes<span class="op">$</span>ER1, 
      BIC.sample.classes<span class="op">$</span>PR1, 
      BIC.sample.classes<span class="op">$</span>Her2)</code></pre></div>
<pre><code>, ,  = Negative

          
           Negative Positive
  Negative      131        8
  Positive       70      422

, ,  = Positive

          
           Negative Positive
  Negative       41        4
  Positive       25      118</code></pre>
</div>
<div id="projection-acp-des-echantillons" class="section level2">
<h2>Projection ACP des échantillons</h2>
<p>Nous allons réaliser une ACP sans mise à l’échelle.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Transformation ACP
BIC.prcomp &lt;-<span class="st"> </span><span class="kw">prcomp</span>(BIC.expr.transposed, <span class="dt">center =</span> <span class="ot">FALSE</span>, <span class="dt">scale. =</span> <span class="ot">FALSE</span>)
<span class="kw">names</span>(BIC.prcomp) <span class="co"># check the name of the fields of the PCA object</span></code></pre></div>
<pre><code>[1] &quot;sdev&quot;     &quot;rotation&quot; &quot;center&quot;   &quot;scale&quot;    &quot;x&quot;       </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
## Plot de l&#39;écart-type sur les premières composantes
<span class="kw">barplot</span>(BIC.prcomp<span class="op">$</span>sdev[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>], <span class="dt">col =</span> <span class="st">&quot;#FFEEDD&quot;</span>,
        <span class="dt">main =</span> <span class="st">&quot;Standard deviation per PC&quot;</span>)

## Visualisation des données
<span class="kw">plot</span>(BIC.prcomp, <span class="dt">main =</span> <span class="st">&quot;Variance per PC&quot;</span>, <span class="dt">col=</span><span class="st">&quot;#BBDDFF&quot;</span>)</code></pre></div>
<p><img src="figures/TP_pca_var-1.png" width="80%" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</code></pre></div>
<p>Définissez une couleur pour chaque classe, et assignez à chaque échantillon la couleur correspondant à sa classe. Dessinez ensuite un nuage de points avec les coordonnées de chaque échantillon dans les 1ère et 2ème composantes (PC2 vs PC1)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Assign a color to each cancer type</span>
classes &lt;-<span class="st"> </span><span class="kw">unique</span>(BIC.sample.classes<span class="op">$</span>cancer.type)
class.colors &lt;-<span class="st"> </span><span class="kw">rainbow</span>(<span class="dt">n =</span> <span class="kw">length</span>(classes))
<span class="kw">names</span>(class.colors) &lt;-<span class="st"> </span>classes
<span class="kw">data.frame</span>(class.colors)</code></pre></div>
<pre><code>             class.colors
Luminal.A       #FF0000FF
Unclassified    #CCFF00FF
HER2pos         #00FF66FF
Luminal.B       #0066FFFF
Basal.like      #CC00FFFF</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Associate a color to each sample accordinig to its cancer type</span>
sample.colors &lt;-<span class="st"> </span>class.colors[BIC.sample.classes<span class="op">$</span>cancer.type]
<span class="kw">table</span>(BIC.sample.classes<span class="op">$</span>cancer.type, sample.colors)</code></pre></div>
<pre><code>              sample.colors
               #0066FFFF #00FF66FF #CC00FFFF #CCFF00FF #FF0000FF
  Basal.like           0         0         0         0       131
  HER2pos              0         0         0        41         0
  Luminal.A            0       422         0         0         0
  Luminal.B          118         0         0         0         0
  Unclassified         0         0       107         0         0</code></pre>
<p><strong>Question: </strong> comment interprétez-vous les barplots des écarts-types et variances pour les premières comosantes ? A discuter pendant le cours.</p>
</div>
<div id="clustering-hierarchique" class="section level2">
<h2>Clustering hiérarchique</h2>
<div id="calcul-de-la-matrice-de-distance" class="section level3">
<h3>Calcul de la matrice de distance</h3>
<p>Nous allons maintenant calculer la distance entre chaque paire d’échantillon, en utilsiant comme métrique le <strong>coefficient de corrélation de Spearman</strong>, plus adapté à ce type de données que la distance euclidienne utilisée sur les données iris durant le cours</p>
<ol style="list-style-type: decimal">
<li><p>Lisez l’aide de la fonction <code>cor</code>, et utilisez cette fonction pour calculer la matrice de corrélation entre échantillons.</p></li>
<li><p>transformation du corrélation de Spearman en une distance à l’aide de la transformation : <span class="math inline">\(d = 1 - r\)</span></p></li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Compute Spearman correlation coefficient
BIC.cor &lt;-<span class="st"> </span><span class="kw">cor</span>(BIC.expr, <span class="dt">method =</span> <span class="st">&quot;spearman&quot;</span>)

<span class="co"># Check the dimensions of the correlation matrix</span>
<span class="co"># (should be N x N, where N is the number of samples)</span>
<span class="kw">dim</span>(BIC.cor)</code></pre></div>
<pre><code>[1] 819 819</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Derive a dissimilarity value from the corrrelation
BIC.dist &lt;-<span class="st"> </span><span class="kw">as.dist</span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span>BIC.cor)</code></pre></div>
</div>
<div id="hclust" class="section level3">
<h3>hclust</h3>
<p>Faites un premier clustering hiérarchique, avec le critère d’aggrégation par défaut (lisez l’aide de la fonction <code>hclust()</code> pour savoir quelle est ce critère par défaut).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Run hierarchical clustering on the expression data
## (use default parameters)
BIC.hclust.complete &lt;-<span class="st"> </span><span class="kw">hclust</span>(BIC.dist)

## Plot the resulting tree
<span class="kw">plot</span>(BIC.hclust.complete, <span class="dt">labels =</span> F, <span class="dt">hang =</span> <span class="op">-</span><span class="dv">1</span>, <span class="dt">las =</span> <span class="dv">1</span>, <span class="dt">col =</span> <span class="st">&quot;grey&quot;</span>,
     <span class="dt">main =</span> <span class="st">&quot;Spearman dissimilarity, complete linkage&quot;</span>)</code></pre></div>
<p><img src="figures/TP_unnamed-chunk-5-1.png" width="480" style="display: block; margin: auto;" /></p>
<ol start="2" style="list-style-type: decimal">
<li>faire un deuxième clustering hiérarchique, avec le critère d’aggrégation de Ward</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">##  Run hierarchical clustering with Ward agglomeration
BIC.hclust.ward &lt;-<span class="st"> </span><span class="kw">hclust</span>(BIC.dist, <span class="dt">method =</span> <span class="st">&quot;ward.D2&quot;</span>)

## Plot the resulting tree
<span class="kw">plot</span>(BIC.hclust.ward, <span class="dt">labels =</span> F, <span class="dt">hang =</span> <span class="op">-</span><span class="dv">1</span>, <span class="dt">las =</span> <span class="dv">1</span>, <span class="dt">col =</span> <span class="st">&quot;#88BB66&quot;</span>, 
     <span class="dt">main =</span> <span class="st">&quot;Spearman dissimilarity, Ward linkage&quot;</span>)</code></pre></div>
<p><img src="figures/TP_unnamed-chunk-6-1.png" width="480" style="display: block; margin: auto;" /></p>
<ol start="3" style="list-style-type: decimal">
<li>Redessiner les arbres de ces deux résultats de clustering en colorant les échantillons selon la classe de cancer.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ClassDiscovery<span class="op">::</span><span class="kw">plotColoredClusters</span>(
  BIC.hclust.ward, 
  <span class="dt">labs=</span>BIC.sample.classes<span class="op">$</span>cancer.type, <span class="dt">col =</span> <span class="st">&quot;grey&quot;</span>,
  <span class="dt">cols=</span>sample.colors, <span class="dt">cex=</span><span class="fl">0.2</span>)

<span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>,
       <span class="dt">legend=</span><span class="kw">sort</span>(classes), 
       <span class="dt">col =</span> class.colors,
       <span class="dt">text.col =</span> class.colors, <span class="dt">cex=</span><span class="fl">0.9</span>)</code></pre></div>
<p><img src="figures/TP_tree_colored_labels-1.png" width="480" style="display: block; margin: auto;" /></p>
<ol start="3" style="list-style-type: decimal">
<li>Comparer les classifications obtenues avec les règles d’agglomératioin complète et Ward, respectivement, en étudiant l’impact du nombre de clusters.</li>
</ol>
<p><strong>Astuces:</strong></p>
<ul>
<li>Cous pouvez utiliser les commandes <code>rect.hclust</code> et <code>cutree</code> pour visualiser les clusters sur le dendrogramme, puis récupérer les clusters.</li>
</ul>
</div>
<div id="combinaison-dun-arbre-et-dune-carte-de-temperature" class="section level3">
<h3>Combinaison d’un arbre et d’une carte de température</h3>
<p>La fonction R <code>heatmap()</code> permet de représenter à la fois les arbres produit par le clustering hiérarchique, et les profils d’expression.</p>
<p>Par défaut, elle effectue simultanément un clustering hiérarchique sur les lignes et sur les colonnes, ce qui permet de distinguer non seulement les groupes d’échantillons biologiques, mais également ceux de gènes.</p>
<p><strong>Attention:</strong> la fonction <code>heatmap()</code> effectue par défaut un clustering hiérarchique sur les lignes et colonnes de votre matrice d’expression.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">heatmap</span>(<span class="kw">as.matrix</span>(BIC.expr), 
        <span class="dt">distfun =</span> <span class="cf">function</span>(x) <span class="kw">as.dist</span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span><span class="kw">cor</span>(<span class="kw">t</span>(x), <span class="dt">method =</span> <span class="st">&quot;spearman&quot;</span>)), 
        <span class="dt">hclustfun =</span> <span class="cf">function</span>(x) <span class="kw">hclust</span>(x, <span class="dt">method =</span> <span class="st">&quot;ward.D2&quot;</span>),
        <span class="dt">labRow =</span> <span class="ot">NA</span>, <span class="dt">labCol =</span> <span class="ot">NA</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/TP_heatmap-1.png" alt="Heat map of the expression matrix clustered by genes (rows) and samples (columns)." width="80%" />
<p class="caption">
Heat map of the expression matrix clustered by genes (rows) and samples (columns).
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Compute mean value per gene</span>
gene.means &lt;-<span class="st"> </span><span class="kw">apply</span>(BIC.expr, <span class="dv">1</span>, mean)
gene.sd &lt;-<span class="st">  </span><span class="kw">apply</span>(BIC.expr, <span class="dv">1</span>, sd)

<span class="co"># Compute gene-wise centered expression values</span>
BIC.expr.genes.centered &lt;-<span class="st"> </span>BIC.expr <span class="op">-</span><span class="st"> </span>gene.means

<span class="co"># Compute gene-wise centred + scaled expression values</span>
BIC.expr.genes.standardized &lt;-<span class="st"> </span>BIC.expr.genes.centered <span class="op">/</span><span class="st"> </span>gene.sd

<span class="kw">summary</span>(<span class="kw">unlist</span>(BIC.expr.genes.standardized))</code></pre></div>
<pre><code>    Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
-7.00569 -0.63658 -0.03651  0.00000  0.60201 12.86463 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(<span class="kw">unlist</span>(BIC.expr.genes.standardized), <span class="dt">breaks =</span> <span class="dv">100</span>,
     <span class="dt">main =</span> <span class="st">&quot;Gene-wise standardized expression values&quot;</span>,
     <span class="dt">xlab =</span> <span class="st">&quot;z-score&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Number of measures&quot;</span>, 
     <span class="dt">col =</span> <span class="st">&quot;#DDEEFF&quot;</span>, <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">7</span>,<span class="dv">7</span>))

<span class="co"># Draw a vertical grid</span>
<span class="kw">abline</span>(<span class="dt">v =</span> <span class="op">-</span><span class="dv">7</span><span class="op">:</span><span class="dv">7</span>, <span class="dt">col =</span> <span class="st">&quot;#BBBBBB&quot;</span>, <span class="dt">lty =</span> <span class="st">&quot;dotted&quot;</span>)

<span class="co">#Mark center</span>
<span class="kw">abline</span>(<span class="dt">v =</span> <span class="dv">0</span>, <span class="dt">col =</span> <span class="st">&quot;#008800&quot;</span>, <span class="dt">lwd =</span> <span class="dv">2</span>)

<span class="co"># Mark dispersion (mean +- 1 sdev)</span>
<span class="kw">arrows</span>(<span class="dt">x0 =</span> <span class="op">-</span><span class="dv">1</span>, <span class="dt">y0 =</span> <span class="dv">50000</span>, <span class="dt">x1 =</span> <span class="dv">1</span>, <span class="dt">y1 =</span> <span class="dv">50000</span>, <span class="dt">length =</span> <span class="fl">0.05</span>, <span class="dt">angle =</span> <span class="dv">30</span>, <span class="dt">code =</span> <span class="dv">3</span>, <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">col =</span> <span class="st">&quot;#008800&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/TP_standardize_genes-1.png" alt="Histogram of expression values after gene-wise standardization (centering and scaling)." width="80%" />
<p class="caption">
Histogram of expression values after gene-wise standardization (centering and scaling).
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## define a Blue - White - Red palette
frenchflag.palette &lt;-<span class="st"> </span><span class="kw">colorRampPalette</span>(<span class="kw">c</span>(<span class="st">&#39;dark blue&#39;</span>,<span class="st">&#39;white&#39;</span>,<span class="st">&#39;dark red&#39;</span>))

## Define a green - black - red palette
GBR.palette &lt;-<span class="st"> </span><span class="kw">colorRampPalette</span>(<span class="kw">c</span>(<span class="st">&#39;green&#39;</span>,<span class="st">&#39;black&#39;</span>,<span class="st">&#39;red&#39;</span>))

<span class="kw">heatmap</span>(<span class="kw">as.matrix</span>(BIC.expr.genes.standardized),
        <span class="dt">zlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">4</span>,<span class="dv">4</span>),
        <span class="dt">distfun =</span> <span class="cf">function</span>(x) <span class="kw">as.dist</span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span><span class="kw">cor</span>(<span class="kw">t</span>(x), <span class="dt">method =</span> <span class="st">&quot;spearman&quot;</span>)), 
        <span class="dt">hclustfun =</span> <span class="cf">function</span>(x) <span class="kw">hclust</span>(x, <span class="dt">method =</span> <span class="st">&quot;ward.D2&quot;</span>),
        <span class="dt">labRow =</span> <span class="ot">NA</span>, <span class="dt">labCol =</span> <span class="ot">NA</span>,  <span class="co"># DO not print the labels (unreadable anyway)</span>
        <span class="dt">col =</span> <span class="kw">GBR.palette</span>(<span class="dv">100</span>),
        <span class="dt">scale =</span> <span class="st">&quot;none&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/TP_heatmap_frenchflag_palette-1.png" alt="Heat map of the expression matrix clustered by genes (rows) and samples (columns)." width="80%" />
<p class="caption">
Heat map of the expression matrix clustered by genes (rows) and samples (columns).
</p>
</div>
</div>
<div id="draw-a-tree-with-heatmap.2" class="section level3">
<h3>Draw a tree with heatmap.2()</h3>
<p>The function <code>heatmap.2()</code> is derived from <code>heatmap()</code> but offers nice additional formatting options.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Define a color for the heatmap</span>
expr.colors &lt;-<span class="st"> </span><span class="kw">heat.colors</span>(<span class="dt">n =</span> <span class="dv">100</span>)

<span class="co"># Compute dendrogram with the hclust() function</span>
<span class="co"># Choose a custom dissimilirity measure</span>
<span class="co"># and agglomeration rule</span>

<span class="co"># compute dissimilarity between samples</span>
sample.dist &lt;-<span class="st"> </span><span class="kw">as.dist</span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span><span class="kw">cor</span>(BIC.expr, <span class="dt">method =</span> <span class="st">&quot;spearman&quot;</span>)) 

<span class="co"># Run hierarchical clustering on samples</span>
sample.clust =<span class="st"> </span><span class="kw">hclust</span>(sample.dist, <span class="dt">method =</span> <span class="st">&quot;complete&quot;</span>)

<span class="co"># Convert the clustering result in a tree object that can be used with plot() and / or heatmap()</span>
sample.tree &lt;-<span class="st"> </span><span class="kw">as.dendrogram</span>(sample.clust)

<span class="co"># Define colors for the cancer type</span>
cancer.type &lt;-<span class="st"> </span><span class="kw">unique</span>(BIC.sample.classes<span class="op">$</span>cancer.type)
type.cancer.colors &lt;-<span class="st"> </span><span class="kw">rainbow</span>(<span class="dt">n =</span> <span class="kw">length</span>(cancer.type))
<span class="kw">names</span>(type.cancer.colors) &lt;-<span class="st"> </span>cancer.type

<span class="co"># Define colors for ER1 marker</span>
ER1.classes &lt;-<span class="st"> </span><span class="kw">unique</span>(BIC.sample.classes<span class="op">$</span>ER1)
ER1.colors &lt;-<span class="st"> </span><span class="kw">rainbow</span>(<span class="dt">n =</span> <span class="kw">length</span>(ER1.classes))
<span class="kw">names</span>(ER1.colors) &lt;-<span class="st"> </span>ER1.classes
<span class="kw">print</span>(ER1.colors)</code></pre></div>
<pre><code>   Positive    Negative 
&quot;#FF0000FF&quot; &quot;#00FFFFFF&quot; </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Define colors for PR1 marker</span>
PR1.classes &lt;-<span class="st"> </span><span class="kw">unique</span>(BIC.sample.classes<span class="op">$</span>PR1)
PR1.colors &lt;-<span class="st"> </span><span class="kw">heat.colors</span>(<span class="dt">n =</span> <span class="kw">length</span>(PR1.classes))
<span class="kw">names</span>(PR1.colors) &lt;-<span class="st"> </span>PR1.classes
<span class="kw">print</span>(PR1.colors)</code></pre></div>
<pre><code>   Positive    Negative 
&quot;#FF0000FF&quot; &quot;#FFFF00FF&quot; </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Define colors for Her2 marker</span>
Her2.classes &lt;-<span class="st"> </span><span class="kw">unique</span>(BIC.sample.classes<span class="op">$</span>Her2)
Her2.colors &lt;-<span class="st"> </span><span class="kw">topo.colors</span>(<span class="dt">n =</span> <span class="kw">length</span>(Her2.classes))
<span class="kw">names</span>(Her2.colors) &lt;-<span class="st"> </span>Her2.classes

## Detiine annotations for the heatmap
## combining the 3 markers + cancer type
annot.tumeur.column =<span class="st"> </span><span class="kw">HeatmapAnnotation</span>(
  <span class="dt">df =</span> BIC.sample.classes,
  <span class="dt">col =</span> <span class="kw">list</span>(<span class="dt">cancer.type =</span> type.cancer.colors,
             <span class="dt">ER1 =</span> ER1.colors, 
             <span class="dt">PR1 =</span> PR1.colors,
             <span class="dt">Her2 =</span> Her2.colors)
)

<span class="co"># A first heatmap</span>
my.heatmap &lt;-<span class="st"> </span>ComplexHeatmap<span class="op">::</span><span class="kw">Heatmap</span>(
  <span class="kw">as.matrix</span>(BIC.expr.genes.standardized), 
  <span class="dt">name =</span> <span class="st">&quot;TCGA Breast Invasive Cancer&quot;</span>, 
  <span class="co"># col = frenchflag.palette(100),</span>
  <span class="dt">col =</span> <span class="kw">brewer.pal</span>(<span class="dv">11</span>,<span class="st">&quot;RdBu&quot;</span>),
  <span class="dt">column_title =</span> <span class="st">&quot;Samples&quot;</span>,
  <span class="dt">row_title =</span> <span class="st">&quot;Genes&quot;</span>,
  <span class="dt">cluster_columns =</span> sample.tree,
  <span class="dt">show_column_names =</span> <span class="ot">FALSE</span>, 
  <span class="dt">show_row_names =</span> <span class="ot">FALSE</span>,
  <span class="dt">bottom_annotation =</span> annot.tumeur.column
)
<span class="kw">draw</span>(my.heatmap)</code></pre></div>
<p><img src="figures/TP_complex_heatmap-1.png" width="95%" style="display: block; margin: auto;" /></p>
</div>
<div id="kmeans" class="section level3">
<h3>kmeans</h3>
<ol style="list-style-type: decimal">
<li><p>faire un premier kmeans, par exemple, en prenant le nombre de groupe trouvé sur le <code>hclust</code></p></li>
<li><p>faire une boucle pour trouver le nombre optimal de cluster, en calculant l’inertie intra totale en fonction du nombre de groupe <code>kmeans()$totss</code> [faire une boucle pour i allant de 1 à 10 <code>for (i in 1:10) {}</code>]</p></li>
<li><p>refaire le kmeans avec ce nombre optimal</p></li>
<li><p>visualiser ces groupes par exemple sur une projection des données dans le plan par PCA, à l’aide de la fonction <code>prcomp()</code>.</p></li>
</ol>
<p><strong>Astuce:</strong> dans le résultat de <code>prcomp()</code>, les coordonnées des points se trouvent dans le champs <code>x</code> .</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Run k-means clustering with 20 centers
BIC.kmeans &lt;-<span class="st"> </span><span class="kw">kmeans</span>(BIC.expr.transposed, <span class="dt">centers=</span><span class="dv">20</span>)

## Report the table of the clusters
<span class="kw">table</span>(BIC.kmeans<span class="op">$</span>cluster)</code></pre></div>
<pre><code>
  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20 
115  13  62  57   7  33  16  32   5  23  48  36  70  82  32   9  70  74  10  25 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">T1 =<span class="st"> </span><span class="kw">Sys.time</span>() <span class="co"># take time at the beginning of the task</span>
I.intra =<span class="st"> </span><span class="kw">numeric</span>(<span class="dt">length=</span><span class="dv">20</span>)
I.intra[<span class="dv">1</span>] =<span class="st"> </span><span class="kw">kmeans</span>(BIC.expr.transposed, <span class="dt">centers=</span><span class="dv">2</span>)<span class="op">$</span>totss
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="op">:</span><span class="dv">20</span>) {
  <span class="kw">message</span>(<span class="st">&quot;Running k-means with &quot;</span>, i, <span class="st">&quot; centers&quot;</span>)
  kmi &lt;-<span class="st"> </span><span class="kw">kmeans</span>(BIC.expr.transposed, <span class="dt">centers=</span>i)
  I.intra[i] &lt;-<span class="st"> </span>kmi<span class="op">$</span>tot.withinss
}

<span class="co"># Plot a curve showint the intra-cluster variance as a function of the number of clusters </span>
<span class="kw">plot</span>((<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>)<span class="op">-</span><span class="fl">0.5</span>, I.intra, <span class="dt">type=</span><span class="st">&quot;h&quot;</span>, <span class="dt">lwd=</span><span class="dv">2</span>, <span class="dt">col =</span> <span class="st">&quot;blue&quot;</span>, 
     <span class="dt">xlab =</span> <span class="st">&quot;k&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Intra-cluster variance&quot;</span>,
     <span class="dt">main =</span> <span class="st">&quot;k-mean: impact of k on the intra-cluster variance&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/TP_intra_cluster_variance-1.png" alt="Intra-cluster variance plot. for a series of k-mean clustering with increasing values of k.  " width="768" />
<p class="caption">
Intra-cluster variance plot. for a series of k-mean clustering with increasing values of k.
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Measure elapsed time</span>
T2 =<span class="st"> </span><span class="kw">Sys.time</span>()
Tdiff =<span class="st"> </span><span class="kw">difftime</span>(T2,T1) ## Measure elapsed time

## Run k-means clustering with 2 clusters
BIC.kmeans2 &lt;-<span class="st"> </span><span class="kw">kmeans</span>(BIC.expr.transposed, <span class="dt">centers =</span> <span class="dv">2</span>)
<span class="kw">table</span>(BIC.kmeans2<span class="op">$</span>cluster) ## Cluster sizes</code></pre></div>
<pre><code>
  1   2 
661 158 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Run k-means clustering with 3 clusters
BIC.kmeans3 &lt;-<span class="st"> </span><span class="kw">kmeans</span>(BIC.expr.transposed, <span class="dt">centers =</span> <span class="dv">3</span>)
<span class="kw">table</span>(BIC.kmeans3<span class="op">$</span>cluster) ## Cluster sizes</code></pre></div>
<pre><code>
  1   2   3 
244 435 140 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Run k-means clustering with 10 clusters
BIC.kmeans10 &lt;-<span class="st"> </span><span class="kw">kmeans</span>(BIC.expr.transposed, <span class="dt">centers =</span> <span class="dv">10</span>)
<span class="kw">table</span>(BIC.kmeans10<span class="op">$</span>cluster) ## Cluster sizes</code></pre></div>
<pre><code>
  1   2   3   4   5   6   7   8   9  10 
143  79  16 127  63  36  57 104  58 136 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Cut the tree at some arbitrary levels to get clusters

## 2 clusters to see if the first subdivision corresponds to one of the 3 markers (ER1, PR1, Her2)
BIC.cutree2 &lt;-<span class="st"> </span><span class="kw">cutree</span>(BIC.hclust.ward, <span class="dt">k =</span> <span class="dv">2</span>) 

## 5 clusters to see if the match the cancer types defined by biologists
BIC.cutree5 &lt;-<span class="st"> </span><span class="kw">cutree</span>(BIC.hclust.ward, <span class="dt">k =</span> <span class="dv">5</span>)


## Define sample colors reflecting the cluster membership
hclust.k2.colors &lt;-<span class="st"> </span>BIC.cutree2
kmeans.k2.colors &lt;-<span class="st"> </span>BIC.kmeans2<span class="op">$</span>cluster
kmeans.k10.colors &lt;-<span class="st"> </span>BIC.kmeans10<span class="op">$</span>cluster

## Define characters reflecting markers
pch.cancer.type &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(BIC.sample.classes<span class="op">$</span>cancer.type)
pch.er1 &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(BIC.sample.classes<span class="op">$</span>ER1)
pch.pr1 &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(BIC.sample.classes<span class="op">$</span>PR1)
pch.her2 &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(BIC.sample.classes<span class="op">$</span>Her2)

## Compare clusters and markers on the PC plot, to evaluate whether 
## the components capture relevant information
<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">plot</span>(BIC.prcomp<span class="op">$</span>x[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], 
     <span class="dt">col =</span> hclust.k2.colors, 
     <span class="dt">pch =</span> pch.er1,
     <span class="dt">las =</span> <span class="dv">1</span>, <span class="dt">cex =</span> <span class="fl">0.7</span>, 
     <span class="dt">main =</span> <span class="st">&quot;Hclust 2 clusters versus markers&quot;</span>)
<span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="dt">cex =</span> <span class="fl">0.8</span>, <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&quot;ER1+&quot;</span>, <span class="st">&quot;ER1-&quot;</span>), <span class="dt">pch =</span> )

<span class="kw">plot</span>(BIC.prcomp<span class="op">$</span>x[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], 
     <span class="dt">col =</span> kmeans.k10.colors, 
     <span class="dt">pch =</span> pch.er1,
     <span class="dt">las =</span> <span class="dv">1</span>, <span class="dt">cex =</span> <span class="fl">0.7</span>, 
     <span class="dt">main =</span> <span class="st">&quot;K-means 10 clusters versus markers&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/TP_intra_cluster_variance-2.png" alt="Intra-cluster variance plot. for a series of k-mean clustering with increasing values of k.  " width="768" />
<p class="caption">
Intra-cluster variance plot. for a series of k-mean clustering with increasing values of k.
</p>
</div>
</div>
<div id="comparaisons" class="section level3">
<h3>Comparaisons</h3>
<div id="kmeans-versus-hclust" class="section level4">
<h4>kmeans versus hclust</h4>
<p>Nous allons maintenant comparer les résultats de ces deux méthodes de clustering.</p>
<ol style="list-style-type: decimal">
<li><p>à l’aide de la fonction <code>table</code>, calculez la matrice de confusion de vos deux clustering. Commentez.</p></li>
<li><p>à l’aide de la fonction <code>adjustedRand(clues)</code> calculez le RI et le ARI de vos clustering. Commentez.</p></li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">plot</span>(BIC.hclust.ward, <span class="dt">labels=</span><span class="ot">FALSE</span>, <span class="dt">hang=</span><span class="op">-</span><span class="dv">1</span>, <span class="dt">main =</span> <span class="st">&quot;Ward&quot;</span>)

<span class="kw">rect.hclust</span>(BIC.hclust.ward, <span class="dt">k=</span><span class="dv">3</span>)

BIC.cutree3 &lt;-<span class="st"> </span><span class="kw">cutree</span>(BIC.hclust.ward, <span class="dt">k=</span><span class="dv">3</span>)

<span class="kw">plot</span>(BIC.hclust.ward, <span class="dt">labels =</span> F, <span class="dt">hang =</span> <span class="op">-</span><span class="dv">1</span>, <span class="dt">main =</span> <span class="st">&quot;complete&quot;</span>)

<span class="kw">rect.hclust</span>(BIC.hclust.ward, <span class="dt">k =</span> <span class="dv">10</span>)</code></pre></div>
<p><img src="figures/TP_unnamed-chunk-7-1.png" width="480" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">BIC.cutree10 &lt;-<span class="st"> </span><span class="kw">cutree</span>(BIC.hclust.ward, <span class="dt">k =</span> <span class="dv">10</span>)

<span class="kw">table</span>(BIC.cutree3, BIC.kmeans<span class="op">$</span>cluster)</code></pre></div>
<pre><code>           
BIC.cutree3   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20
          1 114   2  59   0   0   0  16  32   0   0  27  31  70  42   0   0   4  65  10   0
          2   0   0   0   0   7  33   0   0   5  23   0   0   0   0  31   9   0   0   0  24
          3   1  11   3  57   0   0   0   0   0   0  21   5   0  40   1   0  66   9   0   1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(BIC.cutree10, BIC.kmeans<span class="op">$</span>cluster)</code></pre></div>
<pre><code>            
BIC.cutree10   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20
          1  107   0  55   0   0   0   2   5   0   0  21   1   3  40   0   0   4  23   0   0
          2    0   0   0   0   7  33   0   0   0  13   0   0   0   0   3   9   0   0   0   1
          3    0   0   0  51   0   0   0   0   0   0   0   0   0   0   1   0   3   0   0   1
          4    1  11   3   6   0   0   0   0   0   0  21   5   0  40   0   0  63   9   0   0
          5    0   0   0   0   0   0  14   1   0   0   0   0   0   0   0   0   0   1   0   0
          6    0   0   0   0   0   0   0   0   0  10   0   0   0   0  28   0   0   0   0  23
          7    7   0   2   0   0   0   0  26   0   0   6  30   0   2   0   0   0  41   0   0
          8    0   2   2   0   0   0   0   0   0   0   0   0  66   0   0   0   0   0   0   0
          9    0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0  10   0
          10   0   0   0   0   0   0   0   0   5   0   0   0   0   0   0   0   0   0   0   0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))

clues<span class="op">::</span><span class="kw">adjustedRand</span>(BIC.cutree3, BIC.kmeans3<span class="op">$</span>cluster)</code></pre></div>
<pre><code>     Rand        HA        MA        FM   Jaccard 
0.6742643 0.3286149 0.3297773 0.6058162 0.4341984 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clues<span class="op">::</span><span class="kw">adjustedRand</span>(BIC.cutree10, BIC.kmeans10<span class="op">$</span>cluster)</code></pre></div>
<pre><code>     Rand        HA        MA        FM   Jaccard 
0.8481182 0.4189739 0.4228339 0.5141232 0.3377504 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clues<span class="op">::</span><span class="kw">adjustedRand</span>(BIC.cutree3, BIC.kmeans10<span class="op">$</span>cluster)</code></pre></div>
<pre><code>     Rand        HA        MA        FM   Jaccard 
0.6586272 0.2323173 0.2344287 0.4545910 0.2348100 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clues<span class="op">::</span><span class="kw">adjustedRand</span>(BIC.cutree10, BIC.kmeans3<span class="op">$</span>cluster)</code></pre></div>
<pre><code>     Rand        HA        MA        FM   Jaccard 
0.6745688 0.2536610 0.2557085 0.4747715 0.2824229 </code></pre>
</div>
<div id="clustering-versus-statut" class="section level4">
<h4>clustering versus statut</h4>
<p>Nous connaissons les types de cancer des différentes tumeurs, définie en combinant trois marqueurs immunologiques :</p>
<ul>
<li>HER2,</li>
<li>ER1 (récepteur d’œstrogène)</li>
<li>PR1 (récepteur de progestérone)</li>
</ul>
<p>et nous obtenons les classes suivantes :</p>
<ul>
<li>Basal.like</li>
<li>HER2pos</li>
<li>Luminal.A</li>
<li>Luminal.B</li>
</ul>
<p>qqs tumeurs sont non classées</p>
<p>Vous pouvez lire les données concernant le type de cancer grâce à la fonction <code>read.table</code>, la ligne de commande est : <code>mes.classes &lt;- read.table(&quot;../../xxxx/BIC_sample-classes.tsv&quot;, h=T)</code>. A l’aide de la fonction <code>summary</code>, déterminez le nombre de tumeurs pour chaque type de cancer</p>
<ol style="list-style-type: decimal">
<li>comparez vos résultats de clustering avec la réalité
<ul>
<li>par des visualisations</li>
<li>le calcul de la matrice de confusion</li>
<li>le calcul des rand index et adjusted rand index</li>
</ul></li>
<li>Interprétez les résultats suivants (cliquez sur “code” pour afficher le code, et exécutez-le)</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Correspondence between markers
<span class="kw">kable</span>(<span class="kw">table</span>(BIC.sample.classes<span class="op">$</span>ER1, BIC.sample.classes<span class="op">$</span>PR1))</code></pre></div>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">Negative</th>
<th align="right">Positive</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Negative</td>
<td align="right">172</td>
<td align="right">12</td>
</tr>
<tr class="even">
<td>Positive</td>
<td align="right">95</td>
<td align="right">540</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">fisher.test</span>(<span class="kw">table</span>(BIC.sample.classes<span class="op">$</span>ER1, BIC.sample.classes<span class="op">$</span>PR1))</code></pre></div>
<pre><code>
    Fisher&#39;s Exact Test for Count Data

data:  table(BIC.sample.classes$ER1, BIC.sample.classes$PR1)
p-value &lt; 0.00000000000000022
alternative hypothesis: true odds ratio is not equal to 1
95 percent confidence interval:
  42.94805 165.19652
sample estimates:
odds ratio 
  80.90786 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># cut the tree at different levels (2, 5)</span>
<span class="kw">plot</span>(BIC.hclust.ward, <span class="dt">labels =</span> <span class="ot">FALSE</span>, <span class="dt">hang =</span> <span class="op">-</span><span class="dv">1</span>)
<span class="kw">rect.hclust</span>(BIC.hclust.ward, <span class="dt">k =</span> <span class="dv">5</span>, <span class="dt">border =</span> <span class="st">&quot;blue&quot;</span>)
<span class="kw">rect.hclust</span>(BIC.hclust.ward, <span class="dt">k =</span> <span class="dv">2</span>, <span class="dt">border =</span> <span class="st">&quot;red&quot;</span>)
<span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, 
       <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&quot;2 clusters&quot;</span>, <span class="st">&quot;5 clusters&quot;</span>), 
       <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;red&quot;</span>, <span class="st">&quot;blue&quot;</span>), <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">cex =</span> <span class="fl">0.7</span>)</code></pre></div>
<p><img src="figures/TP_markers_vs_clusters-1.png" width="480" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## hclust 2 clusters versus each marker
<span class="kw">table</span>(BIC.cutree2, BIC.sample.classes<span class="op">$</span>Her2)</code></pre></div>
<pre><code>           
BIC.cutree2 Negative Positive
          1      508      179
          2      123        9</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(BIC.cutree2, BIC.sample.classes<span class="op">$</span>ER1)</code></pre></div>
<pre><code>           
BIC.cutree2 Negative Positive
          1       67      620
          2      117       15</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(BIC.cutree2, BIC.sample.classes<span class="op">$</span>PR1)</code></pre></div>
<pre><code>           
BIC.cutree2 Negative Positive
          1      142      545
          2      125        7</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## K-means 2 clusters versus each marker
<span class="kw">table</span>(BIC.kmeans2<span class="op">$</span>cluster, BIC.sample.classes<span class="op">$</span>ER1)</code></pre></div>
<pre><code>   
    Negative Positive
  1       45      616
  2      139       19</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(BIC.kmeans2<span class="op">$</span>cluster, BIC.sample.classes<span class="op">$</span>PR1)</code></pre></div>
<pre><code>   
    Negative Positive
  1      119      542
  2      148       10</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(BIC.kmeans2<span class="op">$</span>cluster, BIC.sample.classes<span class="op">$</span>Her2)</code></pre></div>
<pre><code>   
    Negative Positive
  1      497      164
  2      134       24</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## 5 clusters
<span class="kw">table</span>(BIC.cutree5, BIC.sample.classes<span class="op">$</span>Her2)</code></pre></div>
<pre><code>           
BIC.cutree5 Negative Positive
          1      337       54
          2      123        9
          3       16       40
          4       96       63
          5       59       22</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(BIC.cutree5, BIC.sample.classes<span class="op">$</span>ER1)</code></pre></div>
<pre><code>           
BIC.cutree5 Negative Positive
          1        4      387
          2      117       15
          3       42       14
          4        4      155
          5       17       64</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(BIC.cutree5, BIC.sample.classes<span class="op">$</span>cancer.type)</code></pre></div>
<pre><code>           
BIC.cutree5 Basal.like HER2pos Luminal.A Luminal.B Unclassified
          1          3       0       298        51           39
          2        104       7         1         0           20
          3         11      28         2         5           10
          4          1       2        80        47           29
          5         12       4        41        15            9</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cluster.vs.cancertype &lt;-<span class="st"> </span><span class="kw">table</span>(
  BIC.cutree5, 
  BIC.sample.classes<span class="op">$</span>cancer.type)

<span class="kw">kable</span>(cluster.vs.cancertype, <span class="dt">caption =</span> <span class="st">&quot;Hierarchical clusters (c = 5) versus cancer type. &quot;</span>)</code></pre></div>
<table>
<caption>Hierarchical clusters (c = 5) versus cancer type.</caption>
<thead>
<tr class="header">
<th align="right">Basal.like</th>
<th align="right">HER2pos</th>
<th align="right">Luminal.A</th>
<th align="right">Luminal.B</th>
<th align="right">Unclassified</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">3</td>
<td align="right">0</td>
<td align="right">298</td>
<td align="right">51</td>
<td align="right">39</td>
</tr>
<tr class="even">
<td align="right">104</td>
<td align="right">7</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">20</td>
</tr>
<tr class="odd">
<td align="right">11</td>
<td align="right">28</td>
<td align="right">2</td>
<td align="right">5</td>
<td align="right">10</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="right">2</td>
<td align="right">80</td>
<td align="right">47</td>
<td align="right">29</td>
</tr>
<tr class="odd">
<td align="right">12</td>
<td align="right">4</td>
<td align="right">41</td>
<td align="right">15</td>
<td align="right">9</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## clustering versus cancer.type
## Compute the confusion table and adjusted RAND index (ARI)

## 2 clusters versus each marker
<span class="kw">table</span>(BIC.cutree2, BIC.sample.classes<span class="op">$</span>ER1)</code></pre></div>
<pre><code>           
BIC.cutree2 Negative Positive
          1       67      620
          2      117       15</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clues<span class="op">::</span><span class="kw">adjustedRand</span>(
  BIC.cutree2, 
  <span class="kw">as.numeric</span>(BIC.sample.classes<span class="op">$</span>ER1))</code></pre></div>
<pre><code>     Rand        HA        MA        FM   Jaccard 
0.8195844 0.5810808 0.5813049 0.8707027 0.7688294 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(BIC.cutree2, BIC.sample.classes<span class="op">$</span>PR1)</code></pre></div>
<pre><code>           
BIC.cutree2 Negative Positive
          1      142      545
          2      125        7</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clues<span class="op">::</span><span class="kw">adjustedRand</span>(
  BIC.cutree2, 
  <span class="kw">as.numeric</span>(BIC.sample.classes<span class="op">$</span>PR1))</code></pre></div>
<pre><code>     Rand        HA        MA        FM   Jaccard 
0.7019742 0.3692338 0.3696219 0.7755577 0.6244903 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(BIC.cutree2, BIC.sample.classes<span class="op">$</span>Her2)</code></pre></div>
<pre><code>           
BIC.cutree2 Negative Positive
          1      508      179
          2      123        9</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clues<span class="op">::</span><span class="kw">adjustedRand</span>(
  BIC.cutree2, 
  <span class="kw">as.numeric</span>(BIC.sample.classes<span class="op">$</span>Her2))</code></pre></div>
<pre><code>       Rand          HA          MA          FM     Jaccard 
 0.53388801 -0.07618062 -0.07559919  0.66226224  0.49369937 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## 5 clusters versus cancer type
<span class="kw">table</span>(BIC.cutree5, BIC.sample.classes<span class="op">$</span>cancer.type)</code></pre></div>
<pre><code>           
BIC.cutree5 Basal.like HER2pos Luminal.A Luminal.B Unclassified
          1          3       0       298        51           39
          2        104       7         1         0           20
          3         11      28         2         5           10
          4          1       2        80        47           29
          5         12       4        41        15            9</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clues<span class="op">::</span><span class="kw">adjustedRand</span>(
  BIC.cutree5, 
  <span class="kw">as.numeric</span>(BIC.sample.classes<span class="op">$</span>cancer.type))</code></pre></div>
<pre><code>     Rand        HA        MA        FM   Jaccard 
0.7105600 0.3330860 0.3348268 0.5451923 0.3743410 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Negative control: compute the same stat with randomly permuted values
<span class="kw">table</span>(BIC.cutree5, BIC.sample.classes<span class="op">$</span>cancer.type)</code></pre></div>
<pre><code>           
BIC.cutree5 Basal.like HER2pos Luminal.A Luminal.B Unclassified
          1          3       0       298        51           39
          2        104       7         1         0           20
          3         11      28         2         5           10
          4          1       2        80        47           29
          5         12       4        41        15            9</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clues<span class="op">::</span><span class="kw">adjustedRand</span>(
  <span class="kw">sample</span>(BIC.cutree5), 
  <span class="kw">as.numeric</span>(BIC.sample.classes<span class="op">$</span>cancer.type))</code></pre></div>
<pre><code>      Rand         HA         MA         FM    Jaccard 
0.57235104 0.01463136 0.01720336 0.32763791 0.19572626 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(BIC.kmeans3<span class="op">$</span>cluster, BIC.sample.classes<span class="op">$</span>cancer.type)</code></pre></div>
<pre><code>   
    Basal.like HER2pos Luminal.A Luminal.B Unclassified
  1         22      31       111        41           39
  2          1       0       310        77           47
  3        108      10         1         0           21</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clues<span class="op">::</span><span class="kw">adjustedRand</span>(
  BIC.kmeans3<span class="op">$</span>cluster,
  <span class="kw">as.numeric</span>(BIC.sample.classes<span class="op">$</span>cancer.type))</code></pre></div>
<pre><code>     Rand        HA        MA        FM   Jaccard 
0.6658905 0.2828718 0.2843799 0.5447041 0.3720078 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(BIC.kmeans10<span class="op">$</span>cluster, BIC.sample.classes[,<span class="dv">1</span>])</code></pre></div>
<pre><code>    
     Basal.like HER2pos Luminal.A Luminal.B Unclassified
  1           2       0       112        17           12
  2          12       4        39        15            9
  3           0       0        13         3            0
  4           0       0        99        21            7
  5           0       0        45         8           10
  6          23       1         1         2            9
  7           0       0        36        12            9
  8          85       6         0         0           13
  9           7      30         2         8           11
  10          2       0        75        32           27</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clues<span class="op">::</span><span class="kw">adjustedRand</span>(BIC.kmeans10<span class="op">$</span>cluster,
                    <span class="kw">as.numeric</span>(BIC.sample.classes[,<span class="dv">1</span>]))</code></pre></div>
<pre><code>     Rand        HA        MA        FM   Jaccard 
0.6788707 0.1385942 0.1418883 0.3302313 0.1726480 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Visualisation</span>


## hclust (2 groupes) et HER2+/-
<span class="kw">plot</span>(BIC.prcomp<span class="op">$</span>x[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], 
     <span class="dt">col =</span> hclust.k2.colors,
     <span class="dt">pch =</span> <span class="kw">as.numeric</span>(BIC.sample.classes<span class="op">$</span>ER1), <span class="co"># character indicates ER1 status</span>
     <span class="dt">las =</span> <span class="dv">1</span>, <span class="dt">cex =</span> <span class="fl">0.7</span>, 
     <span class="dt">main =</span> <span class="st">&quot;Clusters versus markers&quot;</span>)
<span class="kw">grid</span>()
texte.legend &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;ER1-&quot;</span>, <span class="st">&quot;ER1+&quot;</span>, <span class="st">&quot;gr1&quot;</span>, <span class="st">&quot;gr2&quot;</span>)
<span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, texte.legend, 
       <span class="dt">col =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">2</span>), 
       <span class="dt">pch =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="ot">NA</span>, <span class="ot">NA</span>), 
       <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">2</span>))</code></pre></div>
<p><img src="figures/TP_unnamed-chunk-9-1.png" width="80%" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Draw a confusion table of cluster versus ER1
<span class="kw">kable</span>(<span class="kw">table</span>(BIC.cutree2, BIC.sample.classes<span class="op">$</span>ER1), <span class="dt">caption =</span> <span class="st">&quot;hclust clusters versus ER1 marker&quot;</span>)</code></pre></div>
<table>
<caption>hclust clusters versus ER1 marker</caption>
<thead>
<tr class="header">
<th align="right">Negative</th>
<th align="right">Positive</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">67</td>
<td align="right">620</td>
</tr>
<tr class="even">
<td align="right">117</td>
<td align="right">15</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="lister-son-environnement" class="section level2">
<h2>Lister son environnement</h2>
<p>A la fin de tout travail d’analyse, il est important de conserver une trace précise et complète de l’environnement précis utilisé pour produire les résultats.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Print the complete list of libraries + versions used in this session
<span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>R version 3.5.2 (2018-12-20)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Mojave 10.14.3

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] grid      stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] ClassDiscovery_3.3.9  oompaBase_3.2.6       cluster_2.0.7-1       ComplexHeatmap_1.20.0 RColorBrewer_1.1-2    clues_0.5.9           FactoMineR_1.41       knitr_1.21           

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.0           magrittr_1.5         MASS_7.3-51.1        leaps_3.0            mclust_5.4.2         scatterplot3d_0.3-41 colorspace_1.4-0     lattice_0.20-38      rjson_0.2.20         highr_0.7            stringr_1.4.0        tools_3.5.2          circlize_0.4.5       xfun_0.5            
[15] htmltools_0.3.6      yaml_2.2.0           digest_0.6.18        oompaData_3.1.1      GlobalOptions_0.1.0  shape_1.4.4          evaluate_0.13        rmarkdown_1.11       stringi_1.3.1        compiler_3.5.2       flashClust_1.01-2    GetoptLong_0.1.7    </code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
