<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Olivier Sand and Jacques van Helden" />

<meta name="date" content="2020-06-02" />

<title>Practical - exploring omics data</title>

<script src="pavkovic2019_exercises_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="pavkovic2019_exercises_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="pavkovic2019_exercises_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="pavkovic2019_exercises_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="pavkovic2019_exercises_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="pavkovic2019_exercises_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="pavkovic2019_exercises_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="pavkovic2019_exercises_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="pavkovic2019_exercises_files/navigation-1.1/tabsets.js"></script>
<script src="pavkovic2019_exercises_files/navigation-1.1/codefolding.js"></script>




<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Practical - exploring omics data</h1>
<h3 class="subtitle">DUBii 2020</h3>
<h4 class="author">Olivier Sand and Jacques van Helden</h4>
<h4 class="date">2020-06-02</h4>

</div>


<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>The goal of this practical is to combine different statistical approaches to explore an omics dataset (transcriptome, proteome).</p>
<ul>
<li>graphical representations of the distribution</li>
<li>impact of the normalisation and standardisation</li>
<li>multidimensional scaling with Principal Component Analysis (PCA)</li>
<li>class discovery by clustering</li>
</ul>
</div>
<div id="normalisation-and-standardisation" class="section level2">
<h2>Normalisation and standardisation</h2>
<p>We prepared a notebook (in R markdown) detailing all the dirty cooking details to download the datasets from the Zenodo repository, and to run some preprocessing steps</p>
<ul>
<li>log2 transformation</li>
<li>filtering out of the undetected features (genes, proteins)</li>
<li>median-based sample-wise centering</li>
<li>IQR-based sample-wise scaling</li>
</ul>
</div>
<div id="data-loading" class="section level2">
<h2>Data loading</h2>
<p>We prepared memory images that enable you to reload the preprocessed datasets.</p>
<p><a href="R/01_reload_data.R">R script to reload data</a></p>
</div>
<div id="instructions" class="section level2">
<h2>Instructions</h2>
<p>The following tasks will be covered in 2 sessions + a bit of personal work.</p>
<ul>
<li>Write the solution of each exercise in a separate R script file.</li>
<li>Take care to properly document the code</li>
<li>Once the code is satisfying, write a report in R markdown that will incoroporate the R files, and comment the results.</li>
</ul>
<p>Alternatively you can immediately write the code inside a R markdown document, as far as it combines a properly documented code and relevant interpretation comments.</p>
</div>
<div id="exercises" class="section level2">
<h2>Exercises</h2>
<ol style="list-style-type: decimal">
<li><p>Choose one of the datasets and load the memory image</p></li>
<li><p>Compute sample-wise and feature-wise descriptive statistics</p>
<ul>
<li>mean,</li>
<li>median,</li>
<li>sd,</li>
<li>var,</li>
<li>IQR,</li>
<li><p>some relevant percentiles (0, 05, 25, 50 , 75, 95, 100)</p>
<p>Use different graphical representations to compare the values and get familiar with your data.</p></li>
</ul>
<p>For example :</p>
<ul>
<li>histogram of all the values</li>
<li>boxplot of the values per sample</li>
<li>feature means versus medians</li>
<li>feature standard deviation versus IQR</li>
<li>mean versus variance plot</li>
</ul></li>
<li><p>Compute the mean value per condition (mean between the replicates)</p>
<p>Draw a dot plot to compare the values between each time point and the control.</p></li>
<li><p>Select the 500 top features according to two different criteria</p>
<ol style="list-style-type: lower-alpha">
<li>highest variance</li>
<li>differential analysis (will be provided by the teachers)</li>
</ol></li>
<li><p>Compute different matrices indicating the relations between samples (columns) of the log2-transformed values</p>
<ul>
<li>covariance,</li>
<li>Pearson correlation</li>
<li>Spearman correlation</li>
<li>Euclidian distance</li>
</ul>
<p>Compute these on the commplete dataset and on the 500 matrix selections</p></li>
<li><p>Use the hclust() method on the seleected features</p>
<ul>
<li>with different dissimlarity metrics (Euclidian, Pearson, Spearman)</li>
<li>with different agglomeration rules (single, average, complete, ward)</li>
</ul>
<p>Plot the feature tree and compare the results obtained with the different choices.</p></li>
<li><p>Draw heatmaps of the feature profiles (do not cluster the samples)</p></li>
<li><p>Run principal component analysis</p>
<ul>
<li><p>Use the visualisation tools from Magali Berland’s course</p></li>
<li><p>Generate PC plots with</p>
<pre><code>- the whole dataset
- the subset of features declared positive in the differential analysis (provided by the teachers)</code></pre></li>
<li><p>Plot the sample positions on the 6 first components. Are they consistent with their time and status?</p></li>
</ul></li>
<li><p>Enrichment analysis</p>
<ul>
<li><p>analyse the features declared positive (provided by the teachers) to the gProfiler function profiling web tool (<a href="https://biit.cs.ut.ee/gprofiler/gost" class="uri">https://biit.cs.ut.ee/gprofiler/gost</a>)</p></li>
<li><p>write a piece of R code that runs the same anlaysis with the <code>gProfileR</code> package</p></li>
<li><p>analyse (either with R or with the Web site) the different gene groups discovered by the hclust approach, and evaluate if the results are more relevant when you submit all the differentially expressed features at once, or when you submit them cluster per cluster</p></li>
</ul></li>
</ol>
<hr />
</div>
<div id="modalités-de-contrôle-des-connaissances" class="section level2">
<h2>Modalités de contrôle des connaissances</h2>
<div id="note-dassiduité-25" class="section level3">
<h3>Note d’assiduité (25%)</h3>
<p>Présence physique et virtuelle, comme à l’école des fans, tous gagnants</p>
</div>
<div id="contrôle-continu-25" class="section level3">
<h3>Contrôle continu (25%)</h3>
<p>QCM dernière séance, sur les concepts de stat de l’ensemble des cours</p>
</div>
<div id="evaluation-finale-50" class="section level3">
<h3>Evaluation finale (50%)</h3>
<p>Choix entre deux sujets:</p>
<ol style="list-style-type: decimal">
<li>Terminer les exercices d’exploration des données</li>
<li>Comparer 2 méthodes de classification supervisée</li>
</ol>
<p>Date de Rendu:</p>
<ul>
<li>Le plus tard possible ? Septembre ? Demander à Bertrand ce qui est possible.</li>
</ul>
<p>Rendu:</p>
<ul>
<li>Un script bien documenté + un rapport d’interprétation (pdf avec figures, tables, texte …)</li>
<li>Optionnellement (pour ceux qui veulent pousser plus loin), un notebook Rmd qui inclut le code + les résultats, et peut être ré-utilisé par les formateurs. Joindre aussi le fichier compilé en pdf et / ou HTML self-contained avec code visible.</li>
</ul>
<p>Le rapport doit être synthétique:</p>
<ul>
<li>Intro / contexte : ~15 lignes</li>
<li>Mat é Méthodes: une explication synthétique (éventuellement graphique, flow chart) du déroulé de l’analyse</li>
<li>Résultats
<ul>
<li>utiliser des représentations graphiques + tableaux représentatifs (max 10)</li>
<li>quelques phrase d’interprétation pour chaque figure ou tableau</li>
</ul></li>
<li>Discussion: conclusions et perspectives</li>
</ul>
<p>Le travail peut être réalisé à titre individuel ou en binôme.</p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
