<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Olivier Sand and Jacques van Helden" />

<meta name="date" content="2021-03-11" />

<title>Practical - exploring omics data</title>

<script src="pavkovic2019_exercises_files/header-attrs-2.5/header-attrs.js"></script>
<script src="pavkovic2019_exercises_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="pavkovic2019_exercises_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="pavkovic2019_exercises_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="pavkovic2019_exercises_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="pavkovic2019_exercises_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="pavkovic2019_exercises_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="pavkovic2019_exercises_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="pavkovic2019_exercises_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="pavkovic2019_exercises_files/navigation-1.1/tabsets.js"></script>
<script src="pavkovic2019_exercises_files/navigation-1.1/codefolding.js"></script>
<link href="pavkovic2019_exercises_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="pavkovic2019_exercises_files/anchor-sections-1.0/anchor-sections.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>




<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Practical - exploring omics data</h1>
<h3 class="subtitle">DUBii 2020</h3>
<h4 class="author">Olivier Sand and Jacques van Helden</h4>
<h4 class="date">2021-03-11</h4>

</div>


<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>The goal of this practical is to combine different statistical approaches to explore an omics dataset (transcriptome, proteome).</p>
<ul>
<li>graphical representations of the distribution</li>
<li>impact of the normalisation and standardisation</li>
<li>multidimensional scaling with Principal Component Analysis (PCA)</li>
<li>class discovery by clustering</li>
</ul>
</div>
<div id="normalisation-and-standardisation" class="section level2">
<h2>Normalisation and standardisation</h2>
<p>We prepared a notebook (in R markdown) detailing all the dirty cooking details to download the datasets from the Zenodo repository, and to run some preprocessing steps</p>
<ul>
<li>log2 transformation</li>
<li>filtering out of the undetected features (genes, proteins)</li>
<li>median-based sample-wise centering</li>
<li>IQR-based sample-wise scaling</li>
</ul>
</div>
<div id="data-loading" class="section level2">
<h2>Data loading</h2>
<p>We prepared memory images that enable you to reload the preprocessed datasets.</p>
<p><a href="R/01_reload_data.R">R script to reload data</a></p>
</div>
<div id="instructions" class="section level2">
<h2>Instructions</h2>
<p>The following tasks will be covered in 2 sessions + a bit of personal work.</p>
<ul>
<li>Write the solution of each exercise in a separate R script file.</li>
<li>Take care to properly document the code</li>
<li>Once the code is satisfying, write a report in R markdown that will incoroporate the R files, and comment the results.</li>
</ul>
<p>Alternatively you can immediately write the code inside a R markdown document, as far as it combines a properly documented code and relevant interpretation comments.</p>
</div>
<div id="exercises" class="section level2">
<h2>Exercises</h2>
<div id="data-loading-1" class="section level3">
<h3>1. Data loading</h3>
<p>Choose one of the datasets and load the memory image</p>
</div>
<div id="descriptive-statistics" class="section level3">
<h3>2. Descriptive statistics</h3>
<p>Compute sample-wise and feature-wise descriptive statistics</p>
<ul>
<li>mean,</li>
<li>median,</li>
<li>sd,</li>
<li>var,</li>
<li>IQR,</li>
<li>some relevant percentiles (0, 05, 25, 50 , 75, 95, 100)</li>
</ul>
<p>Use different graphical representations to compare the values and get familiar with your data. For example :</p>
<ul>
<li>histogram of all the values</li>
<li>boxplot of the values per sample</li>
<li>feature means versus medians</li>
<li>feature standard deviation versus IQR</li>
<li>mean versus variance plot</li>
</ul>
<p>Solution by the teachers: [<a href="https://raw.githubusercontent.com/DU-Bii/module-3-Stat-R/master/stat-R_2020/practicals/pavkovic_2019/R/02_descriptive_stats.R">02_descriptive_stats.R</a>]</p>
</div>
<div id="summary-per-condition" class="section level3">
<h3>3. Summary per condition</h3>
<ul>
<li>Compute the mean value per condition (mean between the replicates)</li>
<li>Draw a dot plot to compare the values between each time point and the control.</li>
</ul>
</div>
<div id="feature-selection" class="section level3">
<h3>4. Feature selection</h3>
<p>Select the 500 top features according to two different criteria</p>
<ol style="list-style-type: lower-alpha">
<li>highest variance</li>
<li>differential analysis (will be provided by the teachers)</li>
</ol>
</div>
<div id="comparison-metrics" class="section level3">
<h3>5. Comparison metrics</h3>
<p>Compute different metrics indicating the relations between samples (columns) of the log2-transformed values</p>
<ul>
<li>covariance</li>
<li>Pearson correlation + derived distance matrix</li>
<li>Spearman correlation + derived distance matrix</li>
<li>Euclidian distance</li>
</ul>
<p>Compute these on the commplete dataset and on the 500 matrix selections</p>
<p>Draw graphical representations of the result with <code>corrplot()</code>.</p>
<p>Comment the results</p>
<ul>
<li>differences between metrics</li>
<li>differences between results with all the features and selected features</li>
</ul>
</div>
<div id="clustering" class="section level3">
<h3>6. Clustering</h3>
<ul>
<li><p>Run hierarchical clustering with <code>hclust()</code> to extract clusters from the dataset with selected variables.</p>
<ul>
<li>with different dissimlarity metrics (Euclidian, Pearson, Spearman)</li>
<li>with different agglomeration rules (single, average, complete, ward)</li>
</ul></li>
<li><p>Plot the feature trees and compare the results obtained with the different choices.</p></li>
<li><p>Plot heatmpas with the different feature trees obtained before, and compare the results. Inactivate the individual clustering (default in the heatmap).</p></li>
</ul>
</div>
<div id="principal-component-analysis-pca" class="section level3">
<h3>8. Principal Component Analysis (PCA)</h3>
<p>We provide hereby</p>
<ul>
<li><p>An R script with the principal component analysis of the expression set: [<a href="https://raw.githubusercontent.com/DU-Bii/module-3-Stat-R/master/stat-R_2020/practicals/pavkovic_2019/R/04_PCA.R">04_PCA.R</a>]</p></li>
<li><p>An R markdown model of scientific report: [<a href="https://raw.githubusercontent.com/DU-Bii/module-3-Stat-R/master/stat-R_2020/templates/YOUR-NAME_REPORT-TOPICS.Rmd">YOUR-NAME_REPORT-TOPICS.Rmd</a>]</p></li>
</ul>
<p>Write a short report that will integrate some pieces of chunks from the R script in the R markdown notebook, add your interpretation of the results, and compile it as a self-contained HTML file.</p>
</div>
<div id="functionnal-enrichment-analysis" class="section level3">
<h3>9. Functionnal enrichment analysis</h3>
<ul>
<li><p>analyse the features declared positive (provided by the teachers) to the gProfiler function profiling web tool (<a href="https://biit.cs.ut.ee/gprofiler/gost" class="uri">https://biit.cs.ut.ee/gprofiler/gost</a>)</p></li>
<li><p>write a piece of R code that runs the same anlaysis with the <code>gProfileR</code> package</p></li>
<li><p>analyse (with either R or the Web site) the different gene groups discovered by the hclust approach, and evaluate if the results are more relevant when you submit all the differentially expressed features at once, or when you submit them cluster per cluster</p></li>
</ul>
<hr />
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
