<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Jacques van Helden" />

<meta name="date" content="2019-03-07" />

<title>Détection de gènes différentiellement exprimés à partir de données RNA-seq</title>

<script src="TP_differential_analysis_RNA-seq_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="TP_differential_analysis_RNA-seq_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="TP_differential_analysis_RNA-seq_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="TP_differential_analysis_RNA-seq_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="TP_differential_analysis_RNA-seq_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="TP_differential_analysis_RNA-seq_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="TP_differential_analysis_RNA-seq_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="TP_differential_analysis_RNA-seq_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="TP_differential_analysis_RNA-seq_files/navigation-1.1/tabsets.js"></script>
<script src="TP_differential_analysis_RNA-seq_files/navigation-1.1/codefolding.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #303030; color: #cccccc; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; }
td.sourceCode { padding-left: 5px; }
pre, code { color: #cccccc; background-color: #303030; }
code > span.kw { color: #f0dfaf; } /* Keyword */
code > span.dt { color: #dfdfbf; } /* DataType */
code > span.dv { color: #dcdccc; } /* DecVal */
code > span.bn { color: #dca3a3; } /* BaseN */
code > span.fl { color: #c0bed1; } /* Float */
code > span.ch { color: #dca3a3; } /* Char */
code > span.st { color: #cc9393; } /* String */
code > span.co { color: #7f9f7f; } /* Comment */
code > span.ot { color: #efef8f; } /* Other */
code > span.al { color: #ffcfaf; } /* Alert */
code > span.fu { color: #efef8f; } /* Function */
code > span.er { color: #c3bf9f; } /* Error */
code > span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */
code > span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code > span.sc { color: #dca3a3; } /* SpecialChar */
code > span.vs { color: #cc9393; } /* VerbatimString */
code > span.ss { color: #cc9393; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { } /* Variable */
code > span.cf { color: #f0dfaf; } /* ControlFlow */
code > span.op { color: #f0efd0; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code > span.at { } /* Attribute */
code > span.do { color: #7f9f7f; } /* Documentation */
code > span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code > span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code > span.in { color: #7f9f7f; font-weight: bold; } /* Information */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<div class="container-fluid main-container">

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Détection de gènes différentiellement exprimés à partir de données RNA-seq</h1>
<h3 class="subtitle"><em>Diplôme Interuniversitaire en Bioinformatique intégrative (DU-Bii 2019)</em></h3>
<h4 class="author"><em>Jacques van Helden</em></h4>
<h4 class="date"><em>2019-03-07</em></h4>

</div>


<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Depuis l’avénement du du séquençage massivement parallèle (NGS, Next Generation Sequencing) en 2007, la détection de gènes différentiellement exprimés (DEG) à partir de données transcriptomiques RNA-seq constitue l’une de ses application les plus populaires.</p>
<p>Le principe est de mesurer, dans différentes conditions, la concentration d’ARN correspondant à chaque gène, et de comparer ces concentrations entre les échantillons de deux (<strong>comparaison à 2 groupes</strong> ou <strong>binaire</strong>) ou plusieurs conditions (comparaison <strong>multi-groupes</strong>).</p>
<p>Dès les premières analyses, les chercheurs se sont rendus compte que les méthodes classiques d’analyse différentielles (tests de Student, ANOVA) ne se prêtaient pas du tout à l’analyse de telles données, car leur nature diffère fondamentalement des hypothèss de travail sous-jacentes, pour différentes raisons.</p>
<ol style="list-style-type: decimal">
<li><p>Le niveau d’expression d’un gène est quantifié sur une échelle discrète (comptage du nombre de lectures alignées qui chevauchent le gène), alors que les tests paramétriques reposent sur une hypoothèse de normalité.</p></li>
<li><p>Les ordres de grandeur des comptages varient fortement d’un gène à l’autre: certains gènes ont une poignée de comptages par échantillons, d’autres des centaines de milliers. Les gènes très très fortement représentés correspondent généralement à des gènes non-codants (par exemple ARN ribosomique) ou à des fonctions cellulaires particulières liées au métabolisme de l’ARN, et on comprend les raisons biologiques de leur sur-représentation. Ils n’en constituent pas moins ce qu’on appelle en statistique des “valeurs aberrantes” (“outliers”).</p></li>
<li><p>Les distributions de comptages comptent généralement un très grand nombre de valeurs nulles (“zero-inflated distributions”).</p></li>
<li><p>Ces particularités posent des problèmes particuliers pour la normalisation des librairies de comptages :</p>
<ul>
<li><p>la présence d’outliers affecte fortement la moyenne, et de façon très instable, ce qui biaise fortement l’estimation de la tendance centrale.</p></li>
<li><p>la stratégie de repli sur des estimateurs robustes, comme la médiane est contestable du fait du très grand nombre de valeurs nulles (dans certains cas, plusn d’un quart voire la moitié des gènes ont une valeur nulle pour un échantillon).</p></li>
</ul></li>
</ol>
<p>Des méthodes spécifiques ont donc été développées dès 2010 pour affronter ces difficultés particulières de la normalisation et de l’analyse différentielle des données de RNA-seq.</p>
</div>
<div id="but-de-ce-tp" class="section level2">
<h2>But de ce TP</h2>
<p>Le but de ce TP est d’effectuer une première exploration de l’analyse différentielle des données d’expression, sur base d’un petit cas d’étude simple: l’analyse transcriptionnelle de mutants de sporulation chez la levure <strong>Saccharomyces cerevisiae</strong>.</p>
</div>
<div id="cas-detude" class="section level2">
<h2>Cas d’étude</h2>
<p>… A ECRIRE: DESCRIPTION DU CAS D’ETUDE</p>
</div>
<div id="sources-des-donnees" class="section level2">
<h2>Sources des données</h2>
</div>
<div id="parametres-de-lanalyse" class="section level2">
<h2>Paramètres de l’analyse</h2>
<p>Nous définissons dans une variable R (de type liste) les paramètres de l’analyse. Ceci nous permettra de reproduire ultérieurement exactement la même succession d’étapes soit en utilisant des données différentes, soit en modifiant les paramètres particuliers (design, seuils, …).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Load libraries
<span class="kw">message</span>(<span class="st">&quot;Loading libraries&quot;</span>)
<span class="kw">library</span>(knitr)
<span class="co">#library(kableExtra) ## Note: kableExtra has some side effect on kable: column padding is null, so all numbers seem to be mixed up</span>
<span class="co">#library(FactoMineR)</span>
<span class="co"># library(clues)</span>
<span class="co">#library(RColorBrewer)</span>
<span class="co"># library(ComplexHeatmap)</span>
<span class="kw">library</span>(vioplot)
<span class="kw">library</span>(DESeq2)
<span class="kw">library</span>(edgeR)

<span class="co"># library(corrplot)</span>
<span class="co"># library(ClassDiscovery)</span>
<span class="co"># library(formattable)</span>

<span class="co"># message(&quot;getwd()\t&quot;, getwd())</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">parameters &lt;-<span class="st"> </span><span class="kw">list</span>(
  <span class="co"># data.folder = &quot;data/GSE89530&quot;, # dossier des données locales</span>
  <span class="dt">data.folder =</span> <span class="st">&quot;/shared/projects/du_bii_2019/data/module3/seance5/GSE89530&quot;</span>, <span class="co"># on the IFB-cluster-core</span>
  <span class="dt">workdir =</span> <span class="st">&quot;~/DU-Bii/m3s5/&quot;</span>, <span class="co"># Directory to export the results</span>
  <span class="dt">counts =</span> <span class="st">&quot;GSE89530_counts.tsv.gz&quot;</span>,
  <span class="dt">sample.descr =</span> <span class="st">&quot;GSE89530_samples.tsv.gz&quot;</span>,
  <span class="dt">alpha =</span> <span class="fl">0.05</span>, <span class="co"># seuil de significativité</span>
  <span class="dt">epsilon =</span> <span class="fl">0.1</span>, <span class="co"># pseudo-comptage pour la transformation log2</span>
  <span class="dt">gene.filter.percent.zeros =</span> <span class="dv">90</span>, <span class="co"># Discard genes having zero values in more than this percentage of samples</span>
  <span class="dt">gene.filter.min.count =</span> <span class="dv">10</span>, <span class="co"># Minimal counts per gene</span>
  
  ## Test and control groups for differential analysis
  <span class="dt">test.group =</span> <span class="st">&quot;bdf1_Y187F_Y354F_mutant_8&quot;</span>,
  <span class="dt">control.group =</span> <span class="st">&quot;Wild_type_8&quot;</span>
)

## Create output directory
<span class="kw">dir.create</span>(parameters<span class="op">$</span>workdir, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>, <span class="dt">recursive =</span> <span class="ot">TRUE</span>)

## Print the parameter values in the report
<span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">data.frame</span>(parameters)), <span class="dt">col.names =</span> <span class="st">&quot;Parameter Value&quot;</span>)</code></pre></div>
<table>
<thead>
<tr class="header">
<th></th>
<th align="left">Parameter Value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>data.folder</td>
<td align="left">/shared/projects/du_bii_2019/data/module3/seance5/GSE89530</td>
</tr>
<tr class="even">
<td>workdir</td>
<td align="left">~/DU-Bii/m3s5/</td>
</tr>
<tr class="odd">
<td>counts</td>
<td align="left">GSE89530_counts.tsv.gz</td>
</tr>
<tr class="even">
<td>sample.descr</td>
<td align="left">GSE89530_samples.tsv.gz</td>
</tr>
<tr class="odd">
<td>alpha</td>
<td align="left">0.05</td>
</tr>
<tr class="even">
<td>epsilon</td>
<td align="left">0.1</td>
</tr>
<tr class="odd">
<td>gene.filter.percent.zeros</td>
<td align="left">90</td>
</tr>
<tr class="even">
<td>gene.filter.min.count</td>
<td align="left">10</td>
</tr>
<tr class="odd">
<td>test.group</td>
<td align="left">bdf1_Y187F_Y354F_mutant_8</td>
</tr>
<tr class="even">
<td>control.group</td>
<td align="left">Wild_type_8</td>
</tr>
</tbody>
</table>
</div>
<div id="telechargement-des-donnees" class="section level2">
<h2>Téléchargement des données</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## List files in the data folder
<span class="co"># list.files(parameters$data.folder)</span>

## Load sample descriptions
<span class="kw">message</span>(<span class="st">&quot;Loading sample descriptions from file</span><span class="ch">\t</span><span class="st">&quot;</span>, parameters<span class="op">$</span>sample.descr)
samples &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="kw">file.path</span>(parameters<span class="op">$</span>data.folder, parameters<span class="op">$</span>sample.descr), <span class="dt">row.names =</span> <span class="dv">1</span>, <span class="dt">header =</span> <span class="ot">TRUE</span>)
<span class="co"># View(samples) # Check visually the content of the samples data frame</span>

## Load counts of reads per gene
<span class="kw">message</span>(<span class="st">&quot;Loading counts from file</span><span class="ch">\t</span><span class="st">&quot;</span>, parameters<span class="op">$</span>counts)
counts &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="kw">file.path</span>(parameters<span class="op">$</span>data.folder, parameters<span class="op">$</span>counts), <span class="dt">row.names =</span> <span class="dv">1</span>, <span class="dt">header =</span> <span class="ot">TRUE</span>)

## Check that the count table contains the same sample IDs as the sample description table. If not, stop everything here.
<span class="cf">if</span> (<span class="op">!</span><span class="kw">setequal</span>(<span class="kw">rownames</span>(samples), <span class="kw">colnames</span>(counts))) {
  <span class="kw">stop</span>(<span class="st">&quot;The sample IDs differ between count table (column names) and sample descriptions (row names). &quot;</span>)
}

## Check that sample IDs are ordered in the same way between sample descriptions (rows) and count table (columns). 
differing.IDs &lt;-<span class="st"> </span><span class="kw">colnames</span>(counts) <span class="op">!=</span><span class="st"> </span><span class="kw">rownames</span>(samples)
<span class="cf">if</span> (<span class="kw">sum</span>(differing.IDs) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) {
  ## If not, re-order columnts of the count table to match the rows of the sample description.  
  <span class="kw">message</span>(<span class="st">&quot;Reordering columns of the count table in order to match the order of sample descrition row names&quot;</span>)
  counts &lt;-<span class="st"> </span>counts[, <span class="kw">rownames</span>(samples)]
  
}</code></pre></div>
<p>La table de comptages comporte 7478 lignes (correspondant aux gènes) et 18 colonnes (correspondant aux échantillons). Nous pouvons afficher un petit morceau de cette table, ens électionnant au hasard quelques lignes et colonnes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">some.genes &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(counts), <span class="dt">size =</span> <span class="dv">10</span>, <span class="dt">replace =</span> <span class="ot">FALSE</span>)
some.samples &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(counts), <span class="dt">size =</span> <span class="dv">4</span>, <span class="dt">replace =</span> <span class="ot">FALSE</span>)

<span class="kw">kable</span>(counts[some.genes, some.samples], <span class="dt">caption =</span> <span class="st">&quot;Random selection of some genes and some columns of the raw count table. &quot;</span>)</code></pre></div>
<table>
<caption>Random selection of some genes and some columns of the raw count table.</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">GSM2375285</th>
<th align="right">GSM2375295</th>
<th align="right">GSM2375282</th>
<th align="right">GSM2375288</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>YLR221C</td>
<td align="right">312</td>
<td align="right">486</td>
<td align="right">307</td>
<td align="right">156</td>
</tr>
<tr class="even">
<td>YJL037W</td>
<td align="right">706</td>
<td align="right">3270</td>
<td align="right">15</td>
<td align="right">12</td>
</tr>
<tr class="odd">
<td>YOR252W</td>
<td align="right">263</td>
<td align="right">477</td>
<td align="right">233</td>
<td align="right">148</td>
</tr>
<tr class="even">
<td>YKL166C</td>
<td align="right">129</td>
<td align="right">156</td>
<td align="right">109</td>
<td align="right">68</td>
</tr>
<tr class="odd">
<td>YJL052W</td>
<td align="right">666</td>
<td align="right">465</td>
<td align="right">856</td>
<td align="right">830</td>
</tr>
<tr class="even">
<td>YLL018C-A</td>
<td align="right">597</td>
<td align="right">450</td>
<td align="right">724</td>
<td align="right">482</td>
</tr>
<tr class="odd">
<td>YOR337W</td>
<td align="right">500</td>
<td align="right">1194</td>
<td align="right">729</td>
<td align="right">489</td>
</tr>
<tr class="even">
<td>YJL122W</td>
<td align="right">274</td>
<td align="right">634</td>
<td align="right">323</td>
<td align="right">117</td>
</tr>
<tr class="odd">
<td>YIR007W</td>
<td align="right">443</td>
<td align="right">169</td>
<td align="right">465</td>
<td align="right">350</td>
</tr>
<tr class="even">
<td>YPL144W</td>
<td align="right">75</td>
<td align="right">63</td>
<td align="right">100</td>
<td align="right">55</td>
</tr>
</tbody>
</table>
<p>Le tableau GSE89530_samples.tsv.gz fournit une description de chaque échantillon.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># print(samples[some.samples, ])</span>
<span class="kw">kable</span>(samples[some.samples, ], 
      <span class="dt">caption =</span> <span class="st">&quot;**Table de comptage de lectures (short reads) par gène.** Sélection arbitraire de quelques gènes (lignes) et échantillons (colonnes)&quot;</span>, 
      <span class="dt">align =</span> <span class="st">&quot;c&quot;</span>)</code></pre></div>
<table>
<caption><strong>Table de comptage de lectures (short reads) par gène.</strong> Sélection arbitraire de quelques gènes (lignes) et échantillons (colonnes)</caption>
<thead>
<tr class="header">
<th></th>
<th align="center">Condition</th>
<th align="center">Replicate</th>
<th align="center">Genotype</th>
<th align="center">Time.point</th>
<th align="center">Label</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>GSM2375285</td>
<td align="center">Wild_type_4</td>
<td align="center">2</td>
<td align="center">Wild-type</td>
<td align="center">4</td>
<td align="center">Sc_WT_4h_2</td>
</tr>
<tr class="even">
<td>GSM2375295</td>
<td align="center">bdf1_Y187F_Y354F_mutant_8</td>
<td align="center">3</td>
<td align="center">bdf1-Y187F-Y354F mutant</td>
<td align="center">8</td>
<td align="center">bd1bd2mut_8h_3</td>
</tr>
<tr class="odd">
<td>GSM2375282</td>
<td align="center">bdf1_Y187F_Y354F_mutant_4</td>
<td align="center">1</td>
<td align="center">bdf1-Y187F-Y354F mutant</td>
<td align="center">4</td>
<td align="center">bd1bd2mut_4h_1</td>
</tr>
<tr class="even">
<td>GSM2375288</td>
<td align="center">bdf1_Y187F_Y354F_mutant_4</td>
<td align="center">2</td>
<td align="center">bdf1-Y187F-Y354F mutant</td>
<td align="center">4</td>
<td align="center">bd1bd2mut_4h_2</td>
</tr>
</tbody>
</table>
</div>
<div id="exploration-des-donnees" class="section level2">
<h2>Exploration des données</h2>
<p>Avant toute autre chose, il convient de mener une exploration préliminaire des données, afin de se familiariser avec leur distribution.</p>
<div id="statistiques-descriptives" class="section level3">
<h3>Statistiques descriptives</h3>
<p>La fonction R <code>summary()</code> calcule des statistiques de base pour chaque colonne d’une matrice ou data frame. Nous imprimons ici un sous-ensemble de ces échantillons.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(counts[, some.samples])</code></pre></div>
<pre><code>   GSM2375285         GSM2375295         GSM2375282       GSM2375288      
 Min.   :     0.0   Min.   :     0.0   Min.   :     0   Min.   :     0.0  
 1st Qu.:    30.2   1st Qu.:    35.2   1st Qu.:    32   1st Qu.:    20.0  
 Median :   254.0   Median :   239.0   Median :   262   Median :   173.0  
 Mean   :   809.9   Mean   :   919.0   Mean   :   903   Mean   :   638.4  
 3rd Qu.:   595.0   3rd Qu.:   643.0   3rd Qu.:   647   3rd Qu.:   435.8  
 Max.   :472819.0   Max.   :511934.0   Max.   :542630   Max.   :425896.0  </code></pre>
<p>Nous pouvons calculer quelques paramètres additionnels, qui nous éclaireront sur les données.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">all.counts.vector &lt;-<span class="st"> </span><span class="kw">unlist</span>(counts) <span class="co"># Regroup all counts in a vevtor to compute statistics</span>

count.stats &lt;-<span class="st"> </span><span class="kw">data.frame</span>(
  <span class="dt">mean =</span> <span class="kw">mean</span>(all.counts.vector),
  <span class="dt">sd =</span> <span class="kw">sd</span>(all.counts.vector),
  <span class="dt">iqr =</span> <span class="kw">IQR</span>(all.counts.vector),
  <span class="dt">min =</span> <span class="kw">min</span>(all.counts.vector),
  <span class="dt">P05 =</span> <span class="kw">quantile</span>(all.counts.vector, <span class="dt">probs =</span> <span class="fl">0.05</span>),
  <span class="dt">Q1 =</span> <span class="kw">quantile</span>(all.counts.vector, <span class="dt">probs =</span> <span class="fl">0.25</span>),
  <span class="dt">median =</span> <span class="kw">quantile</span>(all.counts.vector, <span class="dt">probs =</span> <span class="fl">0.5</span>),
  <span class="dt">Q3 =</span> <span class="kw">quantile</span>(all.counts.vector, <span class="dt">probs =</span> <span class="fl">0.75</span>),
  <span class="dt">P95 =</span> <span class="kw">quantile</span>(all.counts.vector, <span class="dt">probs =</span> <span class="fl">0.95</span>),
  <span class="dt">P99 =</span> <span class="kw">quantile</span>(all.counts.vector, <span class="dt">probs =</span> <span class="fl">0.99</span>),
  <span class="dt">max =</span> <span class="kw">max</span>(all.counts.vector)
  )

<span class="co"># View(count.stat) # for quick check in R interface</span>
<span class="kw">kable</span>(<span class="kw">t</span>(count.stats), 
      <span class="dt">caption =</span> <span class="st">&quot;Statistics on raw count values (all samples together)&quot;</span>, 
      <span class="dt">format.args =</span> <span class="kw">list</span>(<span class="dt">decimal.mark =</span> <span class="st">&#39;.&#39;</span>, <span class="dt">big.mark =</span> <span class="st">&quot;,&quot;</span>), <span class="dt">digits =</span> <span class="dv">2</span>,
      <span class="dt">col.names =</span> <span class="st">&quot;Parameter value&quot;</span>)</code></pre></div>
<table>
<caption>Statistics on raw count values (all samples together)</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">Parameter value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>mean</td>
<td align="right">827.44</td>
</tr>
<tr class="even">
<td>sd</td>
<td align="right">8,891.07</td>
</tr>
<tr class="odd">
<td>iqr</td>
<td align="right">524.00</td>
</tr>
<tr class="even">
<td>min</td>
<td align="right">0.00</td>
</tr>
<tr class="odd">
<td>P05</td>
<td align="right">0.00</td>
</tr>
<tr class="even">
<td>Q1</td>
<td align="right">24.00</td>
</tr>
<tr class="odd">
<td>median</td>
<td align="right">206.00</td>
</tr>
<tr class="even">
<td>Q3</td>
<td align="right">548.00</td>
</tr>
<tr class="odd">
<td>P95</td>
<td align="right">2,321.70</td>
</tr>
<tr class="even">
<td>P99</td>
<td align="right">7,759.73</td>
</tr>
<tr class="odd">
<td>max</td>
<td align="right">722,346.00</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rm</span>(all.counts.vector) ## Free memory space</code></pre></div>
<p>Nous pouvons également appliquer ces méthodes à chaque échantillon séparément, au moyen de la fonction <code>apply()</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Compute sample means using the apply() function
sample.means &lt;-<span class="st"> </span><span class="kw">apply</span>(counts, <span class="dv">2</span>, mean)

## Compute different stats per sample and store them as columns of a data frame
sample.stats &lt;-<span class="st"> </span><span class="kw">data.frame</span>(
  <span class="dt">label =</span> samples<span class="op">$</span>Label,
  <span class="dt">mean =</span> <span class="kw">apply</span>(counts, <span class="dv">2</span>, mean),
  <span class="dt">sd =</span> <span class="kw">apply</span>(counts, <span class="dv">2</span>, sd),
  <span class="dt">zeros =</span> <span class="kw">apply</span>(counts <span class="op">==</span><span class="st"> </span><span class="dv">0</span>, <span class="dv">2</span>, sum),
  <span class="dt">Q1 =</span> <span class="kw">apply</span>(counts, <span class="dv">2</span>, quantile, <span class="dt">probs =</span> <span class="fl">0.75</span>),
  <span class="dt">median =</span> <span class="kw">apply</span>(counts, <span class="dv">2</span>, median),
  <span class="dt">Q3 =</span> <span class="kw">apply</span>(counts, <span class="dv">2</span>, quantile, <span class="dt">probs =</span> <span class="fl">0.75</span>),
  <span class="dt">P95 =</span> <span class="kw">apply</span>(counts, <span class="dv">2</span>, quantile, <span class="dt">probs =</span> <span class="fl">0.95</span>),
  <span class="dt">lib.size.Mb =</span> <span class="kw">apply</span>(counts, <span class="dv">2</span>, sum)<span class="op">/</span><span class="fl">1e6</span>
)

<span class="kw">kable</span>(sample.stats, 
      <span class="dt">caption =</span> <span class="st">&quot;Sample-wise statistics&quot;</span>, 
      <span class="dt">digits =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">1</span>), 
      <span class="dt">format.args =</span> <span class="kw">list</span>(<span class="dt">decimal.mark =</span> <span class="st">&#39;.&#39;</span>, <span class="dt">big.mark =</span> <span class="st">&quot;,&quot;</span>),
      <span class="dt">align =</span> <span class="st">&quot;l&quot;</span>)</code></pre></div>
<table>
<caption>Sample-wise statistics</caption>
<thead>
<tr class="header">
<th></th>
<th align="left">label</th>
<th align="left">mean</th>
<th align="left">sd</th>
<th align="left">zeros</th>
<th align="left">Q1</th>
<th align="left">median</th>
<th align="left">Q3</th>
<th align="left">P95</th>
<th align="left">lib.size.Mb</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>GSM2375281</td>
<td align="left">bd1bd2mut_0h_1</td>
<td align="left">1,007</td>
<td align="left">10,212</td>
<td align="left">957</td>
<td align="left">684</td>
<td align="left">253</td>
<td align="left">684</td>
<td align="left">2,952</td>
<td align="left">7.5</td>
</tr>
<tr class="even">
<td>GSM2375287</td>
<td align="left">bd1bd2mut_0h_2</td>
<td align="left">632</td>
<td align="left">6,591</td>
<td align="left">1,083</td>
<td align="left">412</td>
<td align="left">147</td>
<td align="left">412</td>
<td align="left">1,897</td>
<td align="left">4.7</td>
</tr>
<tr class="odd">
<td>GSM2375293</td>
<td align="left">bd1bd2mut_0h_3</td>
<td align="left">636</td>
<td align="left">6,584</td>
<td align="left">1,078</td>
<td align="left">423</td>
<td align="left">151</td>
<td align="left">423</td>
<td align="left">1,898</td>
<td align="left">4.8</td>
</tr>
<tr class="even">
<td>GSM2375278</td>
<td align="left">Sc_WT_0h_1</td>
<td align="left">1,127</td>
<td align="left">11,725</td>
<td align="left">957</td>
<td align="left">775</td>
<td align="left">297</td>
<td align="left">775</td>
<td align="left">3,385</td>
<td align="left">8.4</td>
</tr>
<tr class="odd">
<td>GSM2375284</td>
<td align="left">Sc_WT_0h_2</td>
<td align="left">521</td>
<td align="left">5,592</td>
<td align="left">1,101</td>
<td align="left">362</td>
<td align="left">135</td>
<td align="left">362</td>
<td align="left">1,567</td>
<td align="left">3.9</td>
</tr>
<tr class="even">
<td>GSM2375290</td>
<td align="left">Sc_WT_0h_3</td>
<td align="left">779</td>
<td align="left">7,821</td>
<td align="left">1,026</td>
<td align="left">548</td>
<td align="left">194</td>
<td align="left">548</td>
<td align="left">2,187</td>
<td align="left">5.8</td>
</tr>
<tr class="odd">
<td>GSM2375282</td>
<td align="left">bd1bd2mut_4h_1</td>
<td align="left">903</td>
<td align="left">9,814</td>
<td align="left">811</td>
<td align="left">647</td>
<td align="left">262</td>
<td align="left">647</td>
<td align="left">2,569</td>
<td align="left">6.8</td>
</tr>
<tr class="even">
<td>GSM2375288</td>
<td align="left">bd1bd2mut_4h_2</td>
<td align="left">638</td>
<td align="left">7,149</td>
<td align="left">942</td>
<td align="left">436</td>
<td align="left">173</td>
<td align="left">436</td>
<td align="left">1,818</td>
<td align="left">4.8</td>
</tr>
<tr class="odd">
<td>GSM2375294</td>
<td align="left">bd1bd2mut_4h_3</td>
<td align="left">732</td>
<td align="left">8,118</td>
<td align="left">850</td>
<td align="left">515</td>
<td align="left">208</td>
<td align="left">515</td>
<td align="left">2,102</td>
<td align="left">5.5</td>
</tr>
<tr class="even">
<td>GSM2375279</td>
<td align="left">Sc_WT_4h_1</td>
<td align="left">980</td>
<td align="left">9,135</td>
<td align="left">878</td>
<td align="left">735</td>
<td align="left">310</td>
<td align="left">735</td>
<td align="left">2,830</td>
<td align="left">7.3</td>
</tr>
<tr class="odd">
<td>GSM2375285</td>
<td align="left">Sc_WT_4h_2</td>
<td align="left">810</td>
<td align="left">8,039</td>
<td align="left">898</td>
<td align="left">595</td>
<td align="left">254</td>
<td align="left">595</td>
<td align="left">2,264</td>
<td align="left">6.1</td>
</tr>
<tr class="even">
<td>GSM2375291</td>
<td align="left">Sc_WT_4h_3</td>
<td align="left">632</td>
<td align="left">5,556</td>
<td align="left">1,013</td>
<td align="left">470</td>
<td align="left">194</td>
<td align="left">470</td>
<td align="left">1,853</td>
<td align="left">4.7</td>
</tr>
<tr class="odd">
<td>GSM2375283</td>
<td align="left">bd1bd2mut_8h_1</td>
<td align="left">804</td>
<td align="left">9,438</td>
<td align="left">835</td>
<td align="left">500</td>
<td align="left">190</td>
<td align="left">500</td>
<td align="left">2,187</td>
<td align="left">6.0</td>
</tr>
<tr class="even">
<td>GSM2375289</td>
<td align="left">bd1bd2mut_8h_2</td>
<td align="left">956</td>
<td align="left">10,819</td>
<td align="left">808</td>
<td align="left">626</td>
<td align="left">248</td>
<td align="left">626</td>
<td align="left">2,608</td>
<td align="left">7.1</td>
</tr>
<tr class="odd">
<td>GSM2375295</td>
<td align="left">bd1bd2mut_8h_3</td>
<td align="left">919</td>
<td align="left">9,672</td>
<td align="left">862</td>
<td align="left">643</td>
<td align="left">239</td>
<td align="left">643</td>
<td align="left">2,689</td>
<td align="left">6.9</td>
</tr>
<tr class="even">
<td>GSM2375280</td>
<td align="left">Sc_WT_8h_1</td>
<td align="left">1,078</td>
<td align="left">10,956</td>
<td align="left">879</td>
<td align="left">607</td>
<td align="left">230</td>
<td align="left">607</td>
<td align="left">2,672</td>
<td align="left">8.1</td>
</tr>
<tr class="odd">
<td>GSM2375286</td>
<td align="left">Sc_WT_8h_2</td>
<td align="left">756</td>
<td align="left">8,022</td>
<td align="left">893</td>
<td align="left">476</td>
<td align="left">189</td>
<td align="left">476</td>
<td align="left">1,942</td>
<td align="left">5.7</td>
</tr>
<tr class="even">
<td>GSM2375292</td>
<td align="left">Sc_WT_8h_3</td>
<td align="left">983</td>
<td align="left">11,193</td>
<td align="left">973</td>
<td align="left">462</td>
<td align="left">166</td>
<td align="left">462</td>
<td align="left">2,219</td>
<td align="left">7.3</td>
</tr>
</tbody>
</table>
</div>
<div id="comparaison-moyenne---variance-par-gene" class="section level3">
<h3>Comparaison moyenne - variance par gène</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gene.stats &lt;-<span class="st"> </span><span class="kw">data.frame</span>(
  <span class="dt">mean =</span> <span class="kw">apply</span>(counts, <span class="dv">1</span>, mean),
  <span class="dt">var =</span> <span class="kw">apply</span>(counts, <span class="dv">1</span>, var),
  <span class="dt">sd =</span> <span class="kw">apply</span>(counts, <span class="dv">1</span>, sd)
)


<span class="kw">plot</span>(<span class="dt">x =</span> gene.stats<span class="op">$</span>mean, <span class="dt">y =</span> gene.stats<span class="op">$</span>var,
     <span class="dt">main =</span> <span class="st">&quot;Mean / variance plot&quot;</span>,
     <span class="dt">xlab =</span> <span class="st">&quot;Mean per gene&quot;</span>,
     <span class="dt">ylab =</span> <span class="st">&quot;Var per gene&quot;</span>, <span class="dt">log=</span><span class="st">&quot;xy&quot;</span>,
     <span class="dt">col =</span> <span class="kw">densCols</span>(<span class="dt">x =</span> gene.stats<span class="op">$</span>mean, <span class="dt">y =</span> gene.stats<span class="op">$</span>var))
<span class="kw">abline</span>()</code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/TP_mean_var_plot-1.png" alt="Mean-Variance plot" width="80%" />
<p class="caption">
Mean-Variance plot
</p>
</div>
</div>
<div id="distribution-des-comptages" class="section level3">
<h3>Distribution des comptages</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Compute some basic statistics

## Display histogram of the raw counts
<span class="kw">hist</span>(<span class="kw">unlist</span>(counts), 
     <span class="dt">breaks =</span> <span class="dv">100</span>, 
     <span class="dt">col =</span> <span class="st">&quot;#FFDDBB&quot;</span>,
     <span class="dt">xlab =</span> <span class="st">&quot;Reads per gene&quot;</span>, 
     <span class="dt">ylab =</span> <span class="st">&quot;Number of observations&quot;</span>, 
     <span class="dt">main =</span> <span class="st">&quot;Distribution of raw counts&quot;</span>)</code></pre></div>
<p><img src="figures/TP_count_distribution-1.png" width="80%" style="display: block; margin: auto;" /></p>
<p>Cet histogramme n’est pas très informatif, car toutes les valeurs sont concentrées dans la première tranche (à l’extrême gauche). Ceci résulte du fiat que les intervalles de classe ont été définis sur base de l’étendue totale, et qu’il existe apparemment une observation qui a une valeur énorme par rapport aux autres (valeur aberrante, “<strong>outlier</strong>”). De fait, la valeur la plus élevée (722346) dépasse de très loin la moyenne (827.4361757)</p>
<p>Afin de visualiser toute l’étendue des observations tout en mettant plus de détail sur les valeurs faibles, nous pouvons effectuer une transformation logarithmique. Nous devons cependant prêter attention à certains détails.</p>
<ul>
<li><p>Conventionnellement, on utlise la transformation log2 pour les données RNA-seq, car elle fournit un découpage plus fin des valeurs couvertes.</p></li>
<li><p>Les données RnA-seq contiennent généralement un bon nombre de valeurs nulles (gènes non détectés), qui posent un problème pour la conversion logarithmique (<span class="math inline">\(log(0) = -\infty\)</span>). On contourne ce problème en ajoutant aux comptages un petit nombre arbitraire, qu’on dénomme <strong>pseudo-comptage</strong> et qu’on symbolise par la lettre grècque epsilon (<span class="math inline">\(\epsilon\)</span>).</p></li>
</ul>
<p>Il est cournant d’ajouter la valeur <span class="math inline">\(1\)</span>, mais nous préférons ajouter une valeur inférieure à 1, pour bien distinguer les comptages réels (qui peuvent de fait prendre une valeur 1) des pseudo-comptages.</p>
<p>Pour cette analyse nous choisissons <span class="math inline">\(\epsilon = 0.1\)</span>.</p>
<p>La conversion donne donc.</p>
<p><span class="math display">\[\text{log2count}= log2(n + \epsilon) = log2(n + 0.1) \]</span> L’histogramme des log2counts est plus informatif que celui des comptages bruts.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">counts.log2 &lt;-<span class="st"> </span><span class="kw">log2</span>(counts <span class="op">+</span><span class="st"> </span>parameters<span class="op">$</span>epsilon)

<span class="kw">hist</span>(<span class="kw">unlist</span>(counts.log2), 
     <span class="dt">xlab =</span> <span class="st">&quot;log2(counts + epsilon)&quot;</span>, 
     <span class="dt">ylab =</span> <span class="st">&quot;Number of observations&quot;</span>, 
     <span class="dt">main =</span> <span class="st">&quot;Distribution of log2(counts)&quot;</span>, 
     <span class="dt">breaks =</span> <span class="dv">100</span>, <span class="dt">col =</span> <span class="st">&quot;#BBDDEE&quot;</span>)</code></pre></div>
<p><img src="figures/TP_log2_conversion-1.png" width="80%" style="display: block; margin: auto;" /></p>
<p>Notons d’emblée le pic très élevé à gauche, qui correspond à toutes les observations nulles. Sur l’axe <span class="math inline">\(X\)</span>, il apparait à la valeur <span class="math inline">\(-3.3\)</span>, qui correspond à <span class="math inline">\(log_2(\epsilon) = log_2(0.1)\)</span>.</p>
</div>
</div>
<div id="boites-a-moustache" class="section level2">
<h2>Boîtes à moustache</h2>
<p>Les boîtes à moustache sont très utilisées pour obtenir une vision d’ensemble de plusieurs distribution.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Associate a specific color to each condition
conditions &lt;-<span class="st"> </span><span class="kw">unique</span>(samples<span class="op">$</span>Condition)
condition.colors &lt;-<span class="st"> </span><span class="kw">rainbow</span>(<span class="dt">n =</span> <span class="kw">length</span>(conditions))
<span class="kw">names</span>(condition.colors) &lt;-<span class="st"> </span>conditions

## Associate a color to each sample according to its condition
samples<span class="op">$</span>color &lt;-<span class="st"> </span>condition.colors[samples<span class="op">$</span>Condition]

par.ori &lt;-<span class="st"> </span><span class="kw">par</span>(<span class="dt">no.readonly =</span> <span class="ot">TRUE</span>) <span class="co"># Store the original value of the graphical parameters before modifying them</span>

<span class="co"># Box plot of counts and log2-transformed counts</span>
<span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="fl">5.1</span>, <span class="fl">8.1</span>, <span class="fl">4.1</span>, <span class="fl">1.1</span>)) <span class="co"># Increase leftmargin for sample labels</span>

<span class="kw">boxplot</span>(counts, <span class="dt">horizontal =</span> <span class="ot">TRUE</span>, <span class="dt">las =</span> <span class="dv">1</span>, <span class="dt">names =</span> samples<span class="op">$</span>Label, <span class="dt">col =</span> samples<span class="op">$</span>color, <span class="dt">xlab =</span> <span class="st">&quot;Raw counts&quot;</span>)

<span class="kw">boxplot</span>(counts.log2, <span class="dt">horizontal =</span> <span class="ot">TRUE</span>, <span class="dt">las =</span> <span class="dv">1</span>, <span class="dt">names =</span> samples<span class="op">$</span>Label, <span class="dt">col =</span> samples<span class="op">$</span>color, <span class="dt">xlab =</span> <span class="st">&quot;log2counts&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/TP_boxplots-1.png" alt="**Boxplot of the counts per gene in each sample.** Left: raw counts. Right: log2_transformed counts. " width="80%" />
<p class="caption">
<strong>Boxplot of the counts per gene in each sample.</strong> Left: raw counts. Right: log2_transformed counts.
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(par.ori)</code></pre></div>
</div>
<div id="violin-plots" class="section level2">
<h2>Violin plots</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Violin Plots</span>
<span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="fl">5.1</span>, <span class="fl">8.1</span>, <span class="fl">4.1</span>, <span class="fl">1.1</span>)) <span class="co"># Increase left margin for sample labels</span>
<span class="kw">par</span>(<span class="dt">las =</span> <span class="dv">1</span>)
<span class="kw">vioplot</span>(<span class="kw">as.list</span>(counts), <span class="dt">horizontal =</span> <span class="ot">TRUE</span>, 
        <span class="dt">col =</span> samples<span class="op">$</span>color, <span class="dt">xlab =</span> <span class="st">&quot;Raw counts&quot;</span>,
        <span class="dt">main =</span> <span class="st">&quot;Violin plot</span><span class="ch">\n</span><span class="st">Raw counts&quot;</span>,
        <span class="dt">names =</span> samples<span class="op">$</span>Label)

<span class="kw">vioplot</span>(<span class="kw">as.list</span>(counts.log2), <span class="dt">horizontal =</span> <span class="ot">TRUE</span>, 
        <span class="dt">col =</span> samples<span class="op">$</span>color, <span class="dt">xlab =</span> <span class="st">&quot;Log2(counts + epsilon)&quot;</span>,
        <span class="dt">main =</span> <span class="st">&quot;Violin plot</span><span class="ch">\n</span><span class="st">log-transformed counts&quot;</span>,
        <span class="dt">names =</span> samples<span class="op">$</span>Label)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/TP_violin_plot-1.png" alt="**Violin plots**. Left: raw counts. Right: log2-transformed counts. " width="80%" />
<p class="caption">
<strong>Violin plots</strong>. Left: raw counts. Right: log2-transformed counts.
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(par.ori)</code></pre></div>
</div>
<div id="filtrage-des-genes" class="section level2">
<h2>Filtrage des gènes</h2>
<div id="genes-a-variance-nulle-valeur-constante" class="section level3">
<h3>Gènes à variance nulle (valeur constante)</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Zero-variance filter
gene.var &lt;-<span class="st"> </span><span class="kw">apply</span>(counts, <span class="dv">1</span>, var)

## Define a Boolean vector (TRUE/FALSE) 
## indicating whether each gene has a null variance.
zero.var.genes &lt;-<span class="st"> </span>gene.var <span class="op">==</span><span class="st"> </span><span class="dv">0</span>

<span class="kw">kable</span>(<span class="kw">table</span>(zero.var.genes))</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">zero.var.genes</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FALSE</td>
<td align="right">6918</td>
</tr>
<tr class="even">
<td align="left">TRUE</td>
<td align="right">560</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Count the number of genes having a zero variance
<span class="co"># sum(gene.var == 0)</span></code></pre></div>
<p>Nous écartons les gènes qui ont une variance nulle (autrement dit, qui ont les mêmes valeurs pour tous les échantillons), puisque leurs moyennes seront forcément identiques entre conditions. Dans notre jeu de données, ceci nous amène à écarter 560</p>
</div>
<div id="pourcentage-de-valeurs-nulles" class="section level3">
<h3>Pourcentage de valeurs nulles</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Discard genes having zeros in at least 95% of samples
<span class="kw">message</span>(<span class="st">&quot;Applying threshold on the percent of non-zero counts per gene: &quot;</span>, parameters<span class="op">$</span>gene.filter.percent.zeros, <span class="st">&quot;%&quot;</span>)

## A Boolean table indicating whether each count is null (TRUE) or not (FALSE)
zero.counts &lt;-<span class="st"> </span>counts <span class="op">==</span><span class="st"> </span><span class="dv">0</span> 

## Count number of zero values per gene (=row of the expression table)
percent.zeros &lt;-<span class="st"> </span><span class="dv">100</span><span class="op">*</span><span class="kw">apply</span>(zero.counts, <span class="dv">1</span>, sum) <span class="op">/</span><span class="st"> </span><span class="kw">ncol</span>(counts)

<span class="co"># Note: we store the histogram values in a variable named &quot;h&quot; for further use</span>
h &lt;-<span class="st"> </span><span class="kw">hist</span>(percent.zeros, 
          <span class="dt">breaks =</span> <span class="dv">20</span>, <span class="dt">col =</span> <span class="st">&quot;#DDDDDD&quot;</span>, 
          <span class="dt">main =</span> <span class="st">&quot;Filter on the percentage of zero counts&quot;</span>, 
          <span class="dt">xlab =</span> <span class="st">&quot;Percent of samples&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Number of genes&quot;</span>, <span class="dt">las =</span> <span class="dv">1</span>)

<span class="co"># View(h)</span>

## Get histogram max counts (y value) to position the arrow
h.max &lt;-<span class="st"> </span><span class="kw">max</span>(h<span class="op">$</span>counts) 
<span class="kw">arrows</span>(<span class="dt">x0 =</span> parameters<span class="op">$</span>gene.filter.percent.zeros, 
       <span class="dt">y0 =</span> h.max <span class="op">*</span><span class="st"> </span><span class="fl">0.5</span>, 
       <span class="dt">x1 =</span> parameters<span class="op">$</span>gene.filter.percent.zeros, 
       <span class="dt">y1 =</span> h.max <span class="op">*</span><span class="st"> </span><span class="fl">0.3</span>, 
       <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">lwd =</span> <span class="dv">2</span>,
       <span class="dt">angle =</span> <span class="dv">30</span>, <span class="dt">length =</span> <span class="fl">0.1</span>)
<span class="kw">arrows</span>(<span class="dt">x0 =</span> parameters<span class="op">$</span>gene.filter.percent.zeros, 
       <span class="dt">y0 =</span> h.max <span class="op">*</span><span class="st"> </span><span class="fl">0.5</span>, 
       <span class="dt">x1 =</span> <span class="dv">100</span>, 
       <span class="dt">y1 =</span> h.max <span class="op">*</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">code =</span> <span class="dv">0</span>,
       <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">lwd =</span> <span class="dv">4</span>,
       <span class="dt">angle =</span> <span class="dv">30</span>, <span class="dt">length =</span> <span class="fl">0.1</span>)
<span class="kw">text</span>(<span class="dt">x =</span> parameters<span class="op">$</span>gene.filter.percent.zeros, 
     <span class="dt">y =</span> h.max<span class="op">*</span><span class="fl">0.5</span>, 
     <span class="dt">labels =</span> <span class="kw">paste</span>(<span class="dt">sep =</span> <span class="st">&quot;&quot;</span>, parameters<span class="op">$</span>gene.filter.percent.zeros, <span class="st">&quot;%&quot;</span>),
     <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">pos =</span> <span class="dv">3</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/TP_gene_filtering_zero_counts-1.png" alt="Frequency of samples with zero counts per gene. Genes exceeding the thresold (red arrow) were filtered out. " width="80%" />
<p class="caption">
Frequency of samples with zero counts per gene. Genes exceeding the thresold (red arrow) were filtered out.
</p>
</div>
</div>
<div id="comptages-minimaux-par-gene" class="section level3">
<h3>Comptages minimaux par gène</h3>
<p>Un autre critère possible est de supprimer les gènes faiblement détectés, en écartant les gènes pour lesquels les comptages sont inférieurs à un seuil donné pour tous les échantillons. Nous fixons ici (arbitrairement) le seuil à 10.</p>
<p>Nous procédont comme suit:</p>
<ul>
<li>Calculer pour chaque gène <span class="math inline">\(i\)</span> le comptage maximal observé parmi tous les échantillons.</li>
</ul>
<p><span class="math display">\[x_{max,i} = \text{max}_{j = 0}^{s} x_{i,j}\]</span></p>
<p>où <span class="math inline">\(i\)</span> est l’indice de lignes (gènes), <span class="math inline">\(j\)</span> l’indice de colonnes (échantillons), et <span class="math inline">\(s\)</span> le nombre d’échantillons.</p>
<ul>
<li>Tester si la valeur maximale observée est inférieure au seuil minimum.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">max.per.gene &lt;-<span class="st"> </span><span class="kw">apply</span>(counts, <span class="dv">1</span>, max)

<span class="co"># View(summary(min.count))</span>
h &lt;-<span class="st"> </span><span class="kw">hist</span>(<span class="kw">log2</span>(max.per.gene <span class="op">+</span><span class="st"> </span>parameters<span class="op">$</span>epsilon), <span class="dt">breaks =</span> <span class="dv">100</span>, <span class="dt">col =</span> <span class="st">&quot;#BBBBFF&quot;</span>,
          <span class="dt">xlab =</span> <span class="st">&quot;log2(max count per gene)&quot;</span>,
          <span class="dt">ylab =</span> <span class="st">&quot;Number of genes&quot;</span>, <span class="dt">las =</span> <span class="dv">1</span>,
          <span class="dt">main =</span> <span class="st">&quot;Filtering on min count per gene&quot;</span>)
h.max &lt;-<span class="st"> </span><span class="kw">max</span>(h<span class="op">$</span>counts)
<span class="kw">arrows</span>(<span class="dt">x0 =</span> <span class="kw">log2</span>(parameters<span class="op">$</span>gene.filter.min.count), 
       <span class="dt">y0 =</span> h.max <span class="op">*</span><span class="st"> </span><span class="fl">0.5</span>, 
       <span class="dt">x1 =</span> <span class="kw">log2</span>(parameters<span class="op">$</span>gene.filter.min.count), 
       <span class="dt">y1 =</span> h.max <span class="op">*</span><span class="st"> </span><span class="fl">0.3</span>, 
       <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">lwd =</span> <span class="dv">2</span>,
       <span class="dt">angle =</span> <span class="dv">30</span>, <span class="dt">length =</span> <span class="fl">0.1</span>)
<span class="kw">arrows</span>(<span class="dt">x0 =</span> <span class="kw">log2</span>(parameters<span class="op">$</span>epsilon), 
       <span class="dt">y0 =</span> h.max <span class="op">*</span><span class="st"> </span><span class="fl">0.5</span>, 
       <span class="dt">x1 =</span> <span class="kw">log2</span>(parameters<span class="op">$</span>gene.filter.min.count), 
       <span class="dt">y1 =</span> h.max <span class="op">*</span><span class="st"> </span><span class="fl">0.5</span>, 
       <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">lwd =</span> <span class="dv">4</span>, <span class="dt">code =</span> <span class="dv">0</span>,
       <span class="dt">angle =</span> <span class="dv">30</span>, <span class="dt">length =</span> <span class="fl">0.1</span>)
<span class="kw">text</span>(<span class="dt">x =</span> <span class="kw">log2</span>(parameters<span class="op">$</span>gene.filter.min.count), 
     <span class="dt">y =</span> h.max <span class="op">*</span><span class="st"> </span><span class="fl">0.5</span>, 
     <span class="dt">labels =</span> <span class="kw">paste</span>(<span class="dt">sep =</span> <span class="st">&quot;&quot;</span>, <span class="st">&quot;log2(&quot;</span>, 
                    parameters<span class="op">$</span>gene.filter.min.count,
                    <span class="st">&quot;) = &quot;</span>, <span class="kw">signif</span>(<span class="dt">digits =</span> <span class="dv">3</span>, <span class="kw">log2</span>(parameters<span class="op">$</span>gene.filter.min.count))),
     <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">pos =</span> <span class="dv">3</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/TP_gene_filtering_min_count-1.png" alt="Distribution of max counts per gene. Genes below the thresold (red arrow) are filtered out. " width="80%" />
<p class="caption">
Distribution of max counts per gene. Genes below the thresold (red arrow) are filtered out.
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Build a data frame with one row per gene and one column per filtering criterion</span>
discarded.genes &lt;-<span class="st"> </span><span class="kw">data.frame</span>(
  <span class="dt">zero.var =</span> zero.var.genes,
  <span class="dt">too.many.zeros =</span> (percent.zeros <span class="op">&gt;</span><span class="st"> </span>parameters<span class="op">$</span>gene.filter.percent.zeros),
  <span class="dt">too.small.counts =</span> max.per.gene <span class="op">&lt;</span><span class="st"> </span>parameters<span class="op">$</span>gene.filter.min.count
)


## Print the discarding criteria for some genes 
<span class="kw">kable</span>(discarded.genes[some.genes, ], <span class="dt">caption =</span> <span class="st">&quot;Gene filtering criteria. TRUE indicates that a gene is discarded. &quot;</span>)</code></pre></div>
<table>
<caption>Gene filtering criteria. TRUE indicates that a gene is discarded.</caption>
<thead>
<tr class="header">
<th></th>
<th align="left">zero.var</th>
<th align="left">too.many.zeros</th>
<th align="left">too.small.counts</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>YLR221C</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td>YJL037W</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td>YOR252W</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td>YKL166C</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td>YJL052W</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td>YLL018C-A</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td>YOR337W</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td>YJL122W</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td>YIR007W</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td>YPL144W</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Draw a venn diagram indicating the number of genes 
## discarded by the different criteria. 
venn.counts &lt;-<span class="st"> </span><span class="kw">vennCounts</span>(discarded.genes)
<span class="kw">vennDiagram</span>(venn.counts, <span class="dt">cex =</span> <span class="fl">0.8</span>,
            <span class="dt">main =</span> <span class="st">&quot;Genes discarded by different criteria&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/TP_gene_filtering-1.png" alt="Number of genes discarded during the filtering step. " width="70%" />
<p class="caption">
Number of genes discarded during the filtering step.
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Genes passing the filters are those for which 
## all the discarding criteria are FALSE, 
## i.e. the sum of the row is 0
filtered.genes &lt;-<span class="st"> </span><span class="kw">apply</span>(discarded.genes, <span class="dv">1</span>, sum) <span class="op">==</span><span class="st"> </span><span class="dv">0</span>

## Select a matrix with the filtered genes, 
## i.e. those not discarded by any criterion
filtered.counts &lt;-<span class="st"> </span>counts[filtered.genes, ]

## Update the set of test genes to avoir NA values
some.genes &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(filtered.counts), <span class="dt">size =</span> <span class="dv">10</span>, <span class="dt">replace =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
</div>
<div id="reproductibilite-des-replicats" class="section level2">
<h2>Reproductibilité des réplicats</h2>
<p>Nous pouvons comparer les réplicats pour une même condition au moyen d’un nuage de points, avec la fonction <code>plot()</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Sélectionner les échantillons de la condition &quot;test&quot;
<span class="kw">message</span>(<span class="st">&quot;Comparing counts between replicates for condition&quot;</span>, parameters<span class="op">$</span>test.group)
selected.samples &lt;-<span class="st"> </span>samples<span class="op">$</span>Condition  <span class="op">==</span><span class="st"> </span>parameters<span class="op">$</span>test.group
<span class="kw">plot</span>(counts.log2[, selected.samples], 
     <span class="dt">col =</span> samples[selected.samples, <span class="st">&quot;color&quot;</span>])</code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/TP_between_replicate_reproducibility-1.png" alt="Between-replicate comparison of counts per gene for the test condition. " width="80%" />
<p class="caption">
Between-replicate comparison of counts per gene for the test condition.
</p>
</div>
</div>
<div id="detection-de-genes-differentiellement-exprimes-avec-edger" class="section level2">
<h2>Détection de gènes différentiellement exprimés avec edgeR</h2>
<div id="comparaison-entre-2-groupes" class="section level3">
<h3>Comparaison entre 2 groupes</h3>
<p>Dans un premier temps, nous allons effectuer une simple comparaison de 2 groupes: sauvage versus mutant au temps 0.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Detection of Differentially Expressed Genes (DEG) with edgeR
<span class="kw">message</span>(<span class="st">&quot;Detecting Differentially Expressed Genes (DEG) with edgeR&quot;</span>)


## Select the subsets of counts (columns) and samples (rows) corresponding to the control and test groups
control.samples &lt;-<span class="st"> </span><span class="kw">which</span>(samples<span class="op">$</span>Condition <span class="op">==</span><span class="st"> </span>parameters<span class="op">$</span>control.group)
test.samples &lt;-<span class="st"> </span><span class="kw">which</span>(samples<span class="op">$</span>Condition <span class="op">==</span><span class="st"> </span>parameters<span class="op">$</span>test.group)
selected.samples &lt;-<span class="st"> </span><span class="kw">c</span>(test.samples, control.samples)

## Select samples for a 2-group

## Build a &quot;model matrix&quot; from the class labels
## This matrix contains one row per sample and one column per class
designMat &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span><span class="kw">as.vector</span>(samples[selected.samples, <span class="st">&quot;Condition&quot;</span>]))
<span class="co"># View(designMat)</span>

## Build an edgeR::DGEList object which is required to run edgeR DE analysis
dgList &lt;-<span class="st"> </span><span class="kw">DGEList</span>(<span class="dt">counts =</span> filtered.counts[, selected.samples])
<span class="co"># class(dgList)</span>
<span class="co"># is(dgList)</span>

## Estimate the dispersion parameters. 
<span class="kw">message</span>(<span class="st">&quot;</span><span class="ch">\t</span><span class="st">edgeR</span><span class="ch">\t</span><span class="st">Estimating dispersion&quot;</span>)
dgList &lt;-<span class="st"> </span><span class="kw">estimateDisp</span>(dgList, <span class="dt">design =</span> designMat)
dgList &lt;-<span class="st"> </span><span class="kw">estimateTagwiseDisp</span>(dgList, <span class="dt">design =</span> designMat)

## Fit edgeR model for differential expression analysis.
## We chose glmQLFit because it is claimed to offer a more accurate control of type I error. 
<span class="kw">message</span>(<span class="st">&quot;</span><span class="ch">\t</span><span class="st">edgeR</span><span class="ch">\t</span><span class="st">model fitting with glmQLFit()&quot;</span>)
fit &lt;-<span class="st"> </span><span class="kw">glmQLFit</span>(dgList, <span class="dt">design =</span> designMat)

## Run test to detect differentially expressed genes
<span class="kw">message</span>(<span class="st">&quot;</span><span class="ch">\t</span><span class="st">edgeR</span><span class="ch">\t</span><span class="st">detecting differentially expressed genes with glmQLFTest()&quot;</span>)
qlf &lt;-<span class="st"> </span><span class="kw">glmQLFTest</span>(fit, <span class="dt">coef =</span> <span class="dv">2</span><span class="op">:</span><span class="kw">ncol</span>(designMat))
qlf.TT &lt;-<span class="st"> </span><span class="kw">topTags</span>(qlf, <span class="dt">n =</span> <span class="kw">nrow</span>(qlf<span class="op">$</span>table), <span class="dt">sort.by =</span> <span class="st">&quot;none&quot;</span>,
                  <span class="dt">adjust.method =</span> <span class="st">&quot;fdr&quot;</span>)


## Select differentially expressed genes
edgeR.DEG.table &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(qlf.TT<span class="op">$</span>table)

## Compute some additional statistics
edgeR.DEG.table<span class="op">$</span>evalue &lt;-<span class="st"> </span>edgeR.DEG.table<span class="op">$</span>PValue <span class="op">*</span><span class="st"> </span><span class="kw">nrow</span>(edgeR.DEG.table)
edgeR.DEG.table<span class="op">$</span>rank &lt;-<span class="st"> </span><span class="kw">rank</span>(edgeR.DEG.table<span class="op">$</span>FDR, <span class="dt">ties.method =</span> <span class="st">&quot;average&quot;</span>)
edgeR.DEG.table<span class="op">$</span>positive &lt;-<span class="st"> </span>edgeR.DEG.table<span class="op">$</span>FDR <span class="op">&lt;</span><span class="st"> </span>parameters<span class="op">$</span>alpha
<span class="co"># table(edgeR.DEG.table$positive)</span>
<span class="co"># names(edgeR.DEG.table)</span>
<span class="co"># View(edgeR.DEG.table)</span>
<span class="co"># sum(egeR.DEG)</span></code></pre></div>
</div>
<div id="exercices-solutions-ci-dessous" class="section level3">
<h3>Exercices (solutions ci-dessous)</h3>
<ol style="list-style-type: decimal">
<li><p>Dessinez un nuage de points (<code>plot()</code>) comparant les P-valeurs ajustées et non-ajustées.</p></li>
<li><p>Quel serait le nombre de faux-positifs attendus sous hypothès nulle si l’on appliquait le seuil <span class="math inline">\(\alpha = 0.05\)</span> à la p-valeur non-ajustée ?</p></li>
<li><p>Comptez le nombre de gènes déclarés positifs selon qu’on applique le seuil <span class="math inline">\(\alpha = 0.05\)</span> à la p-valeur ajustée ou non-ajustée.</p></li>
<li><p>Dessinez un histogramme (<code>hist()</code>) des P-valeurs <em>non ajustées</em>, avec 20 intervalles de classe (breaks = 20). Observez la répartition des gènes dans la moitié inférieure (p-valeur &lt; 0.5) et supérieure (p-valeur &gt; 0.5) de l’histogramme. Interprétez le résultat en 2 ou 3 phrases. A quoi s’attendrait-on si tous les gènes étaient sous hypothèse nulle ?</p></li>
<li><p>Dessinez un nuage de points (<code>plot()</code>) comparant le log2(fold-change) en abscisse et la P-valeur en ordonnée. Interprétez le résultat en 2 ou 3 phrases. Où se trouvent les gènes signicatifs ?</p></li>
<li><p>Dessinez un <strong>volcano plot</strong>, c’est-à-dire un nuage de points (<code>plot()</code>) présentant en abscisse le log2(fold-change), et en ordonnée la significativité, définie comme suit.</p></li>
</ol>
<p><span class="math display">\[\text{sig} = -log_{10}(\text{FDR})\]</span></p>
<p>Interprétez le résultat en 2 ou 3 phrases. Commentez la relation entre le logFC et la significativité. Où se trouvent les gènes signicatifs ?</p>
</div>
<div id="adjusted-versus-non-adjusted-p-values" class="section level3">
<h3>Adjusted versus non-adjusted P-values</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># class(qlf.TT)</span>
<span class="kw">plot</span>(edgeR.DEG.table<span class="op">$</span>PValue, edgeR.DEG.table<span class="op">$</span>FDR, 
     <span class="dt">main =</span> <span class="st">&quot;Multiple testing correction&quot;</span>,
     <span class="dt">xlab =</span> <span class="st">&quot;Nominal p-value&quot;</span>,
     <span class="dt">ylab =</span> <span class="st">&quot;FDR&quot;</span>,
     <span class="dt">log =</span> <span class="st">&quot;xy&quot;</span>, 
     <span class="dt">col =</span> <span class="kw">densCols</span>(<span class="dt">x =</span> edgeR.DEG.table<span class="op">$</span>PValue,
                    <span class="dt">y =</span> edgeR.DEG.table<span class="op">$</span>FDR),
     <span class="dt">panel.first =</span> <span class="kw">grid</span>(), 
     <span class="dt">las =</span> <span class="dv">1</span>)
<span class="kw">abline</span>(<span class="dt">h =</span> parameters<span class="op">$</span>alpha, <span class="dt">col =</span> <span class="st">&quot;#BB0000&quot;</span>)
<span class="kw">abline</span>(<span class="dt">v =</span> parameters<span class="op">$</span>alpha, <span class="dt">col =</span> <span class="st">&quot;#FF0000&quot;</span>, <span class="dt">lty =</span> <span class="st">&quot;dashed&quot;</span>)
<span class="kw">legend</span>(<span class="st">&quot;topleft&quot;</span>,
       <span class="dt">legend =</span> <span class="kw">c</span>(
         <span class="kw">paste</span>(<span class="dt">sep =</span> <span class="st">&quot;&quot;</span>, <span class="st">&quot;p-value &lt; &quot;</span>, parameters<span class="op">$</span>alpha, <span class="st">&quot;: &quot;</span>, 
               <span class="kw">sum</span>(edgeR.DEG.table<span class="op">$</span>PValue <span class="op">&lt;</span><span class="st"> </span>parameters<span class="op">$</span>alpha), <span class="st">&quot; genes&quot;</span>),
         <span class="kw">paste</span>(<span class="dt">sep =</span> <span class="st">&quot;&quot;</span>, <span class="st">&quot;FDR &lt; &quot;</span>, parameters<span class="op">$</span>alpha, <span class="st">&quot;: &quot;</span>, <span class="kw">sum</span>(edgeR.DEG.table<span class="op">$</span>FDR <span class="op">&lt;</span><span class="st"> </span>parameters<span class="op">$</span>alpha), <span class="st">&quot; genes&quot;</span> )))</code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/TP_edger_nominal_vs_adjusted-1.png" alt="**Ajusted (ordinate) versus non-adjusted (abscissa) p-values returned by edger**. Strong colors indicate a higher density of overlapping points. Note the strong impact of the multiple testing correction." width="60%" />
<p class="caption">
<strong>Ajusted (ordinate) versus non-adjusted (abscissa) p-values returned by edger</strong>. Strong colors indicate a higher density of overlapping points. Note the strong impact of the multiple testing correction.
</p>
</div>
</div>
<div id="histogram-of-the-nominal-p-values" class="section level3">
<h3>Histogram of the nominal P-values</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## P-value histogram
<span class="kw">hist</span>(edgeR.DEG.table<span class="op">$</span>PValue, 
     <span class="dt">breaks =</span> <span class="dv">20</span>, <span class="dt">col =</span> <span class="st">&quot;#DDEEFF&quot;</span>, <span class="dt">las =</span> <span class="dv">1</span>,
     <span class="dt">main =</span> <span class="st">&quot;edgeR P-value histogram&quot;</span>,
     <span class="dt">xlab =</span> <span class="st">&quot;Nominal P-value&quot;</span>, 
     <span class="dt">ylab =</span> <span class="st">&quot;Number of genes&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/TP_edger_pvalue_histogram-1.png" alt="**Histogram of edgeR nominal P-values.**" width="80%" />
<p class="caption">
<strong>Histogram of edgeR nominal P-values.</strong>
</p>
</div>
</div>
<div id="volcano-plot" class="section level3">
<h3>Volcano plot</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Quick test: plot the different symbols</span>
<span class="co"># plot(1:30, pch = 1:30)</span>


<span class="co"># Assign a graphical symbol to each gene according to its positive/negative status</span>
edgeR.DEG.table<span class="op">$</span>pch &lt;-<span class="st"> </span><span class="dv">1</span> <span class="co"># circle symbol</span>
edgeR.DEG.table<span class="op">$</span>pch[edgeR.DEG.table<span class="op">$</span>positive] &lt;-<span class="st"> </span><span class="dv">3</span> <span class="co"># + symbol</span>

<span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">plot</span>(<span class="dt">x =</span> edgeR.DEG.table<span class="op">$</span>logFC, 
     <span class="dt">y =</span> <span class="op">-</span><span class="kw">log10</span>(edgeR.DEG.table<span class="op">$</span>PValue),
     <span class="dt">las =</span> <span class="dv">1</span>, 
     <span class="dt">xlab =</span> <span class="st">&quot;log2(fold-change)&quot;</span>,
     <span class="dt">main =</span> <span class="st">&quot;Volcano plot of edgeR unadjusted p-values&quot;</span>,
     <span class="dt">ylab =</span> <span class="st">&quot;-log10(P-value)&quot;</span>,
     <span class="dt">pch =</span> edgeR.DEG.table<span class="op">$</span>pch, 
     <span class="dt">col =</span> <span class="kw">densCols</span>(<span class="dt">x =</span> edgeR.DEG.table<span class="op">$</span>logFC, 
                    <span class="dt">y =</span> <span class="op">-</span><span class="kw">log10</span>(edgeR.DEG.table<span class="op">$</span>PValue)),
     <span class="dt">panel.first =</span> <span class="kw">grid</span>())
<span class="kw">abline</span>(<span class="dt">v =</span> <span class="dv">0</span>)
<span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">0</span>)

<span class="kw">plot</span>(<span class="dt">x =</span> edgeR.DEG.table<span class="op">$</span>logFC, 
     <span class="dt">y =</span> <span class="op">-</span><span class="kw">log10</span>(edgeR.DEG.table<span class="op">$</span>FDR),
     <span class="dt">las =</span> <span class="dv">1</span>, 
     <span class="dt">xlab =</span> <span class="st">&quot;log2(fold-change)&quot;</span>,
     <span class="dt">main =</span> <span class="st">&quot;Volcano plot of edgeR FDR&quot;</span>,
     <span class="dt">ylab =</span> <span class="st">&quot;-log10(P-value)&quot;</span>,
     <span class="dt">pch =</span> edgeR.DEG.table<span class="op">$</span>pch, 
     <span class="dt">col =</span> <span class="kw">densCols</span>(<span class="dt">x =</span> edgeR.DEG.table<span class="op">$</span>logFC, 
                    <span class="dt">y =</span> <span class="op">-</span><span class="kw">log10</span>(edgeR.DEG.table<span class="op">$</span>FDR)),
     <span class="dt">panel.first =</span> <span class="kw">grid</span>())
<span class="kw">abline</span>(<span class="dt">v =</span> <span class="dv">0</span>)
<span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">0</span>)
<span class="kw">abline</span>(<span class="dt">h =</span> <span class="op">-</span><span class="kw">log10</span>(parameters<span class="op">$</span>alpha), <span class="dt">col =</span> <span class="st">&quot;#BB0000&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/TP_edger_volcano-1.png" alt="**Volcano plots of edgeR unadjusted (left) and adjusted (right) P-values**. Multple testing correction method: FDR. " width="95%" />
<p class="caption">
<strong>Volcano plots of edgeR unadjusted (left) and adjusted (right) P-values</strong>. Multple testing correction method: FDR.
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</code></pre></div>
</div>
</div>
<div id="detection-de-genes-differentiellement-exprimes-avec-deseq2" class="section level2">
<h2>Détection de gènes différentiellement exprimés avec DESeq2</h2>
<p>Nous allons maintenant utiliser DESeq2 pour détecter des gènes différentiellement exprimés sur le même jeu de données. Nous comparerons ensuite les résultats obtenus avec edgeR et DESeq2, respectivement.</p>
<div id="comparaison-de-2-groupes" class="section level3">
<h3>Comparaison de 2 groupes</h3>
<p>Dans un premier temps, nous allons effectuer une simple comparaison de 2 groupes: sauvage versus mutant au temps 0.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Use the DESeqDataSetFromMatrix to create a DESeqDataSet object
<span class="kw">message</span>(<span class="st">&quot;Creating DESeq2 dataset&quot;</span>)
dds.2groups &lt;-<span class="st"> </span><span class="kw">DESeqDataSetFromMatrix</span>(
  <span class="dt">countData =</span> filtered.counts[, selected.samples], 
  <span class="dt">colData =</span> samples[selected.samples, ], 
  <span class="dt">design =</span> <span class="op">~</span><span class="st"> </span>Condition)
<span class="co"># View(dds.all.conditions)</span>

## Run differential expression analysis with DESeq2
dds.2groups &lt;-<span class="st"> </span><span class="kw">DESeq</span>(dds.2groups)
<span class="co"># View(dds.all.conditions)</span>

<span class="kw">resultsNames</span>(dds.2groups)</code></pre></div>
<pre><code>[1] &quot;Intercept&quot;                                          &quot;Condition_Wild_type_8_vs_bdf1_Y187F_Y354F_mutant_8&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Get the result table
deseq2.res.dds.2groups &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">results</span>(dds.2groups))
<span class="kw">kable</span>(deseq2.res.dds.2groups[some.genes, ])</code></pre></div>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">baseMean</th>
<th align="right">log2FoldChange</th>
<th align="right">lfcSE</th>
<th align="right">stat</th>
<th align="right">pvalue</th>
<th align="right">padj</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>YBR225W</td>
<td align="right">244.689195</td>
<td align="right">0.4296799</td>
<td align="right">0.2509285</td>
<td align="right">1.7123596</td>
<td align="right">0.0868304</td>
<td align="right">0.1775735</td>
</tr>
<tr class="even">
<td>YGR010W</td>
<td align="right">160.653261</td>
<td align="right">-0.7817281</td>
<td align="right">0.2900742</td>
<td align="right">-2.6949250</td>
<td align="right">0.0070404</td>
<td align="right">0.0244651</td>
</tr>
<tr class="odd">
<td>YKL001C</td>
<td align="right">293.381189</td>
<td align="right">-1.5336306</td>
<td align="right">0.4681641</td>
<td align="right">-3.2758399</td>
<td align="right">0.0010535</td>
<td align="right">0.0051962</td>
</tr>
<tr class="even">
<td>YPR145W</td>
<td align="right">3999.507230</td>
<td align="right">-1.9770166</td>
<td align="right">0.4876586</td>
<td align="right">-4.0540996</td>
<td align="right">0.0000503</td>
<td align="right">0.0003952</td>
</tr>
<tr class="odd">
<td>YFL001W</td>
<td align="right">138.282126</td>
<td align="right">-0.4780824</td>
<td align="right">0.4012022</td>
<td align="right">-1.1916247</td>
<td align="right">0.2334084</td>
<td align="right">0.3700795</td>
</tr>
<tr class="even">
<td>YPR059C</td>
<td align="right">6.445059</td>
<td align="right">-0.2402650</td>
<td align="right">0.7820804</td>
<td align="right">-0.3072127</td>
<td align="right">0.7586815</td>
<td align="right">0.8422931</td>
</tr>
<tr class="odd">
<td>YLL039C</td>
<td align="right">9189.263978</td>
<td align="right">-1.1966941</td>
<td align="right">0.1632193</td>
<td align="right">-7.3318193</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td>YMR291W</td>
<td align="right">476.968986</td>
<td align="right">-0.4244133</td>
<td align="right">0.1941120</td>
<td align="right">-2.1864356</td>
<td align="right">0.0287838</td>
<td align="right">0.0758575</td>
</tr>
<tr class="odd">
<td>YMR022W</td>
<td align="right">540.335918</td>
<td align="right">-0.1484145</td>
<td align="right">0.2328350</td>
<td align="right">-0.6374236</td>
<td align="right">0.5238489</td>
<td align="right">0.6579550</td>
</tr>
<tr class="even">
<td>YDR327W</td>
<td align="right">182.555445</td>
<td align="right">0.2791229</td>
<td align="right">0.2313968</td>
<td align="right">1.2062525</td>
<td align="right">0.2277202</td>
<td align="right">0.3633561</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Sort genes by significance (increasing adjusted pvalue)
deseq2.res.dds.2groups &lt;-<span class="st"> </span>deseq2.res.dds.2groups[<span class="kw">order</span>(deseq2.res.dds.2groups<span class="op">$</span>padj, <span class="dt">decreasing =</span> <span class="ot">FALSE</span>), ]

## Compute the E-value, i.e. the expected number of false positives
deseq2.res.dds.2groups<span class="op">$</span>evalue &lt;-<span class="st"> </span>deseq2.res.dds.2groups<span class="op">$</span>pvalue <span class="op">*</span><span class="st"> </span><span class="kw">nrow</span>(deseq2.res.dds.2groups)

## Flag the genes declared positive 
## (i.e. differentially expressed genes)
deseq2.res.dds.2groups<span class="op">$</span>positive &lt;-<span class="st"> </span>deseq2.res.dds.2groups<span class="op">$</span>padj <span class="op">&lt;</span><span class="st"> </span>parameters<span class="op">$</span>alpha

## Replace NA values by FALSE for the flag
deseq2.res.dds.2groups<span class="op">$</span>positive[<span class="kw">is.na</span>(deseq2.res.dds.2groups<span class="op">$</span>positive)] &lt;-<span class="st"> </span><span class="ot">FALSE</span>


<span class="kw">kable</span>(<span class="kw">head</span>(deseq2.res.dds.2groups, <span class="dt">n=</span><span class="dv">20</span>), <span class="dt">caption =</span> <span class="st">&quot;Most significant genes&quot;</span>)</code></pre></div>
<table>
<caption>Most significant genes</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">baseMean</th>
<th align="right">log2FoldChange</th>
<th align="right">lfcSE</th>
<th align="right">stat</th>
<th align="right">pvalue</th>
<th align="right">padj</th>
<th align="right">evalue</th>
<th align="left">positive</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>YDR403W</td>
<td align="right">5959.4264</td>
<td align="right">6.802686</td>
<td align="right">0.4680482</td>
<td align="right">14.534158</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td>YDL218W</td>
<td align="right">1330.4326</td>
<td align="right">5.967320</td>
<td align="right">0.4179094</td>
<td align="right">14.278979</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td>YDR402C</td>
<td align="right">478.8717</td>
<td align="right">3.829931</td>
<td align="right">0.3565490</td>
<td align="right">10.741668</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td>YLR012C</td>
<td align="right">283.6450</td>
<td align="right">3.701907</td>
<td align="right">0.3546005</td>
<td align="right">10.439656</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td>YDR019C</td>
<td align="right">1656.5499</td>
<td align="right">-1.666358</td>
<td align="right">0.1606152</td>
<td align="right">-10.374847</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td>YHR136C</td>
<td align="right">364.6893</td>
<td align="right">-3.235702</td>
<td align="right">0.3132024</td>
<td align="right">-10.331024</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td>YCL048W-A</td>
<td align="right">1166.8588</td>
<td align="right">3.494313</td>
<td align="right">0.3499982</td>
<td align="right">9.983804</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td>YBR054W</td>
<td align="right">465.8305</td>
<td align="right">1.959048</td>
<td align="right">0.1967399</td>
<td align="right">9.957556</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td>YDR481C</td>
<td align="right">1776.0346</td>
<td align="right">-1.571571</td>
<td align="right">0.1632137</td>
<td align="right">-9.628918</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td>YHL028W</td>
<td align="right">1033.7650</td>
<td align="right">2.896800</td>
<td align="right">0.3057668</td>
<td align="right">9.473886</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td>YKR093W</td>
<td align="right">219.6354</td>
<td align="right">2.245243</td>
<td align="right">0.2438249</td>
<td align="right">9.208421</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td>YPL017C</td>
<td align="right">468.1963</td>
<td align="right">-1.658115</td>
<td align="right">0.1807319</td>
<td align="right">-9.174449</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td>YPR140W</td>
<td align="right">4221.3821</td>
<td align="right">2.969006</td>
<td align="right">0.3266519</td>
<td align="right">9.089204</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td>YKL051W</td>
<td align="right">293.1184</td>
<td align="right">-1.701202</td>
<td align="right">0.1890227</td>
<td align="right">-8.999988</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td>YIL045W</td>
<td align="right">5220.8484</td>
<td align="right">2.885226</td>
<td align="right">0.3221962</td>
<td align="right">8.954872</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td>YJL153C</td>
<td align="right">5320.2053</td>
<td align="right">-1.556167</td>
<td align="right">0.1743461</td>
<td align="right">-8.925733</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td>YLR265C</td>
<td align="right">290.9602</td>
<td align="right">2.397209</td>
<td align="right">0.2684635</td>
<td align="right">8.929364</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td>YEL057C</td>
<td align="right">353.0099</td>
<td align="right">3.669170</td>
<td align="right">0.4146033</td>
<td align="right">8.849834</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td>YKL163W</td>
<td align="right">762.7615</td>
<td align="right">-2.117467</td>
<td align="right">0.2400691</td>
<td align="right">-8.820242</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td>YGR088W</td>
<td align="right">552.0450</td>
<td align="right">3.103978</td>
<td align="right">0.3588942</td>
<td align="right">8.648729</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">TRUE</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## BEWARE, DESeq2 produces NA values for some genes
<span class="kw">kable</span>(<span class="kw">tail</span>(deseq2.res.dds.2groups, <span class="dt">n=</span><span class="dv">10</span>), <span class="dt">caption =</span> <span class="st">&quot;Less significant genes&quot;</span>)</code></pre></div>
<table>
<caption>Less significant genes</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">baseMean</th>
<th align="right">log2FoldChange</th>
<th align="right">lfcSE</th>
<th align="right">stat</th>
<th align="right">pvalue</th>
<th align="right">padj</th>
<th align="right">evalue</th>
<th align="left">positive</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>YGR280C</td>
<td align="right">407.252191</td>
<td align="right">-0.0025888</td>
<td align="right">0.4236952</td>
<td align="right">-0.0061100</td>
<td align="right">0.9951249</td>
<td align="right">0.9966600</td>
<td align="right">6243.414</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td>snR36</td>
<td align="right">6.144165</td>
<td align="right">-0.0038069</td>
<td align="right">0.7130335</td>
<td align="right">-0.0053390</td>
<td align="right">0.9957401</td>
<td align="right">0.9970114</td>
<td align="right">6247.273</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td>YLR004C</td>
<td align="right">431.919635</td>
<td align="right">0.0011678</td>
<td align="right">0.2387877</td>
<td align="right">0.0048905</td>
<td align="right">0.9960980</td>
<td align="right">0.9972106</td>
<td align="right">6249.519</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td>YKL069W</td>
<td align="right">224.639084</td>
<td align="right">0.0007356</td>
<td align="right">0.2364025</td>
<td align="right">0.0031116</td>
<td align="right">0.9975173</td>
<td align="right">0.9983264</td>
<td align="right">6258.424</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td>YNL113W</td>
<td align="right">74.851169</td>
<td align="right">-0.0014150</td>
<td align="right">0.4572433</td>
<td align="right">-0.0030947</td>
<td align="right">0.9975308</td>
<td align="right">0.9983264</td>
<td align="right">6258.508</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td>YLR185W_intron</td>
<td align="right">21.555742</td>
<td align="right">0.0016993</td>
<td align="right">0.6165551</td>
<td align="right">0.0027562</td>
<td align="right">0.9978009</td>
<td align="right">0.9984374</td>
<td align="right">6260.203</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td>YCL026C-B</td>
<td align="right">35.385132</td>
<td align="right">-0.0006591</td>
<td align="right">0.4163923</td>
<td align="right">-0.0015828</td>
<td align="right">0.9987371</td>
<td align="right">0.9992149</td>
<td align="right">6266.077</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td>YOR004W</td>
<td align="right">175.276696</td>
<td align="right">0.0004707</td>
<td align="right">0.4393938</td>
<td align="right">0.0010713</td>
<td align="right">0.9991453</td>
<td align="right">0.9994639</td>
<td align="right">6268.637</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td>YDL166C</td>
<td align="right">98.693288</td>
<td align="right">0.0001988</td>
<td align="right">0.3229337</td>
<td align="right">0.0006157</td>
<td align="right">0.9995088</td>
<td align="right">0.9996681</td>
<td align="right">6270.918</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td>YBR293W</td>
<td align="right">330.930747</td>
<td align="right">-0.0000905</td>
<td align="right">0.2949593</td>
<td align="right">-0.0003069</td>
<td align="right">0.9997551</td>
<td align="right">0.9997551</td>
<td align="right">6272.464</td>
<td align="left">FALSE</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(<span class="kw">table</span>(deseq2.res.dds.2groups<span class="op">$</span>positive), 
      <span class="dt">caption =</span> <span class="st">&quot;Number of genes declared positive&quot;</span>)</code></pre></div>
<table>
<caption>Number of genes declared positive</caption>
<thead>
<tr class="header">
<th align="left">Var1</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FALSE</td>
<td align="right">4138</td>
</tr>
<tr class="even">
<td align="left">TRUE</td>
<td align="right">2136</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Export the results
outfiles &lt;-<span class="st"> </span><span class="kw">vector</span>()</code></pre></div>
</div>
<div id="diagnostic-plots" class="section level3">
<h3>Diagnostic plots</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Assign a symbol to genes according to the test result
deseq2.res.dds.2groups<span class="op">$</span>pch &lt;-<span class="st"> </span><span class="dv">1</span> <span class="co"># circle</span>
deseq2.res.dds.2groups<span class="op">$</span>pch[deseq2.res.dds.2groups<span class="op">$</span>positive] &lt;-<span class="st"> </span><span class="dv">3</span> <span class="co"># + symbol</span>

## Set one color per gene depending on whether it is positive or not
gene.color &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dt">x =</span> <span class="st">&quot;#BBBBBB&quot;</span>, <span class="dt">times =</span> <span class="kw">nrow</span>(deseq2.res.dds.2groups))
gene.color[deseq2.res.dds.2groups<span class="op">$</span>positive] &lt;-<span class="st"> &quot;#FF8844&quot;</span>
gene.pch &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dt">x =</span> <span class="dv">1</span>, <span class="dt">times =</span> <span class="kw">nrow</span>(deseq2.res.dds.2groups))
gene.pch[deseq2.res.dds.2groups<span class="op">$</span>positive] &lt;-<span class="st"> &quot;#+&quot;</span>

<span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))
## MA plot? 
## Note: we explicitly indicate the package name to avoid confusion with limma::plotMA()
DESeq2<span class="op">::</span><span class="kw">plotMA</span>(<span class="dt">object =</span> <span class="kw">results</span>(dds.2groups), <span class="dt">las =</span> <span class="dv">1</span>, 
               <span class="dt">panel.first =</span> <span class="kw">grid</span>(),
               <span class="dt">main =</span> <span class="st">&quot;DESDeq2 : MA plot&quot;</span>)


## Volcano plot of adjusted p-values
<span class="kw">plot</span>(<span class="dt">x =</span> deseq2.res.dds.2groups<span class="op">$</span>log2FoldChange, 
     <span class="dt">y =</span> <span class="op">-</span><span class="kw">log10</span>(deseq2.res.dds.2groups<span class="op">$</span>padj),
     <span class="dt">main =</span> <span class="st">&quot;Volcano plot</span><span class="ch">\n</span><span class="st">Adjusted P-values&quot;</span>,
     <span class="dt">las =</span> <span class="dv">1</span>, <span class="dt">panel.first =</span> <span class="kw">grid</span>(),
     <span class="dt">xlab =</span> <span class="st">&quot;Log2 fold change&quot;</span>,
     <span class="dt">ylab  =</span> <span class="st">&quot;-log10(padj)&quot;</span>,
     <span class="dt">col =</span> <span class="kw">densCols</span>(
       <span class="dt">x =</span> deseq2.res.dds.2groups<span class="op">$</span>log2FoldChange, 
       <span class="dt">y =</span> <span class="op">-</span><span class="kw">log10</span>(deseq2.res.dds.2groups<span class="op">$</span>padj)),
     <span class="dt">pch =</span> deseq2.res.dds.2groups<span class="op">$</span>pch
     )
<span class="kw">abline</span>(<span class="dt">v =</span> <span class="dv">0</span>)
<span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">0</span>)


## P-value histogram
<span class="kw">hist</span>(deseq2.res.dds.2groups<span class="op">$</span>pvalue, 
     <span class="dt">breaks =</span> <span class="dv">20</span>, <span class="dt">col =</span> <span class="st">&quot;#DDEEFF&quot;</span>, 
     <span class="dt">xlab =</span> <span class="st">&quot;Nominal P-value&quot;</span>, 
     <span class="dt">ylab =</span> <span class="st">&quot;Number of genes&quot;</span>)


<span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</code></pre></div>
<p><img src="figures/TP_deg_diagnostic_plots_2groups-1.png" width="95%" style="display: block; margin: auto;" /></p>
</div>
<div id="toutes-conditions-confondues" class="section level3">
<h3>Toutes conditions confondues</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Use the DESeqDataSetFromMatrix to create a DESeqDataSet object
<span class="kw">message</span>(<span class="st">&quot;Creating DESeq2 dataset&quot;</span>)
dds.all.conditions &lt;-<span class="st"> </span><span class="kw">DESeqDataSetFromMatrix</span>(
  <span class="dt">countData =</span> filtered.counts, 
  <span class="dt">colData =</span> samples, 
  <span class="dt">design =</span> <span class="op">~</span><span class="st"> </span>Condition)
<span class="co"># View(dds.all.conditions)</span>

## Run differential expression analysis with DESeq2
dds.all.conditions &lt;-<span class="st"> </span><span class="kw">DESeq</span>(dds.all.conditions)
<span class="co"># View(dds.all.conditions)</span>

<span class="kw">resultsNames</span>(dds.all.conditions)</code></pre></div>
<pre><code>[1] &quot;Intercept&quot;                                                        &quot;Condition_bdf1_Y187F_Y354F_mutant_4_vs_bdf1_Y187F_Y354F_mutant_0&quot; &quot;Condition_bdf1_Y187F_Y354F_mutant_8_vs_bdf1_Y187F_Y354F_mutant_0&quot; &quot;Condition_Wild_type_0_vs_bdf1_Y187F_Y354F_mutant_0&quot;              
[5] &quot;Condition_Wild_type_4_vs_bdf1_Y187F_Y354F_mutant_0&quot;               &quot;Condition_Wild_type_8_vs_bdf1_Y187F_Y354F_mutant_0&quot;              </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Get the result table
deseq2.res.all.conditions &lt;-<span class="st"> </span><span class="kw">results</span>(dds.all.conditions)
deseq2.res.all.conditions.df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(deseq2.res.all.conditions)
<span class="kw">kable</span>(deseq2.res.all.conditions.df[some.genes, ])</code></pre></div>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">baseMean</th>
<th align="right">log2FoldChange</th>
<th align="right">lfcSE</th>
<th align="right">stat</th>
<th align="right">pvalue</th>
<th align="right">padj</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>YBR225W</td>
<td align="right">302.641570</td>
<td align="right">0.2096397</td>
<td align="right">0.2393356</td>
<td align="right">0.8759233</td>
<td align="right">0.3810717</td>
<td align="right">0.4412991</td>
</tr>
<tr class="even">
<td>YGR010W</td>
<td align="right">247.073433</td>
<td align="right">-0.8058750</td>
<td align="right">0.1897339</td>
<td align="right">-4.2473972</td>
<td align="right">0.0000216</td>
<td align="right">0.0000458</td>
</tr>
<tr class="odd">
<td>YKL001C</td>
<td align="right">235.681556</td>
<td align="right">0.3805740</td>
<td align="right">0.4589381</td>
<td align="right">0.8292492</td>
<td align="right">0.4069634</td>
<td align="right">0.4671219</td>
</tr>
<tr class="even">
<td>YPR145W</td>
<td align="right">2772.090602</td>
<td align="right">-0.3093528</td>
<td align="right">0.3590851</td>
<td align="right">-0.8615028</td>
<td align="right">0.3889612</td>
<td align="right">0.4489224</td>
</tr>
<tr class="odd">
<td>YFL001W</td>
<td align="right">185.724353</td>
<td align="right">-0.9781276</td>
<td align="right">0.3134956</td>
<td align="right">-3.1200676</td>
<td align="right">0.0018081</td>
<td align="right">0.0031285</td>
</tr>
<tr class="even">
<td>YPR059C</td>
<td align="right">7.076586</td>
<td align="right">1.5086643</td>
<td align="right">0.8368546</td>
<td align="right">1.8027794</td>
<td align="right">0.0714229</td>
<td align="right">0.0972877</td>
</tr>
<tr class="odd">
<td>YLL039C</td>
<td align="right">8112.265124</td>
<td align="right">0.0082526</td>
<td align="right">0.2028185</td>
<td align="right">0.0406894</td>
<td align="right">0.9675435</td>
<td align="right">0.9731273</td>
</tr>
<tr class="even">
<td>YMR291W</td>
<td align="right">527.551116</td>
<td align="right">-0.8971674</td>
<td align="right">0.1819015</td>
<td align="right">-4.9321615</td>
<td align="right">0.0000008</td>
<td align="right">0.0000020</td>
</tr>
<tr class="odd">
<td>YMR022W</td>
<td align="right">564.104525</td>
<td align="right">0.4024555</td>
<td align="right">0.1988592</td>
<td align="right">2.0238220</td>
<td align="right">0.0429885</td>
<td align="right">0.0608002</td>
</tr>
<tr class="even">
<td>YDR327W</td>
<td align="right">193.958497</td>
<td align="right">0.4798121</td>
<td align="right">0.1960650</td>
<td align="right">2.4472091</td>
<td align="right">0.0143967</td>
<td align="right">0.0221060</td>
</tr>
</tbody>
</table>
</div>
<div id="exercices" class="section level3">
<h3>Exercices</h3>
<ol style="list-style-type: decimal">
<li><p>Dessinez un nuage (<code>plot()</code>) qui compare lles p-valeurs nominales (non-ajustées) retournées respectivement par edgeR et DEseq2.</p></li>
<li><p>Dessinez un nuage (<code>plot()</code>) qui compare lles p-valeurs ajustées retournées respectivement par edgeR et DEseq2.</p></li>
<li><p>Dessinez un diagramme de Venn qui compare les ensemble de gènes déclarés positifs par edgeR et DESeq2.</p></li>
</ol>
</div>
</div>
<div id="controles-negatifs" class="section level2">
<h2>Contrôles négatifs</h2>
<div id="exercice" class="section level3">
<h3>Exercice</h3>
<p>Générez un tableau ayant les mêmes dimensions (lignes, colonnes) que votre tableau d’expresion, en le remplissant de nombres aléatoires tirés selon une distribution de Poisson d’espérance 200 (fonction <code>rpois()</code>).</p>
<p>Effectuez une analyse différentielle sur ces données.</p>
</div>
</div>
<div id="fichiers-de-resultats" class="section level2">
<h2>Fichiers de résultats</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Build a file prefix by concatenating the two compared groups
file.prefix &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="dt">sep =</span> <span class="st">&quot;__&quot;</span>, parameters<span class="op">$</span>test.group, 
                     <span class="st">&quot;vs&quot;</span>, parameters<span class="op">$</span>control.group)


## DESeq2 result table
outfiles[<span class="st">&quot;DESeq2 all genes&quot;</span>] &lt;-<span class="st"> </span><span class="kw">file.path</span>(
  parameters<span class="op">$</span>workdir, 
  <span class="kw">paste</span>(<span class="dt">sep =</span> <span class="st">&quot;&quot;</span>, file.prefix, <span class="st">&quot;_DESeq2_all_genes.tsv&quot;</span>))
<span class="kw">write.table</span>(
  <span class="dt">x =</span> deseq2.res.dds.2groups, 
  <span class="dt">quote =</span> <span class="ot">FALSE</span>, <span class="dt">row.names =</span> <span class="ot">TRUE</span>, <span class="dt">col.names =</span> <span class="ot">NA</span>,
  <span class="dt">file =</span> outfiles[<span class="st">&quot;DESeq2 all genes&quot;</span>])


## Differentially expressed genes
outfiles[<span class="st">&quot;DESeq2 positive genes&quot;</span>] &lt;-<span class="st"> </span><span class="kw">file.path</span>(
  parameters<span class="op">$</span>workdir, 
  <span class="kw">paste</span>(<span class="dt">sep =</span> <span class="st">&quot;&quot;</span>, file.prefix, <span class="st">&quot;_DESeq2_positive.tsv&quot;</span>))
<span class="kw">write.table</span>(
  <span class="dt">x =</span> deseq2.res.dds.2groups[deseq2.res.dds.2groups<span class="op">$</span>positive, ], 
  <span class="dt">quote =</span> <span class="ot">FALSE</span>, <span class="dt">row.names =</span> <span class="ot">TRUE</span>, <span class="dt">col.names =</span> <span class="ot">NA</span>,
  <span class="dt">file =</span> outfiles[<span class="st">&quot;DESeq2 positive genes&quot;</span>] )

## Differentially expressed genes
outfiles[<span class="st">&quot;DESeq2 DEG iDs&quot;</span>] &lt;-<span class="st"> </span><span class="kw">file.path</span>(
  parameters<span class="op">$</span>workdir, <span class="kw">paste</span>(<span class="dt">sep =</span> <span class="st">&quot;&quot;</span>, file.prefix, <span class="st">&quot;_DESeq2_positive_geneIDs.txt&quot;</span>))
<span class="kw">write.table</span>(
  <span class="dt">x =</span> <span class="kw">row.names</span>(deseq2.res.dds.2groups[deseq2.res.dds.2groups<span class="op">$</span>positive, ]), 
  <span class="dt">quote =</span> <span class="ot">FALSE</span>, <span class="dt">row.names =</span> <span class="ot">FALSE</span>, <span class="dt">col.names =</span> <span class="ot">FALSE</span>,
  <span class="dt">file =</span> outfiles[<span class="st">&quot;DESeq2 DEG iDs&quot;</span>] )

<span class="kw">kable</span>(outfiles, <span class="dt">caption =</span> <span class="st">&quot;fichiers de résultats&quot;</span>, <span class="dt">col.names =</span> <span class="st">&quot;Result file&quot;</span>)</code></pre></div>
<table>
<caption>fichiers de résultats</caption>
<thead>
<tr class="header">
<th></th>
<th align="left">Result file</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>DESeq2 all genes</td>
<td align="left">~/DU-Bii/m3s5//bdf1_Y187F_Y354F_mutant_8__vs__Wild_type_8_DESeq2_all_genes.tsv</td>
</tr>
<tr class="even">
<td>DESeq2 positive genes</td>
<td align="left">~/DU-Bii/m3s5//bdf1_Y187F_Y354F_mutant_8__vs__Wild_type_8_DESeq2_positive.tsv</td>
</tr>
<tr class="odd">
<td>DESeq2 DEG iDs</td>
<td align="left">~/DU-Bii/m3s5//bdf1_Y187F_Y354F_mutant_8__vs__Wild_type_8_DESeq2_positive_geneIDs.txt</td>
</tr>
</tbody>
</table>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
