---
title: "Tutorial: exploration of multi-omics data"
author: "Jacques van Helden"
date: '`r Sys.Date()`'
output:
  html_document:
    self_contained: no
    fig_caption: yes
    highlight: zenburn
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
    code_folding: "hide"
  ioslides_presentation:
    slide_level: 2
    self_contained: no
    colortheme: dolphin
    fig_caption: yes
    fig_height: 5
    fig_width: 7
    fonttheme: structurebold
    highlight: tango
    smaller: yes
    toc: yes
    widescreen: yes
  slidy_presentation:
    smart: no
    slide_level: 2
    self_contained: yes
    fig_caption: yes
    fig_height: 5
    fig_width: 7
    highlight: tango
    incremental: no
    keep_md: yes
    smaller: yes
    theme: cerulean
    toc: yes
    widescreen: yes
  beamer_presentation:
    colortheme: dolphin
    fig_caption: yes
    fig_height: 5
    fig_width: 7
    fonttheme: structurebold
    highlight: tango
    incremental: no
    keep_tex: no
    slide_level: 2
    theme: Montpellier
    toc: yes
  revealjs::revealjs_presentation:
    theme: night
    transition: none
    self_contained: true
    css: ../slides.css
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  powerpoint_presentation:
    slide_level: 2
    fig_caption: yes
    fig_height: 5
    fig_width: 7
    toc: yes
font-import: http://fonts.googleapis.com/css?family=Risque
subtitle: DUBii 2020
font-family: Garamond
transition: linear
editor_options: 
  chunk_output_type: console
---




```{r settings, include=FALSE, echo=FALSE, eval=TRUE}

options(width = 300)
# options(encoding = 'UTF-8')
knitr::opts_chunk$set(
  fig.width = 7, fig.height = 5, 
  fig.path = 'figures/pavkovic2019_',
  fig.align = "center", 
  size = "tiny", 
  echo = TRUE, 
  eval = TRUE, 
  warning = FALSE, 
  message = FALSE, 
  results = TRUE, 
  comment = "")

options(scipen = 12) ## Max number of digits for non-scientific notation
# knitr::asis_output("\\footnotesize")

```



```{r libraries, echo=FALSE, eval=TRUE}
requiredLib <- c("knitr")
for (lib in requiredLib) {
  if (!require(lib, character.only = TRUE)) {
    install.packages(lib, )
  }
  require(lib, character.only = TRUE)
}

```


## Goals of this tutorial

This tutorial aims at using the methods taught in the previous courses in order to apply them to an illustrative dataset combining proteome and transcriptome data. 

## Study case : mouse kidney 

As study case for this tutorial, we will use multi-omics data from a study published by [Pavkovic et al. (2019)](https://doi.org/10.1038/s41597-019-0095-5), which combines transctiptomics (RNA-seq) and proteomics approaches to understand the molecular mechanisms underlying the kidney fibrosis pathology. 

The authors applied two commonly used treatments to induce kidney fibrosis in mouse: 

- a reversible chemical-induced injury model, denoted as **FA** for **folic acid** induced nephropathy;  

- an irreversible surgically-induced fibrosis model, denoted as **UUO** for **unilateral uretral obstruction**.

## References and data sources


-  **Reference**: Pavkovic, M., Pantano, L., Gerlach, C.V. et al. Multi omics analysis of fibrotic kidneys in two mouse models. Sci Data 6, 92 (2019) <https://doi.org/10.1038/s41597-019-0095-5>

- **Mouse fibrotic kidney browser:** <http://hbcreports.med.harvard.edu/fmm/>

- Data on Zenodo: <https://zenodo.org/record/2592516>

## Data preparation

A description of the study case can be found in the *[Mus musculus section](https://du-bii.github.io/study-cases/#mus-musculus)* of the the [DUBii study cases repository](https://du-bii.github.io/study-cases/) repository. 

We also provide there a detailed explanation of the [data preparation steps](https://du-bii.github.io/study-cases/Mus_musculus/pavkovic_2019/Rmd/prepare-data_pavkovic_2019.html): 

- downloading the data from its original source repository, 
- exploring the datasets it with various graphical representtions, 
- computing some descriptive statistics on the different samples, 
- pre-processing,
- storing the results in a memory image. 


## Data file naming

We prepared the data from Pavkovic as a text file with tab-separated values (**tsv** files). 

All the files are available on github:
- <https://github.com/DU-Bii/module-3-Stat-R/tree/master/stat-R_2021/data/pavkovic_2019>

The files are named according to the following convention:

- the prefix indicate the data type

    - **fa**: folic acid induced nephropathy (reversible), transcriptome data
    - **pfa**: folic acid induced nephropathy, proteome data
    - **uuo**: unilateral uretral obstruction (irreversible), transcriptome data
    - **puuo**: unilateral uretral obstruction, proteome data

- The suffix indicates the data  normalisation

    - **raw**: transcriptome counts provided by the authors (note: not integer, because their raw data was already somehow normalized)
    - **normalizedÂ¨**: transcriptome counts standardized to achieve the same third quartile for each sample
    - **log2**: log2 transformation for proteome data
    
- the  **metadata** files contain a short description of each sample (one row per sample). Note that the last column contains a sample-specific color specification in [hexadecimal web color code](https://en.wikipedia.org/wiki/Web_colors#Hex_triplet) to facilitate the drawings. Don't hesitate to chose [other colors](https://htmlcolorcodes.com/) according to your personal taste. 

## Approach for this tutorial

This tutorial will consist of exercises that can be realised in a stepwise way, with alternance of working sessions and live demos of the solutions, in order to make sure that all the trainees acquire each step.

## Data exploration

Before computing any descriptive parameter on a dataset, I generallyi attempt to get a picture of the whole distribution.


### Finding a data file on github

We will provide here a magic recipe to download the data from the github repository to your local folder, and to load it in R. 


```{r data_loading}
## specify the base URL from which data files can be downloaded 
urlBase <- "https://github.com/DU-Bii/module-3-Stat-R/raw/master/stat-R_2021/data/pavkovic_2019"


## Choose a specific data file
dataPrefix <- "pfa" ## proteome data of folic-acid treated mouse
dataSuffix <- "model" ## no normalization
fileName <- paste0(dataPrefix, "_", dataSuffix, "_counts.tsv.gz")

## Compose the URL to download the file from github
url <- file.path(urlBase, fileName)

message("URL: ",  url)

```

### Loading a data file directly from github

Now we defined the URL, we can easily load the file directly from github to a data frame in our R environement. 

```{r load_from_github, eval=FALSE}
## this requires to load a specific package
if (!require("data.table")) {
  install.packages("data.table")
}
library(data.table)

pfa <- fread(url, header = TRUE, sep = "\t")
dim(pfa)
names(pfa)
kable(head(pfa))

```



### Downloading a data file and storing it locally once forever

We can now download the data file to a local folder, but we would like to do this only once.

```{r download_only_once}
## Specify the path relative to your home directory (~)
localFolder <- "~/DUBii-m3_data/pavkovic_2019"
localFile <- file.path(localFolder, fileName)

## Create the local data folder if it does not exist
dir.create(localFolder, showWarnings = FALSE, recursive = TRUE)

## Download the file ONLY if it is not already there
if (!file.exists(localFile)) {
  message("Downloading file from github to local file\n\t", 
          localFile)
  download.file(url = url, destfile = localFile)
} else {
  message("Local file already exists, no need to download\n\t", 
          localFile)
}

```


### Loading the local copy of your data file


We will now load the proteome file, with the following parameters

- the first row contains the column headers
- the first columnt contains the protein IDs, and we would like to have them as row.names for the loaded data frame. This is a bit tricky because some protein names are diplicated. We use the **very** convenient function `make.names(x, unique = TRUE)`.


```{r load_local_file}
## Load the data from the local file
pfa <- read.delim(file = localFile, header = TRUE, sep = "\t")

kable(head(pfa), caption = "Data frame just after loading")


## Convert the first colum to row names
row.names(pfa) <- make.names(as.vector(pfa$id), unique = TRUE)
pfa <- pfa[, -1] ## Suppress the ID colimn
kable(head(pfa), caption = "Data frame with row names")

```


### Exercise: writing functions to download and load a data file

Write two functions that will respectively download a file from a remote location to a local folder, but do this only if the local file is not yet present there. 

Use this function to load 
- the raw proteome data from folic acid treatment
- the proteome metadata


```{r exercise_download-load_function}
## Your turn to program

```


### Solution: implement functions to automagically download and load a data file

```{r solution_download-load_functions}
#' @title Download a file only if it is not yet here
#' @author Jacques van Helden
#' @param urlBase base of the URL, that will be prepended to the file name
#' @param fileNAme name of the file (should not contain any path)
#' @param localFolder path of a local folder where the file should be stored
#' @return the function returns the path of the local file, built from localFolder and fileName
downloadOnlyOnce <- function(urlBase, 
                             fileName,
                             localFolder) {

  ## Define the source URL  
  url <- file.path(urlBase, fileName)
  message("Source URL\n\t",  url)

  ## Define the local file
  localFile <- file.path(localFolder, fileName)
  
  ## Create the local data folder if it does not exist
  dir.create(localFolder, showWarnings = FALSE, recursive = TRUE)
  
  ## Download the file ONLY if it is not already there
  if (!file.exists(localFile)) {
    message("Downloading file from source URL to local file\n\t", 
            localFile)
    download.file(url = url, destfile = localFile)
  } else {
    message("Local file already exists, no need to download\n\t", 
            localFile)
  }
  
  return(localFile)
}
```

We can now use our new function `downloadOnlyOnce()` to download the folic acid dataset and store it in a local folder. 


```{r download_fa}
## Specify the basic parameters
pavkovicBase <- "https://github.com/DU-Bii/module-3-Stat-R/raw/master/stat-R_2021/data/pavkovic_2019"
pavkovicFolder <- "~/DUBii-m3_data/pavkovic_2019"

#### Dowload folic acid data and metadata ####

## Transcriptome file
localFAfile <- downloadOnlyOnce(
  urlBase = pavkovicBase, 
  fileName = "fa_raw_counts.tsv.gz",
  localFolder =  pavkovicFolder
)

## Proteome file
localPFAfile <- downloadOnlyOnce(
  urlBase = pavkovicBase, 
  fileName = "pfa_model_counts.tsv.gz",
  localFolder =  pavkovicFolder
)

## Metadata
protMetadataFile <- downloadOnlyOnce(
  urlBase = pavkovicBase, 
  fileName = "proteome_metadata.tsv",
  localFolder =  pavkovicFolder
)

## Metadata
transMetadataFile <- downloadOnlyOnce(
  urlBase = pavkovicBase, 
  fileName = "transcriptome_metadata.tsv",
  localFolder =  pavkovicFolder
)


```


### Loading the files

After having run the chunk of code above, try to re-run it. In principle, you should just receive messages telling you that the files are already there. 

```{r my_easy_load_function}
#' @title Load a tab-separated value file and manually set row ames after having forced them to unique values
#' @param file file path
#' @param header=1 Header is set to 1 by default
#' @param sep="\t" Column separator is set to tab by default
#' @param rownames.col=1 Column containing the row names
#' @param ... all other parameters are passed to read.delim()
#' @return a data frame with the loaded data
myEasyLoad <- function(file, 
                       header = 1, 
                       sep = "\t",
                       rownames.col = 1, 
                       ...) {
  x <- read.delim(file = file, ...)
  rownames(x) <- make.names(x[, rownames.col], unique = TRUE)
  x <- x[, -rownames.col]
  return(x)
}

```


We can now load the data that from our local folder. 

```{r load_fa}

## Load transcriptome data
fa <- read.delim(file = localFAfile, sep = "\t", header = TRUE)
fa <- myEasyLoad(file = localFAfile, rownames.col = 1)
dim(fa)
## Check the first lines of the loaded file
View(head(fa))

## Load proteome data
pfa <- read.delim(file = localPFAfile, sep = "\t", header = TRUE)
dim(pfa)

## Check the first lines of the loaded file
kable(head(pfa))

## Load proteome metadata
protMeta <- read.delim(file = protMetadataFile, sep = "\t", header = TRUE)
kable(protMeta, caption = "Metadata for the proteome dataset")


## Load transcriptome metadata
transMeta <- read.delim(file = transMetadataFile, sep = "\t", header = TRUE)
kable(transMeta, caption = "Metadata for the transcriptome dataset")

```



We can now use this function to download and load the different data files. 


## Graphical exploration of the data

### Histogram

Draw histograms of the transcriptome and proteom data, all samples together. 

```{r fa_hist_ugly}
hist(unlist(as.vector(pfa)))

head(pfa)
```

```{r fa_hist_nice}
hist(pfa)
```



### Box plot


### Scatter plot

## Save a memory image of your session

Use the finction save() to store an image of your sessionn, in an Rdata file? Tthis file will contain all the data that you loaded during this session, and enable you to reload everything without having to re-execute all the steps. 

```{r memory_image}

```



## Save your session info

As usual

```{r session_info}
sessionInfo()
```




