<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jacques van Helden et Olivier Sand" />

<meta name="date" content="2021-03-22" />

<title>Tutorial: exploration of multi-omics data</title>

<script src="tuto_data-exploration_pavkovic_files/header-attrs-2.5/header-attrs.js"></script>
<script src="tuto_data-exploration_pavkovic_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="tuto_data-exploration_pavkovic_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="tuto_data-exploration_pavkovic_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="tuto_data-exploration_pavkovic_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="tuto_data-exploration_pavkovic_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="tuto_data-exploration_pavkovic_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="tuto_data-exploration_pavkovic_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="tuto_data-exploration_pavkovic_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="tuto_data-exploration_pavkovic_files/navigation-1.1/tabsets.js"></script>
<script src="tuto_data-exploration_pavkovic_files/navigation-1.1/codefolding.js"></script>
<link href="tuto_data-exploration_pavkovic_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="tuto_data-exploration_pavkovic_files/anchor-sections-1.0/anchor-sections.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  { color: #cccccc; background-color: #303030; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ffcfaf; } /* Alert */
code span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #dca3a3; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #f0dfaf; } /* ControlFlow */
code span.ch { color: #dca3a3; } /* Char */
code span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code span.co { color: #7f9f7f; } /* Comment */
code span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code span.do { color: #7f9f7f; } /* Documentation */
code span.dt { color: #dfdfbf; } /* DataType */
code span.dv { color: #dcdccc; } /* DecVal */
code span.er { color: #c3bf9f; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #c0bed1; } /* Float */
code span.fu { color: #efef8f; } /* Function */
code span.im { } /* Import */
code span.in { color: #7f9f7f; font-weight: bold; } /* Information */
code span.kw { color: #f0dfaf; } /* Keyword */
code span.op { color: #f0efd0; } /* Operator */
code span.ot { color: #efef8f; } /* Other */
code span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code span.sc { color: #dca3a3; } /* SpecialChar */
code span.ss { color: #cc9393; } /* SpecialString */
code span.st { color: #cc9393; } /* String */
code span.va { } /* Variable */
code span.vs { color: #cc9393; } /* VerbatimString */
code span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Tutorial: exploration of multi-omics data</h1>
<h3 class="subtitle">DUBii 2021</h3>
<h4 class="author">Jacques van Helden et Olivier Sand</h4>
<h4 class="date">2021-03-22</h4>

</div>


<div id="goals-of-this-tutorial" class="section level2">
<h2>Goals of this tutorial</h2>
<p>This tutorial aims at</p>
<ol style="list-style-type: decimal">
<li><p>Learn to load files from remote locations, either directly or by downloading them to a local folder.</p></li>
<li><p>Apply some of the methods taught in the previous courses in order to explore a data set.</p></li>
<li><p>Show some convenient ways of combinig R code and markdown elements within an R markdown document in order to obtain a well-formatted scientific report.</p>
<ul>
<li>configuring the parameters in the yaml header of the R markdown file</li>
<li>organising the code in R chunks</li>
<li>writing, documenting and using an R function</li>
<li>generating an ugly figure, improving it and controlling its incorporation in the report (size, legend)</li>
</ul></li>
</ol>
</div>
<div id="study-case-mouse-kidney" class="section level2">
<h2>Study case : mouse kidney</h2>
<p>As study case for this tutorial, we will use multi-omics data from a study published by <a href="https://doi.org/10.1038/s41597-019-0095-5">Pavkovic et al. (2019)</a>, which combines transctiptomics (RNA-seq) and proteomics approaches to understand the molecular mechanisms underlying the kidney fibrosis pathology.</p>
<p>The authors applied two commonly used treatments to induce kidney fibrosis in mouse:</p>
<ul>
<li><p>a reversible chemical-induced injury model, denoted as <strong>FA</strong> for <strong>folic acid</strong> induced nephropathy;</p></li>
<li><p>an irreversible surgically-induced fibrosis model, denoted as <strong>UUO</strong> for <strong>unilateral uretral obstruction</strong>.</p></li>
</ul>
</div>
<div id="references-and-data-sources" class="section level2">
<h2>References and data sources</h2>
<ul>
<li><p><strong>Reference</strong>: Pavkovic, M., Pantano, L., Gerlach, C.V. et al. Multi omics analysis of fibrotic kidneys in two mouse models. Sci Data 6, 92 (2019) <a href="https://doi.org/10.1038/s41597-019-0095-5" class="uri">https://doi.org/10.1038/s41597-019-0095-5</a></p></li>
<li><p><strong>Mouse fibrotic kidney browser:</strong> <a href="http://hbcreports.med.harvard.edu/fmm/" class="uri">http://hbcreports.med.harvard.edu/fmm/</a></p></li>
<li><p>Data on Zenodo: <a href="https://zenodo.org/record/2592516" class="uri">https://zenodo.org/record/2592516</a></p></li>
</ul>
</div>
<div id="data-preparation" class="section level2">
<h2>Data preparation</h2>
<p>A description of the study case can be found in the <em><a href="https://du-bii.github.io/study-cases/#mus-musculus">Mus musculus section</a></em> of the the <a href="https://du-bii.github.io/study-cases/">DUBii study cases repository</a> repository.</p>
<p>We also provide there a detailed explanation of the <a href="https://du-bii.github.io/study-cases/Mus_musculus/pavkovic_2019/Rmd/prepare-data_pavkovic_2019.html">data preparation steps</a>:</p>
<ul>
<li>downloading the data from its original source repository,</li>
<li>exploring the datasets it with various graphical representtions,</li>
<li>computing some descriptive statistics on the different samples,</li>
<li>pre-processing,</li>
<li>storing the results in a memory image.</li>
</ul>
</div>
<div id="data-file-naming" class="section level2">
<h2>Data file naming</h2>
<p>We prepared the data from Pavkovic as a text file with tab-separated values (<strong>tsv</strong> files).</p>
<p>All the files are available on github: - <a href="https://github.com/DU-Bii/module-3-Stat-R/tree/master/stat-R_2021/data/pavkovic_2019" class="uri">https://github.com/DU-Bii/module-3-Stat-R/tree/master/stat-R_2021/data/pavkovic_2019</a></p>
<p>The files are named according to the following convention:</p>
<ul>
<li><p>the prefix indicate the data type</p>
<ul>
<li><strong>fa</strong>: folic acid induced nephropathy (reversible), transcriptome data</li>
<li><strong>pfa</strong>: folic acid induced nephropathy, proteome data</li>
<li><strong>uuo</strong>: unilateral uretral obstruction (irreversible), transcriptome data</li>
<li><strong>puuo</strong>: unilateral uretral obstruction, proteome data</li>
</ul></li>
<li><p>The suffix indicates the data normalisation</p>
<ul>
<li><strong>raw</strong>: transcriptome counts provided by the authors (note: not integer, because their raw data was already somehow normalized)</li>
<li><strong>normalized¨</strong>: transcriptome counts standardized to achieve the same third quartile for each sample</li>
<li><strong>log2</strong>: log2 transformation for proteome data</li>
</ul></li>
<li><p>the <strong>metadata</strong> files contain a short description of each sample (one row per sample). Note that the last column contains a sample-specific color specification in <a href="https://en.wikipedia.org/wiki/Web_colors#Hex_triplet">hexadecimal web color code</a> to facilitate the drawings. Don’t hesitate to chose <a href="https://htmlcolorcodes.com/">other colors</a> according to your personal taste.</p></li>
</ul>
</div>
<div id="r-style" class="section level2">
<h2>R style</h2>
<p>The R code belows follows the tidyverse styling guide (<a href="https://style.tidyverse.org/" class="uri">https://style.tidyverse.org/</a>).</p>
</div>
<div id="approach-for-this-tutorial" class="section level2">
<h2>Approach for this tutorial</h2>
<p>This tutorial will consist of exercises that can be realised in a stepwise way, with alternance of working sessions and live demos of the solutions, in order to make sure that all the trainees acquire each step.</p>
</div>
<div id="data-exploration" class="section level2">
<h2>Data exploration</h2>
<p>Before computing any descriptive parameter on a dataset, I generallyi attempt to get a picture of the whole distribution.</p>
<div id="finding-a-data-file-on-github" class="section level3">
<h3>Finding a data file on github</h3>
<p>We will provide here a magic recipe to download the data from the github repository to your local folder, and to load it in R.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="do">## specify the base URL from which data files can be downloaded </span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>url_base <span class="ot">&lt;-</span> <span class="st">&quot;https://github.com/DU-Bii/module-3-Stat-R/raw/master/stat-R_2021/data/pavkovic_2019&quot;</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Choose a specific data file</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>data_prefix <span class="ot">&lt;-</span> <span class="st">&quot;pfa&quot;</span> <span class="do">## proteome data of folic-acid treated mouse</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>data_suffix <span class="ot">&lt;-</span> <span class="st">&quot;model&quot;</span> <span class="do">## no normalization</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>file_name <span class="ot">&lt;-</span> <span class="fu">paste0</span>(data_prefix, <span class="st">&quot;_&quot;</span>, data_suffix, <span class="st">&quot;_counts.tsv.gz&quot;</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="do">## Compose the URL to download the file from github</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>url <span class="ot">&lt;-</span> <span class="fu">file.path</span>(url_base, file_name)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="st">&quot;URL: &quot;</span>,  url)</span></code></pre></div>
</div>
<div id="loading-a-data-file-directly-from-github" class="section level3">
<h3>Loading a data file directly from github</h3>
<p>Now we defined the URL, we can easily load the file directly from github to a data frame in our R environement.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="do">## this requires to load a specific package</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">&quot;data.table&quot;</span>)) {</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;data.table&quot;</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>pfa <span class="ot">&lt;-</span> <span class="fu">fread</span>(url, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(pfa)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pfa)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(<span class="fu">head</span>(pfa))</span></code></pre></div>
</div>
<div id="downloading-a-data-file-and-storing-it-locally-once-forever" class="section level3">
<h3>Downloading a data file and storing it locally once forever</h3>
<p>We can now download the data file to a local folder, but we would like to do this only once.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Specify the path relative to your home directory (~)</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>local_folder <span class="ot">&lt;-</span> <span class="st">&quot;~/DUBii-m3_data/pavkovic_2019&quot;</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>local_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(local_folder, file_name)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Create the local data folder if it does not exist</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(local_folder, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="do">## Download the file ONLY if it is not already there</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(local_file)) {</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="st">&quot;Downloading file from github to local file</span><span class="sc">\n\t</span><span class="st">&quot;</span>, </span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>          local_file)</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">download.file</span>(<span class="at">url =</span> url, <span class="at">destfile =</span> local_file)</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="st">&quot;Local file already exists, no need to download</span><span class="sc">\n\t</span><span class="st">&quot;</span>, </span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>          local_file)</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="loading-the-local-copy-of-your-data-file" class="section level3">
<h3>Loading the local copy of your data file</h3>
<p>We will now load the proteome file, with the following parameters</p>
<ul>
<li>the first row contains the column headers</li>
<li>the first columnt contains the protein IDs, and we would like to have them as row.names for the loaded data frame. This is a bit tricky because some protein names are diplicated. We use the <strong>very</strong> convenient function <code>make.names(x, unique = TRUE)</code>.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Load the data from the local file</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>pfa <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="at">file =</span> local_file, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(<span class="fu">head</span>(pfa), <span class="at">caption =</span> <span class="st">&quot;Data frame just after loading&quot;</span>)</span></code></pre></div>
<table>
<caption>Data frame just after loading</caption>
<colgroup>
<col width="16%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="7%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">id</th>
<th align="right">normal_1</th>
<th align="right">normal_2</th>
<th align="right">day1_1</th>
<th align="right">day1_2</th>
<th align="right">day2_1</th>
<th align="right">day2_2</th>
<th align="right">day7_1</th>
<th align="right">day7_2</th>
<th align="right">day14_1</th>
<th align="right">day14_2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ENSMUSG00000037686</td>
<td align="right">531.2680</td>
<td align="right">651.7200</td>
<td align="right">335.5910</td>
<td align="right">334.8460</td>
<td align="right">197.1740</td>
<td align="right">307.194</td>
<td align="right">123.2060</td>
<td align="right">272.6190</td>
<td align="right">93.7247</td>
<td align="right">196.1590</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000027831</td>
<td align="right">221.6020</td>
<td align="right">266.3590</td>
<td align="right">175.4090</td>
<td align="right">159.4190</td>
<td align="right">234.8080</td>
<td align="right">256.927</td>
<td align="right">149.9380</td>
<td align="right">315.0590</td>
<td align="right">110.5880</td>
<td align="right">126.3600</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000039201</td>
<td align="right">26.0723</td>
<td align="right">29.1331</td>
<td align="right">57.7329</td>
<td align="right">45.8475</td>
<td align="right">81.6009</td>
<td align="right">88.870</td>
<td align="right">29.8560</td>
<td align="right">44.5586</td>
<td align="right">13.6292</td>
<td align="right">27.6568</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000031095</td>
<td align="right">4363.0500</td>
<td align="right">4784.0800</td>
<td align="right">4064.4800</td>
<td align="right">3917.2900</td>
<td align="right">4599.0300</td>
<td align="right">5957.030</td>
<td align="right">2806.5200</td>
<td align="right">6792.0900</td>
<td align="right">2022.5100</td>
<td align="right">3226.7500</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000034931</td>
<td align="right">879.2790</td>
<td align="right">1065.3900</td>
<td align="right">914.2870</td>
<td align="right">928.2760</td>
<td align="right">1000.1000</td>
<td align="right">1264.270</td>
<td align="right">738.3520</td>
<td align="right">1362.0700</td>
<td align="right">466.4440</td>
<td align="right">714.5870</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000038208</td>
<td align="right">68.2225</td>
<td align="right">89.8871</td>
<td align="right">57.9041</td>
<td align="right">76.3510</td>
<td align="right">84.8474</td>
<td align="right">105.245</td>
<td align="right">72.8696</td>
<td align="right">138.9810</td>
<td align="right">40.8101</td>
<td align="right">59.2117</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Convert the first colum to row names</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(pfa) <span class="ot">&lt;-</span> <span class="fu">make.names</span>(<span class="fu">as.vector</span>(pfa<span class="sc">$</span>id), <span class="at">unique =</span> <span class="cn">TRUE</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>pfa <span class="ot">&lt;-</span> pfa[, <span class="sc">-</span><span class="dv">1</span>] <span class="do">## Suppress the ID colimn</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(<span class="fu">head</span>(pfa), <span class="at">caption =</span> <span class="st">&quot;Data frame with row names&quot;</span>)</span></code></pre></div>
<table>
<caption>Data frame with row names</caption>
<colgroup>
<col width="16%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="7%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">normal_1</th>
<th align="right">normal_2</th>
<th align="right">day1_1</th>
<th align="right">day1_2</th>
<th align="right">day2_1</th>
<th align="right">day2_2</th>
<th align="right">day7_1</th>
<th align="right">day7_2</th>
<th align="right">day14_1</th>
<th align="right">day14_2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ENSMUSG00000037686</td>
<td align="right">531.2680</td>
<td align="right">651.7200</td>
<td align="right">335.5910</td>
<td align="right">334.8460</td>
<td align="right">197.1740</td>
<td align="right">307.194</td>
<td align="right">123.2060</td>
<td align="right">272.6190</td>
<td align="right">93.7247</td>
<td align="right">196.1590</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000027831</td>
<td align="right">221.6020</td>
<td align="right">266.3590</td>
<td align="right">175.4090</td>
<td align="right">159.4190</td>
<td align="right">234.8080</td>
<td align="right">256.927</td>
<td align="right">149.9380</td>
<td align="right">315.0590</td>
<td align="right">110.5880</td>
<td align="right">126.3600</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000039201</td>
<td align="right">26.0723</td>
<td align="right">29.1331</td>
<td align="right">57.7329</td>
<td align="right">45.8475</td>
<td align="right">81.6009</td>
<td align="right">88.870</td>
<td align="right">29.8560</td>
<td align="right">44.5586</td>
<td align="right">13.6292</td>
<td align="right">27.6568</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000031095</td>
<td align="right">4363.0500</td>
<td align="right">4784.0800</td>
<td align="right">4064.4800</td>
<td align="right">3917.2900</td>
<td align="right">4599.0300</td>
<td align="right">5957.030</td>
<td align="right">2806.5200</td>
<td align="right">6792.0900</td>
<td align="right">2022.5100</td>
<td align="right">3226.7500</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000034931</td>
<td align="right">879.2790</td>
<td align="right">1065.3900</td>
<td align="right">914.2870</td>
<td align="right">928.2760</td>
<td align="right">1000.1000</td>
<td align="right">1264.270</td>
<td align="right">738.3520</td>
<td align="right">1362.0700</td>
<td align="right">466.4440</td>
<td align="right">714.5870</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000038208</td>
<td align="right">68.2225</td>
<td align="right">89.8871</td>
<td align="right">57.9041</td>
<td align="right">76.3510</td>
<td align="right">84.8474</td>
<td align="right">105.245</td>
<td align="right">72.8696</td>
<td align="right">138.9810</td>
<td align="right">40.8101</td>
<td align="right">59.2117</td>
</tr>
</tbody>
</table>
</div>
<div id="writing-a-function-to-download-a-file-only-once" class="section level3">
<h3>Writing a function to download a file only once</h3>
<p>Write a functions that will download a file from a remote location to a local folder, but do this only if the local file is not yet present there.</p>
<p>Note that we use the <a href="https://kbroman.org/pkg_primer/pages/docs.html">roxygen2</a> format to write the documentation of this function. In any programming language, a function should always be documented in order to enable other people to use it, and the doc is also very useful for a developer to reuse her/his own code. The documentation becomes particularly interesting when you start building your own R packages, since it will automatically generate the help pages.</p>
<p>The documentation of a function should include - a description of what it does - the author name and a way to contact her/him - a description of each parameter (argument) of the function - a description of the return value</p>
<p>Roxygen2 provides is a very convenient way of documenting a function, because - the formalism is very simple - the doc comes together with the code of the function (by default, R functions are documented in a separate file)</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @title Download a file only if it is not yet here</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @author Jacques van Helden email{Jacques.van-Helden@@france-bioinformatique.fr}</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @param url_base base of the URL, that will be prepended to the file name</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @param file_name name of the file (should not contain any path)</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @param local_folder path of a local folder where the file should be stored</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @return the function returns the path of the local file, built from local_folder and file_name</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @export</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>download_only_once <span class="ot">&lt;-</span> <span class="cf">function</span>(url_base, </span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>                             file_name,</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>                             local_folder) {</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Define the source URL  </span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>  url <span class="ot">&lt;-</span> <span class="fu">file.path</span>(url_base, file_name)</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="st">&quot;Source URL</span><span class="sc">\n\t</span><span class="st">&quot;</span>,  url)</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Define the local file</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>  local_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(local_folder, file_name)</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Create the local data folder if it does not exist</span></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir.create</span>(local_folder, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Download the file ONLY if it is not already there</span></span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(local_file)) {</span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">message</span>(<span class="st">&quot;Downloading file from source URL to local file</span><span class="sc">\n\t</span><span class="st">&quot;</span>, </span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a>            local_file)</span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">download.file</span>(<span class="at">url =</span> url, <span class="at">destfile =</span> local_file)</span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">message</span>(<span class="st">&quot;Local file already exists, no need to download</span><span class="sc">\n\t</span><span class="st">&quot;</span>, </span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true" tabindex="-1"></a>            local_file)</span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb6-32"><a href="#cb6-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(local_file)</span>
<span id="cb6-33"><a href="#cb6-33" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>We can now use our new function <code>download_only_once()</code> to download the files from the folic acid dataset and store them in a local folder. We will download successively :</p>
<ul>
<li>transcriptome data (<code>fa</code>)</li>
<li>transcriptome medata</li>
<li>proteome data (<code>pfa</code>)</li>
<li>proteome medata</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Specify the basic parameters</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>pavkovic_base <span class="ot">&lt;-</span> <span class="st">&quot;https://github.com/DU-Bii/module-3-Stat-R/raw/master/stat-R_2021/data/pavkovic_2019&quot;</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>pavkovic_folder <span class="ot">&lt;-</span> <span class="st">&quot;~/DUBii-m3_data/pavkovic_2019&quot;</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="do">#### Dowload folic acid data and metadata ####</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Transcriptome data table</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>local_fa_file <span class="ot">&lt;-</span> <span class="fu">download_only_once</span>(</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">url_base =</span> pavkovic_base, </span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">file_name =</span> <span class="st">&quot;fa_raw_counts.tsv.gz&quot;</span>,</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">local_folder =</span>  pavkovic_folder</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="do">## Normalised transcriptome data table</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>local_fa_norm_file <span class="ot">&lt;-</span> <span class="fu">download_only_once</span>(</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">url_base =</span> pavkovic_base, </span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">file_name =</span> <span class="st">&quot;fa_normalized_counts.tsv.gz&quot;</span>,</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">local_folder =</span>  pavkovic_folder</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a><span class="do">## Transcriptome metadata</span></span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>fa_metadata_file <span class="ot">&lt;-</span> <span class="fu">download_only_once</span>(</span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">url_base =</span> pavkovic_base, </span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">file_name =</span> <span class="st">&quot;fa_transcriptome_metadata.tsv&quot;</span>,</span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">local_folder =</span>  pavkovic_folder</span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a><span class="do">## FA proteome data table</span></span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a>local_pfa_file <span class="ot">&lt;-</span> <span class="fu">download_only_once</span>(</span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">url_base =</span> pavkovic_base, </span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a>  <span class="at">file_name =</span> <span class="st">&quot;pfa_model_counts.tsv.gz&quot;</span>,</span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a>  <span class="at">local_folder =</span>  pavkovic_folder</span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true" tabindex="-1"></a><span class="do">## FA proteome metadata</span></span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true" tabindex="-1"></a>pfa_metadata_file <span class="ot">&lt;-</span> <span class="fu">download_only_once</span>(</span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true" tabindex="-1"></a>  <span class="at">url_base =</span> pavkovic_base, </span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true" tabindex="-1"></a>  <span class="at">file_name =</span> <span class="st">&quot;pfa_proteome_metadata.tsv&quot;</span>,</span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true" tabindex="-1"></a>  <span class="at">local_folder =</span>  pavkovic_folder</span>
<span id="cb7-41"><a href="#cb7-41" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb7-42"><a href="#cb7-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-43"><a href="#cb7-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-44"><a href="#cb7-44" aria-hidden="true" tabindex="-1"></a><span class="do">## UUO proteome data table</span></span>
<span id="cb7-45"><a href="#cb7-45" aria-hidden="true" tabindex="-1"></a>local_puuo_file <span class="ot">&lt;-</span> <span class="fu">download_only_once</span>(</span>
<span id="cb7-46"><a href="#cb7-46" aria-hidden="true" tabindex="-1"></a>  <span class="at">url_base =</span> pavkovic_base, </span>
<span id="cb7-47"><a href="#cb7-47" aria-hidden="true" tabindex="-1"></a>  <span class="at">file_name =</span> <span class="st">&quot;puuo_model_counts.tsv.gz&quot;</span>,</span>
<span id="cb7-48"><a href="#cb7-48" aria-hidden="true" tabindex="-1"></a>  <span class="at">local_folder =</span>  pavkovic_folder</span>
<span id="cb7-49"><a href="#cb7-49" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>After having run the chunk of code above, try to re-run it. In principle, you should just receive messages telling you that the files are already there.</p>
</div>
<div id="loading-the-files" class="section level3">
<h3>Loading the files</h3>
<p>We now write a function <code>load_fix_row_names()</code> that loads a data file and takes a specified column as row names, whilst automatically fixing potential problems due to duplicate labels in this column.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @title Load a tab-separated value file and manually set row ames after having forced them to unique values</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @author Jacques van Helden email{Jacques.van-Helden@@france-bioinformatique.fr}</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @param file file path</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @param header=1 Header is set to 1 by default</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @param sep=&quot;\t&quot; Column separator is set to tab by default</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @param rownames.col=1 Column containing the row names</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @param ... all other parameters are passed to read.delim()</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; @return a data frame with the loaded data</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>load_fix_row_names <span class="ot">&lt;-</span> <span class="cf">function</span>(file, </span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>                       <span class="at">header =</span> <span class="dv">1</span>, </span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>                       <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>,</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>                       <span class="at">rownames.col =</span> <span class="dv">1</span>, </span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>                       ...) {</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="at">file =</span> file, ...)</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(x) <span class="ot">&lt;-</span> <span class="fu">make.names</span>(x[, rownames.col], <span class="at">unique =</span> <span class="cn">TRUE</span>)</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> x[, <span class="sc">-</span>rownames.col]</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(x)</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>We can now load the data that from our local folder.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Load transcriptome data (raw counts)</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>fa_by_read_delim <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="at">file =</span> local_fa_file, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Check the first lines of the loaded file</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co"># dim(fa)</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="st">&quot;Loaded raw counts of FA transcriptome raw counts with read_delim(): &quot;</span>, </span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>        <span class="fu">nrow</span>(fa_by_read_delim), <span class="st">&quot; rows x &quot;</span>, <span class="fu">ncol</span>(fa_by_read_delim), <span class="st">&quot; columns&quot;</span>)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(<span class="fu">head</span>(fa_by_read_delim), <span class="at">caption =</span> <span class="st">&quot;FA transcriptome loaded with read.delim()&quot;</span>)</span></code></pre></div>
<table style="width:100%;">
<caption>FA transcriptome loaded with read.delim()</caption>
<colgroup>
<col width="8%" />
<col width="4%" />
<col width="5%" />
<col width="5%" />
<col width="4%" />
<col width="4%" />
<col width="5%" />
<col width="5%" />
<col width="4%" />
<col width="4%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="4%" />
<col width="4%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">rowname</th>
<th align="right">day1_1</th>
<th align="right">day1_2</th>
<th align="right">day1_3</th>
<th align="right">day14_1</th>
<th align="right">day14_2</th>
<th align="right">day14_3</th>
<th align="right">day2_1</th>
<th align="right">day2_2</th>
<th align="right">day2_3</th>
<th align="right">day3_1</th>
<th align="right">day3_2</th>
<th align="right">day3_3</th>
<th align="right">day7_1</th>
<th align="right">day7_2</th>
<th align="right">day7_3</th>
<th align="right">normal_1</th>
<th align="right">normal_2</th>
<th align="right">normal_3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ENSMUSG00000000001</td>
<td align="right">2278.80022</td>
<td align="right">1786.498848</td>
<td align="right">2368.618959</td>
<td align="right">627.758017</td>
<td align="right">559.156031</td>
<td align="right">611.4338605</td>
<td align="right">2145.223456</td>
<td align="right">262.454849</td>
<td align="right">745.843632</td>
<td align="right">987.1850529</td>
<td align="right">1077.645231</td>
<td align="right">1335.116771</td>
<td align="right">1096.075988</td>
<td align="right">1035.8458456</td>
<td align="right">1090.0375905</td>
<td align="right">483.2298191</td>
<td align="right">1842.14841</td>
<td align="right">475.696800</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000003</td>
<td align="right">0.00000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.0000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.0000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.00000</td>
<td align="right">0.000000</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000028</td>
<td align="right">36.27547</td>
<td align="right">22.147861</td>
<td align="right">39.484949</td>
<td align="right">14.470759</td>
<td align="right">10.167813</td>
<td align="right">31.6910193</td>
<td align="right">300.558779</td>
<td align="right">4.771672</td>
<td align="right">123.896184</td>
<td align="right">51.8555945</td>
<td align="right">8.434511</td>
<td align="right">69.936866</td>
<td align="right">6.665195</td>
<td align="right">6.9552273</td>
<td align="right">42.5783251</td>
<td align="right">7.3515305</td>
<td align="right">11.19676</td>
<td align="right">1.034465</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000031</td>
<td align="right">13.18853</td>
<td align="right">7.151932</td>
<td align="right">1.115304</td>
<td align="right">0.867429</td>
<td align="right">0.000000</td>
<td align="right">0.0000000</td>
<td align="right">1.711944</td>
<td align="right">0.000000</td>
<td align="right">5.260747</td>
<td align="right">0.8022308</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.8489214</td>
<td align="right">1.7106814</td>
<td align="right">0.8603067</td>
<td align="right">0.00000</td>
<td align="right">0.000000</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000037</td>
<td align="right">0.00000</td>
<td align="right">27.903213</td>
<td align="right">6.897842</td>
<td align="right">5.692254</td>
<td align="right">1.901719</td>
<td align="right">0.6549762</td>
<td align="right">57.382077</td>
<td align="right">0.000000</td>
<td align="right">38.898261</td>
<td align="right">8.9308534</td>
<td align="right">6.966606</td>
<td align="right">0.000000</td>
<td align="right">7.939323</td>
<td align="right">101.6481244</td>
<td align="right">0.6500858</td>
<td align="right">32.0575944</td>
<td align="right">10.42322</td>
<td align="right">0.000000</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000049</td>
<td align="right">30.86001</td>
<td align="right">4.861367</td>
<td align="right">51.466810</td>
<td align="right">26.152649</td>
<td align="right">1.968290</td>
<td align="right">55.8319868</td>
<td align="right">10.796186</td>
<td align="right">2.747397</td>
<td align="right">1.805883</td>
<td align="right">0.9468345</td>
<td align="right">26.954566</td>
<td align="right">7.666032</td>
<td align="right">32.235700</td>
<td align="right">6.7001171</td>
<td align="right">33.9132330</td>
<td align="right">27.7647071</td>
<td align="right">38.42445</td>
<td align="right">15.903837</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Load the same data with load_fix_row_names</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>fa <span class="ot">&lt;-</span> <span class="fu">load_fix_row_names</span>(<span class="at">file =</span> local_fa_file, <span class="at">rownames.col =</span> <span class="dv">1</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="st">&quot;Loaded raw counts of FA transcriptome raw counts with load_fix_row_names(): &quot;</span>, </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">nrow</span>(fa), <span class="st">&quot; rows x &quot;</span>, <span class="fu">ncol</span>(fa), <span class="st">&quot; columns&quot;</span>)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(<span class="fu">head</span>(fa), <span class="at">caption =</span> <span class="st">&quot;FA transcriptome loaded with load_fix_row_names()&quot;</span>)</span></code></pre></div>
<table style="width:100%;">
<caption>FA transcriptome loaded with load_fix_row_names()</caption>
<colgroup>
<col width="8%" />
<col width="4%" />
<col width="5%" />
<col width="5%" />
<col width="4%" />
<col width="4%" />
<col width="5%" />
<col width="5%" />
<col width="4%" />
<col width="4%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="4%" />
<col width="4%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">day1_1</th>
<th align="right">day1_2</th>
<th align="right">day1_3</th>
<th align="right">day14_1</th>
<th align="right">day14_2</th>
<th align="right">day14_3</th>
<th align="right">day2_1</th>
<th align="right">day2_2</th>
<th align="right">day2_3</th>
<th align="right">day3_1</th>
<th align="right">day3_2</th>
<th align="right">day3_3</th>
<th align="right">day7_1</th>
<th align="right">day7_2</th>
<th align="right">day7_3</th>
<th align="right">normal_1</th>
<th align="right">normal_2</th>
<th align="right">normal_3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ENSMUSG00000000001</td>
<td align="right">2278.80022</td>
<td align="right">1786.498848</td>
<td align="right">2368.618959</td>
<td align="right">627.758017</td>
<td align="right">559.156031</td>
<td align="right">611.4338605</td>
<td align="right">2145.223456</td>
<td align="right">262.454849</td>
<td align="right">745.843632</td>
<td align="right">987.1850529</td>
<td align="right">1077.645231</td>
<td align="right">1335.116771</td>
<td align="right">1096.075988</td>
<td align="right">1035.8458456</td>
<td align="right">1090.0375905</td>
<td align="right">483.2298191</td>
<td align="right">1842.14841</td>
<td align="right">475.696800</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000003</td>
<td align="right">0.00000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.0000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.0000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.00000</td>
<td align="right">0.000000</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000028</td>
<td align="right">36.27547</td>
<td align="right">22.147861</td>
<td align="right">39.484949</td>
<td align="right">14.470759</td>
<td align="right">10.167813</td>
<td align="right">31.6910193</td>
<td align="right">300.558779</td>
<td align="right">4.771672</td>
<td align="right">123.896184</td>
<td align="right">51.8555945</td>
<td align="right">8.434511</td>
<td align="right">69.936866</td>
<td align="right">6.665195</td>
<td align="right">6.9552273</td>
<td align="right">42.5783251</td>
<td align="right">7.3515305</td>
<td align="right">11.19676</td>
<td align="right">1.034465</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000031</td>
<td align="right">13.18853</td>
<td align="right">7.151932</td>
<td align="right">1.115304</td>
<td align="right">0.867429</td>
<td align="right">0.000000</td>
<td align="right">0.0000000</td>
<td align="right">1.711944</td>
<td align="right">0.000000</td>
<td align="right">5.260747</td>
<td align="right">0.8022308</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.8489214</td>
<td align="right">1.7106814</td>
<td align="right">0.8603067</td>
<td align="right">0.00000</td>
<td align="right">0.000000</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000037</td>
<td align="right">0.00000</td>
<td align="right">27.903213</td>
<td align="right">6.897842</td>
<td align="right">5.692254</td>
<td align="right">1.901719</td>
<td align="right">0.6549762</td>
<td align="right">57.382077</td>
<td align="right">0.000000</td>
<td align="right">38.898261</td>
<td align="right">8.9308534</td>
<td align="right">6.966606</td>
<td align="right">0.000000</td>
<td align="right">7.939323</td>
<td align="right">101.6481244</td>
<td align="right">0.6500858</td>
<td align="right">32.0575944</td>
<td align="right">10.42322</td>
<td align="right">0.000000</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000049</td>
<td align="right">30.86001</td>
<td align="right">4.861367</td>
<td align="right">51.466810</td>
<td align="right">26.152649</td>
<td align="right">1.968290</td>
<td align="right">55.8319868</td>
<td align="right">10.796186</td>
<td align="right">2.747397</td>
<td align="right">1.805883</td>
<td align="right">0.9468345</td>
<td align="right">26.954566</td>
<td align="right">7.666032</td>
<td align="right">32.235700</td>
<td align="right">6.7001171</td>
<td align="right">33.9132330</td>
<td align="right">27.7647071</td>
<td align="right">38.42445</td>
<td align="right">15.903837</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># hist(unlist(fa), breaks = 100)</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Load normalised data with load_fix_row_names</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>fa_norm <span class="ot">&lt;-</span> <span class="fu">load_fix_row_names</span>(<span class="at">file =</span> local_fa_norm_file, <span class="at">rownames.col =</span> <span class="dv">1</span>)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="st">&quot;Loaded raw counts of FA transcriptome normalised counts with load_fix_row_names(): &quot;</span>, </span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">nrow</span>(fa_norm), <span class="st">&quot; rows x &quot;</span>, <span class="fu">ncol</span>(fa_norm), <span class="st">&quot; columns&quot;</span>)</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="co"># dim(fa_norm)</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(<span class="fu">head</span>(fa_norm), <span class="at">caption =</span> <span class="st">&quot;First lines of the loaded normalised counts of FA transcriptome with load_fix_row_names()&quot;</span>)</span></code></pre></div>
<table>
<caption>First lines of the loaded normalised counts of FA transcriptome with load_fix_row_names()</caption>
<colgroup>
<col width="8%" />
<col width="4%" />
<col width="5%" />
<col width="5%" />
<col width="4%" />
<col width="4%" />
<col width="4%" />
<col width="5%" />
<col width="4%" />
<col width="4%" />
<col width="5%" />
<col width="5%" />
<col width="4%" />
<col width="4%" />
<col width="5%" />
<col width="5%" />
<col width="4%" />
<col width="5%" />
<col width="4%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">day1_1</th>
<th align="right">day1_2</th>
<th align="right">day1_3</th>
<th align="right">day14_1</th>
<th align="right">day14_2</th>
<th align="right">day14_3</th>
<th align="right">day2_1</th>
<th align="right">day2_2</th>
<th align="right">day2_3</th>
<th align="right">day3_1</th>
<th align="right">day3_2</th>
<th align="right">day3_3</th>
<th align="right">day7_1</th>
<th align="right">day7_2</th>
<th align="right">day7_3</th>
<th align="right">normal_1</th>
<th align="right">normal_2</th>
<th align="right">normal_3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ENSMUSG00000000001</td>
<td align="right">2711.22417</td>
<td align="right">1094.745589</td>
<td align="right">1331.5880610</td>
<td align="right">728.668808</td>
<td align="right">603.846446</td>
<td align="right">646.124076</td>
<td align="right">1185.658698</td>
<td align="right">1239.4695</td>
<td align="right">829.596362</td>
<td align="right">928.1069573</td>
<td align="right">1110.908359</td>
<td align="right">951.071644</td>
<td align="right">953.722863</td>
<td align="right">1008.0265279</td>
<td align="right">1001.4667413</td>
<td align="right">719.032852</td>
<td align="right">1247.299971</td>
<td align="right">809.375515</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000003</td>
<td align="right">0.00000</td>
<td align="right">0.000000</td>
<td align="right">0.0000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.0000</td>
<td align="right">0.000000</td>
<td align="right">0.0000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000028</td>
<td align="right">42.82759</td>
<td align="right">13.485108</td>
<td align="right">21.9214582</td>
<td align="right">16.244209</td>
<td align="right">10.802262</td>
<td align="right">33.839559</td>
<td align="right">166.379146</td>
<td align="right">23.6540</td>
<td align="right">137.895374</td>
<td align="right">48.8972257</td>
<td align="right">8.244218</td>
<td align="right">49.868925</td>
<td align="right">6.091296</td>
<td align="right">6.8109901</td>
<td align="right">39.5074036</td>
<td align="right">10.420766</td>
<td align="right">7.448588</td>
<td align="right">1.700369</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000031</td>
<td align="right">15.46552</td>
<td align="right">4.290716</td>
<td align="right">0.5620887</td>
<td align="right">1.160301</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">1.105509</td>
<td align="right">0.0000</td>
<td align="right">5.560297</td>
<td align="right">0.9403313</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.9729986</td>
<td align="right">1.8375537</td>
<td align="right">1.488681</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000000037</td>
<td align="right">0.00000</td>
<td align="right">17.162865</td>
<td align="right">3.9346207</td>
<td align="right">6.961804</td>
<td align="right">2.160452</td>
<td align="right">1.057486</td>
<td align="right">31.507014</td>
<td align="right">0.0000</td>
<td align="right">43.370319</td>
<td align="right">8.4629814</td>
<td align="right">7.213691</td>
<td align="right">0.000000</td>
<td align="right">6.961481</td>
<td align="right">99.2458551</td>
<td align="right">0.9187768</td>
<td align="right">47.637787</td>
<td align="right">6.771444</td>
<td align="right">0.000000</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000000049</td>
<td align="right">36.87931</td>
<td align="right">3.064797</td>
<td align="right">28.6665222</td>
<td align="right">30.167817</td>
<td align="right">2.160452</td>
<td align="right">59.219228</td>
<td align="right">6.080301</td>
<td align="right">14.1924</td>
<td align="right">2.224119</td>
<td align="right">0.9403313</td>
<td align="right">27.824235</td>
<td align="right">5.699306</td>
<td align="right">27.845923</td>
<td align="right">6.8109901</td>
<td align="right">31.2384121</td>
<td align="right">41.683064</td>
<td align="right">25.731487</td>
<td align="right">27.205900</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Load FA proteome data</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>pfa <span class="ot">&lt;-</span> <span class="fu">load_fix_row_names</span>(<span class="at">file =</span> local_pfa_file, <span class="at">rownames.col =</span> <span class="dv">1</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="st">&quot;Loaded raw counts of FA proteome (pfa)  with load_fix_row_names(): &quot;</span>, </span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">nrow</span>(pfa), <span class="st">&quot; rows x &quot;</span>, <span class="fu">ncol</span>(pfa), <span class="st">&quot; columns&quot;</span>)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="co"># dim(pfa)</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Load UUO proteome data</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>puuo <span class="ot">&lt;-</span> <span class="fu">load_fix_row_names</span>(<span class="at">file =</span> local_puuo_file, <span class="at">rownames.col =</span> <span class="dv">1</span>)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="st">&quot;Loaded raw counts of UUO proteome (puuo)  with load_fix_row_names(): &quot;</span>, </span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>        <span class="fu">nrow</span>(puuo), <span class="st">&quot; rows x &quot;</span>, <span class="fu">ncol</span>(puuo), <span class="st">&quot; columns&quot;</span>)</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a><span class="co"># dim(pfa)</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a><span class="do">## Check the first lines of the loaded file</span></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(<span class="fu">head</span>(pfa), <span class="at">caption =</span> <span class="st">&quot;First lines of the proteome table&quot;</span>)</span></code></pre></div>
<table>
<caption>First lines of the proteome table</caption>
<colgroup>
<col width="16%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="7%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">normal_1</th>
<th align="right">normal_2</th>
<th align="right">day1_1</th>
<th align="right">day1_2</th>
<th align="right">day2_1</th>
<th align="right">day2_2</th>
<th align="right">day7_1</th>
<th align="right">day7_2</th>
<th align="right">day14_1</th>
<th align="right">day14_2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ENSMUSG00000037686</td>
<td align="right">531.2680</td>
<td align="right">651.7200</td>
<td align="right">335.5910</td>
<td align="right">334.8460</td>
<td align="right">197.1740</td>
<td align="right">307.194</td>
<td align="right">123.2060</td>
<td align="right">272.6190</td>
<td align="right">93.7247</td>
<td align="right">196.1590</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000027831</td>
<td align="right">221.6020</td>
<td align="right">266.3590</td>
<td align="right">175.4090</td>
<td align="right">159.4190</td>
<td align="right">234.8080</td>
<td align="right">256.927</td>
<td align="right">149.9380</td>
<td align="right">315.0590</td>
<td align="right">110.5880</td>
<td align="right">126.3600</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000039201</td>
<td align="right">26.0723</td>
<td align="right">29.1331</td>
<td align="right">57.7329</td>
<td align="right">45.8475</td>
<td align="right">81.6009</td>
<td align="right">88.870</td>
<td align="right">29.8560</td>
<td align="right">44.5586</td>
<td align="right">13.6292</td>
<td align="right">27.6568</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000031095</td>
<td align="right">4363.0500</td>
<td align="right">4784.0800</td>
<td align="right">4064.4800</td>
<td align="right">3917.2900</td>
<td align="right">4599.0300</td>
<td align="right">5957.030</td>
<td align="right">2806.5200</td>
<td align="right">6792.0900</td>
<td align="right">2022.5100</td>
<td align="right">3226.7500</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000034931</td>
<td align="right">879.2790</td>
<td align="right">1065.3900</td>
<td align="right">914.2870</td>
<td align="right">928.2760</td>
<td align="right">1000.1000</td>
<td align="right">1264.270</td>
<td align="right">738.3520</td>
<td align="right">1362.0700</td>
<td align="right">466.4440</td>
<td align="right">714.5870</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000038208</td>
<td align="right">68.2225</td>
<td align="right">89.8871</td>
<td align="right">57.9041</td>
<td align="right">76.3510</td>
<td align="right">84.8474</td>
<td align="right">105.245</td>
<td align="right">72.8696</td>
<td align="right">138.9810</td>
<td align="right">40.8101</td>
<td align="right">59.2117</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Load proteome metadata</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>pfa_metadata <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="at">file =</span> pfa_metadata_file, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(pfa_metadata, <span class="at">caption =</span> <span class="st">&quot;Metadata for the FA proteome dataset&quot;</span>)</span></code></pre></div>
<table>
<caption>Metadata for the FA proteome dataset</caption>
<thead>
<tr class="header">
<th align="left">dataType</th>
<th align="left">sampleName</th>
<th align="left">condition</th>
<th align="right">sampleNumber</th>
<th align="left">color</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">proteome</td>
<td align="left">normal_1</td>
<td align="left">normal</td>
<td align="right">1</td>
<td align="left">#BBFFBB</td>
</tr>
<tr class="even">
<td align="left">proteome</td>
<td align="left">normal_2</td>
<td align="left">normal</td>
<td align="right">2</td>
<td align="left">#BBFFBB</td>
</tr>
<tr class="odd">
<td align="left">proteome</td>
<td align="left">day1_1</td>
<td align="left">day1</td>
<td align="right">1</td>
<td align="left">#FFFFDD</td>
</tr>
<tr class="even">
<td align="left">proteome</td>
<td align="left">day1_2</td>
<td align="left">day1</td>
<td align="right">2</td>
<td align="left">#FFFFDD</td>
</tr>
<tr class="odd">
<td align="left">proteome</td>
<td align="left">day2_1</td>
<td align="left">day2</td>
<td align="right">1</td>
<td align="left">#BBD7FF</td>
</tr>
<tr class="even">
<td align="left">proteome</td>
<td align="left">day2_2</td>
<td align="left">day2</td>
<td align="right">1</td>
<td align="left">#BBD7FF</td>
</tr>
<tr class="odd">
<td align="left">proteome</td>
<td align="left">day3_1</td>
<td align="left">day3</td>
<td align="right">1</td>
<td align="left">#F0BBFF</td>
</tr>
<tr class="even">
<td align="left">proteome</td>
<td align="left">day3_2</td>
<td align="left">day3</td>
<td align="right">2</td>
<td align="left">#F0BBFF</td>
</tr>
<tr class="odd">
<td align="left">proteome</td>
<td align="left">day7_1</td>
<td align="left">day7</td>
<td align="right">1</td>
<td align="left">#FFDD88</td>
</tr>
<tr class="even">
<td align="left">proteome</td>
<td align="left">day7_2</td>
<td align="left">day7</td>
<td align="right">2</td>
<td align="left">#FFDD88</td>
</tr>
<tr class="odd">
<td align="left">proteome</td>
<td align="left">day14_1</td>
<td align="left">day14</td>
<td align="right">1</td>
<td align="left">#FF4400</td>
</tr>
<tr class="even">
<td align="left">proteome</td>
<td align="left">day14_2</td>
<td align="left">day14</td>
<td align="right">2</td>
<td align="left">#FF4400</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Load transcriptome metadata</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>fa_metadata <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="at">file =</span> fa_metadata_file, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(fa_metadata, <span class="at">caption =</span> <span class="st">&quot;Metadata for the transcriptome dataset&quot;</span>)</span></code></pre></div>
<table>
<caption>Metadata for the transcriptome dataset</caption>
<thead>
<tr class="header">
<th align="left">dataType</th>
<th align="left">sampleName</th>
<th align="left">condition</th>
<th align="right">sampleNumber</th>
<th align="left">color</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">1 transcriptome</td>
<td align="left">day14_1</td>
<td align="left">day14</td>
<td align="right">1</td>
<td align="left">#FF4400</td>
</tr>
<tr class="even">
<td align="left">2 transcriptome</td>
<td align="left">day14_2</td>
<td align="left">day14</td>
<td align="right">2</td>
<td align="left">#FF4400</td>
</tr>
<tr class="odd">
<td align="left">3 transcriptome</td>
<td align="left">day14_3</td>
<td align="left">day14</td>
<td align="right">3</td>
<td align="left">#FF4400</td>
</tr>
<tr class="even">
<td align="left">4 transcriptome</td>
<td align="left">day1_1</td>
<td align="left">day1</td>
<td align="right">1</td>
<td align="left">#BBD7FF</td>
</tr>
<tr class="odd">
<td align="left">5 transcriptome</td>
<td align="left">day1_2</td>
<td align="left">day1</td>
<td align="right">2</td>
<td align="left">#BBD7FF</td>
</tr>
<tr class="even">
<td align="left">6 transcriptome</td>
<td align="left">day1_3</td>
<td align="left">day1</td>
<td align="right">3</td>
<td align="left">#BBD7FF</td>
</tr>
<tr class="odd">
<td align="left">7 transcriptome</td>
<td align="left">day2_1</td>
<td align="left">day1</td>
<td align="right">1</td>
<td align="left">#F0BBFF</td>
</tr>
<tr class="even">
<td align="left">8 transcriptome</td>
<td align="left">day2_2</td>
<td align="left">day1</td>
<td align="right">2</td>
<td align="left">#F0BBFF</td>
</tr>
<tr class="odd">
<td align="left">9 transcriptome</td>
<td align="left">day2_3</td>
<td align="left">day1</td>
<td align="right">3</td>
<td align="left">#F0BBFF</td>
</tr>
<tr class="even">
<td align="left">10 transcriptome</td>
<td align="left">day3_1</td>
<td align="left">day3</td>
<td align="right">1</td>
<td align="left">#FFFFDD</td>
</tr>
<tr class="odd">
<td align="left">11 transcriptome</td>
<td align="left">day3_2</td>
<td align="left">day3</td>
<td align="right">2</td>
<td align="left">#FFFFDD</td>
</tr>
<tr class="even">
<td align="left">12 transcriptome</td>
<td align="left">day3_3</td>
<td align="left">day3</td>
<td align="right">3</td>
<td align="left">#FFFFDD</td>
</tr>
<tr class="odd">
<td align="left">13 transcriptome</td>
<td align="left">day7_1</td>
<td align="left">day7</td>
<td align="right">1</td>
<td align="left">#FFDD88</td>
</tr>
<tr class="even">
<td align="left">14 transcriptome</td>
<td align="left">day7_2</td>
<td align="left">day7</td>
<td align="right">2</td>
<td align="left">#FFDD88</td>
</tr>
<tr class="odd">
<td align="left">15 transcriptome</td>
<td align="left">day7_3</td>
<td align="left">day7</td>
<td align="right">3</td>
<td align="left">#FFDD88</td>
</tr>
<tr class="even">
<td align="left">16 transcriptome</td>
<td align="left">normal_1</td>
<td align="left">normal</td>
<td align="right">1</td>
<td align="left">#BBFFBB</td>
</tr>
<tr class="odd">
<td align="left">17 transcriptome</td>
<td align="left">normal_2</td>
<td align="left">normal</td>
<td align="right">2</td>
<td align="left">#BBFFBB</td>
</tr>
<tr class="even">
<td align="left">18 transcriptome</td>
<td align="left">normal_3</td>
<td align="left">normal</td>
<td align="right">3</td>
<td align="left">#BBFFBB</td>
</tr>
</tbody>
</table>
<p>We can now use this function to download and load the different data files.</p>
</div>
</div>
<div id="graphical-exploration-of-the-data" class="section level2">
<h2>Graphical exploration of the data</h2>
<div id="histogram" class="section level3">
<h3>Histogram</h3>
<p>Draw histograms of the transcriptome and proteom data, all samples together.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>pfa_vector <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">as.vector</span>(pfa))</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(pfa_vector)</span></code></pre></div>
<p><img src="figures/pavkovic2019_fa_hist_ugly-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Let us now improve the histogram to get an intuition of our data. A first step is to increase the number of bins, with the</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(pfa_vector, <span class="at">breaks =</span> <span class="dv">200</span>)</span></code></pre></div>
<p><img src="figures/pavkovic2019_fa_hist_breaks-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The distribution is stronly right-skewed, and its most representative part is “compressed” on the left side of the graph. As shown below, a log2 transformation provides a more informative view of the data.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Plot the histogram of log2-transformed proteome data</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Note: we add an epsilon of 0.1 in order to avoid problems with null values</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">log2</span>(pfa_vector <span class="sc">+</span> <span class="fl">0.1</span>), <span class="at">breaks =</span> <span class="dv">100</span>)</span></code></pre></div>
<p><img src="figures/pavkovic2019_fa_hist_log2-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="enhancing-the-figure" class="section level3">
<h3>Enhancing the figure</h3>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Improve the histogram</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="do">## - add a title, axis labels, coloring, ...</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="do">## - increase readaility by writing each parameter on a separate line</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">log2</span>(pfa_vector <span class="sc">+</span> <span class="fl">0.1</span>), </span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="sc">-</span><span class="dv">4</span>, <span class="at">to =</span> <span class="dv">20</span>, <span class="at">by =</span> <span class="fl">0.1</span>),</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="st">&quot;#BBDDFF&quot;</span>,</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">border =</span> <span class="st">&quot;#88AAFF&quot;</span>,</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">las =</span> <span class="dv">1</span>, <span class="co"># I like to read the labels</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">&quot;log2(x)&quot;</span>, </span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">&quot;Number of observations&quot;</span>, </span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">&quot;Proteome, folic-acid treated samples&quot;</span>)</span></code></pre></div>
<p><img src="figures/pavkovic2019_fa_hist_log2_nicer-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="controlling-the-insertion-of-images-in-an-r-markdown-report" class="section level3">
<h3>Controlling the insertion of images in an R markdown report</h3>
<p>We will now add a some options to the header of the R chunk in the R markdown, in order to control the way this figure is inserted in the report. Note that the chunk header must come on a single line.</p>
<p><code>```{r fa_hist_log2_sized, out.width="75%", fig.width=7, fig.height=4, fig.cap="Distribution of log2-transformed values for the proteome of folic acid-treated samples (data from Pavcovic et al., 2019)"}</code></p>
<ul>
<li><p><code>fig.height</code> and <code>fig.width</code> specify the size of the figure generated by R. These parameters are convenient to ensure proper proportions between height and widths, but also to incresae the readability of the labels (if you increasa the size, the fonts will look smaller).</p></li>
<li><p><code>out.width</code> enables you to control the width occupied by your figure in the HTML or pdf report.</p></li>
<li><p><code>fig.cap</code> enables you to add a caption to the figure</p></li>
</ul>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Improve the histogram</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="do">## - add a title, axis labels, coloring, ...</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="do">## - increase readaility by writing each parameter on a separate line</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">log2</span>(pfa_vector), </span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">0</span>, <span class="at">to =</span> <span class="dv">20</span>, <span class="at">by =</span> <span class="fl">0.1</span>),</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="st">&quot;#BBDDFF&quot;</span>,</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">border =</span> <span class="st">&quot;#88AAFF&quot;</span>,</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">las =</span> <span class="dv">1</span>, <span class="co"># I like to read the labels</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">&quot;log2(x)&quot;</span>, </span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">&quot;Number of observations&quot;</span>, </span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">&quot;Proteome, folic-acid treated samples&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/pavkovic2019_fa_hist_log2_sized-1.png" alt="Distribution of log2-transformed values for the proteome of folic acid-treated samples (data from Pavcovic et al., 2019)" width="90%" />
<p class="caption">
Distribution of log2-transformed values for the proteome of folic acid-treated samples (data from Pavcovic et al., 2019)
</p>
</div>
</div>
</div>
<div id="box-plots" class="section level2">
<h2>Box plots</h2>
<p>Here is an example of code for a log2 boxplot ofthe fa data that uses the metadata tables to set up the colors.</p>
<div id="proteome-data" class="section level3">
<h3>Proteome data</h3>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Box plot of the transcriptome data before and after normalisation</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>par_ori <span class="ot">&lt;-</span> <span class="fu">par</span>(<span class="at">no.readonly =</span> <span class="cn">TRUE</span>) <span class="do">## Store the original parameters in order to restore them afterwards</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">6</span>, <span class="dv">5</span>, <span class="dv">1</span>)) <span class="do">## Adapt the margins of the boxplot in order to have enough space for sample labels</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)) <span class="do">## Two side-by-side panes</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Compute the log2 of the proteome data</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Note: we add an epsilon of 0.1 to avoid problems with null values</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>log2pfa <span class="ot">&lt;-</span> <span class="fu">log2</span>(pfa <span class="sc">+</span> <span class="fl">0.1</span>)</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(pfa, </span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> pfa_metadata<span class="sc">$</span>color,</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">horizontal =</span> <span class="cn">TRUE</span>, </span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">las =</span> <span class="dv">1</span>, </span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">&quot;Proteome, original values&quot;</span>, </span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">&quot;value&quot;</span>)</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(log2pfa, </span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> pfa_metadata<span class="sc">$</span>color,</span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">horizontal =</span> <span class="cn">TRUE</span>, </span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">las =</span> <span class="dv">1</span>, </span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">&quot;Proteome, log2-transformed&quot;</span>, </span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">&quot;log2(value)&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/pavkovic2019_proteome_boxplots-1.png" alt="Box plots of proteome data before (left) and after (right) log2 transformation. " width="90%" />
<p class="caption">
Box plots of proteome data before (left) and after (right) log2 transformation.
</p>
</div>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(par_ori)</span></code></pre></div>
<p>The box plots show the distribution of the data before (left) and after (right) log2 transformation.</p>
</div>
<div id="transcriptome-data" class="section level3">
<h3>Transcriptome data</h3>
<p>The transcriptome data is providedd</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Box plot of the transcriptome data before and after normalisation</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>par_ori <span class="ot">&lt;-</span> <span class="fu">par</span>(<span class="at">no.readonly =</span> <span class="cn">TRUE</span>) <span class="do">## Store the original parameters in order to restore them afterwards</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">6</span>, <span class="dv">5</span>, <span class="dv">1</span>)) <span class="do">## Adapt the margins of the boxplot in order to have enough space for sample labels</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>)) <span class="do">## 4 panels, with 2 rows and 2 columns</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Counts before and after normalisation, but no log2 transformation</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a><span class="do">## Raw counts</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(fa, </span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> fa_metadata<span class="sc">$</span>color,</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">horizontal =</span> <span class="cn">TRUE</span>, </span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">las =</span> <span class="dv">1</span>, </span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">&quot;FA transcriptome</span><span class="sc">\n</span><span class="st">raw counts&quot;</span>, </span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a><span class="do">## Normalised transcriptome data</span></span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(fa_norm, </span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> fa_metadata<span class="sc">$</span>color,</span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">horizontal =</span> <span class="cn">TRUE</span>, </span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">las =</span> <span class="dv">1</span>, </span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">&quot;FA transcriptome</span><span class="sc">\n</span><span class="st">normalised counts&quot;</span>, </span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-24"><a href="#cb22-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-25"><a href="#cb22-25" aria-hidden="true" tabindex="-1"></a><span class="do">## Compute the log2 of the transcriptome data</span></span>
<span id="cb22-26"><a href="#cb22-26" aria-hidden="true" tabindex="-1"></a><span class="do">## Note: we add an epsilon of 0.1 to avoid problems with null values</span></span>
<span id="cb22-27"><a href="#cb22-27" aria-hidden="true" tabindex="-1"></a>log2fa <span class="ot">&lt;-</span> <span class="fu">log2</span>(fa <span class="sc">+</span> <span class="fl">0.1</span>)</span>
<span id="cb22-28"><a href="#cb22-28" aria-hidden="true" tabindex="-1"></a>log2fa_norm <span class="ot">&lt;-</span> <span class="fu">log2</span>(fa_norm <span class="sc">+</span> <span class="fl">0.1</span>)</span>
<span id="cb22-29"><a href="#cb22-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-30"><a href="#cb22-30" aria-hidden="true" tabindex="-1"></a><span class="do">## Raw counts</span></span>
<span id="cb22-31"><a href="#cb22-31" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(log2fa, </span>
<span id="cb22-32"><a href="#cb22-32" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> fa_metadata<span class="sc">$</span>color,</span>
<span id="cb22-33"><a href="#cb22-33" aria-hidden="true" tabindex="-1"></a>        <span class="at">horizontal =</span> <span class="cn">TRUE</span>, </span>
<span id="cb22-34"><a href="#cb22-34" aria-hidden="true" tabindex="-1"></a>        <span class="at">las =</span> <span class="dv">1</span>, </span>
<span id="cb22-35"><a href="#cb22-35" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">&quot;FA transcriptome</span><span class="sc">\n</span><span class="st">log2raw (counts)&quot;</span>, </span>
<span id="cb22-36"><a href="#cb22-36" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">&quot;log2(counts)&quot;</span>)</span>
<span id="cb22-37"><a href="#cb22-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-38"><a href="#cb22-38" aria-hidden="true" tabindex="-1"></a><span class="do">## Normalised transcriptome data</span></span>
<span id="cb22-39"><a href="#cb22-39" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(log2fa_norm, </span>
<span id="cb22-40"><a href="#cb22-40" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> fa_metadata<span class="sc">$</span>color,</span>
<span id="cb22-41"><a href="#cb22-41" aria-hidden="true" tabindex="-1"></a>        <span class="at">horizontal =</span> <span class="cn">TRUE</span>, </span>
<span id="cb22-42"><a href="#cb22-42" aria-hidden="true" tabindex="-1"></a>        <span class="at">las =</span> <span class="dv">1</span>, </span>
<span id="cb22-43"><a href="#cb22-43" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">&quot;FA transcriptome</span><span class="sc">\n</span><span class="st">log2(normalised counts)&quot;</span>, </span>
<span id="cb22-44"><a href="#cb22-44" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">&quot;log2(normalised counts)&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/pavkovic2019_transcriptome_boxplots-1.png" alt="Box plots of transcriptome data before (left) and after (right) normalisation. All values are log2 transformed. " width="90%" />
<p class="caption">
Box plots of transcriptome data before (left) and after (right) normalisation. All values are log2 transformed.
</p>
</div>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(par_ori)</span></code></pre></div>
<div id="observations" class="section level4">
<h4>Observations</h4>
<ol style="list-style-type: decimal">
<li><p>The raw and normalised counts (top) show a strongly right-skewed distribution due to the presence of outliers. In each sample, a very few genes have very high values (25 miilion counts !). Such outliers can be really problematic because they will exert a very strong effect on the subsequent analyses (PCA, clustering, or any other analysis that relies on distances between samples). For PCA, we recommend to use the log2-transformed values.</p></li>
<li><p>The log2 transformation (bottom) mittigates the impact of these outliers. The bottom right panel shows that all the samples have the same third quartile (right side of the boxplot rectangle), thereby indicating the method used to standardise this method.</p></li>
<li><p>For all samples, the third quartile equals the minimal value, indicating that at least 25% of the genes had a null count.</p></li>
<li><p>The sample <em>day2_2</em> has a median value equal to its minimal value, indicating that at least 50% of the genes have null value in this sample.</p></li>
</ol>
</div>
</div>
</div>
<div id="exercise-pca-of-pavkovicz-data" class="section level2">
<h2>Exercise: PCA of Pavkovicz data</h2>
<p>As an exercise, we ask you to insert here your analysis of Pavkovicz data with the PCA methods presented in the session 4 of this course. Try to generate nice figures that will give you insight into the structure of the 4 datasets : transcriptome and proteome for the two respective treatments (folic acid and surgery, respectively).</p>
<p>For this, you should first download a copy of the R markdown source, install it in a local folder, and run knitr to compile an HTML or pdf report. If this works, you will start adding your own chunks of code and interpretation of the results.</p>
<p>Tips: - play with the scale options of the PC transformation - test if the 3rd and 4th components provide additional information than the 1st and 2nd components - compare the results obtained with the raw and normalised data sets, respectively - test the PCA with and without log2 transformation</p>
<p>It is not necessary to show all the results in this report, only select the most informative plots.</p>
<p>At the end, we expect to find here the most relevant figures resulting from this exploration, with a short explanation of - the parameters you chose for the analysis (normalised or not, log2 transformed or not, PC scaling or not, …) - some interpretation of the results in term of positioning of the samples on the components.</p>
<p>Don’t hesitate to call the team in case of trouble.</p>
</div>
<div id="save-a-memory-image-of-your-session" class="section level2">
<h2>Save a memory image of your session</h2>
<p>Use the function <code>save.image()</code> to store an image of your session, in a file <code>pavkovic_memory_image.Rdata</code> in the local folder specified above. This file will contain all the data that you loaded during this session, and enable you to reload everything without having to re-execute all the steps.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Define the path to the memory image file</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>memory_image_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(local_folder, <span class="st">&quot;pavkovic_memory_image.Rdata&quot;</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Save the memory image</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="fu">save.image</span>(<span class="at">file =</span> memory_image_file)</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="st">&quot;Memory image saved in file</span><span class="sc">\n\t</span><span class="st">&quot;</span>, memory_image_file)</span></code></pre></div>
<p>For a future session, you will be able to reload all the data with a single command:</p>
<p><code>load([path_to_your_memory_image])</code></p>
<p>You will need to replace <code>[path_to_your_memory_image]</code> by your actual path. In my case, the command becomes:</p>
<p><code>load("~/DUBii-m3_data/pavkovic_2019/pavkovic_memory_image.Rdata")</code></p>
</div>
<div id="save-your-session-info" class="section level2">
<h2>Save your session info</h2>
<p>For the sake of traceability, store the specifications of your R environment in the report, with the command <code>sessionInfo()</code>. This will indicate the version of R as wella sof all the libraries used in this notebook.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>R version 4.0.2 (2020-06-22)
Platform: x86_64-conda_cos6-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /shared/ifbstor1/software/miniconda/envs/r-4.0.2/lib/libopenblasp-r0.3.10.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] knitr_1.30

loaded via a namespace (and not attached):
 [1] compiler_4.0.2    magrittr_2.0.1    tools_4.0.2       htmltools_0.5.1.1 yaml_2.2.1        stringi_1.5.3     rmarkdown_2.5     highr_0.8         stringr_1.4.0     xfun_0.20         digest_0.6.27     rlang_0.4.10      evaluate_0.14    </code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
